{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1AfSrrXBI8oi",
        "outputId": "3b20254b-e106-40b9-8f89-a0aa03fba087"
      },
      "outputs": [],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 156,
      "metadata": {
        "id": "UWSHSonXlTeG"
      },
      "outputs": [],
      "source": [
        "# packages\n",
        "import torch\n",
        "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
        "import time"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BuXWNP8jcVkz"
      },
      "source": [
        "# 1. Pubtator API (TODO-Partially done)\n",
        "- using the input folder, check if the file(s) have been pre-annotated by checking if the following APIs return something\n",
        "```\n",
        "https://www.ncbi.nlm.nih.gov/research/pubtator3-api/publications/export/biocxml?pmids={id}\n",
        "or\n",
        "https://www.ncbi.nlm.nih.gov/research/pubtator3-api/publications/pmc_export/biocxml?pmcids=PMC{id}\n",
        "```\n",
        "  - if annotated, retrieve it and put it in the output folder\n",
        "  - if not, use the [example code](https://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/tmTools/download/ExampleCode.Python.zip) they provide to send it to the server to be annotated (Done)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0sl2nmCPFsQq"
      },
      "source": [
        "Using colab:\n",
        "\n",
        "```\n",
        "mkdir input\n",
        "```\n",
        "\n",
        "to make a input folder"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "edkmlyhFFifa",
        "outputId": "304ffdaa-8b04-4d0b-de90-f3caa4e5f27f"
      },
      "outputs": [],
      "source": [
        "!bash"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EQIruerqbU4Q",
        "outputId": "da090b16-ca20-4798-a26d-7bce4bc4090a"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
            "To disable this warning, you can either:\n",
            "\t- Avoid using `tokenizers` before the fork if possible\n",
            "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n"
          ]
        },
        {
          "ename": "UnboundLocalError",
          "evalue": "local variable 'child' referenced before assignment",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py:151\u001b[0m, in \u001b[0;36mProcessHandler.system\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    150\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 151\u001b[0m     child \u001b[38;5;241m=\u001b[39m \u001b[43mpexpect\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspawn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msh\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-c\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcmd\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# Vanilla Pexpect\u001b[39;00m\n\u001b[1;32m    152\u001b[0m flush \u001b[38;5;241m=\u001b[39m sys\u001b[38;5;241m.\u001b[39mstdout\u001b[38;5;241m.\u001b[39mflush\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:205\u001b[0m, in \u001b[0;36mspawn.__init__\u001b[0;34m(self, command, args, timeout, maxread, searchwindowsize, logfile, cwd, env, ignore_sighup, echo, preexec_fn, encoding, codec_errors, dimensions, use_poll)\u001b[0m\n\u001b[1;32m    204\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 205\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_spawn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcommand\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpreexec_fn\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdimensions\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    206\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39muse_poll \u001b[38;5;241m=\u001b[39m use_poll\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:303\u001b[0m, in \u001b[0;36mspawn._spawn\u001b[0;34m(self, command, args, preexec_fn, dimensions)\u001b[0m\n\u001b[1;32m    300\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39margs \u001b[38;5;241m=\u001b[39m [a \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(a, \u001b[38;5;28mbytes\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m a\u001b[38;5;241m.\u001b[39mencode(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mencoding)\n\u001b[1;32m    301\u001b[0m                  \u001b[38;5;28;01mfor\u001b[39;00m a \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39margs]\n\u001b[0;32m--> 303\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mptyproc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_spawnpty\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43menv\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menv\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    304\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mcwd\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcwd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    306\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mptyproc\u001b[38;5;241m.\u001b[39mpid\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:315\u001b[0m, in \u001b[0;36mspawn._spawnpty\u001b[0;34m(self, args, **kwargs)\u001b[0m\n\u001b[1;32m    314\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m'''Spawn a pty and return an instance of PtyProcess.'''\u001b[39;00m\n\u001b[0;32m--> 315\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mptyprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mPtyProcess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspawn\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ptyprocess/ptyprocess.py:315\u001b[0m, in \u001b[0;36mPtyProcess.spawn\u001b[0;34m(cls, argv, cwd, env, echo, preexec_fn, dimensions, pass_fds)\u001b[0m\n\u001b[1;32m    314\u001b[0m os\u001b[38;5;241m.\u001b[39mclose(exec_err_pipe_write)\n\u001b[0;32m--> 315\u001b[0m exec_err_data \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexec_err_pipe_read\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m4096\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    316\u001b[0m os\u001b[38;5;241m.\u001b[39mclose(exec_err_pipe_read)\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
            "\nDuring handling of the above exception, another exception occurred:\n",
            "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[311], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# SubmitText_request.py\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msystem\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpip install unidecode\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mrequests\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mos\u001b[39;00m\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py:657\u001b[0m, in \u001b[0;36mZMQInteractiveShell.system_piped\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    655\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39muser_ns[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_exit_code\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m system(cmd)\n\u001b[1;32m    656\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 657\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39muser_ns[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_exit_code\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43msystem\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar_expand\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcmd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdepth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py:167\u001b[0m, in \u001b[0;36mProcessHandler.system\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    162\u001b[0m         out_size \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(child\u001b[38;5;241m.\u001b[39mbefore)\n\u001b[1;32m    163\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m    164\u001b[0m     \u001b[38;5;66;03m# We need to send ^C to the process.  The ascii code for '^C' is 3\u001b[39;00m\n\u001b[1;32m    165\u001b[0m     \u001b[38;5;66;03m# (the character is known as ETX for 'End of Text', see\u001b[39;00m\n\u001b[1;32m    166\u001b[0m     \u001b[38;5;66;03m# curses.ascii.ETX).\u001b[39;00m\n\u001b[0;32m--> 167\u001b[0m     \u001b[43mchild\u001b[49m\u001b[38;5;241m.\u001b[39msendline(\u001b[38;5;28mchr\u001b[39m(\u001b[38;5;241m3\u001b[39m))\n\u001b[1;32m    168\u001b[0m     \u001b[38;5;66;03m# Read and print any more output the program might produce on its\u001b[39;00m\n\u001b[1;32m    169\u001b[0m     \u001b[38;5;66;03m# way out.\u001b[39;00m\n\u001b[1;32m    170\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
            "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'child' referenced before assignment"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "--- Logging error ---\n",
            "Traceback (most recent call last):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py\", line 151, in system\n",
            "    child = pexpect.spawn(self.sh, args=['-c', cmd])  # Vanilla Pexpect\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py\", line 205, in __init__\n",
            "    self._spawn(command, args, preexec_fn, dimensions)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py\", line 303, in _spawn\n",
            "    self.ptyproc = self._spawnpty(self.args, env=self.env,\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py\", line 315, in _spawnpty\n",
            "    return ptyprocess.PtyProcess.spawn(args, **kwargs)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ptyprocess/ptyprocess.py\", line 289, in spawn\n",
            "    os.execv(command, argv)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py\", line 743, in new_execv\n",
            "    send_process_about_to_be_replaced()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py\", line 985, in send_process_about_to_be_replaced\n",
            "    py_db.send_process_about_to_be_replaced()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/pydevd.py\", line 2043, in send_process_about_to_be_replaced\n",
            "    time.sleep(0.05)\n",
            "KeyboardInterrupt\n",
            "\n",
            "During handling of the above exception, another exception occurred:\n",
            "\n",
            "Traceback (most recent call last):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n",
            "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
            "  File \"/var/folders/s7/c_3jrk0j61l1nkjwz_167y9r0000gn/T/ipykernel_41327/1190397474.py\", line 2, in <module>\n",
            "    get_ipython().system('pip install unidecode')\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 657, in system_piped\n",
            "    self.user_ns[\"_exit_code\"] = system(self.var_expand(cmd, depth=1))\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py\", line 167, in system\n",
            "    child.sendline(chr(3))\n",
            "UnboundLocalError: local variable 'child' referenced before assignment\n",
            "\n",
            "During handling of the above exception, another exception occurred:\n",
            "\n",
            "Traceback (most recent call last):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/logging/__init__.py\", line 1103, in emit\n",
            "    stream.write(msg + self.terminator)\n",
            "OSError: [Errno 9] Bad file descriptor\n",
            "Call stack:\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n",
            "    return _run_code(code, main_globals, None,\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/runpy.py\", line 86, in _run_code\n",
            "    exec(code, run_globals)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
            "    app.launch_new_instance()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
            "    app.start()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
            "    self.io_loop.start()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
            "    self.asyncio_loop.run_forever()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py\", line 600, in run_forever\n",
            "    self._run_once()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py\", line 1896, in _run_once\n",
            "    handle._run()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/asyncio/events.py\", line 80, in _run\n",
            "    self._context.run(self._callback, *self._args)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
            "    await self.process_one()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
            "    await dispatch(*args)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
            "    await result\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
            "    await super().execute_request(stream, ident, parent)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
            "    reply_content = await reply_content\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
            "    res = shell.run_cell(\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
            "    return super().run_cell(*args, **kwargs)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n",
            "    result = self._run_cell(\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n",
            "    result = runner(coro)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
            "    coro.send(None)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n",
            "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n",
            "    if await self.run_code(code, result, async_=asy):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3599, in run_code\n",
            "    self.showtraceback(running_compiled_code=True)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 2179, in showtraceback\n",
            "    self._showtraceback(etype, value, stb)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 569, in _showtraceback\n",
            "    dh.session.send(  # type:ignore[attr-defined]\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/jupyter_client/session.py\", line 831, in send\n",
            "    get_logger().warning(\"WARNING: attempted to send message from fork\\n%s\", msg)\n",
            "Message: 'WARNING: attempted to send message from fork\\n%s'\n",
            "Arguments: {'header': {'msg_id': '9135c61b-d20a3a8c8e9f0be36432a3fe_43609_2740', 'msg_type': 'error', 'username': 'crystalcho', 'session': '9135c61b-d20a3a8c8e9f0be36432a3fe', 'date': datetime.datetime(2025, 4, 3, 9, 36, 25, 503664, tzinfo=datetime.timezone.utc), 'version': '5.3'}, 'msg_id': '9135c61b-d20a3a8c8e9f0be36432a3fe_43609_2740', 'msg_type': 'error', 'parent_header': {'date': datetime.datetime(2025, 4, 3, 9, 36, 21, 895000, tzinfo=tzutc()), 'msg_id': '606dc216-4fe7-423d-b07f-b4ffd201bace', 'msg_type': 'execute_request', 'session': 'e3a89fb4-0c8b-4ecc-a37c-6312a14df9e6', 'username': '60edfaf0-327c-463c-a35d-0c9f29c77a14', 'version': '5.2'}, 'content': {'traceback': ['\\x1b[0;31m---------------------------------------------------------------------------\\x1b[0m', '\\x1b[0;31mKeyboardInterrupt\\x1b[0m                         Traceback (most recent call last)', \"File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py:151\\x1b[0m, in \\x1b[0;36mProcessHandler.system\\x1b[0;34m(self, cmd)\\x1b[0m\\n\\x1b[1;32m    150\\x1b[0m \\x1b[38;5;28;01melse\\x1b[39;00m:\\n\\x1b[0;32m--> 151\\x1b[0m     child \\x1b[38;5;241m=\\x1b[39m \\x1b[43mpexpect\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mspawn\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43msh\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43margs\\x1b[49m\\x1b[38;5;241;43m=\\x1b[39;49m\\x1b[43m[\\x1b[49m\\x1b[38;5;124;43m'\\x1b[39;49m\\x1b[38;5;124;43m-c\\x1b[39;49m\\x1b[38;5;124;43m'\\x1b[39;49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43mcmd\\x1b[49m\\x1b[43m]\\x1b[49m\\x1b[43m)\\x1b[49m  \\x1b[38;5;66;03m# Vanilla Pexpect\\x1b[39;00m\\n\\x1b[1;32m    152\\x1b[0m flush \\x1b[38;5;241m=\\x1b[39m sys\\x1b[38;5;241m.\\x1b[39mstdout\\x1b[38;5;241m.\\x1b[39mflush\\n\", 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:205\\x1b[0m, in \\x1b[0;36mspawn.__init__\\x1b[0;34m(self, command, args, timeout, maxread, searchwindowsize, logfile, cwd, env, ignore_sighup, echo, preexec_fn, encoding, codec_errors, dimensions, use_poll)\\x1b[0m\\n\\x1b[1;32m    204\\x1b[0m \\x1b[38;5;28;01melse\\x1b[39;00m:\\n\\x1b[0;32m--> 205\\x1b[0m     \\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43m_spawn\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43mcommand\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43margs\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43mpreexec_fn\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43mdimensions\\x1b[49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m    206\\x1b[0m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39muse_poll \\x1b[38;5;241m=\\x1b[39m use_poll\\n', 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:303\\x1b[0m, in \\x1b[0;36mspawn._spawn\\x1b[0;34m(self, command, args, preexec_fn, dimensions)\\x1b[0m\\n\\x1b[1;32m    300\\x1b[0m     \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39margs \\x1b[38;5;241m=\\x1b[39m [a \\x1b[38;5;28;01mif\\x1b[39;00m \\x1b[38;5;28misinstance\\x1b[39m(a, \\x1b[38;5;28mbytes\\x1b[39m) \\x1b[38;5;28;01melse\\x1b[39;00m a\\x1b[38;5;241m.\\x1b[39mencode(\\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39mencoding)\\n\\x1b[1;32m    301\\x1b[0m                  \\x1b[38;5;28;01mfor\\x1b[39;00m a \\x1b[38;5;129;01min\\x1b[39;00m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39margs]\\n\\x1b[0;32m--> 303\\x1b[0m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39mptyproc \\x1b[38;5;241m=\\x1b[39m \\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43m_spawnpty\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43margs\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43menv\\x1b[49m\\x1b[38;5;241;43m=\\x1b[39;49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43menv\\x1b[49m\\x1b[43m,\\x1b[49m\\n\\x1b[1;32m    304\\x1b[0m \\x1b[43m                             \\x1b[49m\\x1b[43mcwd\\x1b[49m\\x1b[38;5;241;43m=\\x1b[39;49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mcwd\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[38;5;241;43m*\\x1b[39;49m\\x1b[38;5;241;43m*\\x1b[39;49m\\x1b[43mkwargs\\x1b[49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m    306\\x1b[0m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39mpid \\x1b[38;5;241m=\\x1b[39m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39mptyproc\\x1b[38;5;241m.\\x1b[39mpid\\n', \"File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:315\\x1b[0m, in \\x1b[0;36mspawn._spawnpty\\x1b[0;34m(self, args, **kwargs)\\x1b[0m\\n\\x1b[1;32m    314\\x1b[0m \\x1b[38;5;250m\\x1b[39m\\x1b[38;5;124;03m'''Spawn a pty and return an instance of PtyProcess.'''\\x1b[39;00m\\n\\x1b[0;32m--> 315\\x1b[0m \\x1b[38;5;28;01mreturn\\x1b[39;00m \\x1b[43mptyprocess\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mPtyProcess\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mspawn\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43margs\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[38;5;241;43m*\\x1b[39;49m\\x1b[38;5;241;43m*\\x1b[39;49m\\x1b[43mkwargs\\x1b[49m\\x1b[43m)\\x1b[49m\\n\", 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ptyprocess/ptyprocess.py:289\\x1b[0m, in \\x1b[0;36mPtyProcess.spawn\\x1b[0;34m(cls, argv, cwd, env, echo, preexec_fn, dimensions, pass_fds)\\x1b[0m\\n\\x1b[1;32m    288\\x1b[0m \\x1b[38;5;28;01mif\\x1b[39;00m env \\x1b[38;5;129;01mis\\x1b[39;00m \\x1b[38;5;28;01mNone\\x1b[39;00m:\\n\\x1b[0;32m--> 289\\x1b[0m     \\x1b[43mos\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mexecv\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43mcommand\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43margv\\x1b[49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m    290\\x1b[0m \\x1b[38;5;28;01melse\\x1b[39;00m:\\n', 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py:743\\x1b[0m, in \\x1b[0;36mcreate_execv.<locals>.new_execv\\x1b[0;34m(path, args)\\x1b[0m\\n\\x1b[1;32m    742\\x1b[0m     send_process_created_message()\\n\\x1b[0;32m--> 743\\x1b[0m     \\x1b[43msend_process_about_to_be_replaced\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m    745\\x1b[0m \\x1b[38;5;28;01mreturn\\x1b[39;00m \\x1b[38;5;28mgetattr\\x1b[39m(os, original_name)(path, args)\\n', 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py:985\\x1b[0m, in \\x1b[0;36msend_process_about_to_be_replaced\\x1b[0;34m()\\x1b[0m\\n\\x1b[1;32m    984\\x1b[0m \\x1b[38;5;28;01mif\\x1b[39;00m py_db \\x1b[38;5;129;01mis\\x1b[39;00m \\x1b[38;5;129;01mnot\\x1b[39;00m \\x1b[38;5;28;01mNone\\x1b[39;00m:\\n\\x1b[0;32m--> 985\\x1b[0m     \\x1b[43mpy_db\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43msend_process_about_to_be_replaced\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43m)\\x1b[49m\\n', 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/pydevd.py:2043\\x1b[0m, in \\x1b[0;36mPyDB.send_process_about_to_be_replaced\\x1b[0;34m(self)\\x1b[0m\\n\\x1b[1;32m   2042\\x1b[0m \\x1b[38;5;28;01mwhile\\x1b[39;00m \\x1b[38;5;129;01mnot\\x1b[39;00m sent[\\x1b[38;5;241m0\\x1b[39m]:\\n\\x1b[0;32m-> 2043\\x1b[0m     \\x1b[43mtime\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43msleep\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;241;43m0.05\\x1b[39;49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m   2045\\x1b[0m     \\x1b[38;5;28;01mif\\x1b[39;00m (time\\x1b[38;5;241m.\\x1b[39mtime() \\x1b[38;5;241m-\\x1b[39m initial_time) \\x1b[38;5;241m>\\x1b[39m timeout:\\n', '\\x1b[0;31mKeyboardInterrupt\\x1b[0m: ', '\\nDuring handling of the above exception, another exception occurred:\\n', '\\x1b[0;31mUnboundLocalError\\x1b[0m                         Traceback (most recent call last)', \"Cell \\x1b[0;32mIn[311], line 2\\x1b[0m\\n\\x1b[1;32m      1\\x1b[0m \\x1b[38;5;66;03m# SubmitText_request.py\\x1b[39;00m\\n\\x1b[0;32m----> 2\\x1b[0m \\x1b[43mget_ipython\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43m)\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43msystem\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;124;43m'\\x1b[39;49m\\x1b[38;5;124;43mpip install unidecode\\x1b[39;49m\\x1b[38;5;124;43m'\\x1b[39;49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m      4\\x1b[0m \\x1b[38;5;28;01mimport\\x1b[39;00m\\x1b[38;5;250m \\x1b[39m\\x1b[38;5;21;01mrequests\\x1b[39;00m\\n\\x1b[1;32m      5\\x1b[0m \\x1b[38;5;28;01mimport\\x1b[39;00m\\x1b[38;5;250m \\x1b[39m\\x1b[38;5;21;01mos\\x1b[39;00m\\n\", 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py:657\\x1b[0m, in \\x1b[0;36mZMQInteractiveShell.system_piped\\x1b[0;34m(self, cmd)\\x1b[0m\\n\\x1b[1;32m    655\\x1b[0m         \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39muser_ns[\\x1b[38;5;124m\"\\x1b[39m\\x1b[38;5;124m_exit_code\\x1b[39m\\x1b[38;5;124m\"\\x1b[39m] \\x1b[38;5;241m=\\x1b[39m system(cmd)\\n\\x1b[1;32m    656\\x1b[0m \\x1b[38;5;28;01melse\\x1b[39;00m:\\n\\x1b[0;32m--> 657\\x1b[0m     \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39muser_ns[\\x1b[38;5;124m\"\\x1b[39m\\x1b[38;5;124m_exit_code\\x1b[39m\\x1b[38;5;124m\"\\x1b[39m] \\x1b[38;5;241m=\\x1b[39m \\x1b[43msystem\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mvar_expand\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43mcmd\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43mdepth\\x1b[49m\\x1b[38;5;241;43m=\\x1b[39;49m\\x1b[38;5;241;43m1\\x1b[39;49m\\x1b[43m)\\x1b[49m\\x1b[43m)\\x1b[49m\\n', \"File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py:167\\x1b[0m, in \\x1b[0;36mProcessHandler.system\\x1b[0;34m(self, cmd)\\x1b[0m\\n\\x1b[1;32m    162\\x1b[0m         out_size \\x1b[38;5;241m=\\x1b[39m \\x1b[38;5;28mlen\\x1b[39m(child\\x1b[38;5;241m.\\x1b[39mbefore)\\n\\x1b[1;32m    163\\x1b[0m \\x1b[38;5;28;01mexcept\\x1b[39;00m \\x1b[38;5;167;01mKeyboardInterrupt\\x1b[39;00m:\\n\\x1b[1;32m    164\\x1b[0m     \\x1b[38;5;66;03m# We need to send ^C to the process.  The ascii code for '^C' is 3\\x1b[39;00m\\n\\x1b[1;32m    165\\x1b[0m     \\x1b[38;5;66;03m# (the character is known as ETX for 'End of Text', see\\x1b[39;00m\\n\\x1b[1;32m    166\\x1b[0m     \\x1b[38;5;66;03m# curses.ascii.ETX).\\x1b[39;00m\\n\\x1b[0;32m--> 167\\x1b[0m     \\x1b[43mchild\\x1b[49m\\x1b[38;5;241m.\\x1b[39msendline(\\x1b[38;5;28mchr\\x1b[39m(\\x1b[38;5;241m3\\x1b[39m))\\n\\x1b[1;32m    168\\x1b[0m     \\x1b[38;5;66;03m# Read and print any more output the program might produce on its\\x1b[39;00m\\n\\x1b[1;32m    169\\x1b[0m     \\x1b[38;5;66;03m# way out.\\x1b[39;00m\\n\\x1b[1;32m    170\\x1b[0m     \\x1b[38;5;28;01mtry\\x1b[39;00m:\\n\", \"\\x1b[0;31mUnboundLocalError\\x1b[0m: local variable 'child' referenced before assignment\"], 'ename': 'UnboundLocalError', 'evalue': \"local variable 'child' referenced before assignment\"}, 'metadata': {}}\n"
          ]
        }
      ],
      "source": [
        "# SubmitText_request.py\n",
        "!pip install unidecode\n",
        "\n",
        "import requests\n",
        "import os\n",
        "import sys\n",
        "from unidecode import unidecode\n",
        "import unicodedata\n",
        "import re\n",
        "\n",
        "def submit_text_request(input_folder, bioconcept, output_file_session_number):\n",
        "    pattern = r'[^0-9a-zA-Z\\!\\@\\#\\$\\%\\^\\&\\*\\(\\)\\_\\+\\{\\}\\|\\:\\\"\\<\\>\\?\\-\\=\\[\\]\\\\;\\'\\,\\.\\/ \\t\\n\\r]'\n",
        "    unicode_to_regular = {}\n",
        "    with open('lib/unicode.txt', 'r', encoding='utf-8') as input_file:\n",
        "        for line in input_file:\n",
        "            line = line.strip()\n",
        "            parts = line.split(\"\\t\")\n",
        "            if len(parts) == 2:\n",
        "                uni, reg = parts\n",
        "                if reg == '-whitespace-':\n",
        "                    unicode_to_regular[uni] = ' '\n",
        "                else:\n",
        "                    unicode_to_regular[uni] = reg\n",
        "\n",
        "    with open(output_file_session_number, 'w', encoding='utf-8') as output_file:\n",
        "        for filename in os.listdir(input_folder):\n",
        "            if filename.startswith('.'):\n",
        "                continue  # Skip hidden files\n",
        "            text_str = ''\n",
        "            with open(os.path.join(input_folder, filename), 'r', encoding='utf-8') as file_input:\n",
        "                for line in file_input:\n",
        "                    normalized_line = unicodedata.normalize('NFC', line)\n",
        "                    for uni, reg in unicode_to_regular.items():\n",
        "                        normalized_line = normalized_line.replace(uni, reg)\n",
        "                    normalized_line = unidecode(normalized_line)  # Assuming unidecode does similar work to Perl's version\n",
        "                    text_str += normalized_line\n",
        "            text_str = re.sub(pattern, ' ', text_str)\n",
        "            url = \"https://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/RESTful/request.cgi\"\n",
        "            response = requests.post(url, data={'text': text_str, 'bioconcept': bioconcept})\n",
        "\n",
        "            if response.status_code == 200:\n",
        "                session_number = response.json().get('id', '')\n",
        "                print(f\"Thanks for your submission. The session number is: {session_number}\")\n",
        "                output_file.write(f\"{session_number}\\t{filename}\\n\")\n",
        "            else:\n",
        "                print(f\"Error: HTTP {response.status_code} for {filename}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 158,
      "metadata": {
        "id": "KccYUHYJcEfG"
      },
      "outputs": [],
      "source": [
        "# SubmitText_retreive.py\n",
        "import requests\n",
        "import os\n",
        "import sys\n",
        "\n",
        "def submit_text_retrieve(input_folder, inputfile_session_number, output_folder):\n",
        "    # Ensure output directory exists\n",
        "    if not os.path.exists(output_folder):\n",
        "        os.makedirs(output_folder)\n",
        "\n",
        "    sn_hash = {}\n",
        "    with open(inputfile_session_number, 'r') as input_file:\n",
        "        for line in input_file:\n",
        "            line = line.strip()\n",
        "            if line:\n",
        "                session_number, input_filename = line.split(\"\\t\")\n",
        "                sn_hash[session_number] = input_filename\n",
        "\n",
        "    base_url = \"https://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/RESTful/retrieve.cgi\"\n",
        "\n",
        "    for session_number, input_filename in sn_hash.items():\n",
        "        output_file_path = os.path.join(output_folder, input_filename)\n",
        "        if os.path.exists(output_file_path):\n",
        "            print(f\"{output_file_path} - finished\")\n",
        "            continue\n",
        "\n",
        "        retrieve_url = f\"{base_url}?id={session_number}\"\n",
        "        response = requests.get(retrieve_url)\n",
        "        if response.status_code == 200:\n",
        "            with open(output_file_path, 'w', encoding='utf-8') as output_file:\n",
        "                output_file.write(response.text)\n",
        "            print(f\"{session_number} : Result is retrieved.\")\n",
        "        else:\n",
        "            print(f\"{session_number} : Error retrieving results - {response.status_code}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 159,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8_lpKfmLcaQ-",
        "outputId": "756447fc-8c07-43f6-f6a6-4db2fcb68e62"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Thanks for your submission. The session number is: A3E22D4AB6ACE4972C67\n",
            "Thanks for your submission. The session number is: 5F7EB4CF8B3CA805725B\n",
            "Thanks for your submission. The session number is: 78C52AEE0E16B061871A\n",
            "Thanks for your submission. The session number is: 6E147BA18018E5008A36\n",
            "Thanks for your submission. The session number is: 6079846D8D32CB31A6EB\n",
            "Thanks for your submission. The session number is: FD2A549E7A4A5A784569\n"
          ]
        }
      ],
      "source": [
        "input_folder = \"old_annotated\"       # pull xml,json, or pubtator formatted files here to be annotated\n",
        "bioconcept = \"Gene\"                                         # we only want the genes\n",
        "output_file_session_number = \"Session_Number.txt\"  # output file for session number which will be used to retrieve the results\n",
        "submit_text_request(input_folder, bioconcept, output_file_session_number)    # submit the text for annotation"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "02JFZ15QfHqt"
      },
      "source": [
        "## Running the code below takes time, so it might say error 400. Just rerun it a bit later."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 163,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "mmQYNjmJdzwO",
        "outputId": "0284e0d6-cf18-4d40-85f9-b49dd3e7890a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "A3E22D4AB6ACE4972C67 : Error retrieving results - 400\n",
            "5F7EB4CF8B3CA805725B : Error retrieving results - 400\n",
            "78C52AEE0E16B061871A : Error retrieving results - 400\n",
            "6E147BA18018E5008A36 : Error retrieving results - 400\n",
            "6079846D8D32CB31A6EB : Error retrieving results - 400\n",
            "FD2A549E7A4A5A784569 : Error retrieving results - 400\n"
          ]
        }
      ],
      "source": [
        "input_folder = \"old_annotated\"\n",
        "inputfile_session_number = \"Session_Number.txt\"\n",
        "output_folder = \"old_output\"\n",
        "submit_text_retrieve(input_folder, inputfile_session_number, output_folder)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "duMZk-ubdK0K"
      },
      "source": [
        "# 2. Relevance (Done)\n",
        "- This can be integrated with Gene Chunking. We can just check if the paper has genes (i.e check if there is an item in the dictionary @ line 171)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S-Gak1R_RF8Y"
      },
      "source": [
        "*3. Gene Chunking*\n",
        "=============\n",
        "Given a **preannotated** article (check input folder) analyze the **XML** file via `parse_xml_file(xml_file_path)` which returns a dictionary of \"Gene Class\" from that file  \n",
        "\n",
        "The Gene Class has the following information:\n",
        "- self.gene_id: the gene id (from the xml file)\n",
        "- self.occurences: A list of occurences of the gene in the text. Each occurence is **3-sentences long**.\n",
        "    - The second sentence contains the explicit mention of the gene.\n",
        "    - The first and third sentences are the context of the gene mention.\n",
        "- self.symbol: the gene symbol (from NCBI. Initialized to None)\n",
        "- self.organism: the organism of the gene (from NCBI. Initialized to None)\n",
        "- self.full_name: the full name of the gene (from NCBI. Initialized to None. Set to name_from_article if not found in NCBI)\n",
        "- self.also_known_as: a list of other names of the gene (from NCBI. Initialized to None. Set to name_from_article if not found in NCBI)\n",
        "- self.name_from_article: the name of the gene as found in the article\n",
        "\n",
        "Steps:\n",
        "1. define the class Gene\n",
        "2. Parse the xml file\n",
        "3. update the Gene class with the information from NCBI\n",
        "Now you have a dictionary of Gene classes.~\n",
        "\n",
        "TIPS:\n",
        "- control + F \"sentence_buffer\" to find where you can adjust the buffer"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 556,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "M9sSOV2PXqOJ",
        "outputId": "6722699e-967f-4dee-c51c-fce74a88168e"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
            "To disable this warning, you can either:\n",
            "\t- Avoid using `tokenizers` before the fork if possible\n",
            "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
            "1322870.84s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: biopython in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (1.85)\n",
            "Requirement already satisfied: numpy in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from biopython) (2.2.2)\n",
            "\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip available: \u001b[0m\u001b[31;49m22.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
            "There are 34 genes in the paper.\n",
            "Gene(2638)\n",
            "  Symbol          : GC\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : GC vitamin D binding protein\n",
            "  Also Known As   : DBP, DBP-maf, DBP/GC, GRD3, Gc-MAF, GcMAF, HEL-S-51, VDB, VDBG, VDBP\n",
            "  In-text Name    : vitamin D binding protein\n",
            "  Occurrences     : ['Recent research activities have provided new insights in vitamin D metabolism in various conditions. Furthermore, substantial progress has been made in the analysis of vitamin D metabolites and related biomarkers, such as vitamin D binding protein. Liquid chromatography tandem mass spectrometric (LC-MS/MS) methods are capable of accurately measuring multiple vitamin D metabolites in parallel.', 'When interpreted together, they seem to provide a functional measure of vitamin D metabolism beyond the analysis of 25(OH)D alone. The determination of VDBP, free and bioavailable 25(OH)D is compromised by unresolved analytical issues, lacking reference intervals and insufficient clinical data. Therefore, future research activities should focus on analytical standardization and exploration of their clinical value.', 'Furthermore, vitamin D deficiency has been linked to a broad range of clinical conditions including cardiovascular disease, malignancies, autoimmune diseases, neuropsychiatric diseases and endocrinopathies. Substantial progress has also been made in the analysis of vitamin D metabolites and related biomarkers, such as vitamin D binding protein (VDBP). Mass spectrometric methods are capable of measuring multiple vitamin D metabolites in parallel.', 'However, recent evidence from experimental and clinical studies merits a closer look into the utility and the analytical standards of traditional and emerging biomarkers of vitamin D metabolism. The aim of this review is to provide structured information on 25(OH)D, 1,25(OH)2D, 24,25(OH)2D, VDBP and the derived parameters vitamin D metabolite ratio (VMR), free (free-25(OH)D) and bioavailable (bio-25(OH)D) 25(OH)D. Particular focus will be put on the clinical utility, preanalytical, analytical and postanalytical aspects, and interfering factors.', 'Vitamin D degradation is predominantly driven by CYP24A1, which metabolises 25(OH)D to 24,25-dihydroxy-vitamin D [24,25(OH2D)] and 1,25(OH)2D to 1,24,25-trihydroxy-vitamin D [1,24,25(OH)3D]. In the circulation, all vitamin D metabolites are bound to vitamin D binding protein (DBP), albumin and lipoproteins.', 'The role of vitamin D 3-epimers in health and disease is still a matter of debate as most of our knowledge is derived from in vitro studies. Apparently, 3-epi-25(OH)D has low affinity for VDBP (36-46% compared with 25(OH)D) and VDR (2-3%), which explains at least partly the low serum concentration. At present it is believed that 3-epi-25(OH)D does not adequately reflect vitamin D status and thus its separate measurement is not recommended.', 'Furthermore, lower serum 25(OH)D concentrations in blacks are not associated with higher PTH concentrations, lower BMD or increased fracture risk. This conundrum is at least partly explained by a high prevalence (> 90%) of the GC1F haplotype of VDBP (rs7041-T/rs4588-C) in blacks, whereas the GC1S haplotype of VDBP (rs7041-G/rs4588-C) is dominant in Caucasians. While carriers of the GC1F and the GC1S haplotypes show comparable VDBP concentrations they differ in their affinity for most vitamin D species, which contributes to the different 25(OH)D concentrations.', 'However, the availability of free vitamin D for metabolism is not affected.', '25(OH)D is a challenging analyte due to its strong binding to VDBP and other carriers, the need of measuring 25(OH)D2 and 25(OH)D3 in an equimolar fashion, the coexistence of multiple chemically related molecules that may cross-react, and common matrix effects, such as heterophilic antibodies or changes in protein composition. Medical laboratories employ different methods for the measurement of 25(OH)D in serum and plasma with automated immunoassays, ELISAs and LC-MS/MS being the most frequently used.', 'The strategies that are used by immunoassay manufacturers to separate 25(OH)D from its carriers are optimized for the expected matrix composition. In situations where the matrix is altered, such as in pregnant women, patients with chronic kidney disease or individuals with a polymorphic variant of VDBP, these approaches may be less efficient and thus introduce analytical bias. In contrast, the organic solvents used in LC-MS/MS are strong enough to precipitate all proteins and detach all vitamin D metabolites from their carriers.', 'Most 25(OH)D in serum is bound to VDBP (85%) and albumin (15%). Variations in the serum concentrations of these carriers lead to corresponding changes of 25(OH)D.', 'Pregnancy causes a small increase in 25(OH)D which is at least partly driven by an increase in VDBP. However, not all studies found a significant changes of serum 25(OH)D in pregnant women.', 'Also, in black individuals, the parallel determination of both metabolites does better reflect the individual vitamin D status than the measurement of 25(OH)D alone. On average, blacks have a 40% lower 25(OH)D serum concentration, which is at least partially due to ethnic differences in the prevalence of common genetic VDBP polymorphisms. Recent evidence from a cohort study of elderly community-dwelling individuals has shown that the VMR was associated with the risk of hip fracture whereas 25(OH)D was not.', 'Measurement of Vitamin D Binding Protein (VDBP) and Calculation of Free and Bioavailable 25(OH)D', 'VDBP is a highly polymorphic protein with over 100 isoforms that is encoded by the VDBP gene on chromosome 4q12-q13. Structurally, the protein is closely related to albumin and alpha-fetoprotein. It is mainly produced in the liver, but animal studies suggest that it can also be expressed by other tissues, such as kidney, testis, and adipose tissue.', 'The expression of VDBP is regulated by estrogens, which show higher concentrations during pregnancy and upon oral contraception.', 'As mentioned above, VDBP binds approximately 85% and albumin 15% of all circulating vitamin D metabolites. Less than 0.1% of all 25(OH)D in the plasma of normal individuals is free. Similar to other steroid hormones, it is believed that only the free fraction can enter the cells to bind the cytosolic VDR.', 'The free hormone hypothesis is supported by the observation that lacking VDBP does not necessarily cause functional vitamin D deficiency despite very low serum concentrations of 25(OH)D and 1,25(OH)2D. In addition to its transport function of vitamin D metabolites, VDBP has also other functions including the renal megalin/cubilin-mediated reuptake of 25(OH)D in the kidneys,the scavenging of actin after muscular damage, neutrophil recruitment, complement 5a-mediated chemotaxis, fatty acid binding and formation of VDBP-Macrophage Activating Factor V (DBP-MAF). For example, significant amounts of actin are released from damaged cells in response to trauma, sepsis, liver trauma, acute lung injury, preeclampsia, surgery, and burns.', 'Upon polymerization, filamentous F-actin in combination with coagulation factor Va can promote disseminated intravascular coagulation and multiorgan failure. Through binding to VDBP, actin is rapidly cleared by the liver, lungs and spleen, thus preventing polymerization.', 'In contrast to other steroid hormones, the active metabolite 1,25(OH)2D is feedback regulated and the concentration of free hormone is so low that it may be insufficient for passive diffusion across the cell membrane and effective receptor binding. In addition, it cannot be excluded that the megalin-dependent uptake of the VDBP-25(OH)D complex into cells is limited to the proximal renal tubule. For example, a vitamin D-dependent expression megalin and cubilin has been found in several other epithelia including placenta, parathyroid glands, pneumocytes, and epididymal epithelial cells.', 'Considering the much lower affinity of 25(OH)D to albumin (Ka = 7 x 108/M) than to VDBP (Ka = 6 x 105/M), it is believed that albumin-bound 25(OH)D can easily dissociate from the protein and is thus available for metabolism. This concept has led to the term bioavailable 25(OH)D, which refers to all the circulating 25(OH)D that is not tightly bound to VDBP. Several studies have used bioavailable 25(OH)D rather than free 25(OH)D.', 'Free and bioavailable 25(OH)D are calculated from total 25(OH)D, albumin and VDBP:', 'For accuracy, the measurements of 25(OH)D and VDBP are best performed by LC-MS/MS. The polymorphic nature of VDBP results in numerous variants of the protein with Gc1f and Gc1s and Gc2 being most common. These variants are believed to differ in their affinity for 25(OH)D.', 'The lowest kVDBP has been reported for the 1f/1f variant with 3.6 x 108/M, whereas 2/2 has supposedly the highest kVDBP with 11.2 x 108/M. However, different affinities of VDBP variants to 25(OH)D are a matter of ongoing scientific debate. Nevertheless, VDBP haplotypes impact the concentrations of total 25(OH)D, free 25(OH)D, and VDBP.', 'The 2/2 haplotype is associated with the lowest total and free 25(OH)D concentrations.', 'Furthermore, the different Gc alleles affect the response to vitamin D supplementation. The clinical significance of these VDBP variants is insufficiently understood. Existing studies do not show differences in fracture rate.', 'VDBP is relatively stable. Storing serum for 48 h at room temperature or refrigerated introduces only a small positive bias of < 10%. Also, up to 2 freeze-thaw-cycles have no major impact on VDBP.', 'Regarding LC-MS/MS methods, autosampler stability is also acceptable with < 16% loss during 24 h in a 96-well plate. In order to minimize the impact of preanalytical factors, samples should be processed rapidly after collection. The distribution of VDBP variants differs substantially between black and white individuals, which has important analytical implications (see below) and impacts free 25(OH)D levels.', 'Haemolysis and lipaemia have also been reported to interfere with the measurement.', 'The measurement of VDBP is challenging and LC-MS/MS is the preferred method for this analyte due to superior sensitivity and specificity. For example, only LC-MS/MS can detect VDBP isoforms. However, LC-MS/MS analyses require experienced staff and expensive equipment, which are not widely available.', 'Immunoassays have been developed as easy-to-use alternative that allow a relatively high throughput. However, the use of immunoassays for the quantitation of a highly polymorphic protein, such as VDBP, is problematic, especially when the target epitope(s) are located near the polymorphic region. For example, Henderson CM et al.', 'have shown that assays with a monoclonal capture antibody may react differently with common variants, and thus deliver misleading results. With a monoclonal immunoassay, blacks, which carry predominantly the Gc1f variant, have been reported to have substantially lower VDBP concentrations than whites. When using a validated LC-MS/MS method that can distinguish between VDBP variants, no ethnic differences in VDBP concentration were found.', 'Although polyclonal assays are less influenced by isoforms and other genetic variations, they still deliver largely discrepant results when compared to LC-MS/MS with a median bias of + 50%. Such differences have great impact on the calculation of free and bioavailable 25(OH)D. Therefore, free and bioavailable 25(OH)D concentrations should be determined on the basis of accurate VDBP and total 25(OH)D results obtained with validated LC-MS/MS methods.', 'Even sensitive LC-MS/MS systems are not sufficient to accurately measure such low concentrations. Therefore, most studies calculate the free 25(OH)D concentration from measurements of total 25(OH)D, VDBP and albumin. However, the result is strongly dependent on the accuracy of the methods used for total 25(OH)D and VDBP.', 'A comparison of calculated free 25(OH)D results obtained with four different VDBP immunoassays has shown differences of up to 50%. Direct methods for the measurement of free 25(OH)D have also been developed.', 'To date, reliable reference values for free 25(OH)D and VDBP have not yet been established, and several ranges have been proposed. Due to the lack of standardization, reference intervals have to be method-specific. In addition, differences between black and white individuals have been reported for some methods, but not for others.', 'When measuring free 25(OH)D by direct immunoassay in 279 healthy controls with a median age of 36.6 years, 95% of the results ranged between 0.5 and 8.1 pg/ml, with VDBP concentrations measured by a polyclonal immunodiffusion assay between 190.8 and 395.2 microg/ml. Amongst 2085 participants of the Healthy Aging in Neighbourhoods of Diversity across the Life Span cohort, blacks and whites showed mean +- standard error concentrations of 168 +- 3 microg/ml and 337 +- 5 microg/ml, respectively. Bioavailable 25(OH)D-was 2.9 +- 0.1 ng/ml in blacks and 3.1 +- 0.1 ng/ml in whites.', 'However, VDBP was measured by a monoclonal immunoassay that suffers from variable detection of different VDBP variants. In 173 healthy women, Peris et al.', 'Currently, routine measurement of VDBP, free and bioavailable 25(OH)D are not recommended due to unresolved analytical issues and the lack clinical evidence that supports an additional value beyond the measurement of total 25(OH)D. Potential areas of interest for the use of these markers are patients with liver or renal disease, which cause a decreased synthesis or loss of VDBP. Moreover, determining free 25(OH)D could also be useful for pregnant women or women on hormone therapy, because estrogens can upregulate VDBP levels.', 'In the same line, elderly people could also benefit from assessing this marker since VDBP levels can be altered in this group due to a multifactorial effect. However, future clinical studies have to clarify the potential value of VDBP, free and bioavailable 25(OH)D in such patients.', 'Influence of VDBP in Pathological Conditions', 'Different variants of VDBP with different affinities for 25(OH)D influence the measurement of VDBP as well as the determination of free 25(OH)D. As the distribution of VDBP variants changes between blacks and whites, ethnicity needs to be considered when assessing these markers. Based on the method used, differences in VDBP and free 25(OH)D have been found or not.', 'Changes in estrogen levels, such as in pregnancy, alter the expression of VDBP. Between the second and third trimester, VDBP levels increase up to two-fold. However, these variations are not linked to changes in free 25(OH)D, which remains stable or slightly decreases.', 'Critically ill patients with organ dysfunction also show substantial variation in serum VDBP when compared with normal individuals. Liver diseases, especially at late stages, often cause impaired protein synthesis, resulting in decreasing levels of VDBP and albumin. In turn, free 25(OH)D is increased in these patients compared to controls.', 'These changes are most pronounced in cirrhotic patients. Renal disease also decreases VDBP levels, via proteinuria and saturation of the maximal transport capacity of the megalin/cubilin system. Consistently, nephrotic patients present low total and free 25(OH)D levels.', 'However, in dialysis patients VDBP levels are comparable with healthy individuals. Despite normal VDBP levels, immunoassays and LC-MS/MS methods vary in the quantitation of 25(OH)D in dialysis patients. Potential differences may be due to increased urea concentrations that interfere with the release of 25(OH)D from VDBP, the binding equilibrium or the reagents of immunoassays.', 'In contrast, LC-MS/MS methods are not affected from such interferences.', 'The existence of multiple vitamin D metabolites with different biological activities and their binding to VDBP and other carriers makes the assessment of vitamin D status a challenging task. Despite promising results for alternative markers, such as 24,25(OH)D, VMR, free and bioavailable 25(OH)D, 25(OH)D remains the preferred analyte for this purpose.', 'sarcoidosis, tuberculosis). The determination of VDBP, free and bioavailable 25(OH)D is compromised by unresolved analytical issues, the lack of reference intervals and insufficient clinical evidence that justifies their use. Future research should address analytical standardization and explore the clinical value of novel markers of vitamin D metabolism.', 'The causal effect of vitamin D binding protein (DBP) levels on calcemic and cardiometabolic diseases: a Mendelian randomization study', 'Vitamin D binding protein, total and free vitamin D levels in different physiological and pathophysiological conditions', 'Longitudinal changes in serum vitamin D binding protein and free 25-hydroxyvitamin D in a multiracial cohort of pregnant adolescents', 'Vitamin D binding protein: genomic structure, functional domains, and mRNA expression in tissues', 'Common genetic variants of the vitamin D binding protein (DBP) predict differences in response of serum 25-hydroxyvitamin D [25(OH)D] to vitamin D supplementation', 'The associations of 25-hydroxyvitamin D levels, vitamin D binding protein gene polymorphisms, and race with risk of incident fracture-related hospitalization: twenty-year follow-up in a bi-ethnic cohort (the ARIC Study)', 'Common variants of the vitamin D binding protein gene and adverse health outcomes', 'Biological and clinical aspects of the vitamin D binding protein (Gc-globulin) and its polymorphism', 'Recommendations on the measurement and the clinical use of vitamin D metabolites and vitamin D binding protein:a position paper from the IFCC Committee on bone metabolism', 'Accuracy of 6 routine 25-hydroxyvitamin D assays: influence of vitamin D binding protein concentration']\n",
            "  Document ID     : PMC8892115\n",
            "There are 59 occurrence(s) (each occurrence is at most 3 sentences long) for gene 2638 : vitamin D binding protein.\n",
            "\n",
            "Gene(120227)\n",
            "  Symbol          : CYP2R1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : cytochrome P450 family 2 subfamily R member 1\n",
            "  Also Known As   : CYP2R1\n",
            "  In-text Name    : CYP2R1\n",
            "  Occurrences     : ['Circulating vitamin D2 and D3 are activated by two hydroxylation reactions that occur in the liver and the kidneys. The hepatic cytochromes P450 CYP2R1 (microsomal) and CYP27A1 (mitochondrial) hydroxylate vitamin D2 and D3 at carbon 25 resulting in the production of 25(OH)D, the most abundant vitamin D metabolite in blood, which is still inactive. Renal CYP27B1 attaches the second hydroxygroup in position 1 forming active 1,25-(OH)2D.', 'The inactive prohormone 25(OH)D represents the main reservoir and transport form of vitamin D. The generation of 25(OH)D from pro-vitamin D3 is catalyzed by CYP27A, a 25-hydroxylase encoded by the CYP2R1 gene. As the most abundant vitamin D metabolite in blood, 25(OH)D includes the two forms 25(OH) D2 and 25(OH)D3.', 'However, recent studies suggest that both age and metabolism can modulate CYP27A activity. Aged humans and animals are characterized by a lower expression of CYP2R1 than their young counterparts and thus they are less responsive to vitamin D supplementation. As vitamin D can be sequestered in adipose tissue, it is not surprising that nutrition and body composition also modulate CYP2R activity.', 'Women starting pregnancy during winter showed rising 25(OH)D concentrations, whereas women who fell pregnant in summer exhibited no increase. Interestingly, the placenta possesses the principal enzymes of vitamin D metabolism including CYP2R1, CYP27B1 and CYP24A1 and may thus be involved in the regulation of circulating 25(OH)D levels. Maternal 25(OH)D can crosses the placenta and is the exclusive vitamin D source of the foetus reaching cord blood levels between 75 and100% of the maternal value.', 'Mutations of genes involved in 1,25(OH)2D metabolism also influence circulating 1,25(OH)2D levels (reviewed in). These mutations cause rare hereditary metabolic bone conditions, like hereditary vitamin D-resistant rickets (VDR), vitamin-D-dependent rickets type 1A (CYP27B1), type 1B (CYP2R1) or idiopathic infantile hypercalcemia (CYP24A1). Mutations of the cell surface metalloproteinase PHEX cause X-linked hypophosphataemia (XLH), which is characterized by low-normal 1,25(OH)2D concentrations.', 'Decreased serum 25-hydroxyvitamin D in aging male mice is associated with reduced hepatic Cyp2r1 abundance']\n",
            "  Document ID     : PMC8892115\n",
            "There are 6 occurrence(s) (each occurrence is at most 3 sentences long) for gene 120227 : CYP2R1.\n",
            "\n",
            "Gene(1593)\n",
            "  Symbol          : CYP27A1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : cytochrome P450 family 27 subfamily A member 1\n",
            "  Also Known As   : CP27, CTX, CYP27\n",
            "  In-text Name    : CYP27A1\n",
            "  Occurrences     : ['Circulating vitamin D2 and D3 are activated by two hydroxylation reactions that occur in the liver and the kidneys. The hepatic cytochromes P450 CYP2R1 (microsomal) and CYP27A1 (mitochondrial) hydroxylate vitamin D2 and D3 at carbon 25 resulting in the production of 25(OH)D, the most abundant vitamin D metabolite in blood, which is still inactive. Renal CYP27B1 attaches the second hydroxygroup in position 1 forming active 1,25-(OH)2D.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1593 : CYP27A1.\n",
            "\n",
            "Gene(1594)\n",
            "  Symbol          : CYP27B1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : cytochrome P450 family 27 subfamily B member 1\n",
            "  Also Known As   : CP2B, CYP1, CYP1alpha, CYP27B, P450c1, PDDR, VDD1, VDDR, VDDRI, VDR\n",
            "  In-text Name    : CYP27B1\n",
            "  Occurrences     : ['The hepatic cytochromes P450 CYP2R1 (microsomal) and CYP27A1 (mitochondrial) hydroxylate vitamin D2 and D3 at carbon 25 resulting in the production of 25(OH)D, the most abundant vitamin D metabolite in blood, which is still inactive. Renal CYP27B1 attaches the second hydroxygroup in position 1 forming active 1,25-(OH)2D. In addition to the kidneys, CYP27B1 is also expressed in many other cell types so that 1,25-(OH)2D can be produced by most extra-renal tissues, where it has primarily autocrine or paracrine function.', 'However, this extra-renal 1,25-(OH)2D synthesis contributes little to the circulating concentration of this metabolite.', 'As the most abundant vitamin D metabolite in blood, 25(OH)D includes the two forms 25(OH) D2 and 25(OH)D3. Hydroxylation of 25(OH)D in position 1 is mediated by CYP27B1 and results in the formation of active 1,25(OH)2D. However, this alpha-hydroxylation occurs on demand, primarily in the kidneys, under the control of parathyroid hormone (PTH).', \"Consequently, the serum concentration of 1,25(OH)2D is kept within the reference range over a wide concentration range of 25(OH)D and does not reflect vitamin D stores. In fact, individuals with a low serum 25(OH)D concentration often have a high-normal or raised 1,25(OH)2D concentration due to a compensatory induction of CYP27B1. Therefore, current guidelines unanimously recommend 25(OH)D as the preferred indicator of the body's vitamin D stores.\", 'Women starting pregnancy during winter showed rising 25(OH)D concentrations, whereas women who fell pregnant in summer exhibited no increase. Interestingly, the placenta possesses the principal enzymes of vitamin D metabolism including CYP2R1, CYP27B1 and CYP24A1 and may thus be involved in the regulation of circulating 25(OH)D levels. Maternal 25(OH)D can crosses the placenta and is the exclusive vitamin D source of the foetus reaching cord blood levels between 75 and100% of the maternal value.', '1,25(OH)2D is the active form of vitamin D that is produced through 1alpha-hydroxylation of 25(OH)D. This hydroxylation step is catalysed by CYP27B1 (1alpha-hydroxylase) and occurs mainly in the kidneys. Consequently, the circulating 1,25(OH)2D concentration primarily reflects renal production.', 'Based on these results, it has been hypothesized that butyrate-producing bacteria could stimulate the local production of 1,25(OH)2D by dendritic cells in the colon. In this context, a previous animal model lacking VDR and CYP27B1 also showed impaired levels of 1,25(OH)2D and reduced number of Firmicutes.', 'Mutations of genes involved in 1,25(OH)2D metabolism also influence circulating 1,25(OH)2D levels (reviewed in). These mutations cause rare hereditary metabolic bone conditions, like hereditary vitamin D-resistant rickets (VDR), vitamin-D-dependent rickets type 1A (CYP27B1), type 1B (CYP2R1) or idiopathic infantile hypercalcemia (CYP24A1). Mutations of the cell surface metalloproteinase PHEX cause X-linked hypophosphataemia (XLH), which is characterized by low-normal 1,25(OH)2D concentrations.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 8 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1594 : CYP27B1.\n",
            "\n",
            "Gene(1591)\n",
            "  Symbol          : CYP24A1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : cytochrome P450 family 24 subfamily A member 1\n",
            "  Also Known As   : CP24, CYP24, HCAI, HCINF1, P450-CC24\n",
            "  In-text Name    : CYP24A1\n",
            "  Occurrences     : ['However, this extra-renal 1,25-(OH)2D synthesis contributes little to the circulating concentration of this metabolite. Vitamin D degradation is predominantly driven by CYP24A1, which metabolises 25(OH)D to 24,25-dihydroxy-vitamin D [24,25(OH2D)] and 1,25(OH)2D to 1,24,25-trihydroxy-vitamin D [1,24,25(OH)3D]. In the circulation, all vitamin D metabolites are bound to vitamin D binding protein (DBP), albumin and lipoproteins.', 'Women starting pregnancy during winter showed rising 25(OH)D concentrations, whereas women who fell pregnant in summer exhibited no increase. Interestingly, the placenta possesses the principal enzymes of vitamin D metabolism including CYP2R1, CYP27B1 and CYP24A1 and may thus be involved in the regulation of circulating 25(OH)D levels. Maternal 25(OH)D can crosses the placenta and is the exclusive vitamin D source of the foetus reaching cord blood levels between 75 and100% of the maternal value.', 'Anti-epileptic drugs, such as phenobarbital, carbamazepine, phenytoin and valproate, are known to accelerate vitamin D catabolism and thus lower the serum 25(OH)D concentration. Binding of these drugs to the pregnane X receptor increases 25(OH)D catabolism by an induction of CYP3A4 and CYP24A1. The preganane X receptor is a nuclear receptor that shares homologies with the VDR.', 'Mutations of genes involved in 1,25(OH)2D metabolism also influence circulating 1,25(OH)2D levels (reviewed in). These mutations cause rare hereditary metabolic bone conditions, like hereditary vitamin D-resistant rickets (VDR), vitamin-D-dependent rickets type 1A (CYP27B1), type 1B (CYP2R1) or idiopathic infantile hypercalcemia (CYP24A1). Mutations of the cell surface metalloproteinase PHEX cause X-linked hypophosphataemia (XLH), which is characterized by low-normal 1,25(OH)2D concentrations.', '24,25(OH)2D is the principal catabolite of 25(OH)D that is formed through the action of CYP24A1 when sufficient amounts of 25(OH)D are available. Under physiological conditions the serum concentration of 24,25(OH)2D ranges at approximately 10% of the 25(OH)D concentration.', 'Therefore, the isolated measurement of 24,25(OH)2 does not provide superior information when compared to 25(OH)D. However, the simultaneous determination of both metabolites has been found useful to identify patients with reduced CYP24A1 activity. Affected individuals are characterized by a reduced ratio between 24,25(OH)2D and 25(OH)D.', 'This ratio is also known as vitamin D metabolite ratio (VMR). CYP24A1 loss-of-function mutations cause infantile hypercalcemia (IIH) and represent a genetic risk factor for serious adverse effects in response to vitamin D supplementation. In hypercalcaemic patients, the parallel measurement of both vitamin D metabolites and calculation of the VMR allows distinguishing CYP24A1 deficiency from other causes of hypercalcaemia, such as vitamin D intoxication, granulomatous disease, Williams-Beuren syndrome or mutations of the SLC34A1 gene.', 'Simultaneous measurement of 25(OH)-vitamin D and 24,25(OH)2-vitamin D to define cut-offs for CYP24A1 mutation and vitamin D deficiency in a population of 1200 young subjects', 'Mutations in CYP24A1 and idiopathic infantile hypercalcemia', 'LC-MS/MS for identifying patients with CYP24A1 mutations']\n",
            "  Document ID     : PMC8892115\n",
            "There are 10 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1591 : CYP24A1.\n",
            "\n",
            "Gene(1628)\n",
            "  Symbol          : DBP\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : D-box binding PAR bZIP transcription factor\n",
            "  Also Known As   : DABP, taxREB302\n",
            "  In-text Name    : DBP\n",
            "  Occurrences     : ['Vitamin D degradation is predominantly driven by CYP24A1, which metabolises 25(OH)D to 24,25-dihydroxy-vitamin D [24,25(OH2D)] and 1,25(OH)2D to 1,24,25-trihydroxy-vitamin D [1,24,25(OH)3D]. In the circulation, all vitamin D metabolites are bound to vitamin D binding protein (DBP), albumin and lipoproteins.', 'The causal effect of vitamin D binding protein (DBP) levels on calcemic and cardiometabolic diseases: a Mendelian randomization study', 'Affinity differences for vitamin D metabolites associated with the genetic isoforms of the human serum carrier protein (DBP)', 'Common genetic variants of the vitamin D binding protein (DBP) predict differences in response of serum 25-hydroxyvitamin D [25(OH)D] to vitamin D supplementation']\n",
            "  Document ID     : PMC8892115\n",
            "There are 4 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1628 : DBP.\n",
            "\n",
            "Gene(213)\n",
            "  Symbol          : ALB\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : albumin\n",
            "  Also Known As   : FDAHT, HSA, PRO0883, PRO0903, PRO1341\n",
            "  In-text Name    : albumin\n",
            "  Occurrences     : ['Vitamin D degradation is predominantly driven by CYP24A1, which metabolises 25(OH)D to 24,25-dihydroxy-vitamin D [24,25(OH2D)] and 1,25(OH)2D to 1,24,25-trihydroxy-vitamin D [1,24,25(OH)3D]. In the circulation, all vitamin D metabolites are bound to vitamin D binding protein (DBP), albumin and lipoproteins.', 'Most 25(OH)D in serum is bound to VDBP (85%) and albumin (15%). Variations in the serum concentrations of these carriers lead to corresponding changes of 25(OH)D.', 'VDBP is a highly polymorphic protein with over 100 isoforms that is encoded by the VDBP gene on chromosome 4q12-q13. Structurally, the protein is closely related to albumin and alpha-fetoprotein. It is mainly produced in the liver, but animal studies suggest that it can also be expressed by other tissues, such as kidney, testis, and adipose tissue.', 'As mentioned above, VDBP binds approximately 85% and albumin 15% of all circulating vitamin D metabolites. Less than 0.1% of all 25(OH)D in the plasma of normal individuals is free.', 'Considering the much lower affinity of 25(OH)D to albumin (Ka = 7 x 108/M) than to VDBP (Ka = 6 x 105/M), it is believed that albumin-bound 25(OH)D can easily dissociate from the protein and is thus available for metabolism. This concept has led to the term bioavailable 25(OH)D, which refers to all the circulating 25(OH)D that is not tightly bound to VDBP.', 'Free and bioavailable 25(OH)D are calculated from total 25(OH)D, albumin and VDBP:', 'Even sensitive LC-MS/MS systems are not sufficient to accurately measure such low concentrations. Therefore, most studies calculate the free 25(OH)D concentration from measurements of total 25(OH)D, VDBP and albumin. However, the result is strongly dependent on the accuracy of the methods used for total 25(OH)D and VDBP.', 'Critically ill patients with organ dysfunction also show substantial variation in serum VDBP when compared with normal individuals. Liver diseases, especially at late stages, often cause impaired protein synthesis, resulting in decreasing levels of VDBP and albumin. In turn, free 25(OH)D is increased in these patients compared to controls.', 'Assessment of the free fraction of 25-hydroxyvitamin d in serum and its regulation by albumin and the vitamin d-binding protein']\n",
            "  Document ID     : PMC8892115\n",
            "There are 9 occurrence(s) (each occurrence is at most 3 sentences long) for gene 213 : albumin.\n",
            "\n",
            "Gene(7421)\n",
            "  Symbol          : VDR\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : vitamin D receptor\n",
            "  Also Known As   : NR1I1, PPP1R163\n",
            "  In-text Name    : VDR\n",
            "  Occurrences     : ['The role of vitamin D 3-epimers in health and disease is still a matter of debate as most of our knowledge is derived from in vitro studies. Apparently, 3-epi-25(OH)D has low affinity for VDBP (36-46% compared with 25(OH)D) and VDR (2-3%), which explains at least partly the low serum concentration. At present it is believed that 3-epi-25(OH)D does not adequately reflect vitamin D status and thus its separate measurement is not recommended.', 'Existing knowledge indicates that 1 alpha,25(OH)2D3 harbours a lower biological activity and has a markedly lower affinity to the VDR than 1alpha,25(OH)2D3. Also, its ability to stimulate intestinal calcium absorption is significantly reduced.', 'A low biological variability in combination with rather high concentrations in the nM range and a long in vitro half-life make 25(OH)D a robust biomarker that can reliably be measured by clinical laboratories. A recent study from Denmark has shown that the in vivo half life depends on the individual 25(OH)D concentration and the vitamin D receptor (VDR) gene polymorphism rs2228570. In individuals with a 25(OH)D start concentration between 68 and 213 nmol/L, mean half-life was 89 days, whereas lower concentrations were associated with a half-life between 149 and 199 days.', 'The longer half-live at lower 25(OH)D concentrations may be caused by storage mobilisation, changed catabolism or increased intestinal absorption. In addition, the VDR gene polymorphism rs2228570 may explain up to 88% of the observed variation with the genotype GG having a 120 days shorter half-life than the genotype AA/AG. Stability studies have shown that 25(OH)D is stable in whole blood for up to 24 h when kept at room temperature (RT).', 'Binding of these drugs to the pregnane X receptor increases 25(OH)D catabolism by an induction of CYP3A4 and CYP24A1. The preganane X receptor is a nuclear receptor that shares homologies with the VDR. In order to mitigate the risk of vitamin D deficiency, reduced BMD and increased fracture risk, patients on long-term treatment with anti-epileptic drugs should receive a prophylactic vitamin D supplementation and have their serum 25(OH)D concentration monitored.', 'However, 1,25(OH)2D that is produced by extra-renal tissues has mainly auto- and paracrine functions and adds little to the serum level. The biological activity of 1,25(OH)2D is mediated by binding to the VDR, a nuclear receptor that binds specific DNA sequences located at different genomic regions (vitamin D response elements), to subsequently modulate the expression of the respective target genes.', 'Based on these results, it has been hypothesized that butyrate-producing bacteria could stimulate the local production of 1,25(OH)2D by dendritic cells in the colon. In this context, a previous animal model lacking VDR and CYP27B1 also showed impaired levels of 1,25(OH)2D and reduced number of Firmicutes.', 'Pregnancy increases the serum 1,25(OH)2D concentration due to an induction of renal 1alpha-hydroxylase activity. In addition, 1alpha-hydroxylase and the VDR are both expressed in the placenta. However, the function of placental 1,25(OH)2D synthesis is poorly understood.', 'Mutations of genes involved in 1,25(OH)2D metabolism also influence circulating 1,25(OH)2D levels (reviewed in). These mutations cause rare hereditary metabolic bone conditions, like hereditary vitamin D-resistant rickets (VDR), vitamin-D-dependent rickets type 1A (CYP27B1), type 1B (CYP2R1) or idiopathic infantile hypercalcemia (CYP24A1). Mutations of the cell surface metalloproteinase PHEX cause X-linked hypophosphataemia (XLH), which is characterized by low-normal 1,25(OH)2D concentrations.', 'Less than 0.1% of all 25(OH)D in the plasma of normal individuals is free. Similar to other steroid hormones, it is believed that only the free fraction can enter the cells to bind the cytosolic VDR. The free hormone hypothesis is supported by the observation that lacking VDBP does not necessarily cause functional vitamin D deficiency despite very low serum concentrations of 25(OH)D and 1,25(OH)2D.', 'The half-life of 25(OH)D after UVB exposure depends on gender and vitamin D receptor polymorphism but mainly on the start level', 'A perspective on how the vitamin D sterol/vitamin D receptor (VDR) conformational ensemble model can potentially be used to understand the structure-function results of A-ring modified Vitamin D sterols', '1,25-Dihydroxyvitamin D3 controls a cohort of vitamin D receptor target genes in the proximal intestine that is enriched for calcium-regulating components', '1,25-Dihydroxyvitamin D protects against age-related osteoporosis by a novel VDR-Ezh2-p16 signal axis']\n",
            "  Document ID     : PMC8892115\n",
            "There are 14 occurrence(s) (each occurrence is at most 3 sentences long) for gene 7421 : VDR.\n",
            "\n",
            "Gene(5741)\n",
            "  Symbol          : PTH\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : parathyroid hormone\n",
            "  Also Known As   : FIH1, PTH1\n",
            "  In-text Name    : PTH\n",
            "  Occurrences     : ['Also, its ability to stimulate intestinal calcium absorption is significantly reduced. PTH suppression, however, was detected at similar rates as 1,25(OH)2D3 (reviewed in).', 'Hydroxylation of 25(OH)D in position 1 is mediated by CYP27B1 and results in the formation of active 1,25(OH)2D. However, this alpha-hydroxylation occurs on demand, primarily in the kidneys, under the control of parathyroid hormone (PTH). Consequently, the serum concentration of 1,25(OH)2D is kept within the reference range over a wide concentration range of 25(OH)D and does not reflect vitamin D stores.', 'People with dark skin have 30-40% lower 25(OH)D serum concentrations than Caucasians but comparable or higher bone mineral density (BMD) and lower fracture risk. Furthermore, lower serum 25(OH)D concentrations in blacks are not associated with higher PTH concentrations, lower BMD or increased fracture risk. This conundrum is at least partly explained by a high prevalence (> 90%) of the GC1F haplotype of VDBP (rs7041-T/rs4588-C) in blacks, whereas the GC1S haplotype of VDBP (rs7041-G/rs4588-C) is dominant in Caucasians.', 'Berg et al. have shown that blacks and whites with comparable serum PTH concentrations also have a comparable VMR despite substantially lower 25(OH)D levels. Other factors that contribute to lower 25(OH)D concentrations in blacks include adiposity and skin pigmentation.', 'Valcour et al. have shown that PTH continuously decreases with increasing concentrations of 25(OH)D and that there is no threshold above which this relationship plateaus. The same applies to other conditions, such as stroke, cardiovascular disease, diabetes mellitus and autoimmune disease.', 'However, the management of vitamin D status in nephrology is still a matter of ongoing debate. In any case, CKD patients with 25(OH)D deficiency should always be treated, regardless of PTH. Vitamin D deficiency in Stages 1 to 3a follow general population recommendations.', 'To prevent adverse skeletal effects, vitamin D deficiency should be diagnosed and corrected prior to bariatric surgery. However, it may be difficult to normalize 25(OH)D and PTH in these patients.', 'The main role of 1,25(OH)2D is regulating blood calcium levels through intestinal absorption, renal reabsorption and release from bone stores. When the calcium intake is low, PTH increases the synthesis of 1,25(OH)2D, which subsequently activates the transport of calcium from the intestinal lumen to the blood by upregulating TRPV6 and other genes, such as S100G, ATP2B1, and CLDN2, in the intestine. When calcium intake increases, PTH secretion is reduced leading to a decrease in 1,25(OH)2D and activation of a paracellular route, a non-saturable and passive transport through the intercellular space between adjacent cells, which is also modulated by 1,25(OH)2D (reviewed in).', 'Therefore, 1,25(OH)2D is a key regulator of bone mineralisation.', 'To date, the clinical utility of 1,25(OH)2D is insufficiently understood, which limits its clinical use. Serum levels of 1,25(OH)2D have little or no relationship to vitamin D stores but rather are regulated by PTH. Measurement of 1,25(OH)2D is helpful in the investigation of patients with unexplained hypercalcaemia, sarcoidosis, granulomatous disorders, pseudo-vitamin D deficiency, rickets, tumour-induced osteomalacia, hyperparatiroidism, and CYP24A1 deficiency.', 'However, the function of placental 1,25(OH)2D synthesis is poorly understood. Serum 1,25(OH)2D levels are also higher in African individuals of all age-groups, which may be due to increased PTH levels in this population. However, the differences in the metabolism of vitamin, calcium and phosphate are insufficiently understood.', 'The Seattle Kidney Study has shown that patients with a 24,25(OH)2D serum concentration below the cohort median of 2.4 ng/ml had a significantly lower eGFR and an increased risk of mortality. Moreover, the inverse correlation of 24,25(OH)2D with PTH was stronger than that of 25(OH)D or 1,25(OH)2D.', 'Vitamin D status and its relationship with parathyroid hormone and bone mineral status in older adolescents', 'Effects of age and serum 25-OH-vitamin D on serum parathyroid hormone levels', 'Age-related changes in the 25-hydroxyvitamin D versus parathyroid hormone relationship suggest a different reason why older adults require more vitamin D', 'Physiology of the calcium-parathyroid hormone-vitamin D axis']\n",
            "  Document ID     : PMC8892115\n",
            "There are 16 occurrence(s) (each occurrence is at most 3 sentences long) for gene 5741 : PTH.\n",
            "\n",
            "Gene(8856)\n",
            "  Symbol          : NR1I2\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : nuclear receptor subfamily 1 group I member 2\n",
            "  Also Known As   : BXR, ONR1, PAR, PAR1, PAR2, PARq, PRR, PXR, SAR, SXR\n",
            "  In-text Name    : pregnane X receptor\n",
            "  Occurrences     : ['Anti-epileptic drugs, such as phenobarbital, carbamazepine, phenytoin and valproate, are known to accelerate vitamin D catabolism and thus lower the serum 25(OH)D concentration. Binding of these drugs to the pregnane X receptor increases 25(OH)D catabolism by an induction of CYP3A4 and CYP24A1. The preganane X receptor is a nuclear receptor that shares homologies with the VDR.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 8856 : pregnane X receptor.\n",
            "\n",
            "Gene(1576)\n",
            "  Symbol          : CYP3A4\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : cytochrome P450 family 3 subfamily A member 4\n",
            "  Also Known As   : CP33, CP34, CYP3A, CYP3A3, CYPIIIA3, CYPIIIA4, HLP, NF-25, P450C3, P450PCN1, VDDR3\n",
            "  In-text Name    : CYP3A4\n",
            "  Occurrences     : ['Anti-epileptic drugs, such as phenobarbital, carbamazepine, phenytoin and valproate, are known to accelerate vitamin D catabolism and thus lower the serum 25(OH)D concentration. Binding of these drugs to the pregnane X receptor increases 25(OH)D catabolism by an induction of CYP3A4 and CYP24A1. The preganane X receptor is a nuclear receptor that shares homologies with the VDR.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1576 : CYP3A4.\n",
            "\n",
            "Gene(8074)\n",
            "  Symbol          : FGF23\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : fibroblast growth factor 23\n",
            "  Also Known As   : ADHR, FGFN, HFTC2, HPDR2, HYPF, PHPTC\n",
            "  In-text Name    : FGF-23\n",
            "  Occurrences     : ['Similarly, 60-80% of pre-dialysis children with CKD also have low levels of 25(OH)D. In addition to low serum 25(OH)D levels, CKD patients are also characterized by elevated phosphate and FGF-23 concentrations. The adverse impact of CKD on calcium and phosphate metabolism is associated with secondary hyperparathyroidism.', 'Therefore, 1,25(OH)2D is a key regulator of bone mineralisation. The release of calcium from bone is accompanied by an increased secretion of fibroblast growth factor 23 (FGF23) by osteocytes, which, upon binding to Klotho, inhibits 1-alpha hydroxylase and ultimately reduces the renal synthesis of 1,25(OH)2D.', 'Measurement of 1,25(OH)2D is helpful in the investigation of patients with unexplained hypercalcaemia, sarcoidosis, granulomatous disorders, pseudo-vitamin D deficiency, rickets, tumour-induced osteomalacia, hyperparatiroidism, and CYP24A1 deficiency. Moreover, measurement of this vitamin D metabolite can be helpful to differentiate between FGF23-dependent and -independent phosphopenic rickets. Levels of 1,25(OH)2D are usually low in CKD, but its measurement has only been recommended when patients present severe and progressive hyperparathyroidism.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 3 occurrence(s) (each occurrence is at most 3 sentences long) for gene 8074 : FGF-23.\n",
            "\n",
            "Gene(55503)\n",
            "  Symbol          : TRPV6\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : transient receptor potential cation channel subfamily V member 6\n",
            "  Also Known As   : ABP/ZF, CAT1, CATL, ECAC2, HRPTTN, HSA277909, LP6728, ZFAB\n",
            "  In-text Name    : TRPV6\n",
            "  Occurrences     : ['The main role of 1,25(OH)2D is regulating blood calcium levels through intestinal absorption, renal reabsorption and release from bone stores. When the calcium intake is low, PTH increases the synthesis of 1,25(OH)2D, which subsequently activates the transport of calcium from the intestinal lumen to the blood by upregulating TRPV6 and other genes, such as S100G, ATP2B1, and CLDN2, in the intestine. When calcium intake increases, PTH secretion is reduced leading to a decrease in 1,25(OH)2D and activation of a paracellular route, a non-saturable and passive transport through the intercellular space between adjacent cells, which is also modulated by 1,25(OH)2D (reviewed in).']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 55503 : TRPV6.\n",
            "\n",
            "Gene(795)\n",
            "  Symbol          : S100G\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : S100 calcium binding protein G\n",
            "  Also Known As   : CABP, CABP1, CABP9K, CALB3\n",
            "  In-text Name    : S100G\n",
            "  Occurrences     : ['The main role of 1,25(OH)2D is regulating blood calcium levels through intestinal absorption, renal reabsorption and release from bone stores. When the calcium intake is low, PTH increases the synthesis of 1,25(OH)2D, which subsequently activates the transport of calcium from the intestinal lumen to the blood by upregulating TRPV6 and other genes, such as S100G, ATP2B1, and CLDN2, in the intestine. When calcium intake increases, PTH secretion is reduced leading to a decrease in 1,25(OH)2D and activation of a paracellular route, a non-saturable and passive transport through the intercellular space between adjacent cells, which is also modulated by 1,25(OH)2D (reviewed in).']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 795 : S100G.\n",
            "\n",
            "Gene(490)\n",
            "  Symbol          : ATP2B1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : ATPase plasma membrane Ca2+ transporting 1\n",
            "  Also Known As   : MRD66, PMCA1, PMCA1kb\n",
            "  In-text Name    : ATP2B1\n",
            "  Occurrences     : ['The main role of 1,25(OH)2D is regulating blood calcium levels through intestinal absorption, renal reabsorption and release from bone stores. When the calcium intake is low, PTH increases the synthesis of 1,25(OH)2D, which subsequently activates the transport of calcium from the intestinal lumen to the blood by upregulating TRPV6 and other genes, such as S100G, ATP2B1, and CLDN2, in the intestine. When calcium intake increases, PTH secretion is reduced leading to a decrease in 1,25(OH)2D and activation of a paracellular route, a non-saturable and passive transport through the intercellular space between adjacent cells, which is also modulated by 1,25(OH)2D (reviewed in).']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 490 : ATP2B1.\n",
            "\n",
            "Gene(9075)\n",
            "  Symbol          : CLDN2\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : claudin 2\n",
            "  Also Known As   : OAZON, claudin-2\n",
            "  In-text Name    : CLDN2\n",
            "  Occurrences     : ['The main role of 1,25(OH)2D is regulating blood calcium levels through intestinal absorption, renal reabsorption and release from bone stores. When the calcium intake is low, PTH increases the synthesis of 1,25(OH)2D, which subsequently activates the transport of calcium from the intestinal lumen to the blood by upregulating TRPV6 and other genes, such as S100G, ATP2B1, and CLDN2, in the intestine. When calcium intake increases, PTH secretion is reduced leading to a decrease in 1,25(OH)2D and activation of a paracellular route, a non-saturable and passive transport through the intercellular space between adjacent cells, which is also modulated by 1,25(OH)2D (reviewed in).']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 9075 : CLDN2.\n",
            "\n",
            "Gene(9365)\n",
            "  Symbol          : KL\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : klotho\n",
            "  Also Known As   : HFTC3, KLA\n",
            "  In-text Name    : Klotho\n",
            "  Occurrences     : ['Therefore, 1,25(OH)2D is a key regulator of bone mineralisation. The release of calcium from bone is accompanied by an increased secretion of fibroblast growth factor 23 (FGF23) by osteocytes, which, upon binding to Klotho, inhibits 1-alpha hydroxylase and ultimately reduces the renal synthesis of 1,25(OH)2D.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 9365 : Klotho.\n",
            "\n",
            "Gene(12097)\n",
            "  Symbol          : Bglap2\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : bone gamma-carboxyglutamate protein 2\n",
            "  Also Known As   : BGP2, Bglap1, Bgp, Og2, mOC-B\n",
            "  In-text Name    : osteocalcin\n",
            "  Occurrences     : ['In addition to the regulation of calcium homeostasis, 1,25(OH)2D also supports bone formation by increasing the osteoblastic expression of alkaline phosphatase, osteocalcin and ostepontin. Furthermore, it stimulates bone resorption by increasing osteoclastogenesis and osteoclast activity.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 12097 : osteocalcin.\n",
            "\n",
            "Gene(13088)\n",
            "  Symbol          : Cyp2b10\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : cytochrome P450, family 2, subfamily b, polypeptide 10\n",
            "  Also Known As   : Cyp2b, Cyp2b20, p16\n",
            "  In-text Name    : p16\n",
            "  Occurrences     : ['1,25(OH)2D is also involved in the regulation of cell proliferation, differentiation and apoptosis. Experimental evidence from 1alpha-hydrolase knockout mice suggests that low levels of 1,25(OH)2D increase age-related bone loss by activating senescence pathways, like p16/p19. An adequate supply with 1,25(OH)2D could thus preserve bone health by activating an anti-ageing mechanism.', '1,25-Dihydroxyvitamin D protects against age-related osteoporosis by a novel VDR-Ezh2-p16 signal axis', '1,25-Dihydroxyvitamin D exerts an antiaging role by activation of Nrf2-antioxidant signaling and inactivation of p16/p53-senescence signaling']\n",
            "  Document ID     : PMC8892115\n",
            "There are 3 occurrence(s) (each occurrence is at most 3 sentences long) for gene 13088 : p16.\n",
            "\n",
            "Gene(83430)\n",
            "  Symbol          : Il23a\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : interleukin 23, alpha subunit p19\n",
            "  Also Known As   : IL-23, p19\n",
            "  In-text Name    : p19\n",
            "  Occurrences     : ['1,25(OH)2D is also involved in the regulation of cell proliferation, differentiation and apoptosis. Experimental evidence from 1alpha-hydrolase knockout mice suggests that low levels of 1,25(OH)2D increase age-related bone loss by activating senescence pathways, like p16/p19. An adequate supply with 1,25(OH)2D could thus preserve bone health by activating an anti-ageing mechanism.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 83430 : p19.\n",
            "\n",
            "Gene(3586)\n",
            "  Symbol          : IL10\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : interleukin 10\n",
            "  Also Known As   : CSIF, GVHDS, IL-10, IL10A, TGIF\n",
            "  In-text Name    : IL-10\n",
            "  Occurrences     : ['In addition to its role in bone, 1,25(OH)2D also attenuates the inflammatory response of monocytes and macrophages and increases the expression of anti-inflammatory compounds like IL-10. In contrast, the secretion of pro-inflammatory cytokines like IL-1b, IL-6, TNFalpha, RANKL and COX2 is reduced by 1,25(OH)2D.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 3586 : IL-10.\n",
            "\n",
            "Gene(3553)\n",
            "  Symbol          : IL1B\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : interleukin 1 beta\n",
            "  Also Known As   : IL-1, IL1-BETA, IL1F2, IL1beta\n",
            "  In-text Name    : IL-1b\n",
            "  Occurrences     : ['In addition to its role in bone, 1,25(OH)2D also attenuates the inflammatory response of monocytes and macrophages and increases the expression of anti-inflammatory compounds like IL-10. In contrast, the secretion of pro-inflammatory cytokines like IL-1b, IL-6, TNFalpha, RANKL and COX2 is reduced by 1,25(OH)2D. These immunomodulatory effects may explain the 1,25(OH)2D-mediated risk reduction of autoimmune diseases like multiple sclerosis.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 3553 : IL-1b.\n",
            "\n",
            "Gene(3569)\n",
            "  Symbol          : IL6\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : interleukin 6\n",
            "  Also Known As   : BSF-2, BSF2, CDF, HGF, HSF, IFN-beta-2, IFNB2, IL-6\n",
            "  In-text Name    : IL-6\n",
            "  Occurrences     : ['In addition to its role in bone, 1,25(OH)2D also attenuates the inflammatory response of monocytes and macrophages and increases the expression of anti-inflammatory compounds like IL-10. In contrast, the secretion of pro-inflammatory cytokines like IL-1b, IL-6, TNFalpha, RANKL and COX2 is reduced by 1,25(OH)2D. These immunomodulatory effects may explain the 1,25(OH)2D-mediated risk reduction of autoimmune diseases like multiple sclerosis.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 3569 : IL-6.\n",
            "\n",
            "Gene(7124)\n",
            "  Symbol          : TNF\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : tumor necrosis factor\n",
            "  Also Known As   : DIF, IMD127, TNF-alpha, TNFA, TNFSF2, TNLG1F\n",
            "  In-text Name    : TNFalpha\n",
            "  Occurrences     : ['In addition to its role in bone, 1,25(OH)2D also attenuates the inflammatory response of monocytes and macrophages and increases the expression of anti-inflammatory compounds like IL-10. In contrast, the secretion of pro-inflammatory cytokines like IL-1b, IL-6, TNFalpha, RANKL and COX2 is reduced by 1,25(OH)2D. These immunomodulatory effects may explain the 1,25(OH)2D-mediated risk reduction of autoimmune diseases like multiple sclerosis.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 7124 : TNFalpha.\n",
            "\n",
            "Gene(8600)\n",
            "  Symbol          : TNFSF11\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : TNF superfamily member 11\n",
            "  Also Known As   : CD254, ODF, OPGL, OPTB2, RANKL, TNLG6B, TRANCE, hRANKL2, sOdf\n",
            "  In-text Name    : RANKL\n",
            "  Occurrences     : ['In addition to its role in bone, 1,25(OH)2D also attenuates the inflammatory response of monocytes and macrophages and increases the expression of anti-inflammatory compounds like IL-10. In contrast, the secretion of pro-inflammatory cytokines like IL-1b, IL-6, TNFalpha, RANKL and COX2 is reduced by 1,25(OH)2D. These immunomodulatory effects may explain the 1,25(OH)2D-mediated risk reduction of autoimmune diseases like multiple sclerosis.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 8600 : RANKL.\n",
            "\n",
            "Gene(4513)\n",
            "  Symbol          : MT-CO2\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : mitochondrially encoded cytochrome c oxidase II\n",
            "  Also Known As   : COII, MTCO2\n",
            "  In-text Name    : COX2\n",
            "  Occurrences     : ['In addition to its role in bone, 1,25(OH)2D also attenuates the inflammatory response of monocytes and macrophages and increases the expression of anti-inflammatory compounds like IL-10. In contrast, the secretion of pro-inflammatory cytokines like IL-1b, IL-6, TNFalpha, RANKL and COX2 is reduced by 1,25(OH)2D. These immunomodulatory effects may explain the 1,25(OH)2D-mediated risk reduction of autoimmune diseases like multiple sclerosis.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 4513 : COX2.\n",
            "\n",
            "Gene(3630)\n",
            "  Symbol          : INS\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : insulin\n",
            "  Also Known As   : IDDM, IDDM1, IDDM2, ILPR, IRDN, MODY10, PNDM4\n",
            "  In-text Name    : insulin\n",
            "  Occurrences     : ['These immunomodulatory effects may explain the 1,25(OH)2D-mediated risk reduction of autoimmune diseases like multiple sclerosis. Blood pressure, insulin secretion and insulin sensitivity are also modified by calcitriol. These associations could, at least partially, be due to the regulatory role of 1,25(OH)2D in calcium homeostasis.', 'Intravenous calcitriol normalizes insulin sensitivity in uremic patients', 'Role of vitamin D in insulin secretion and insulin sensitivity for glucose homeostasis']\n",
            "  Document ID     : PMC8892115\n",
            "There are 3 occurrence(s) (each occurrence is at most 3 sentences long) for gene 3630 : insulin.\n",
            "\n",
            "Gene(5972)\n",
            "  Symbol          : REN\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : renin\n",
            "  Also Known As   : ADTKD4, HNFJ2, RTD\n",
            "  In-text Name    : renin\n",
            "  Occurrences     : ['These associations could, at least partially, be due to the regulatory role of 1,25(OH)2D in calcium homeostasis. Also, the renin-angiotensin axis is influenced by 1,25(OH)2D, which may explain the association with hypertension (reviewed in).']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 5972 : renin.\n",
            "\n",
            "Gene(5251)\n",
            "  Symbol          : PHEX\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : phosphate regulating endopeptidase X-linked\n",
            "  Also Known As   : HPDR, HPDR1, HYP, HYP1, LXHR, PEX, XLH\n",
            "  In-text Name    : PHEX\n",
            "  Occurrences     : ['These mutations cause rare hereditary metabolic bone conditions, like hereditary vitamin D-resistant rickets (VDR), vitamin-D-dependent rickets type 1A (CYP27B1), type 1B (CYP2R1) or idiopathic infantile hypercalcemia (CYP24A1). Mutations of the cell surface metalloproteinase PHEX cause X-linked hypophosphataemia (XLH), which is characterized by low-normal 1,25(OH)2D concentrations.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 5251 : PHEX.\n",
            "\n",
            "Gene(6569)\n",
            "  Symbol          : SLC34A1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : solute carrier family 34 member 1\n",
            "  Also Known As   : FRTS2, HCINF2, NAPI-3, NPHLOP1, NPT2, NPTIIa, SLC11, SLC17A2\n",
            "  In-text Name    : SLC34A1\n",
            "  Occurrences     : ['CYP24A1 loss-of-function mutations cause infantile hypercalcemia (IIH) and represent a genetic risk factor for serious adverse effects in response to vitamin D supplementation. In hypercalcaemic patients, the parallel measurement of both vitamin D metabolites and calculation of the VMR allows distinguishing CYP24A1 deficiency from other causes of hypercalcaemia, such as vitamin D intoxication, granulomatous disease, Williams-Beuren syndrome or mutations of the SLC34A1 gene. Therefore, 24,25(OH)2D and 25(OH)D can be helpful in guiding genetic testing in patients with inherited hypercalcaemia.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 6569 : SLC34A1.\n",
            "\n",
            "Gene(174)\n",
            "  Symbol          : AFP\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : alpha fetoprotein\n",
            "  Also Known As   : AFPD, FETA, HPAFP\n",
            "  In-text Name    : alpha-fetoprotein\n",
            "  Occurrences     : ['VDBP is a highly polymorphic protein with over 100 isoforms that is encoded by the VDBP gene on chromosome 4q12-q13. Structurally, the protein is closely related to albumin and alpha-fetoprotein. It is mainly produced in the liver, but animal studies suggest that it can also be expressed by other tissues, such as kidney, testis, and adipose tissue.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 174 : alpha-fetoprotein.\n",
            "\n",
            "Gene(4036)\n",
            "  Symbol          : LRP2\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : LDL receptor related protein 2\n",
            "  Also Known As   : DBS, GP330, LRP-2\n",
            "  In-text Name    : megalin\n",
            "  Occurrences     : ['The free hormone hypothesis is supported by the observation that lacking VDBP does not necessarily cause functional vitamin D deficiency despite very low serum concentrations of 25(OH)D and 1,25(OH)2D. In addition to its transport function of vitamin D metabolites, VDBP has also other functions including the renal megalin/cubilin-mediated reuptake of 25(OH)D in the kidneys,the scavenging of actin after muscular damage, neutrophil recruitment, complement 5a-mediated chemotaxis, fatty acid binding and formation of VDBP-Macrophage Activating Factor V (DBP-MAF). For example, significant amounts of actin are released from damaged cells in response to trauma, sepsis, liver trauma, acute lung injury, preeclampsia, surgery, and burns.', 'In contrast to other steroid hormones, the active metabolite 1,25(OH)2D is feedback regulated and the concentration of free hormone is so low that it may be insufficient for passive diffusion across the cell membrane and effective receptor binding. In addition, it cannot be excluded that the megalin-dependent uptake of the VDBP-25(OH)D complex into cells is limited to the proximal renal tubule. For example, a vitamin D-dependent expression megalin and cubilin has been found in several other epithelia including placenta, parathyroid glands, pneumocytes, and epididymal epithelial cells.', 'These changes are most pronounced in cirrhotic patients. Renal disease also decreases VDBP levels, via proteinuria and saturation of the maximal transport capacity of the megalin/cubilin system. Consistently, nephrotic patients present low total and free 25(OH)D levels.', 'Tissue distribution of human gp330/megalin, a putative Ca(2+)-sensing protein']\n",
            "  Document ID     : PMC8892115\n",
            "There are 4 occurrence(s) (each occurrence is at most 3 sentences long) for gene 4036 : megalin.\n",
            "\n",
            "Gene(8029)\n",
            "  Symbol          : CUBN\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : cubilin\n",
            "  Also Known As   : IFCR, IGS, IGS1, MGA1, gp280\n",
            "  In-text Name    : cubilin\n",
            "  Occurrences     : ['The free hormone hypothesis is supported by the observation that lacking VDBP does not necessarily cause functional vitamin D deficiency despite very low serum concentrations of 25(OH)D and 1,25(OH)2D. In addition to its transport function of vitamin D metabolites, VDBP has also other functions including the renal megalin/cubilin-mediated reuptake of 25(OH)D in the kidneys,the scavenging of actin after muscular damage, neutrophil recruitment, complement 5a-mediated chemotaxis, fatty acid binding and formation of VDBP-Macrophage Activating Factor V (DBP-MAF). For example, significant amounts of actin are released from damaged cells in response to trauma, sepsis, liver trauma, acute lung injury, preeclampsia, surgery, and burns.', 'In addition, it cannot be excluded that the megalin-dependent uptake of the VDBP-25(OH)D complex into cells is limited to the proximal renal tubule. For example, a vitamin D-dependent expression megalin and cubilin has been found in several other epithelia including placenta, parathyroid glands, pneumocytes, and epididymal epithelial cells.', 'These changes are most pronounced in cirrhotic patients. Renal disease also decreases VDBP levels, via proteinuria and saturation of the maximal transport capacity of the megalin/cubilin system. Consistently, nephrotic patients present low total and free 25(OH)D levels.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 3 occurrence(s) (each occurrence is at most 3 sentences long) for gene 8029 : cubilin.\n",
            "\n",
            "Gene(83733)\n",
            "  Symbol          : SLC25A18\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : solute carrier family 25 member 18\n",
            "  Also Known As   : GC2\n",
            "  In-text Name    : Gc2\n",
            "  Occurrences     : ['For accuracy, the measurements of 25(OH)D and VDBP are best performed by LC-MS/MS. The polymorphic nature of VDBP results in numerous variants of the protein with Gc1f and Gc1s and Gc2 being most common. These variants are believed to differ in their affinity for 25(OH)D.']\n",
            "  Document ID     : PMC8892115\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 83733 : Gc2.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!pip install biopython\n",
        "import xml.etree.ElementTree as ET\n",
        "from Bio import Entrez\n",
        "import re\n",
        "import time\n",
        "\n",
        "class Gene:\n",
        "    def __init__(self, gene_id):\n",
        "        self.gene_id = gene_id\n",
        "        self.occurrences = []  # list to store snippet(s) where the gene was mentioned\n",
        "        self.symbol = None\n",
        "        self.organism = None\n",
        "        self.full_name = None\n",
        "        self.also_known_as = None\n",
        "        self.name_from_article = None\n",
        "        self.doc_id = None\n",
        "\n",
        "    def add_occurrence(self, snippet):\n",
        "        if snippet not in self.occurrences:  # Avoid duplicates\n",
        "            self.occurrences.append(snippet)\n",
        "\n",
        "    def set_name_from_article(self, name_from_article):\n",
        "        \"\"\"Sets the temporary name of the gene. This is the name according to the article\"\"\"\n",
        "        self.name_from_article = name_from_article\n",
        "\n",
        "    def get_name_from_article(self):\n",
        "        return self.name_from_article\n",
        "\n",
        "    def get_occurrences(self):\n",
        "        return self.occurrences\n",
        "\n",
        "    def get_also_known_as(self):\n",
        "        return self.also_known_as\n",
        "\n",
        "    def get_gene_id(self):\n",
        "        return self.gene_id\n",
        "    \n",
        "    def get_organsim(self):\n",
        "        return self.organism\n",
        "    \n",
        "    def get_doc_id(self):\n",
        "        return self.doc_id\n",
        "    \n",
        "    def set_doc_id(self, doc_id):\n",
        "        self.doc_id = doc_id\n",
        "\n",
        "    def update_info(self, symbol, organism, full_name, also_known_as):\n",
        "        self.symbol = symbol\n",
        "        self.organism = organism\n",
        "        self.full_name = full_name\n",
        "        self.also_known_as = also_known_as\n",
        "\n",
        "    def __repr__(self):\n",
        "        return (f\"Gene({self.gene_id})\\n\"\n",
        "                f\"  Symbol          : {self.symbol}\\n\"\n",
        "                f\"  Organism        : {self.organism}\\n\"\n",
        "                f\"  Full Name       : {self.full_name}\\n\"\n",
        "                f\"  Also Known As   : {self.also_known_as}\\n\"\n",
        "                f\"  In-text Name    : {self.name_from_article}\\n\"\n",
        "                f\"  Occurrences     : {self.occurrences}\\n\"\n",
        "                f\"  Document ID     : {self.doc_id}\")\n",
        "\n",
        "def parse_xml_file(xml_path):\n",
        "    \"\"\"Parses the XML file and returns a gene dictionary keyed by gene ID.\"\"\"\n",
        "    tree = ET.parse(xml_path)\n",
        "    root = tree.getroot()\n",
        "    gene_dict = {}\n",
        "\n",
        "    for document in root.findall('document'):\n",
        "        \n",
        "        # get the document ID (pmid or pmcid)\n",
        "        doc_id_elem = document.find('id')\n",
        "        if doc_id_elem is not None:\n",
        "            doc_id = doc_id_elem.text\n",
        "        else:\n",
        "            doc_id = None\n",
        "\n",
        "\n",
        "        # get the gene annotations in the document from the passages\n",
        "        for passage in document.findall('passage'):       \n",
        "            # Also check for 'article-id_pmc'\n",
        "            article_id_pmc_elem = passage.find(\"infon[@key='article-id_pmc']\")\n",
        "            if article_id_pmc_elem is not None:\n",
        "                pmc_value = article_id_pmc_elem.text.strip()\n",
        "                if not pmc_value.startswith(\"PMC\"):\n",
        "                    pmc_value = \"PMC\" + pmc_value\n",
        "                doc_id = pmc_value     \n",
        "            section_type_elem = passage.find(\"infon[@key='section_type']\")\n",
        "            if section_type_elem is not None and section_type_elem.text and section_type_elem.text.upper() in (\"METHODS\", \"FIG\", \"TABLE\"):\n",
        "                continue  # Skip passages under METHODS\n",
        "            # Get the full passage text.\n",
        "            passage_text_elem = passage.find(\"text\")\n",
        "            passage_text = passage_text_elem.text if passage_text_elem is not None else \"\"\n",
        "            # Determine the starting offset for this passage.\n",
        "            passage_offset_elem = passage.find(\"offset\")\n",
        "            passage_offset = int(passage_offset_elem.text) if passage_offset_elem is not None else 0\n",
        "            # Split the passage into sentences using regex.\n",
        "            sentences = re.split(r'(?<=[.!?])\\s+', passage_text)\n",
        "            # Compute start indices for each sentence within the passage text.\n",
        "            start_indices = []\n",
        "            current_index = passage_offset\n",
        "            for sentence in sentences:\n",
        "                start_indices.append(current_index)\n",
        "                current_index += len(sentence) + 1  # account for the delimiter space\n",
        "\n",
        "            # Dictionary to store annotations per gene in this passage.\n",
        "            # For each gene_id, we collect the sentence indices and a temporary in-text name.\n",
        "            gene_annotations = {}\n",
        "            for annotation in passage.findall('annotation'):\n",
        "                ann_type = annotation.find(\"infon[@key='type']\")\n",
        "                if ann_type is not None and ann_type.text == \"Gene\":\n",
        "                    # Check for both 'identifier' and 'NCBI Gene' keys.\n",
        "                    gene_id_elem = annotation.find(\"infon[@key='identifier']\")\n",
        "                    if gene_id_elem is None:\n",
        "                        gene_id_elem = annotation.find(\"infon[@key='NCBI Gene']\")\n",
        "                    gene_id = gene_id_elem.text if gene_id_elem is not None else None\n",
        "\n",
        "                    # Handle multiple IDs by keeping only the first if needed.\n",
        "                    if gene_id and \";\" in gene_id:\n",
        "                        gene_id = gene_id.split(\";\")[0].strip()\n",
        "\n",
        "                    # Extract the gene name from the annotation text.\n",
        "                    in_text_gene_name_elem = annotation.find(\"text\")\n",
        "                    in_text_gene_name = in_text_gene_name_elem.text if in_text_gene_name_elem is not None else None\n",
        "\n",
        "                    # Extract the annotation location (offset) to find the sentence.\n",
        "                    location_elem = annotation.find(\"location\")\n",
        "                    ann_offset = int(location_elem.attrib.get('offset', 0)) if location_elem is not None else 0\n",
        "\n",
        "                    # Determine which sentence contains the annotation based on its offset.\n",
        "                    sentence_index = None\n",
        "                    for i, start in enumerate(start_indices):\n",
        "                        if start <= ann_offset < start + len(sentences[i]):\n",
        "                            sentence_index = i\n",
        "                            break\n",
        "\n",
        "                    if sentence_index is not None and gene_id:\n",
        "                        if gene_id not in gene_annotations:\n",
        "                            gene_annotations[gene_id] = {'indices': [], 'name': in_text_gene_name}\n",
        "                        gene_annotations[gene_id]['indices'].append(sentence_index)\n",
        "\n",
        "            # For each gene in this passage, process the annotation indices.\n",
        "            for gene_id, data in gene_annotations.items():\n",
        "                # Remove duplicate indices and sort.\n",
        "                indices = sorted(set(data['indices']))\n",
        "                candidate_windows = []\n",
        "                # For each annotation occurrence, form a candidate window:\n",
        "                # one sentence before and one sentence after the annotated sentence.\n",
        "                for idx in indices:\n",
        "                    start_sentence = max(0, idx - 1)\n",
        "                    end_sentence = min(len(sentences), idx + 2)  # end index is non-inclusive\n",
        "                    candidate_windows.append((start_sentence, end_sentence))\n",
        "                # Merge overlapping candidate windows.\n",
        "                merged_windows = []\n",
        "                for window in sorted(candidate_windows):\n",
        "                    if not merged_windows:\n",
        "                        merged_windows.append(window)\n",
        "                    else:\n",
        "                        last = merged_windows[-1]\n",
        "                        # If windows overlap or touch, merge them.\n",
        "                        if window[0] <= last[1]:\n",
        "                            merged_windows[-1] = (last[0], max(last[1], window[1]))\n",
        "                        else:\n",
        "                            merged_windows.append(window)\n",
        "                # For each merged window, if it spans more than 3 sentences, split it into chunks of 3 sentences maximum.\n",
        "                for window in merged_windows:\n",
        "                    start, end = window\n",
        "                    window_size = end - start\n",
        "                    if window_size <= 3:\n",
        "                        snippet = \" \".join(sentences[start:end])\n",
        "                        if gene_id in gene_dict:\n",
        "                            gene_dict[gene_id].add_occurrence(snippet)\n",
        "                        else:\n",
        "                            gene_obj = Gene(gene_id)\n",
        "                            gene_obj.add_occurrence(snippet)\n",
        "                            gene_obj.set_doc_id(doc_id)\n",
        "                            gene_obj.set_name_from_article(data['name'])\n",
        "                            gene_dict[gene_id] = gene_obj\n",
        "                    else:\n",
        "                        # Split the merged window into non-overlapping chunks of at most 3 sentences.\n",
        "                        for i in range(start, end, 3):\n",
        "                            chunk_end = min(i + 3, end)\n",
        "                            snippet = \" \".join(sentences[i:chunk_end])\n",
        "                            if snippet.strip():\n",
        "                                if gene_id in gene_dict:\n",
        "                                    gene_dict[gene_id].add_occurrence(snippet)\n",
        "                                else:\n",
        "                                    gene_obj = Gene(gene_id)\n",
        "                                    gene_obj.set_doc_id(doc_id)\n",
        "                                    gene_obj.add_occurrence(snippet)\n",
        "                                    gene_obj.set_name_from_article(data['name'])\n",
        "                                    gene_dict[gene_id] = gene_obj\n",
        "    return gene_dict\n",
        "\n",
        "def fetch_and_update_gene_info(gene_dict):\n",
        "    \"\"\"Retrieves gene information from NCBI and updates the genes in gene_dict.\"\"\"\n",
        "    gene_ids = list(gene_dict.keys())\n",
        "    if gene_ids:\n",
        "        # Post the gene IDs to NCBI.\n",
        "        handle = Entrez.epost(db=\"gene\", id=\",\".join(gene_ids))\n",
        "        result = Entrez.read(handle)\n",
        "        handle.close()\n",
        "\n",
        "        webenv = result[\"WebEnv\"]\n",
        "        query_key = result[\"QueryKey\"]\n",
        "\n",
        "        handle = Entrez.esummary(db=\"gene\", webenv=webenv, query_key=query_key)\n",
        "        record = Entrez.read(handle)\n",
        "        handle.close()\n",
        "\n",
        "        for docsum in record[\"DocumentSummarySet\"][\"DocumentSummary\"]:\n",
        "            gene_id = docsum.attributes[\"uid\"]\n",
        "            symbol = docsum.get('NomenclatureSymbol', 'No symbol')\n",
        "            organism = docsum.get('Organism', {}).get('ScientificName', 'No organism')\n",
        "            full_name = docsum.get('NomenclatureName', gene_dict[gene_id].get_name_from_article())\n",
        "            also_known_as = docsum.get('OtherAliases', gene_dict[gene_id].get_name_from_article())\n",
        "\n",
        "            if full_name == '':\n",
        "                full_name = gene_dict[gene_id].get_name_from_article()\n",
        "            if also_known_as == '':\n",
        "                also_known_as = gene_dict[gene_id].get_name_from_article()\n",
        "\n",
        "            if gene_id in gene_dict:\n",
        "                gene_dict[gene_id].update_info(symbol, organism, full_name, also_known_as)\n",
        "            time.sleep(0.5)  # Pause briefly to avoid overwhelming NCBI servers.\n",
        "\n",
        "# Set your email (and API key if available)\n",
        "Entrez.email = \"email here\"\n",
        "# Entrez.api_key = \"your_api_key\"\n",
        "\n",
        "xml_path = \"./input/PMC8892115.xml\"\n",
        "gene_dict = parse_xml_file(xml_path)\n",
        "fetch_and_update_gene_info(gene_dict)  # Optionally fetch gene info from NCBI\n",
        "\n",
        "print(\"There are {} genes in the paper.\".format(len(gene_dict)))\n",
        "if len(gene_dict) == 0:\n",
        "    print(\"No genes found in the paper.\")\n",
        "else:\n",
        "    for gene in gene_dict.values():\n",
        "        print(gene)\n",
        "        print(f\"There are {len(gene.get_occurrences())} occurrence(s) (each occurrence is at most 3 sentences long) for gene {gene.get_gene_id()} : {gene.get_name_from_article()}.\")\n",
        "        print(\"\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "33A7yYBLXZes"
      },
      "source": [
        "# 3.5 Prompt Engineering\n",
        "- Creates the prompt using the Gene dictionary (from Gene chunking)\n",
        "- batch_size: allows you to have more than 1 Gene in the prompt such that you have something like the following. Ideally keep it at batch_size=1 to ensure that the GO terms generated are exclusive to the Gene being analyzed.\n",
        "\n",
        "```\n",
        "# batch_size > 1 has the basic format:\n",
        "Basic Prompt\n",
        "Gene 1\n",
        "Gene 2\n",
        "...\n",
        "# batch_size = 1 has the basic format:\n",
        "Basic Prompt\n",
        "Gene 1\n",
        "```\n",
        "- batch_index: allows you to iterate through the gene dictionary so that if you have a paper with x amount of genes, you can just increment  the index and run the inference x amount of times.\n",
        "- direct: ignore this. This is for the future.\n",
        "- customized_prompt: ignore this as well. This is for the future."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UsRz3Wb5YTQl",
        "outputId": "dc245a37-b770-4af2-f071-f0b973bda774"
      },
      "outputs": [],
      "source": [
        "def make_go_term_prompt(gene_dict, batch_size=1, batch_index=0, direct=False, customized_prompt=None):\n",
        "    \"\"\"\n",
        "    Create a prompt for Llama to output Gene Ontology (GO) terms (without IDs) for a batch of genes\n",
        "    from the gene dictionary. The output is expected to be categorized into three sections for each gene:\n",
        "      1. Biological Processes\n",
        "      2. Cellular Components\n",
        "      3. Molecular Functions\n",
        "\n",
        "    For each gene, the output should follow the format below without mixing results across different genes.\n",
        "\n",
        "    :param gene_dict: A dictionary where each key is a gene ID and each value is a Gene object that\n",
        "                      contains attributes such as symbol, organism, full_name, and occurrences.\n",
        "    :param batch_size: The number of genes to include in this prompt batch.\n",
        "    :param batch_index: The index (0-based) of the batch to process. For instance, if batch_size=1,\n",
        "                        batch_index=2 will include the 3rd gene in the dictionary.\n",
        "    :param direct: If True, use a more direct instruction for GO term extraction.\n",
        "    :param customized_prompt: If provided, use this custom prompt text in place of the default.\n",
        "    :return: A string containing the complete prompt.\n",
        "    \"\"\"\n",
        "\n",
        "    # Instruction blocks\n",
        "\n",
        "    # context = \"\"\"You are an efficient and insightful assistant to a geneticist.\"\"\"\n",
        "\n",
        "    general_instructions = \"\"\"Analyze the following gene information and text snippets under \"Occurrences\".\n",
        "Identify all potentially relevant Gene Ontology (GO) terms using only evidence explicitly stated in the text.\n",
        "Do not include any GO terms based on inferences or assumptions not explicitly supported by the text.\n",
        "\n",
        "After completing your analysis, for each Gene Ontology (GO) term you list, provide a short one-line justification referencing the text as well as a relationship between the gene and the term.\n",
        "If the text does not explicitly support a GO term (even if it might be inferred), do not include it. If no GO term is supported for a particular gene, say “None found.”\n",
        "\n",
        "Be concise, do not use unnecessary words.\n",
        "Be factual, do not editorialize.\n",
        "Be specific, avoid overly general statements like \"it is involved in many cellular processes\".\n",
        "\n",
        "\"\"\"\n",
        "\n",
        "    example = \"\"\"To help you in your work, I am providing an example of gene information and the corresponding example analysis output structured in JSON.\n",
        "\n",
        "Example gene information is:\n",
        "Gene: myogenin\n",
        "Organism: Mus musculus\n",
        "Occurrences:\n",
        "- Activation of NF-kappaB increases the expression of the inducible nitric oxide synthase (iNOS) in muscle cells that sequester HuR (RNA-binding protein) by preventing transcription of MyoD. Denervation-induced atrophy is related to the upregulation of specific histone deacetylases (HDAC), able to repress a negative regulator of myogenin, with an increase in MuRF1 expression and muscle wasting. AMPK (protein kinase activated by 5'adenosine monophosphate) is also involved in muscle atrophy.\n",
        "\n",
        "Example anaylysis output is:\n",
        "[\n",
        "  {\n",
        "    \"gene\": \"myogenin\",\n",
        "    \"go_data\": [\n",
        "      {\n",
        "        \"term\": \"DNA-binding transcription activator activity\",\n",
        "        \"relationship\": \"enables\",\n",
        "        \"namespace\": \"molecular_function\",\n",
        "        \"justification\": \"Myogenin is a key transcription factor in muscle differentiation that binds to E-box sequences in muscle-specific genes and activates their transcription through RNA polymerase II.\"\n",
        "      },\n",
        "      {\n",
        "        \"term\": \"skeletal muscle cell differentiation\",\n",
        "        \"relationship\": \"involved_in\",\n",
        "        \"namespace\": \"biological_process\",\n",
        "        \"justification\": \"Myogenin is one of the four myogenic regulatory factors (MRFs) that drive skeletal muscle differentiation by promoting the transition from myoblasts to myotubes.\"\n",
        "      },\n",
        "      {\n",
        "        \"term\": \"nucleus\",\n",
        "        \"relationship\": \"located_in\",\n",
        "        \"namespace\": \"cellular_component\",\n",
        "        \"justification\": \"As a transcription factor, myogenin localizes to the nucleus to regulate the expression of muscle-specific genes.\"\n",
        "      }\n",
        "    ]\n",
        "  }\n",
        "]\n",
        "\"\"\"\n",
        "\n",
        "\n",
        "    # Assemble the prompt text based on the chosen mode\n",
        "    if direct:\n",
        "        # prompt_text = context\n",
        "        # prompt_text += direct_instructions\n",
        "        # prompt_text += format_instructions\n",
        "        # prompt_text += example_output\n",
        "        pass\n",
        "    elif customized_prompt:\n",
        "        # prompt_text = context\n",
        "        # prompt_text += customized_prompt\n",
        "        # prompt_text += format_instructions\n",
        "        # prompt_text += example_output\n",
        "        pass\n",
        "    else:\n",
        "        prompt_text = general_instructions\n",
        "        prompt_text += example\n",
        "        prompt_text += \"\\nHere is the gene information:\\n\"\n",
        "\n",
        "    # Convert the gene dictionary to a list and determine the batch slice.\n",
        "    gene_items = list(gene_dict.items())\n",
        "    start_index = batch_index * batch_size\n",
        "    end_index = start_index + batch_size\n",
        "    batch_genes = gene_items[start_index:end_index]\n",
        "\n",
        "    # Add details and occurrence context for each gene in the selected batch.\n",
        "    for gene_id, gene_obj in batch_genes:\n",
        "        # prompt_text += f\"\\nGene ID: {gene_id}\\n\"\n",
        "        # prompt_text += f\"Symbol: {gene_obj.symbol}\\n\"\n",
        "        # prompt_text += f\"Organism: {gene_obj.organism}\\n\"\n",
        "        # prompt_text += f\"Full Name: {gene_obj.full_name}\\n\"\n",
        "        prompt_text += f\"\\nGene: {gene_obj.get_name_from_article()}\\n\"\n",
        "        if gene_obj.get_organsim() != None:\n",
        "          # add organism if available\n",
        "          prompt_text += f\"Organism: {gene_obj.get_organsim()}\\n\"\n",
        "        prompt_text += \"Occurrences:\\n\"\n",
        "        for occ in gene_obj.occurrences:\n",
        "            prompt_text += f\"- {occ}\\n\"\n",
        "\n",
        "    return prompt_text\n",
        "\n",
        "# Example usage:\n",
        "# Assuming gene_dict is defined and contains Gene objects with attributes: symbol, organism, full_name, and occurrences.\n",
        "# To process only one gene at a time:\n",
        "# prompt = make_go_term_prompt(gene_dict, batch_size=1, batch_index=0)\n",
        "# print(prompt)\n",
        "\n",
        "# prompt = make_go_term_prompt(gene_dict, batch_size=116, batch_index=0)\n",
        "# print(prompt)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6oaKSo6iXGOc"
      },
      "source": [
        "# 4. Inference (Local)\n",
        "The two cells below rely on local inference meaning that you must have the model locally. Skip them if youre using the API like GROQ"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hzFESXHRXBHG"
      },
      "outputs": [],
      "source": [
        "print(\"Loading LLaMA model.\")\n",
        "# model_dir = \"/content/drive/MyDrive/Llama 3.2-3B-Instruct-model\"\n",
        "model_dir = \"/content/drive/MyDrive/GOLLM/Llama 3.2-3B-Instruct-model\"\n",
        "device = torch.device(\"cuda\")\n",
        "print(torch.cuda.is_available())\n",
        "\n",
        "tokenizer = AutoTokenizer.from_pretrained(model_dir, local_files_only=True)\n",
        "model = AutoModelForCausalLM.from_pretrained(model_dir, local_files_only=True).to(device)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 262,
      "metadata": {
        "id": "IzBYJajjYYly"
      },
      "outputs": [
        {
          "ename": "NameError",
          "evalue": "name 'test_prompt' is not defined",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[262], line 69\u001b[0m\n\u001b[1;32m     67\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m3\u001b[39m):\n\u001b[1;32m     68\u001b[0m     start_time \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[0;32m---> 69\u001b[0m     output \u001b[38;5;241m=\u001b[39m run_inference(\u001b[43mtest_prompt\u001b[49m)\n\u001b[1;32m     70\u001b[0m     \u001b[38;5;28mprint\u001b[39m(output)\n\u001b[1;32m     72\u001b[0m     \u001b[38;5;66;03m# with open(f'/content/drive/MyDrive/GOLLM/Inference Outputs/output_{i+1}.txt', 'w') as file:\u001b[39;00m\n\u001b[1;32m     73\u001b[0m     \u001b[38;5;66;03m#     file.write(output)\u001b[39;00m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'test_prompt' is not defined"
          ]
        }
      ],
      "source": [
        "def run_inference(prompt):\n",
        "    \"\"\"\n",
        "    Run inference on the provided prompt using the specified model and tokenizer.\n",
        "\n",
        "    :param prompt: The prompt string to send to the model.\n",
        "    :return: The generated text.\n",
        "    \"\"\"\n",
        "\n",
        "    # Encode the prompt and move to the appropriate device\n",
        "    inputs = tokenizer(prompt, return_tensors=\"pt\").to(\"cuda\")\n",
        "    attention_mask = inputs[\"attention_mask\"]\n",
        "\n",
        "    # Greedy or minimal variation == When: You want consistent, short answers, and you don’t mind if it’s a bit “blunt.”\n",
        "    # outputs = model.generate(\n",
        "    #     input_ids=inputs[\"input_ids\"],\n",
        "    #     attention_mask=attention_mask,\n",
        "    #     max_new_tokens=150,\n",
        "    #     do_sample=False,     # no sampling\n",
        "    #     num_beams=1,         # purely greedy\n",
        "    #     no_repeat_ngram_size=3,\n",
        "    #     pad_token_id=tokenizer.eos_token_id\n",
        "    # )\n",
        "\n",
        "    # Beam Search (Higher Quality / Less Randomness) == You want a more “global optimum” text.\n",
        "    # outputs = model.generate(\n",
        "    #     input_ids=inputs[\"input_ids\"],\n",
        "    #     attention_mask=attention_mask,\n",
        "    #     max_new_tokens=200,\n",
        "    #     do_sample=False,\n",
        "    #     num_beams=4,      # search multiple beams\n",
        "    #     length_penalty=1.0,  # see if you want to encourage or discourage long outputs\n",
        "    #     no_repeat_ngram_size=3,\n",
        "    #     pad_token_id=tokenizer.eos_token_id\n",
        "    # )\n",
        "\n",
        "    # Beam SearchV2\n",
        "    outputs = model.generate(\n",
        "        input_ids=inputs[\"input_ids\"],\n",
        "        attention_mask=attention_mask,\n",
        "        max_new_tokens=400,       # Enough tokens for a concise but thorough answer\n",
        "        do_sample=False,          # Turn off sampling; we want a more deterministic, stable answer\n",
        "        num_beams=3,              # Explore multiple beams for higher-quality completions\n",
        "        length_penalty=0.1,       # 1.0 means \"neutral\" length preference (>=1.0 encourages longer outputs)\n",
        "        no_repeat_ngram_size=4,   # Helps avoid repeating the same phrase\n",
        "        early_stopping=True,      # Stops as soon as the best beam is complete\n",
        "        pad_token_id=tokenizer.eos_token_id,\n",
        "        temperature=None,\n",
        "        top_p=None,\n",
        "    )\n",
        "    # Sampling (More Creative / Less Deterministic)\n",
        "    # outputs = model.generate(\n",
        "    #     input_ids=inputs[\"input_ids\"],\n",
        "    #     attention_mask=attention_mask,\n",
        "    #     max_new_tokens=200,\n",
        "    #     do_sample=True,       # sampling\n",
        "    #     temperature=0.7,      # moderate creativity\n",
        "    #     top_p=0.9,            # nucleus sampling\n",
        "    #     no_repeat_ngram_size=3,\n",
        "    #     pad_token_id=tokenizer.eos_token_id\n",
        "    # )\n",
        "\n",
        "    # Decode the output and return the generated text\n",
        "    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)\n",
        "    completion = generated_text[len(prompt):].strip()\n",
        "    return completion\n",
        "\n",
        "for i in range(3):\n",
        "    start_time = time.time()\n",
        "    output = run_inference(test_prompt)\n",
        "    print(output)\n",
        "\n",
        "    # with open(f'/content/drive/MyDrive/GOLLM/Inference Outputs/output_{i+1}.txt', 'w') as file:\n",
        "    #     file.write(output)\n",
        "\n",
        "    print(f\"-->Model loaded on {device} named {torch.cuda.get_device_name(0)} and it took {time.time()-start_time:.2f} seconds.\")\n",
        "    print(\"-\"*50)\n",
        "\n",
        "# prompt = \"what is 2+2\"\n",
        "# print(\"Running inference...\")\n",
        "# start_time = time.time()\n",
        "# output = run_inference(prompt)\n",
        "# print(f\"Model loaded on {device} in {time.time()-start_time:.2f} seconds.\")\n",
        "# print(output)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gSz3gKoaY7yu"
      },
      "source": [
        "# 4v2. GROQ API Inference\n",
        "- need API Key\n",
        "- uses llama 3.3-70b-versatile"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "J_k_yNbOwpkb",
        "outputId": "16a5c4d1-e0bd-4c8b-a736-77a7dd2e2130"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
            "To disable this warning, you can either:\n",
            "\t- Avoid using `tokenizers` before the fork if possible\n",
            "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
            "1320122.50s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: Groq in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (0.18.0)\n",
            "Requirement already satisfied: pydantic<3,>=1.9.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (2.10.6)\n",
            "Requirement already satisfied: anyio<5,>=3.5.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (4.8.0)\n",
            "Requirement already satisfied: httpx<1,>=0.23.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (0.28.1)\n",
            "Requirement already satisfied: distro<2,>=1.7.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (1.9.0)\n",
            "Requirement already satisfied: sniffio in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (1.3.1)\n",
            "Requirement already satisfied: typing-extensions<5,>=4.10 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (4.12.2)\n",
            "Requirement already satisfied: idna>=2.8 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from anyio<5,>=3.5.0->Groq) (3.10)\n",
            "Requirement already satisfied: exceptiongroup>=1.0.2 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from anyio<5,>=3.5.0->Groq) (1.2.2)\n",
            "Requirement already satisfied: httpcore==1.* in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from httpx<1,>=0.23.0->Groq) (1.0.7)\n",
            "Requirement already satisfied: certifi in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from httpx<1,>=0.23.0->Groq) (2024.12.14)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from httpcore==1.*->httpx<1,>=0.23.0->Groq) (0.14.0)\n",
            "Requirement already satisfied: pydantic-core==2.27.2 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from pydantic<3,>=1.9.0->Groq) (2.27.2)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from pydantic<3,>=1.9.0->Groq) (0.7.0)\n",
            "\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip available: \u001b[0m\u001b[31;49m22.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
          ]
        }
      ],
      "source": [
        "# using GROQ API\n",
        "!pip install Groq\n",
        "from groq import Groq\n",
        "\n",
        "client = Groq(\n",
        "    api_key=\"\"\n",
        ")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pmOgXgUc_DAC"
      },
      "source": [
        "## Run it again if it stops. It will skip documents that already exists."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 558,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6eOnOkZsztbK",
        "outputId": "f9dbfdd0-0f1e-4238-cfb7-7cca62af4790"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Skipping gene 2638: output file already exists.\n",
            "Skipping gene 120227: output file already exists.\n",
            "Skipping gene 1593: output file already exists.\n",
            "Skipping gene 1594: output file already exists.\n",
            "Skipping gene 1591: output file already exists.\n",
            "Skipping gene 1628: output file already exists.\n",
            "Skipping gene 213: output file already exists.\n",
            "Skipping gene 7421: output file already exists.\n",
            "Skipping gene 5741: output file already exists.\n",
            "Skipping gene 8856: output file already exists.\n",
            "Skipping gene 1576: output file already exists.\n",
            "Skipping gene 8074: output file already exists.\n",
            "Skipping gene 55503: output file already exists.\n",
            "Skipping gene 795: output file already exists.\n",
            "Skipping gene 490: output file already exists.\n",
            "Skipping gene 9075: output file already exists.\n",
            "Skipping gene 9365: output file already exists.\n",
            "Skipping gene 12097: output file already exists.\n",
            "Skipping gene 13088: output file already exists.\n",
            "Skipping gene 83430: output file already exists.\n",
            "Skipping gene 3586: output file already exists.\n",
            "Skipping gene 3553: output file already exists.\n",
            "Skipping gene 3569: output file already exists.\n",
            "Skipping gene 7124: output file already exists.\n",
            "Skipping gene 8600: output file already exists.\n",
            "Skipping gene 4513: output file already exists.\n",
            "Skipping gene 3630: output file already exists.\n",
            "Skipping gene 5972: output file already exists.\n",
            "Skipping gene 5251: output file already exists.\n",
            "Skipping gene 6569: output file already exists.\n",
            "Skipping gene 174: output file already exists.\n",
            "Skipping gene 4036: output file already exists.\n",
            "Skipping gene 8029: output file already exists.\n",
            "Skipping gene 83733: output file already exists.\n",
            "Finished processing 34 genes. Final merged JSON saved to /Users/crystalcho/GOLLM/Llama3.2-3B-Instruct-Inference/llama_inferences/llama_output_PMC8892115.json\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "import json\n",
        "import tempfile\n",
        "import shutil \n",
        "\n",
        "# Create a temporary directory to store individual JSON outputs.\n",
        "temp_dir = os.path.join(os.getcwd(), \"temp_inferences\")\n",
        "if not os.path.exists(temp_dir):\n",
        "    os.makedirs(temp_dir)\n",
        "# Create a  directory to store individual llama JSON outputs.\n",
        "llama_dir = os.path.join(os.getcwd(), \"llama_inferences\")\n",
        "if not os.path.exists(llama_dir):\n",
        "    os.makedirs(llama_dir)\n",
        "\n",
        "# retrieve the doc_id from the gene_dict of any gene (since all genes in the gene_dict are from the same document)\n",
        "doc_id = None\n",
        "for gene in gene_dict.values():\n",
        "    if gene.get_doc_id() is not None:\n",
        "        doc_id = gene.get_doc_id()\n",
        "        break\n",
        "\n",
        "# Create a subdirectory under temp_inferences named after doc_id.\n",
        "# If doc_id is None, fall back to a placeholder name, e.g. \"unknown_doc_id\".\n",
        "temp_doc_dir = os.path.join(temp_dir, doc_id if doc_id else \"unknown_doc_id\")\n",
        "if not os.path.exists(temp_doc_dir):\n",
        "    os.makedirs(temp_doc_dir)\n",
        "\n",
        "# List of gene items from gene_dict\n",
        "gene_items = list(gene_dict.items())\n",
        "final_results = []\n",
        "\n",
        "# Process each gene one-by-one (batch_size=1)\n",
        "for idx, (gene_id, gene_obj) in enumerate(gene_items):\n",
        "    output_filename = os.path.join(temp_doc_dir, f\"gene_{gene_id}.json\")\n",
        "\n",
        "    # Check if the gene's output file already exists\n",
        "    if os.path.exists(output_filename):\n",
        "        print(f\"Skipping gene {gene_id}: output file already exists.\")\n",
        "        try:\n",
        "            with open(output_filename, \"r\", encoding=\"utf-8\") as f:\n",
        "                gene_output = f.read()\n",
        "            final_results.append(json.loads(gene_output))\n",
        "        except json.JSONDecodeError:\n",
        "            print(f\"Warning: Existing output for gene {gene_id} is not valid JSON.\")\n",
        "        continue\n",
        "\n",
        "    # Generate a prompt for a single gene.\n",
        "    prompt = make_go_term_prompt(gene_dict, batch_size=1, batch_index=idx)\n",
        "    # Use your GROQ API client to request the inference.\n",
        "    chat_completion = client.chat.completions.create(\n",
        "        messages=[\n",
        "            {\"role\": \"user\", \"content\": prompt},\n",
        "        ],\n",
        "        model=\"llama-3.3-70b-versatile\",\n",
        "        temperature=0,\n",
        "        max_completion_tokens=4096,\n",
        "        top_p=0.5,\n",
        "        stream=False,\n",
        "        response_format={\"type\": \"json_object\"},\n",
        "        stop=None,\n",
        "    )\n",
        "\n",
        "    # Extract the JSON output from the response.\n",
        "    gene_output = chat_completion.choices[0].message.content\n",
        "\n",
        "    # Add the output to the final results list.\n",
        "    try:\n",
        "        model_json = json.loads(gene_output)\n",
        "\n",
        "        # Construct a new dict to ensure key order: gene, geneID, go_data\n",
        "        ordered_model_json = {\n",
        "            \"gene\": model_json.get(\"gene\", \"\"),\n",
        "            \"geneID\": gene_id,\n",
        "            \"go_data\": model_json.get(\"go_data\", [])\n",
        "        }\n",
        "\n",
        "        # If go_data is not a list, set it to an empty list\n",
        "        if not isinstance(ordered_model_json[\"go_data\"], list):\n",
        "            ordered_model_json[\"go_data\"] = []\n",
        "\n",
        "        # Add reference(s) to each GO entry\n",
        "        for entry in ordered_model_json[\"go_data\"]:\n",
        "            # if the term is None found, skip adding the reference\n",
        "            if entry[\"term\"] == \"None found\":\n",
        "                ordered_model_json[\"go_data\"] = []\n",
        "            entry[\"reference(s)\"] = [gene_obj.get_doc_id()]\n",
        "\n",
        "        # Save the updated JSON in the temp folder\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            json.dump(ordered_model_json, f, indent=2)\n",
        "\n",
        "        final_results.append(ordered_model_json)\n",
        "    except json.JSONDecodeError:\n",
        "        print(f\"Warning: Output for gene {gene_id} is not valid JSON.\")\n",
        "\n",
        "# Merge all outputs into one finalized JSON file, using doc_id in the filename\n",
        "llama_output_path = os.path.join(llama_dir, f\"llama_output_{doc_id}.json\")\n",
        "with open(llama_output_path, \"w\", encoding=\"utf-8\") as f:\n",
        "    json.dump(final_results, f, indent=2)\n",
        "\n",
        "print(f\"Finished processing {len(gene_items)} genes. Final merged JSON saved to {llama_output_path}\")\n",
        "\n",
        "# Remove the temporary directory\n",
        "# shutil.rmtree(temp_dir)\n",
        "# print(f\"Removed temporary folder: {temp_dir}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "P58Ns1lwdbbg"
      },
      "source": [
        "# 5. Normalize\n",
        "- use an embedding model (SAPBERT) to ensure that the terms generated by Llama are actual GO terms. You will need the Go.obo file from the [Gene Ontology Consortium](https://geneontology.org/docs/download-ontology/)\n",
        "  - Essentially, for each go term created by Llama, Sapbert will do a similarity search against the Go.obo and retrieve the actual term + GO:ID"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 454,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "n5lKdeAZ1Ssh",
        "outputId": "31f1d594-43c0-4f83-e46c-4873128724d0"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
            "To disable this warning, you can either:\n",
            "\t- Avoid using `tokenizers` before the fork if possible\n",
            "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
            "1310753.89s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: sentence-transformers in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (3.4.1)\n",
            "Requirement already satisfied: transformers<5.0.0,>=4.41.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (4.48.1)\n",
            "Requirement already satisfied: scikit-learn in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (1.6.1)\n",
            "Requirement already satisfied: scipy in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (1.15.1)\n",
            "Requirement already satisfied: torch>=1.11.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (2.5.1)\n",
            "Requirement already satisfied: huggingface-hub>=0.20.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (0.27.1)\n",
            "Requirement already satisfied: tqdm in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (4.67.1)\n",
            "Requirement already satisfied: Pillow in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (11.1.0)\n",
            "Requirement already satisfied: pyyaml>=5.1 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (6.0.2)\n",
            "Requirement already satisfied: packaging>=20.9 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (24.2)\n",
            "Requirement already satisfied: requests in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (2.32.3)\n",
            "Requirement already satisfied: typing-extensions>=3.7.4.3 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (4.12.2)\n",
            "Requirement already satisfied: filelock in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (3.17.0)\n",
            "Requirement already satisfied: fsspec>=2023.5.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (2024.12.0)\n",
            "Requirement already satisfied: sympy==1.13.1 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from torch>=1.11.0->sentence-transformers) (1.13.1)\n",
            "Requirement already satisfied: networkx in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from torch>=1.11.0->sentence-transformers) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from torch>=1.11.0->sentence-transformers) (3.1.5)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sympy==1.13.1->torch>=1.11.0->sentence-transformers) (1.3.0)\n",
            "Requirement already satisfied: safetensors>=0.4.1 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (0.5.2)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (2024.11.6)\n",
            "Requirement already satisfied: numpy>=1.17 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (2.2.2)\n",
            "Requirement already satisfied: tokenizers<0.22,>=0.21 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (0.21.0)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from scikit-learn->sentence-transformers) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from scikit-learn->sentence-transformers) (3.5.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from jinja2->torch>=1.11.0->sentence-transformers) (3.0.2)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (2024.12.14)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (2.3.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (3.10)\n",
            "\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip available: \u001b[0m\u001b[31;49m22.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
            "To disable this warning, you can either:\n",
            "\t- Avoid using `tokenizers` before the fork if possible\n",
            "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
            "1310760.28s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: obonet in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (1.1.0)\n",
            "Requirement already satisfied: networkx in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from obonet) (3.4.2)\n",
            "\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip available: \u001b[0m\u001b[31;49m22.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
          ]
        }
      ],
      "source": [
        "# Install required packages (run these in your environment if not already installed)\n",
        "!pip install sentence-transformers\n",
        "!pip install obonet"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 455,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8TPMPZKF1P1X",
        "outputId": "80689670-c815-40d7-8bb8-d202d1fc88c4"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Loaded 40267 GO terms from /Users/crystalcho/Downloads/go.obo\n"
          ]
        }
      ],
      "source": [
        "import obonet\n",
        "import json\n",
        "from sentence_transformers import SentenceTransformer, util\n",
        "import numpy as np\n",
        "\n",
        "##########################################\n",
        "# 1. Load and parse the GO.obo file\n",
        "##########################################\n",
        "\n",
        "def load_go_ontology(obo_filepath):\n",
        "    \"\"\"\n",
        "    Parse the GO.obo file using obonet.\n",
        "    Returns a dictionary mapping GO IDs to their term names.\n",
        "    \"\"\"\n",
        "    graph = obonet.read_obo(obo_filepath)\n",
        "    go_terms = {}\n",
        "    for go_id, data in graph.nodes(data=True):\n",
        "        # Only include nodes that have a 'name' field\n",
        "        if 'name' in data:\n",
        "            go_terms[go_id] = data\n",
        "    return go_terms\n",
        "\n",
        "# set the path to your local go.obo (or something else later) file\n",
        "obo_file = \"/Users/crystalcho/Downloads/go.obo\"\n",
        "go_terms_dict = load_go_ontology(obo_file)\n",
        "print(f\"Loaded {len(go_terms_dict)} GO terms from {obo_file}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 456,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 340,
          "referenced_widgets": [
            "e4718c03421b41a2a665c26293acd697",
            "7d97b347b7cb40df81041035931d7f27",
            "b85f6bb59d024282b5552ff665da4548",
            "f1219528a46a4a8e94bd1c8de9e6e16a",
            "1b8c25318a2b4b64bab3496083679c85",
            "53bb7f0fbad34a818e266d0ea44ffaf1",
            "bc435a026437452bb820c14d1ce5d3ee",
            "47571c7570f14ab58f96e5d30129da1b",
            "ec42f776af1b48a982636f0c122ab7e6",
            "de6e3ef0670042378384c41ecd695f1a",
            "ec7200455fa34863bec2ae60f9f4aad5",
            "618fd6a5462e4c5296b560f888b5d638",
            "331dbfdb6d7849c6a7b6d43bd33a7325",
            "54a0cadcc9ef406db05d36b690d7891c",
            "2443ed122977428c98248537123b674b",
            "c3911d8aca764dd38fad4c1dded8198f",
            "8eb65674477e477091f2501704df7eb8",
            "6d29b58c64cc4f869d6944699149a60e",
            "c939f017049643b18d5142514937af73",
            "6c463c9a4bef4b5194113acf7ced731e",
            "f7dcd748d0944da09948f4372366e8a8",
            "466f8ef6ba644d31a8586abaf9830448",
            "6ac927f5e2a2453d9aae4b5c474cf96d",
            "7593152124e04332b709dd30ec87c5a5",
            "4a743c94312c4b5db25087a747f6b879",
            "59dc6c4216cb4c5aad8c3764cf3d3f6d",
            "3117723c389b4e47a97d6e582078e5c9",
            "c6b5cf46fbeb41adb8da935949b808c1",
            "d4b8ff5330d14128ba304020f7fe8412",
            "ea731973ab9c44caa479fb7934fd3b4f",
            "097046dc7c4f46f5866068c82b706aca",
            "cc7573c531c84e61999e2161710173f3",
            "9775c0cd37bc43dcb7953f8b15343332",
            "1675d70c8a5e426691d0af5e6ba333f3",
            "784d7bf8c3be48b8ad5a3b6e6954690f",
            "730fe49f91204e2cbd3b58d24442be0a",
            "792eb1f8daf34e2590fda2f60e3ff684",
            "3e0f9cdf58ef4af1a0b16f9795747dfc",
            "1845ddb7ff634d969fa2538209fc4e0a",
            "9b03af45365c42419bb39779b82be47c",
            "d598fdb3d94a4095ba5a950cef321f1b",
            "39794838a2164151a5edf298a63f8147",
            "e7eaa149a903445db37dfdf719e14f9e",
            "4af857d8e5a94c278ae0fb3d7e092ac5",
            "769b37c671334690b6a7e5ba9059343e",
            "bdf351e0ccbc4aa18e168f593d1659cf",
            "8901a8884437441c934c53a98a1dac06",
            "32ef3e232304493d8b7e129c679f4f6f",
            "77472246b6b0497fbe67b8a3529040e5",
            "61de99be2c2e4afe90ff06e3bb30a2b7",
            "873ea6832a0447a19938fe764cb7ac58",
            "b483173c5fa74a72bb0a88206d3206a3",
            "8039b35f8f654a5e93ab69a0d1d8f3ea",
            "b1c01cfdd57e40e7bc4de9baf81d9f45",
            "82a3774cc5cd4acfb1822d9cb74d6f45"
          ]
        },
        "id": "zfuUI3PF1boL",
        "outputId": "e257d393-bdbb-4695-d2ac-d8f2ee4ecbea"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "No sentence-transformers model found with name cambridgeltl/SapBERT-from-PubMedBERT-fulltext. Creating a new one with mean pooling.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Loaded embedding model: cambridgeltl/SapBERT-from-PubMedBERT-fulltext\n"
          ]
        }
      ],
      "source": [
        "##########################################\n",
        "# 2. Load the SAPBERT embedding model (or an alternative)\n",
        "##########################################\n",
        "\n",
        "# Here we use a SAPBERT model available from Hugging Face.\n",
        "# may choose another model (e.g., \"sentence-transformers/all-MiniLM-L6-v2\")\n",
        "model_name = \"cambridgeltl/SapBERT-from-PubMedBERT-fulltext\"  # SAPBERT model fine-tuned on UMLS\n",
        "sapbert_model = SentenceTransformer(model_name)\n",
        "print(f\"Loaded embedding model: {model_name}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 563,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OM1So4sX2JLT",
        "outputId": "8477c5ed-2546-409c-8284-f1d957453bde"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Embedded GO terms with additional metadata.\n"
          ]
        }
      ],
      "source": [
        "def embed_go_terms(go_terms, model):\n",
        "    \"\"\"\n",
        "    Given a dictionary of GO terms (GO ID -> term data) and an embedding model,\n",
        "    returns:\n",
        "      - a list of GO IDs,\n",
        "      - a list of GO term names (in the same order),\n",
        "      - a tensor of embeddings for the GO term names,\n",
        "      - a list of GO namespaces,\n",
        "      - a list of GO relationships (if available, otherwise default to 'unknown')\n",
        "    \"\"\"\n",
        "    go_ids = []\n",
        "    go_names = []\n",
        "    go_namespaces = []\n",
        "    go_relationships = []\n",
        "\n",
        "    for go_id, data in go_terms.items():\n",
        "        if 'name' in data:\n",
        "            go_ids.append(go_id)\n",
        "            go_names.append(data['name'])\n",
        "            go_namespaces.append(data.get('namespace', 'unknown'))\n",
        "            # Some GO entries might not have a \"relationship\" field; default to 'unknown'\n",
        "            go_relationships.append(data.get('relationship', 'unknown'))\n",
        "\n",
        "    go_embeddings = model.encode(go_names, convert_to_tensor=True)\n",
        "    return go_ids, go_names, go_embeddings, go_namespaces, go_relationships\n",
        "\n",
        "# Reload GO ontology using the updated load_go_ontology:\n",
        "# go_terms_dict = load_go_ontology(obo_file) # already loaded in the previous cell\n",
        "go_ids, go_names, go_embeddings, go_namespaces, go_relationships = embed_go_terms(go_terms_dict, sapbert_model)\n",
        "print(\"Embedded GO terms with additional metadata.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 564,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "G0yR2eF6228J",
        "outputId": "3bfeca3b-8998-4598-f8f4-56505af53917"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Example: Llama term 'muscle cell differentiation' matched with 'muscle cell differentiation' (GO:0042692) with score 1.00\n"
          ]
        }
      ],
      "source": [
        "import torch\n",
        "\n",
        "def get_best_matching_go_term(llama_term, go_ids, go_names, go_embeddings, model, threshold=0.6):\n",
        "    \"\"\"\n",
        "    Given a GO term from Llama, compute its embedding and find the best matching GO term\n",
        "    from the provided GO ontology embeddings.\n",
        "\n",
        "    Parameters:\n",
        "      - llama_term: The GO term string generated by Llama.\n",
        "      - go_ids, go_names, go_embeddings: Outputs from embed_go_terms().\n",
        "      - model: The embedding model.\n",
        "      - threshold: Minimum cosine similarity required to accept the match.\n",
        "\n",
        "    Returns:\n",
        "      - best_go_id: The GO:ID of the best match (or None if score is below threshold)\n",
        "      - best_go_name: The GO term name of the best match.\n",
        "      - best_score: The cosine similarity score.\n",
        "    \"\"\"\n",
        "    # Encode the query using the model\n",
        "    query_embedding = model.encode(llama_term, convert_to_tensor=True)\n",
        "    # Compute cosine similarities between the query and all GO term embeddings\n",
        "    cos_scores = torch.squeeze(util.cos_sim(query_embedding, go_embeddings))\n",
        "    # Use torch.argmax to find the best matching index\n",
        "    best_idx = torch.argmax(cos_scores)\n",
        "    best_score = cos_scores[best_idx].item()\n",
        "\n",
        "    if best_score >= threshold:\n",
        "        return go_ids[best_idx], go_names[best_idx], best_score, best_idx\n",
        "    else:\n",
        "        return None, None, best_score, None\n",
        "\n",
        "# Test the matching function with an example Llama-generated term:\n",
        "example_llama_term = \"muscle cell differentiation\"\n",
        "best_go_id, best_go_name, score, best_idx = get_best_matching_go_term(example_llama_term, go_ids, go_names, go_embeddings, sapbert_model)\n",
        "print(f\"Example: Llama term '{example_llama_term}' matched with '{best_go_name}' ({best_go_id}) with score {score:.2f}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 565,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zp0nhP0_3pHz",
        "outputId": "331efcf4-a0ca-4146-ceab-abf18abd3981"
      },
      "outputs": [],
      "source": [
        "import json\n",
        "\n",
        "def parse_llama_output(llama_output_path):\n",
        "    \"\"\"\n",
        "    Parses Llama's output JSON file.\n",
        "    Returns the parsed JSON data (a list of gene entries).\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(llama_output_path, 'r') as f:\n",
        "            data = json.load(f)\n",
        "        # If the JSON is a single dictionary, convert it to a list\n",
        "        if isinstance(data, dict):\n",
        "            data = [data]\n",
        "        return data\n",
        "    except Exception as e:\n",
        "        print(f\"Error parsing JSON: {e}\")\n",
        "        return []\n",
        "\n",
        "# print(parse_llama_output(\"./llama_inferences/llama_output_22118460))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 566,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rpyVuCmM_EtU",
        "outputId": "c8d869f2-162d-4661-cf94-91f28390fa6e"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "there are 34 genes with matched GO terms.\n",
            "Gene: vitamin D binding protein (2638)\n",
            "  Llama Term: vitamin D binding\n",
            "  Matched GO Term: vitamin D binding (GO:0005499)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: VDBP binds approximately 85% of all circulating vitamin D metabolites.\n",
            "\n",
            "  Llama Term: vitamin D transport\n",
            "  Matched GO Term: vitamin transport (GO:0051180)\n",
            "  Similarity: 0.79\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: VDBP is responsible for transporting vitamin D metabolites in the circulation.\n",
            "\n",
            "  Llama Term: blood microparticle\n",
            "  Matched GO Term: blood microparticle (GO:0072562)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005615']\n",
            "  Namespace: cellular_component\n",
            "  Justification: VDBP is found in the blood, where it binds to vitamin D metabolites.\n",
            "\n",
            "  Llama Term: actin binding\n",
            "  Matched GO Term: actin binding (GO:0003779)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: VDBP binds to actin, preventing its polymerization and promoting its clearance.\n",
            "\n",
            "  Llama Term: neutrophil chemotaxis\n",
            "  Matched GO Term: neutrophil chemotaxis (GO:0030593)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: VDBP is involved in neutrophil recruitment and chemotaxis.\n",
            "\n",
            "  Llama Term: fatty acid binding\n",
            "  Matched GO Term: fatty acid binding (GO:0005504)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: VDBP has been shown to bind fatty acids.\n",
            "\n",
            "Gene: CYP2R1 (120227)\n",
            "  Llama Term: vitamin D biosynthetic process\n",
            "  Matched GO Term: vitamin D biosynthetic process (GO:0042368)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: CYP2R1 hydroxylates vitamin D2 and D3 at carbon 25 resulting in the production of 25(OH)D.\n",
            "\n",
            "  Llama Term: oxidoreductase activity\n",
            "  Matched GO Term: oxidoreductase activity (GO:0016491)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CYP2R1 catalyzes the generation of 25(OH)D from pro-vitamin D3 as a 25-hydroxylase.\n",
            "\n",
            "  Llama Term: endoplasmic reticulum\n",
            "  Matched GO Term: endoplasmic reticulum (GO:0005783)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005737', 'part_of GO:0012505']\n",
            "  Namespace: cellular_component\n",
            "  Justification: CYP2R1 is a microsomal enzyme, indicating it is located in the endoplasmic reticulum.\n",
            "\n",
            "  Llama Term: liver\n",
            "  Matched GO Term: liver development (GO:0001889)\n",
            "  Similarity: 0.72\n",
            "  Relationship: ['part_of GO:0061008']\n",
            "  Namespace: biological_process\n",
            "  Justification: CYP2R1 is expressed in the liver where it hydroxylates vitamin D2 and D3.\n",
            "\n",
            "  Llama Term: placenta\n",
            "  Matched GO Term: placenta development (GO:0001890)\n",
            "  Similarity: 0.76\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: The placenta possesses CYP2R1, which may be involved in the regulation of circulating 25(OH)D levels.\n",
            "\n",
            "Gene: CYP27A1 (1593)\n",
            "  Llama Term: vitamin D metabolic process\n",
            "  Matched GO Term: vitamin D metabolic process (GO:0042359)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: CYP27A1 hydroxylates vitamin D2 and D3 at carbon 25 resulting in the production of 25(OH)D.\n",
            "\n",
            "  Llama Term: mitochondrion\n",
            "  Matched GO Term: mitochondrion (GO:0005739)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005737']\n",
            "  Namespace: cellular_component\n",
            "  Justification: CYP27A1 is a mitochondrial cytochrome P450.\n",
            "\n",
            "  Llama Term: monooxygenase activity\n",
            "  Matched GO Term: monooxygenase activity (GO:0004497)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CYP27A1 hydroxylates vitamin D2 and D3, which is a monooxygenase reaction.\n",
            "\n",
            "  Llama Term: liver\n",
            "  Matched GO Term: liver development (GO:0001889)\n",
            "  Similarity: 0.72\n",
            "  Relationship: ['part_of GO:0061008']\n",
            "  Namespace: biological_process\n",
            "  Justification: CYP27A1 is active in the liver where it hydroxylates vitamin D2 and D3.\n",
            "\n",
            "Gene: CYP27B1 (1594)\n",
            "  Llama Term: vitamin D hydroxylation\n",
            "  Matched GO Term: vitamin D 24-hydroxylase activity (GO:0070576)\n",
            "  Similarity: 0.80\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CYP27B1 attaches the second hydroxygroup in position 1 forming active 1,25-(OH)2D.\n",
            "\n",
            "  Llama Term: kidney\n",
            "  Matched GO Term: renal system process (GO:0003014)\n",
            "  Similarity: 0.74\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Renal CYP27B1 attaches the second hydroxygroup in position 1 forming active 1,25-(OH)2D.\n",
            "\n",
            "  Llama Term: heme binding\n",
            "  Matched GO Term: heme binding (GO:0020037)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CYP27B1 is a cytochrome P450 enzyme, which typically binds heme.\n",
            "\n",
            "  Llama Term: iron ion binding\n",
            "  Matched GO Term: iron ion binding (GO:0005506)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CYP27B1 is a cytochrome P450 enzyme, which typically binds iron ions via heme.\n",
            "\n",
            "  Llama Term: monooxygenase activity\n",
            "  Matched GO Term: monooxygenase activity (GO:0004497)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CYP27B1 catalyzes the 1alpha-hydroxylation of 25(OH)D.\n",
            "\n",
            "Gene: CYP24A1 (1591)\n",
            "  Llama Term: vitamin D metabolic process\n",
            "  Matched GO Term: vitamin D metabolic process (GO:0042359)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: CYP24A1 metabolises 25(OH)D to 24,25-dihydroxy-vitamin D and 1,25(OH)2D to 1,24,25-trihydroxy-vitamin D.\n",
            "\n",
            "  Llama Term: oxidoreductase activity\n",
            "  Matched GO Term: oxidoreductase activity (GO:0016491)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CYP24A1 is involved in the metabolism of vitamin D, which requires oxidoreductase activity.\n",
            "\n",
            "  Llama Term: endoplasmic reticulum\n",
            "  Matched GO Term: endoplasmic reticulum (GO:0005783)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005737', 'part_of GO:0012505']\n",
            "  Namespace: cellular_component\n",
            "  Justification: As a cytochrome P450 enzyme, CYP24A1 is expected to be located in the endoplasmic reticulum.\n",
            "\n",
            "  Llama Term: calcium ion homeostasis\n",
            "  Matched GO Term: calcium ion homeostasis (GO:0055074)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Mutations in CYP24A1 cause infantile hypercalcemia, indicating a role in calcium ion homeostasis.\n",
            "\n",
            "Gene: DBP (1628)\n",
            "  Llama Term: vitamin D binding\n",
            "  Matched GO Term: vitamin D binding (GO:0005499)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: DBP binds to vitamin D metabolites in the circulation.\n",
            "\n",
            "  Llama Term: blood circulation\n",
            "  Matched GO Term: blood circulation (GO:0008015)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: DBP is found in the circulation, binding vitamin D metabolites.\n",
            "\n",
            "  Llama Term: response to vitamin D\n",
            "  Matched GO Term: response to vitamin D (GO:0033280)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Genetic variants of DBP predict differences in response of serum 25-hydroxyvitamin D to vitamin D supplementation.\n",
            "\n",
            "Gene: albumin (213)\n",
            "  Llama Term: vitamin D binding\n",
            "  Matched GO Term: vitamin D binding (GO:0005499)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Albumin binds approximately 15% of all circulating vitamin D metabolites.\n",
            "\n",
            "  Llama Term: blood circulation\n",
            "  Matched GO Term: blood circulation (GO:0008015)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Albumin is found in the circulation, where it binds vitamin D metabolites.\n",
            "\n",
            "  Llama Term: extracellular region\n",
            "  Matched GO Term: extracellular region (GO:0005576)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: Albumin is found in the serum, which is part of the extracellular region.\n",
            "\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Albumin binds to vitamin D metabolites.\n",
            "\n",
            "Gene: VDR (7421)\n",
            "  Llama Term: steroid hormone receptor activity\n",
            "  Matched GO Term: nuclear steroid receptor activity (GO:0003707)\n",
            "  Similarity: 0.96\n",
            "  Relationship: ['part_of GO:0030518']\n",
            "  Namespace: molecular_function\n",
            "  Justification: The biological activity of 1,25(OH)2D is mediated by binding to the VDR, a nuclear receptor.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: The VDR is a nuclear receptor that binds specific DNA sequences located at different genomic regions.\n",
            "\n",
            "  Llama Term: regulation of gene expression\n",
            "  Matched GO Term: regulation of gene expression (GO:0010468)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0010467']\n",
            "  Namespace: biological_process\n",
            "  Justification: The VDR binds specific DNA sequences to subsequently modulate the expression of the respective target genes.\n",
            "\n",
            "  Llama Term: vitamin D receptor signaling pathway\n",
            "  Matched GO Term: vitamin D receptor signaling pathway (GO:0070561)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0071305']\n",
            "  Namespace: biological_process\n",
            "  Justification: The VDR mediates the biological activity of 1,25(OH)2D.\n",
            "\n",
            "  Llama Term: cellular response to vitamin D\n",
            "  Matched GO Term: cellular response to vitamin D (GO:0071305)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: The VDR is involved in the response to vitamin D.\n",
            "\n",
            "Gene: PTH (5741)\n",
            "  Llama Term: hormone activity\n",
            "  Matched GO Term: hormone activity (GO:0005179)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: PTH increases the synthesis of 1,25(OH)2D, which regulates blood calcium levels.\n",
            "\n",
            "  Llama Term: calcium ion homeostasis\n",
            "  Matched GO Term: calcium ion homeostasis (GO:0055074)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: PTH increases the synthesis of 1,25(OH)2D, which regulates blood calcium levels through intestinal absorption, renal reabsorption and release from bone stores.\n",
            "\n",
            "  Llama Term: parathyroid gland\n",
            "  Matched GO Term: parathyroid gland development (GO:0060017)\n",
            "  Similarity: 0.72\n",
            "  Relationship: ['part_of GO:0035270']\n",
            "  Namespace: biological_process\n",
            "  Justification: Parathyroid hormone (PTH) is controlled in the kidneys and secreted by the parathyroid gland.\n",
            "\n",
            "Gene: pregnane X receptor (8856)\n",
            "  Llama Term: steroid hormone receptor activity\n",
            "  Matched GO Term: nuclear steroid receptor activity (GO:0003707)\n",
            "  Similarity: 0.96\n",
            "  Relationship: ['part_of GO:0030518']\n",
            "  Namespace: molecular_function\n",
            "  Justification: The pregnane X receptor is a nuclear receptor that shares homologies with the VDR, a known steroid hormone receptor.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: The pregnane X receptor is described as a nuclear receptor.\n",
            "\n",
            "  Llama Term: regulation of vitamin D metabolic process\n",
            "  Matched GO Term: regulation of vitamin D biosynthetic process (GO:0060556)\n",
            "  Similarity: 0.89\n",
            "  Relationship: ['regulates GO:0042368']\n",
            "  Namespace: biological_process\n",
            "  Justification: Binding of anti-epileptic drugs to the pregnane X receptor increases 25(OH)D catabolism by an induction of CYP3A4 and CYP24A1.\n",
            "\n",
            "  Llama Term: drug binding\n",
            "  Matched GO Term: binding (GO:0005488)\n",
            "  Similarity: 0.90\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: The pregnane X receptor binds to anti-epileptic drugs such as phenobarbital, carbamazepine, phenytoin and valproate.\n",
            "\n",
            "Gene: CYP3A4 (1576)\n",
            "  Llama Term: oxidoreductase activity\n",
            "  Matched GO Term: oxidoreductase activity (GO:0016491)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CYP3A4 is involved in the catabolism of vitamin D, indicating its role as an oxidoreductase.\n",
            "\n",
            "  Llama Term: vitamin D catabolic process\n",
            "  Matched GO Term: vitamin D catabolic process (GO:0042369)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: CYP3A4 induction increases 25(OH)D catabolism.\n",
            "\n",
            "  Llama Term: endoplasmic reticulum\n",
            "  Matched GO Term: endoplasmic reticulum (GO:0005783)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005737', 'part_of GO:0012505']\n",
            "  Namespace: cellular_component\n",
            "  Justification: As a cytochrome P450 enzyme, CYP3A4 is typically located in the endoplasmic reticulum.\n",
            "\n",
            "Gene: FGF-23 (8074)\n",
            "  Llama Term: regulation of phosphate metabolic process\n",
            "  Matched GO Term: regulation of phosphate metabolic process (GO:0019220)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0006796']\n",
            "  Namespace: biological_process\n",
            "  Justification: CKD patients have elevated phosphate and FGF-23 concentrations, indicating FGF-23's role in phosphate metabolism.\n",
            "\n",
            "  Llama Term: negative regulation of vitamin D biosynthetic process\n",
            "  Matched GO Term: negative regulation of vitamin D biosynthetic process (GO:0010957)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['negatively_regulates GO:0042368']\n",
            "  Namespace: biological_process\n",
            "  Justification: FGF23 inhibits 1-alpha hydroxylase, reducing the renal synthesis of 1,25(OH)2D.\n",
            "\n",
            "  Llama Term: osteocyte differentiation\n",
            "  Matched GO Term: osteoblast differentiation (GO:0001649)\n",
            "  Similarity: 0.90\n",
            "  Relationship: ['part_of GO:0001503']\n",
            "  Namespace: biological_process\n",
            "  Justification: Osteocytes secrete FGF23, suggesting FGF-23's involvement in osteocyte function.\n",
            "\n",
            "  Llama Term: cell surface\n",
            "  Matched GO Term: cell surface (GO:0009986)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: FGF23 binds to Klotho, implying it interacts at the cell surface.\n",
            "\n",
            "Gene: TRPV6 (55503)\n",
            "  Llama Term: calcium ion transmembrane transport\n",
            "  Matched GO Term: calcium ion transmembrane transport (GO:0070588)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: TRPV6 is upregulated by 1,25(OH)2D to transport calcium from the intestinal lumen to the blood.\n",
            "\n",
            "  Llama Term: integral component of plasma membrane\n",
            "  Matched GO Term: plasma membrane (GO:0005886)\n",
            "  Similarity: 0.83\n",
            "  Relationship: ['part_of GO:0071944']\n",
            "  Namespace: cellular_component\n",
            "  Justification: TRPV6 is involved in the transport of calcium from the intestinal lumen to the blood, implying it is a membrane-bound protein.\n",
            "\n",
            "Gene: S100G (795)\n",
            "  Llama Term: intestinal absorption\n",
            "  Matched GO Term: intestinal absorption (GO:0050892)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: S100G is upregulated by 1,25(OH)2D to facilitate the transport of calcium from the intestinal lumen to the blood.\n",
            "\n",
            "  Llama Term: calcium ion homeostasis\n",
            "  Matched GO Term: calcium ion homeostasis (GO:0055074)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: S100G is involved in regulating blood calcium levels through intestinal absorption.\n",
            "\n",
            "  Llama Term: enterocyte differentiation\n",
            "  Matched GO Term: enterocyte differentiation (GO:1903703)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: None found\n",
            "\n",
            "Gene: ATP2B1 (490)\n",
            "  Llama Term: calcium ion transport\n",
            "  Matched GO Term: calcium ion transport (GO:0006816)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: ATP2B1 is upregulated by 1,25(OH)2D to transport calcium from the intestinal lumen to the blood.\n",
            "\n",
            "  Llama Term: plasma membrane\n",
            "  Matched GO Term: plasma membrane (GO:0005886)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0071944']\n",
            "  Namespace: cellular_component\n",
            "  Justification: As a calcium-transporting ATPase, ATP2B1 is expected to be located in the plasma membrane.\n",
            "\n",
            "Gene: CLDN2 (9075)\n",
            "  Llama Term: intestinal absorption\n",
            "  Matched GO Term: intestinal absorption (GO:0050892)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: CLDN2 is upregulated by 1,25(OH)2D to facilitate the transport of calcium from the intestinal lumen to the blood.\n",
            "\n",
            "  Llama Term: calcium ion transport\n",
            "  Matched GO Term: calcium ion transport (GO:0006816)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: CLDN2 is upregulated by 1,25(OH)2D to facilitate the transport of calcium from the intestinal lumen to the blood.\n",
            "\n",
            "  Llama Term: tight junction\n",
            "  Matched GO Term: tight junction (GO:0070160)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: CLDN2 is a tight junction protein.\n",
            "\n",
            "Gene: Klotho (9365)\n",
            "  Llama Term: fibroblast growth factor receptor signaling pathway\n",
            "  Matched GO Term: fibroblast growth factor receptor signaling pathway (GO:0008543)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0044344']\n",
            "  Namespace: biological_process\n",
            "  Justification: Klotho binds to FGF23, which inhibits 1-alpha hydroxylase and reduces the renal synthesis of 1,25(OH)2D.\n",
            "\n",
            "  Llama Term: cell surface\n",
            "  Matched GO Term: cell surface (GO:0009986)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: Klotho binds to FGF23, implying it is located on the cell surface to interact with FGF23.\n",
            "\n",
            "Gene: osteocalcin (12097)\n",
            "  Llama Term: bone formation\n",
            "  Matched GO Term: bone development (GO:0060348)\n",
            "  Similarity: 0.87\n",
            "  Relationship: ['part_of GO:0001501']\n",
            "  Namespace: biological_process\n",
            "  Justification: 1,25(OH)2D increases osteoblastic expression of osteocalcin, supporting bone formation.\n",
            "\n",
            "  Llama Term: osteoblast differentiation\n",
            "  Matched GO Term: osteoblast differentiation (GO:0001649)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0001503']\n",
            "  Namespace: biological_process\n",
            "  Justification: Osteocalcin expression is increased by 1,25(OH)2D in osteoblasts, indicating a role in osteoblast differentiation.\n",
            "\n",
            "Gene: p16 (13088)\n",
            "  Llama Term: negative regulation of cell proliferation\n",
            "  Matched GO Term: negative regulation of cell population proliferation (GO:0008285)\n",
            "  Similarity: 0.91\n",
            "  Relationship: ['negatively_regulates GO:0008283']\n",
            "  Namespace: biological_process\n",
            "  Justification: p16 is involved in senescence pathways, which suggests its role in regulating cell proliferation.\n",
            "\n",
            "  Llama Term: cellular senescence\n",
            "  Matched GO Term: cellular senescence (GO:0090398)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0033554']\n",
            "  Namespace: biological_process\n",
            "  Justification: p16 is activated in senescence pathways, indicating its involvement in cellular senescence.\n",
            "\n",
            "  Llama Term: regulation of cellular aging\n",
            "  Matched GO Term: regulation of cellular senescence (GO:2000772)\n",
            "  Similarity: 0.86\n",
            "  Relationship: ['regulates GO:0090398']\n",
            "  Namespace: biological_process\n",
            "  Justification: p16 is part of the p16/p53-senescence signaling, which is inactivated by 1,25-Dihydroxyvitamin D to exert an antiaging role.\n",
            "\n",
            "Gene: p19 (83430)\n",
            "  Llama Term: negative regulation of cell proliferation\n",
            "  Matched GO Term: negative regulation of cell population proliferation (GO:0008285)\n",
            "  Similarity: 0.91\n",
            "  Relationship: ['negatively_regulates GO:0008283']\n",
            "  Namespace: biological_process\n",
            "  Justification: p19 is involved in senescence pathways, which suggests its role in regulating cell proliferation.\n",
            "\n",
            "  Llama Term: cellular senescence\n",
            "  Matched GO Term: cellular senescence (GO:0090398)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0033554']\n",
            "  Namespace: biological_process\n",
            "  Justification: p19 is activated in senescence pathways, indicating its involvement in cellular senescence.\n",
            "\n",
            "  Llama Term: regulation of apoptosis\n",
            "  Matched GO Term: regulation of apoptotic process (GO:0042981)\n",
            "  Similarity: 0.98\n",
            "  Relationship: ['regulates GO:0006915']\n",
            "  Namespace: biological_process\n",
            "  Justification: p19 is mentioned alongside the regulation of cell proliferation, differentiation, and apoptosis, suggesting its involvement in apoptosis regulation.\n",
            "\n",
            "Gene: IL-10 (3586)\n",
            "  Llama Term: anti-inflammatory response\n",
            "  Matched GO Term: negative regulation of inflammatory response (GO:0050728)\n",
            "  Similarity: 0.85\n",
            "  Relationship: ['negatively_regulates GO:0006954']\n",
            "  Namespace: biological_process\n",
            "  Justification: IL-10 is described as an anti-inflammatory compound.\n",
            "\n",
            "  Llama Term: cytokine production\n",
            "  Matched GO Term: cytokine production (GO:0001816)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: IL-10 is mentioned as a compound whose expression is increased.\n",
            "\n",
            "  Llama Term: immune system\n",
            "  Matched GO Term: immune system process (GO:0002376)\n",
            "  Similarity: 0.91\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: IL-10 is mentioned in the context of attenuating the inflammatory response of monocytes and macrophages.\n",
            "\n",
            "Gene: IL-1b (3553)\n",
            "  Llama Term: inflammatory response\n",
            "  Matched GO Term: inflammatory response (GO:0006954)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: IL-1b is described as a pro-inflammatory cytokine.\n",
            "\n",
            "  Llama Term: cytokine activity\n",
            "  Matched GO Term: cytokine activity (GO:0005125)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: IL-1b is referred to as a cytokine, implying it has cytokine activity.\n",
            "\n",
            "Gene: IL-6 (3569)\n",
            "  Llama Term: inflammatory response\n",
            "  Matched GO Term: inflammatory response (GO:0006954)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: IL-6 is described as a pro-inflammatory cytokine.\n",
            "\n",
            "  Llama Term: cytokine activity\n",
            "  Matched GO Term: cytokine activity (GO:0005125)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: IL-6 is referred to as a cytokine, implying it has cytokine activity.\n",
            "\n",
            "Gene: TNFalpha (7124)\n",
            "  Llama Term: inflammatory response\n",
            "  Matched GO Term: inflammatory response (GO:0006954)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: TNFalpha is described as a pro-inflammatory cytokine.\n",
            "\n",
            "  Llama Term: cytokine activity\n",
            "  Matched GO Term: cytokine activity (GO:0005125)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TNFalpha is referred to as a cytokine.\n",
            "\n",
            "Gene: RANKL (8600)\n",
            "  Llama Term: cytokine activity\n",
            "  Matched GO Term: cytokine activity (GO:0005125)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: RANKL is mentioned alongside other cytokines like IL-1b, IL-6, and TNFalpha, indicating it has cytokine activity.\n",
            "\n",
            "  Llama Term: inflammatory response\n",
            "  Matched GO Term: inflammatory response (GO:0006954)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: RANKL secretion is reduced by 1,25(OH)2D, which attenuates the inflammatory response of monocytes and macrophages.\n",
            "\n",
            "Gene: COX2 (4513)\n",
            "  Llama Term: inflammatory response\n",
            "  Matched GO Term: inflammatory response (GO:0006954)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: The secretion of COX2 is reduced by 1,25(OH)2D, which attenuates the inflammatory response of monocytes and macrophages.\n",
            "\n",
            "  Llama Term: prostaglandin biosynthetic process\n",
            "  Matched GO Term: prostaglandin biosynthetic process (GO:0001516)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: COX2 is involved in the production of pro-inflammatory compounds, implying a role in prostaglandin biosynthesis.\n",
            "\n",
            "Gene: insulin (3630)\n",
            "  Llama Term: insulin secretion\n",
            "  Matched GO Term: insulin secretion (GO:0030073)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: insulin secretion is modified by calcitriol\n",
            "\n",
            "  Llama Term: glucose homeostasis\n",
            "  Matched GO Term: glucose homeostasis (GO:0042593)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: insulin plays a role in glucose homeostasis\n",
            "\n",
            "  Llama Term: cellular response to insulin stimulus\n",
            "  Matched GO Term: cellular response to insulin stimulus (GO:0032869)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: insulin sensitivity is modified by calcitriol\n",
            "\n",
            "Gene: renin (5972)\n",
            "  Llama Term: peptidase activity\n",
            "  Matched GO Term: peptidase activity (GO:0008233)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Renin is an enzyme that catalyzes the conversion of angiotensinogen to angiotensin I.\n",
            "\n",
            "  Llama Term: blood vessel morphogenesis\n",
            "  Matched GO Term: blood vessel morphogenesis (GO:0048514)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0001568']\n",
            "  Namespace: biological_process\n",
            "  Justification: The renin-angiotensin axis influences blood pressure regulation and vascular development.\n",
            "\n",
            "  Llama Term: circulatory system\n",
            "  Matched GO Term: circulatory system process (GO:0003013)\n",
            "  Similarity: 0.94\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Renin plays a role in the regulation of blood pressure through the renin-angiotensin-aldosterone system.\n",
            "\n",
            "Gene: PHEX (5251)\n",
            "  Llama Term: metalloendopeptidase activity\n",
            "  Matched GO Term: metalloendopeptidase activity (GO:0004222)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: PHEX is described as a cell surface metalloproteinase.\n",
            "\n",
            "  Llama Term: cell surface\n",
            "  Matched GO Term: cell surface (GO:0009986)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: PHEX is described as a cell surface metalloproteinase.\n",
            "\n",
            "  Llama Term: regulation of phosphate transport\n",
            "  Matched GO Term: regulation of phosphate transport (GO:0010966)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0006817']\n",
            "  Namespace: biological_process\n",
            "  Justification: Mutations of PHEX cause X-linked hypophosphataemia, characterized by phosphate transport issues.\n",
            "\n",
            "Gene: SLC34A1 (6569)\n",
            "  Llama Term: calcium ion homeostasis\n",
            "  Matched GO Term: calcium ion homeostasis (GO:0055074)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Mutations of the SLC34A1 gene are associated with hypercalcaemia.\n",
            "\n",
            "Gene: alpha-fetoprotein (174)\n",
            "  Llama Term: liver\n",
            "  Matched GO Term: liver development (GO:0001889)\n",
            "  Similarity: 0.72\n",
            "  Relationship: ['part_of GO:0061008']\n",
            "  Namespace: biological_process\n",
            "  Justification: Alpha-fetoprotein is mainly produced in the liver.\n",
            "\n",
            "Gene: megalin (4036)\n",
            "  Llama Term: receptor activity\n",
            "  Matched GO Term: peptide receptor activity (GO:0001653)\n",
            "  Similarity: 0.81\n",
            "  Relationship: ['has_part GO:0042277']\n",
            "  Namespace: molecular_function\n",
            "  Justification: Megalin is involved in the reuptake of 25(OH)D in the kidneys.\n",
            "\n",
            "  Llama Term: proximal tubule cell\n",
            "  Matched GO Term: pyrenoid tubule (GO:0160223)\n",
            "  Similarity: 0.68\n",
            "  Relationship: ['part_of GO:1990732']\n",
            "  Namespace: cellular_component\n",
            "  Justification: Megalin-dependent uptake of the VDBP-25(OH)D complex into cells is limited to the proximal renal tubule.\n",
            "\n",
            "  Llama Term: endocytosis\n",
            "  Matched GO Term: endocytosis (GO:0006897)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['has_part GO:0006900', 'has_part GO:0010324']\n",
            "  Namespace: biological_process\n",
            "  Justification: Megalin is involved in the renal megalin/cubilin-mediated reuptake of 25(OH)D in the kidneys.\n",
            "\n",
            "  Llama Term: placenta development\n",
            "  Matched GO Term: placenta development (GO:0001890)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Vitamin D-dependent expression of megalin and cubilin has been found in the placenta.\n",
            "\n",
            "Gene: cubilin (8029)\n",
            "  Llama Term: receptor activity\n",
            "  Matched GO Term: peptide receptor activity (GO:0001653)\n",
            "  Similarity: 0.81\n",
            "  Relationship: ['has_part GO:0042277']\n",
            "  Namespace: molecular_function\n",
            "  Justification: cubilin is involved in the renal megalin/cubilin-mediated reuptake of 25(OH)D in the kidneys.\n",
            "\n",
            "  Llama Term: proximal tubule cell\n",
            "  Matched GO Term: pyrenoid tubule (GO:0160223)\n",
            "  Similarity: 0.68\n",
            "  Relationship: ['part_of GO:1990732']\n",
            "  Namespace: cellular_component\n",
            "  Justification: the megalin-dependent uptake of the VDBP-25(OH)D complex into cells is limited to the proximal renal tubule.\n",
            "\n",
            "  Llama Term: vitamin D transport\n",
            "  Matched GO Term: vitamin transport (GO:0051180)\n",
            "  Similarity: 0.79\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: cubilin is involved in the renal megalin/cubilin-mediated reuptake of 25(OH)D in the kidneys.\n",
            "\n",
            "  Llama Term: endocytosis\n",
            "  Matched GO Term: endocytosis (GO:0006897)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['has_part GO:0006900', 'has_part GO:0010324']\n",
            "  Namespace: biological_process\n",
            "  Justification: cubilin is involved in the megalin-dependent uptake of the VDBP-25(OH)D complex into cells.\n",
            "\n",
            "Gene: Gc2 (83733)\n",
            "  Llama Term: vitamin D binding\n",
            "  Matched GO Term: vitamin D binding (GO:0005499)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Gc2 is a variant of the vitamin D-binding protein (VDBP) that binds 25(OH)D.\n",
            "\n",
            "  Llama Term: blood microparticle\n",
            "  Matched GO Term: blood microparticle (GO:0072562)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005615']\n",
            "  Namespace: cellular_component\n",
            "  Justification: VDBP, including the Gc2 variant, is found in the bloodstream, often associated with blood microparticles.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "def auto_match_go_terms(llama_data, go_ids, go_names, go_embeddings, go_namespaces, go_relationships, model, threshold=0.6):\n",
        "    \"\"\"\n",
        "    Given parsed Llama JSON data, match each Llama-generated GO term to the best official GO term.\n",
        "    \"\"\"\n",
        "    matched_go_terms = {}\n",
        "\n",
        "    # If the JSON is nested under \"genes\", extract that list.\n",
        "    if isinstance(llama_data, list) and len(llama_data) > 0 and \"genes\" in llama_data[0]:\n",
        "        gene_entries = llama_data[0][\"genes\"]\n",
        "    else:\n",
        "        gene_entries = llama_data\n",
        "\n",
        "    for gene_entry in gene_entries:\n",
        "        gene_name = gene_entry.get(\"gene\", \"UnknownGene\")\n",
        "        gene_id = gene_entry.get(\"geneID\", \"UnknownID\")     # using id instead of name for the dictionary since some genes have the same name\n",
        "        go_data = gene_entry.get(\"go_data\", [])\n",
        "        unique_matches = {}\n",
        "\n",
        "        for entry in go_data:\n",
        "            llama_term = entry.get(\"term\")\n",
        "            if not llama_term:\n",
        "                continue\n",
        "\n",
        "            best_go_id, best_go_name, best_score, best_idx = get_best_matching_go_term(\n",
        "                llama_term, go_ids, go_names, go_embeddings, model, threshold=threshold\n",
        "            )\n",
        "\n",
        "            if best_go_id and best_idx is not None:\n",
        "                official_relationship = go_relationships[best_idx]\n",
        "                official_namespace = go_namespaces[best_idx]\n",
        "\n",
        "                if best_go_id in unique_matches:\n",
        "                    if best_score > unique_matches[best_go_id][\"similarity\"]:\n",
        "                        unique_matches[best_go_id] = {\n",
        "                            \"llama_term\": llama_term,\n",
        "                            \"matched_go_id\": best_go_id,\n",
        "                            \"matched_go_name\": best_go_name,\n",
        "                            \"similarity\": best_score,\n",
        "                            \"relationship\": official_relationship,\n",
        "                            \"namespace\": official_namespace,\n",
        "                            \"justification\": entry.get(\"justification\")\n",
        "                        }\n",
        "                else:\n",
        "                    unique_matches[best_go_id] = {\n",
        "                        \"llama_term\": llama_term,\n",
        "                        \"matched_go_id\": best_go_id,\n",
        "                        \"matched_go_name\": best_go_name,\n",
        "                        \"similarity\": best_score,\n",
        "                        \"relationship\": official_relationship,\n",
        "                        \"namespace\": official_namespace,\n",
        "                        \"justification\": entry.get(\"justification\")\n",
        "                    }\n",
        "\n",
        "        matched_go_terms[gene_id] = list(unique_matches.values())\n",
        "\n",
        "    return matched_go_terms\n",
        "\n",
        "def replace_terms_with_go(llama_data, matched_go_terms, similarity_threshold=0.65):\n",
        "    \"\"\"\n",
        "    Replaces each llama_data entry's GO term with the matched GO term info from matched_go_terms.\n",
        "    \"\"\"\n",
        "    # If llama_data is nested under \"genes\", re-point gene_entries\n",
        "    if isinstance(llama_data, list) and len(llama_data) > 0 and \"genes\" in llama_data[0]:\n",
        "        gene_entries = llama_data[0][\"genes\"]\n",
        "    else:\n",
        "        gene_entries = llama_data\n",
        "\n",
        "    for gene_entry in gene_entries:\n",
        "        gene_id = gene_entry.get(\"geneID\")\n",
        "        if gene_id in matched_go_terms:\n",
        "            new_go_data = []\n",
        "            for match in matched_go_terms[gene_id]:\n",
        "                # Skip if the match similarity is below threshold\n",
        "                if match[\"similarity\"] < similarity_threshold:\n",
        "                    continue\n",
        "                new_entry = {\n",
        "                    \"term\": match[\"matched_go_name\"],      # replace Llama term\n",
        "                    \"go_id\": match[\"matched_go_id\"],       # new field\n",
        "                    \"relationship\": match[\"relationship\"], # official relationship\n",
        "                    \"namespace\": match[\"namespace\"],\n",
        "                    \"justification\": match[\"justification\"],\n",
        "                    # carry over references if you want:\n",
        "                    \"reference(s)\": gene_entry[\"go_data\"][0].get(\"reference(s)\", [])\n",
        "                }\n",
        "                new_go_data.append(new_entry)\n",
        "            gene_entry[\"go_data\"] = new_go_data\n",
        "    return llama_data\n",
        "\n",
        "# Create a  directory to store PROCESSED llama JSON outputs.\n",
        "processed_dir = os.path.join(os.getcwd(), \"processed_inferences\")\n",
        "if not os.path.exists(processed_dir):\n",
        "    os.makedirs(processed_dir)\n",
        "\n",
        "# Usage example:\n",
        "# 1. Load the Llama output JSON file and parse it\n",
        "llama_output_path = f\"./llama_inferences/llama_output_{doc_id}.json\"\n",
        "llama_data = parse_llama_output(llama_output_path)\n",
        "\n",
        "# 2. Match the Llama terms to official GO terms\n",
        "matched_terms = auto_match_go_terms(\n",
        "    llama_data, go_ids, go_names, go_embeddings, go_namespaces, go_relationships, sapbert_model, threshold=0.6\n",
        ")\n",
        "\n",
        "# 3. Replace llama terms with the official GO terms\n",
        "updated_data = replace_terms_with_go(llama_data, matched_terms)\n",
        "\n",
        "# 4. save the processed JSON output to a new file\n",
        "procesed_output_path = os.path.join(processed_dir, f\"processed_output_{doc_id}.json\")\n",
        "with open(procesed_output_path, \"w\", encoding=\"utf-8\") as f:\n",
        "    json.dump(updated_data, f, indent=2)\n",
        "\n",
        "print(\"there are {} genes with matched GO terms.\".format(len(matched_terms)))\n",
        "for gene_id, matches in matched_terms.items():\n",
        "    print(f\"Gene: {gene_dict[gene_id].get_name_from_article()} ({gene_id})\")\n",
        "    for match in matches:\n",
        "        print(f\"  Llama Term: {match['llama_term']}\")\n",
        "        print(f\"  Matched GO Term: {match['matched_go_name']} ({match['matched_go_id']})\")\n",
        "        print(f\"  Similarity: {match['similarity']:.2f}\")\n",
        "        print(f\"  Relationship: {match['relationship']}\")\n",
        "        print(f\"  Namespace: {match['namespace']}\")\n",
        "        print(f\"  Justification: {match['justification']}\\n\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "52hJV2NbJMzp"
      },
      "source": [
        "# 6. Evaluate"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6DBtiTOtJgFO"
      },
      "source": [
        "6.1. Counting and Plotting Similarity Distributions"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 493,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 487
        },
        "id": "sPcTb_-UJLvq",
        "outputId": "4754b2f1-b823-432a-f485-1a874be690ee"
      },
      "outputs": [
        {
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAqYAAAHWCAYAAAClsUvDAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAANARJREFUeJzt3QeYVNX9N/CDKGADC6iIKBq7iMYao0aJGmI3/mPUWIgt0aixpRFjj6IxliQYyz8qMRFbbIlRVFQ0iQ3BErvYwIodsYCBeZ/fed/Zd3fZhWVZmAP7+TzPKDNzZ+6ZO3d2vvO755zboVKpVBIAANTYArVuAAAABMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFFrg5JNPTh06dJgr69p6663zpWrkyJF53X/961/nyvq/973vpT59+qSSTZo0KR188MFpueWWy9vm6KOPrnWT2qU58bloav+LdcS62tLQoUPz877yyitt+rzA7BFMaXeqX0jVS5cuXdLyyy+fBgwYkH73u9+ljz/+uE3W88Ybb+Qv08ceeyyVpuS2tcQZZ5yR38fDDjss/fnPf0777bffDJefNm1auuKKK9J2222XunfvnhZaaKG0zDLLpG984xvpkksuSZMnT57uMZ988kk67bTTUr9+/dIiiyySunXrlrbccsv8PDM7k3Pjfay5S6k/ACL4n3TSSalv375p0UUXTUsvvXRaf/3101FHHZX3nfnVH/7wh/zetbV33nknb7s111wzLbzwwnnf22STTdLPfvazvK2B/69DZWZ/YWE+E188BxxwQDr11FPTyiuvnL744ov01ltv5crknXfemVZcccX0t7/9LQeSqv/+97/5EiG2pR555JG08cYbp8svvzxXgVpqypQp+f+dOnXK/4929e/fP1133XXp29/+9iy91ta0LbZHBLnOnTunUn3lK19JCy64YPrXv/4102U/++yz9K1vfSvdfvvt6atf/Wraeeed07LLLpvef//9dO+996Zbb701DRw4MF166aV1j3n77bfTNttsk5555pm01157pa222ip9/vnn6frrr0/33Xdf2nPPPdOVV16ZOnbs2OQ6X3rppXT//fc3uC0qvBFGvv/979fdtthii6XddtstlSTe/0033TQ9++yzebtEII3w9NRTT6W///3veT+sVvRb87loyfob738R4iMot2XVdOrUqXldsZ5q1TeCePxwic9cW4n97Mtf/nKaOHFiOvDAA3M4fe+999ITTzyRbrnllvz/Un+gQC0sWJO1QgG23377tNFGG9VdHzRoULr77rvTTjvtlHbZZZccSqK6ESIExWVO+vTTT3NlrhpIayWqiaWbMGFCWnvttVu07DHHHJND6fnnn5+rVvUdd9xx6YUXXsg/SOqLQBbv/4033pj3haof/ehH6Sc/+Un6zW9+k8NGVLyassoqq+RLfYceemi+bd99902zK0Jy7CcLLND2B71uuumm9Oijj+bg/d3vfne69VZ/OM2pz8Wc3v+iEh5V4PhR0dwPi7YUP3jGjRuX/v3vf+cfRvVFWJ2bn/fqa4eiRcUU2pPLL788jhJURo0a1eT9Z5xxRr7/kksuqbvtpJNOyrfVd8cdd1Q233zzSrdu3SqLLrpoZfXVV68MGjQo33fPPffk5RtfYt1hq622qqyzzjqVRx55pLLllltWFl544cpRRx1Vd19cqqrPdfXVV+fnX3bZZSuLLLJIZeedd66MGzeuQZtWWmmlysCBA6d7TfWfc2Zti8fH89Q3adKkyrHHHltZYYUVKp06dcqv9eyzz65MmzatwXLxPIcffnjlxhtvzK8vll177bUrt912W4vem7fffrty4IEHVpZZZplK586dK/369asMHTp0um3R+PLyyy83+XyxfTp27Fj55je/WWmpBx54ID9ntKMpX3zxRWW11VarLLnkkpVPP/20xc8b+0jj9+a1116rHHDAAfn1VrfVpZde2mCZ6mu+6qqrKscff3xl+eWXr3To0KHywQcf5OeL53311VcrO+64Y/533D9kyJD82CeeeKLSv3//vL+suOKKlSuvvHKm7Rw8eHBe3yuvvDLTZZv6XFT3gWuvvbay1lprVbp06VL5yle+ktsSLrroosqXvvSl/P7GPtn4vWtq/4vnjHVVRdsOO+ywvB/G8y+11FKVb3/729M9V/WzPnLkyLx8jx49KksssUSD+6qPiXU23q+ifS+++GL+97nnnjvd6//3v/+d7xs2bFiz2+gHP/hB3genTp1aaYkHH3ywsv322+d2xvu27rrrVs4///wGy9x1112VLbbYIt8ff3922WWXytNPP93ke/PUU09V9t577/x866+/ft39f/7znysbbLBB3n6xL++5557T/T15/vnnK7vvvnv+mxPvV69evfJyH374YYteC7SGiik0Ev0Vf/GLX6Q77rgjHXLIIU0uE4c1o7Iah/ujS0AcDhw7dmyuioS11lor337iiSfmQ7fRNzHUr5jE4byo2sah4qiixeHlGTn99NPzIceo0kXFMCqA2267be4nWq3stkRL2lZf5IKoGt5zzz3poIMOyod2owIZlcPXX389nXfeeQ2Wj8PrN9xwQ/rhD3+YFl988dxv93/+539y1Sj6Ks7okHscIo7teMQRR+RuFnHYOLoafPjhh7naGW2PPqVRBV1hhRVyxTP06NGjyee87bbb8iHbWalSxuHqsP/++zd5f1QIo5J4yimn5Pc73oPWiO4C0SUh3tN4vfEaor2xjaOS1nhAV/R3jeraj3/849wntlppi9cX+9HXvva19Otf/zpXOuP5ojJ2/PHHp3322Sftvvvu6aKLLsqvabPNNsvbtjkrrbRS/n/0pf3lL3/ZqsFN//znP3N3mMMPPzxfHzx4cP68/PSnP839OGPf+OCDD3J74/B2HKmYFaNGjcpdJeKzE/tBDGC68MIL8/7z9NNP5yMP9cX6YvvGPh9Vw6bE5+nII4/M3Stiu4X4TEaVe/PNN8/bNfa7+uK22Md33XXXGW7PeI9iv41K/IxE5T62U8+ePfP+HoP7onIfh/yr1f4RI0bk9zvaFV0b4nPz+9//PrdxzJgx03UL2GOPPdJqq62W+2VXe+7F35ITTjghfec738ldTKIPbDxH7ENRLV9iiSVyZTz63ce+Ftsl2hKf92hLfB6jzzXMEa2KszAfV0xDVCG+/OUvN1sZOu+88/L1d955p9nniOevX4msLyoxcV9Uj5q6r6mKaVQrJk6cWHd7VKTi9t/+9rezVDGdWdsaV6xuuummvOyvfvWrBstFhSoqd2PHjq27LZaLyl/92x5//PF8++9///vKjERVKJb7y1/+UnfblClTKptttlllscUWa/Dao31RIZyZY445Jj/nY4891uD2yZMn5/euenn33Xfr7tttt93yY6Ii2ZwbbrghL/O73/2u0tqK6UEHHVTp2bNng3WHvfbaK+9/1Wps9f1fZZVVpqvQxvPFfVHlr4p2RwU+3puoslc9++yz01UemxLrWGONNfKysZ2/973v5SpuVLNbWjGN6lr96uXFF1+cb19uueUavI9xBKBxxbslFdOmKtXVSvcVV1wx3Wc9qov//e9/GyzfuGIaospf/3PSuP3PPPNMg32ze/fuTX7e6nvrrbdypTYev+aaa1YOPfTQXGFtXHWM9q288sr5tTfe9+ofmYiqZ1TY33vvvQafsQUWWKCy//77T/feRLW0vqg2RwX39NNPb3D7f/7zn8qCCy5Yd/ujjz6aH3/dddfN8PVBWzMqH5oQVZMZjc6PikK4+eab80CN1ogqawzCaqmodkV1pioGQkVlJQbvzEnx/NEXL/pX1hfVysgMUeWrLyqIX/rSl+quR1W5a9eueUDQzNYTVZm99967QX/DWG8MvomBSrMqKo/V97PxuqKCVr1Uq4Sh+r7X39aNVe+rPv+siu0WA6liIFb8+9133627RJXqo48+ytWv+qLa1lxlPKpe9ffNNdZYI1dMoyJWFbfFfTN7H2IdDz30UK6IVwcLRhU39rWonDU1g0FjMXCsfuUuBlOFqJzX367V22fWpqbaWBUDmOLow6qrrppfX+PtFuLIx+z0J43tGAO8okJaFUcN4v2aWTU+qq6PP/547mMcVeKoXEfFPUbmRxW8WsWMSuXLL7+cK+XVvy9V1ar1m2++mY+QxFGEpZZaqsFnLGacaOpvQay3vjiaEX+z4jXV3+/isxeV1TgyEqoV0Xid0f8d5hbBFJoQQWhGwSRGZcehswgE8cUThxSvvfbaWQqpvXr1mqWBD/Gl0fjLKr6M5/Q8jK+++mqeTqvx9ojD6tX764tZDRpbcskl85fyzNYTr7HxgJ7m1tMS1TY3npIn3rs4bBqXmDKqqcfM6IdJS8LrjMSh0zgcGlNV1Q/Ican+WInuGvU1d/g9AlPjrgwRKuIQd+PD8HH7zN6H6nJxmD32rbjEAJ4ItkOGDMlhamYa7wPVkNO7d+8mb29Jm+qLw9dxWD6eL37gxUj62AaxTSPUNzajrgstEUExfkQMGzas7rYIqfEZ/vrXvz7Tx0eoj64GESyfe+653L2l2rWgOhvEiy++WDczQHOqn4F4LxqLz0kEzMZdFRq/9hjsF2E4PmuN973oNlDd7+Jxxx57bPrjH/+Yt2/8YLrgggua3L7QlvQxhUZee+21/Mc3Qt+MKjYxbVBUF/7xj3+k4cOHp2uuuSZ/SUXf1JZUZ2alX2hLNdcfMPq4zY0RyKG59dRiZrqYmic8+eSTab311qu7Pb6Eq31D//KXv0z3BR8j02Man+hz15S4L7R0ZoDGqj9gotrWXL/D+tOVzWh/aW57t9X7ENXk6AcaU25Fv8YIZL/61a9m+Jg53aao3MZUZ1FdjD6zEXBj348fiE39OGyLz1ocsYg+z9G3dd111819aKPv6qzMjBBtXH311fNlxx13zOEwtmf9indba/zaY/tEO+JIR1PvR/2jC+ecc06uzsaRofi7Fkcvor/wgw8+mH/4wJwgmEIjMUghRIVgRuILKQ5ZxuXcc8/Ngwti0ESE1Qg9bX1GnKh0NP4yj4FC9QNMVCajatRUpaX+9EWz0rYIJjHgIqqE9SuEMc9l9f62EM8TgS++OOt/2c/OemKQSHz5xpd/DAJqiRh8El++MfinqWAaIT8qZ7Gto/LaGhGMY1vGc7V28NTcFq83umhEyK+1OAtaBPoITvWnsmpq358VM/pcfPOb38zvW+xL0QUhDm/P7MQOMxKfx9imUUUN1e4vsX2b2yeqn4GoujYWn5OobM5sOqhYT/ztiIpoBOSZiRAelxgIF6E89vnojjCzHyfQWg7lQz0xOjgOVcYf7RkFmZg0u7EYrR6qffCqXxCz+2VZFUGp/uHl+HKOL7UIX/W/dKKaUX+uyRhFO378+AbPNStt22GHHXKAisO49cVo/Pgir7/+2RHriRMdROW5KiZvj9HCUcWJSe5nVRxSjmpfVIcat7+5al3MThDBICpyse0aix8fzz//fB5h3tpKXITl6G8Z/UybCnpxqL9Woj9kHBJu6sdNjHhv6jDy3Bbbr/H7FvtJ7KezIz4XzX0mYjaG6P8cXXai322EtcZV7aZEf92mZgJ4+OGHc9/Y6vbcYIMN8t+dmB2gcRuqrzW6BMTfmT/96U8Nlol9KCqa8RmamZihIbZfzCrReBvG9WhTtf90fP7qi9ccPxpb0s8YWkvFlHYrwkpUGeKPb0zdE6E0+hxGVSIO083obDYx3VIcyo/DcbF89MuKaXDi8NYWW2xRFxKjb1pUF6I6Fl96UWlpbX+3GOwQzx19EKO98QUW3Q3qT2kVhwQjsEZ1JwY3RL+1OFRdfzDSrLYt+tbFmacikEV/wzgkHl+CcXgvDqU2fu7WiqmrLr744nzocPTo0XnwTLyWmJIpXmtr+3PGY2NQSRz+vfrqq/PriYEnEb7iuWN6qMZhK34ERCU8pgGKgSoxpVZ8GcfAkTgrUPQxrg4Oaq0zzzwzV9dju8d7GN0C4gdPDN6JCnVTP37mhvgMxFmWYoqwmM4qfhTE4KTLLrssb4O2Pmd9a0RVO45sxCH82G4PPPBA3mYzmo6sJTbccMPcFzSqgfHZiv2kfh/SOJwf/UPjfTvrrLNa9JzRzqiyRleIeP7oVx59OWN7xt+YmJouROCLdcf+GeEzPucRRONvVExPF4OQwtlnn51/DEYXhhiUVp0uKrZFS96b+LzG64sTisTnOc48Fp+t+IzECSXicxhTksXfw5h2LKabispq/J2M11L9UQVzTJuP84fCVaeJqV5ieqOYxma77bbLUy/Vn86muWlxYoLrXXfdNU9mHo+P/8e0LDEhdX0333xznjQ9pmFpaoL9pjQ3XVRMsB7T68RUMTEdUEyXFBOrN3bOOefkqaViyp44AUBM4t/4OWfUtqam6/n444/z1EvxOhdaaKE8wfyMJthvrLlprBqLKYliwvmYhie2a0wu3tSUVi2dLqr+VDzxPF//+tfzZOzxmmMd22yzTZ6y67PPPpvuMfGaTz755Pw+xfZefPHF8/aMCf8bv+7WTrAfrze2V+/evfN2jf0w2lT/5A7V97+paXuqE+w31tz+1ZLt9tJLL1VOPPHEPCl+7GuxrWK6o3jc3Xff3eIJ9uuLKZni9thn6mvqtbVkuqiYTqm6n8RUYgMGDMjTYTXez2Y0NVxT00XF1E7xOuO9rk6w31hs15iaKU6O0BJxYoGf/OQneTL76r4X04TtsccelTFjxky3/L/+9a/8tyjaEO9tnGSi8VRrI0aMyPti7Jddu3bNJ9toboL95qa0u/766/M0WrGOuMRUVvG+Pffcc3X7QZxkIk6GUD2JQZysIdYNc1KH+M+ci70AMP+IU9HG0Yu77rqr1k2B+ZI+pgDQAo888kieR7S5s4IBs0/FFABmIAYXRb/nmAUg+iZHn9sZ9UEHWk/FFABmIAbhxWCkOMvUVVddJZTCHKRiCgBAEVRMAQAogmAKAEAR5ukJ9uPUhW+88UaeHLitT/8IAMDsi16jcebC5ZdfvsEpp+e7YBqhtHfv3rVuBgAAMxGnx44zJM63wbR6isJ4oV27dq11cwAAaGTixIm5kNiSU0vP08G0evg+QqlgCgBQrpZ0uzT4CQCAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCDUPpq+//nrad99909JLL50WXnjhtO6666ZHHnmk1s0CAGAuWzDV0AcffJA233zz1L9//3TbbbelHj16pBdeeCEtueSStWwWAADtLZieddZZqXfv3unyyy+vu23llVeuZZMAAGiPwfRvf/tbGjBgQNpjjz3Svffem3r16pV++MMfpkMOOaTJ5SdPnpwvVRMnTpyLrQWAtjFu3Lj07rvv1roZtGPdu3dPK664YipNTYPpSy+9lC688MJ07LHHpl/84hdp1KhR6Uc/+lHq1KlTGjhw4HTLDx48OJ1yyik1aSsAtFUoXWPNtdLnn31a66bQjnVZeJH03LPPFBdOO1QqlUqtVh4BdKONNkr3339/3W0RTCOgPvDAAy2qmEZXgI8++ih17dp1rrUbAFprzJgxacMNN0xL73RcWmjp3rVuDu3QF++NT+/dck4aPXp02mCDDeb4+iKvdevWrUV5raYV0549e6a11167wW1rrbVWuv7665tcvnPnzvkCAPO6CKWdl1u11s2AotR0uqgYkf/cc881uO35559PK620Us3aBABAOwymxxxzTHrwwQfTGWeckcaOHZuGDRuWLrnkknT44YfXslkAALS3YLrxxhunG2+8MV111VWpb9++6bTTTkvnn39+2meffWrZLAAAaqCmfUzDTjvtlC8AALRvNT8lKQAABMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgBQBMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgBQBMEUAIAi1DSYnnzyyalDhw4NLmuuuWYtmwQAQI0smGpsnXXWSSNGjKi7vuCCNW8SAAA1UPMUGEF0ueWWq3UzAABo731MX3jhhbT88sunVVZZJe2zzz5p3LhxzS47efLkNHHixAYXAADmDzUNpptuumkaOnRoGj58eLrwwgvTyy+/nLbccsv08ccfN7n84MGDU7du3eouvXv3nuttBgBgPgym22+/fdpjjz1Sv3790oABA9Ktt96aPvzww3Tttdc2ufygQYPSRx99VHcZP378XG8zAADzaR/T+pZYYom0+uqrp7FjxzZ5f+fOnfMFAID5T837mNY3adKk9OKLL6aePXvWuikAALSnYPrjH/843XvvvemVV15J999/f/rWt76VOnbsmPbee+9aNgsAgPZ2KP+1117LIfS9995LPXr0SFtssUV68MEH878BAGhfahpMr7766lquHgCAghTVxxQAgPZLMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgBQBMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgBQBMEUAIAiCKYAABRBMAUAoAjFBNMzzzwzdejQIR199NG1bgoAADVQRDAdNWpUuvjii1O/fv1q3RQAANprMJ00aVLaZ5990v/+7/+mJZdcstbNAQCgvQbTww8/PO24445p2223nemykydPThMnTmxwAQBg/rBgLVd+9dVXpzFjxuRD+S0xePDgdMopp8zxdgEA0I4qpuPHj09HHXVUuvLKK1OXLl1a9JhBgwaljz76qO4SzwEAwPyhZhXT0aNHpwkTJqQNNtig7rapU6em++67Lw0ZMiQftu/YsWODx3Tu3DlfAACY/9QsmG6zzTbpP//5T4PbDjjggLTmmmumn/3sZ9OFUgAA5m81C6aLL7546tu3b4PbFl100bT00ktPdzsAAPO/mo/KBwCAmo/Kb2zkyJG1bgIAADWiYgoAQBEEUwAAiiCYAgBQBMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgAw7wbTVVZZJb333nvT3f7hhx/m+wAAYK4E01deeSVNnTp1utsnT56cXn/99dY8JQAA7dyCs7Lw3/72t7p/33777albt2511yOo3nXXXalPnz5t20IAANqFWQqmu+22W/5/hw4d0sCBAxvct9BCC+VQes4557RtCwEAaBdmKZhOmzYt/3/llVdOo0aNSt27d59T7QIAoJ2ZpWBa9fLLL7d9SwAAaNdaFUxD9CeNy4QJE+oqqVWXXXZZW7QNAIB2pFXB9JRTTkmnnnpq2mijjVLPnj1zn1MAAJjrwfSiiy5KQ4cOTfvtt99srRwAAGZrHtMpU6akr371q615KAAAtF0wPfjgg9OwYcNa81AAAGi7Q/mff/55uuSSS9KIESNSv3798hym9Z177rmteVoAANqxVgXTJ554Iq2//vr5308++WSD+wyEAgBgrgXTe+65p1UrAwCANu1jCgAARVRM+/fvP8ND9nfffffstAkAgHaoVcG02r+06osvvkiPPfZY7m86cODAtmobAADtSKuC6Xnnndfk7SeffHKaNGnS7LYJAIB2qE37mO67777psssua8unBACgnWjTYPrAAw+kLl26tOVTAgDQTrTqUP7uu+/e4HqlUklvvvlmeuSRR9IJJ5zQVm0DAKAdaVUw7datW4PrCyywQFpjjTXSqaeemr7xjW+0VdsAAGhHWhVML7/88rZvCQAA7dps9TEdPXp0+stf/pIvjz766Cw//sILL0z9+vVLXbt2zZfNNtss3XbbbbPTJAAA2lPFdMKECWmvvfZKI0eOTEsssUS+7cMPP8wT71999dWpR48eLXqeFVZYIZ155plptdVWy/1U//SnP6Vdd901h9x11lmnNU0DAKA9VUyPPPLI9PHHH6ennnoqvf/++/kSk+tPnDgx/ehHP2rx8+y8885phx12yMF09dVXT6effnpabLHF0oMPPtiaZgEA0N4qpsOHD08jRoxIa621Vt1ta6+9drrgggtaPfhp6tSp6brrrkuffPJJPqTflMmTJ+dLVQRhAADaccV02rRpaaGFFpru9rgt7psV//nPf3KVtHPnzunQQw9NN954Yw65TRk8eHCeEaB66d27d2uaDwDA/BJMv/71r6ejjjoqvfHGG3W3vf766+mYY45J22yzzSw9V0wz9dhjj6WHHnooHXbYYWngwIHp6aefbnLZQYMGpY8++qjuMn78+NY0HwCA+eVQ/pAhQ9Iuu+yS+vTpU1e1jJDYt2/fPEJ/VnTq1Cmtuuqq+d8bbrhhGjVqVPrtb3+bLr744umWjapqXAAAmP+0KphGGB0zZkzuZ/rss8/m26K/6bbbbjvbDYquAPX7kQIA0D7MUjC9++670xFHHJFHzce8o9ttt12+hDi0HlM8XXTRRWnLLbds0fPFofntt98+rbjiinmU/7Bhw/IUVLfffnvrXg0AAO0jmJ5//vnpkEMOyaG0sRiM9IMf/CCde+65LQ6mMR/q/vvvn9588838+JhsP0JpNewCANB+zFIwffzxx9NZZ53V7P0xVdRvfvObFj/fpZdeOiurBwBgPjZLo/LffvvtJqeJqlpwwQXTO++80xbtAgCgnZmlYNqrV698hqfmPPHEE6lnz55t0S4AANqZWQqmcfrQE044IX3++efT3ffZZ5+lk046Ke20005t2T4AANqJWepj+stf/jLdcMMN+bz2MTo/JscPMWVUnI40Tit6/PHHz6m2AgAwH5ulYLrsssum+++/P5+hKaZ6qlQq+fYOHTqkAQMG5HAaywAAwByfYH+llVZKt956a/rggw/S2LFjczhdbbXV0pJLLjnLKwcAgNk681OIILrxxhu39uEAAND6wU8AADCnCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgBQBMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFqGkwHTx4cNp4443T4osvnpZZZpm02267peeee66WTQIAoD0G03vvvTcdfvjh6cEHH0x33nln+uKLL9I3vvGN9Mknn9SyWQAA1MCCqYaGDx/e4PrQoUNz5XT06NHpa1/7Ws3aBQBAOwumjX300Uf5/0sttVST90+ePDlfqiZOnDjX2gYAQDsZ/DRt2rR09NFHp8033zz17du32T6p3bp1q7v07t17rrcTAID5PJhGX9Mnn3wyXX311c0uM2jQoFxVrV7Gjx8/V9sIAMB8fij/iCOOSLfccku677770gorrNDscp07d84XAADmPzUNppVKJR155JHpxhtvTCNHjkwrr7xyLZsDAEB7DaZx+H7YsGHp5ptvznOZvvXWW/n26D+68MIL17JpAAC0pz6mF154Ye4ruvXWW6eePXvWXa655ppaNgsAgPZ4KB8AAIoalQ8AQPsmmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgBQBMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgBQBMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUISaBtP77rsv7bzzzmn55ZdPHTp0SDfddFMtmwMAQHsNpp988klab7310gUXXFDLZgAAUIAFa7ny7bffPl8AAKCmwXRWTZ48OV+qJk6cONfbMG7cuPTuu+/O9fVCVffu3dOKK65Y62YAQPsOpoMHD06nnHJKzdYfoXSNNddKn3/2ac3aAF0WXiQ99+wzwikA8515KpgOGjQoHXvssQ0qpr17955r649KaYTSpXc6Li209NxbL1R98d749N4t5+R9UTAFYH4zTwXTzp0750utRSjtvNyqtW4GAMB8xTymAAAUoaYV00mTJqWxY8fWXX/55ZfTY489lpZaaimHKQEA2pmaBtNHHnkk9e/fv+56tf/owIED09ChQ2vYMgAA2lUw3XrrrVOlUqllEwAAKIQ+pgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgBQBMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgBQBMEUAIAiCKYAABRBMAUAoAiCKQAARRBMAQAogmAKAEARBFMAAIogmAIAUATBFACAIgimAAAUQTAFAKAIgikAAEUQTAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCEUE0wsuuCD16dMndenSJW266abp4YcfrnWTAABob8H0mmuuSccee2w66aST0pgxY9J6662XBgwYkCZMmFDrpgEA0J6C6bnnnpsOOeSQdMABB6S11147XXTRRWmRRRZJl112Wa2bBgDAXLRgqqEpU6ak0aNHp0GDBtXdtsACC6Rtt902PfDAA9MtP3ny5Hyp+uijj/L/J06cOFfaO2nSpP/bjrfGpmlTPp8r64T6vnj/tbp9cW7t90Db8l1Ce/sumfj/1lGpVMoOpu+++26aOnVqWnbZZRvcHtefffbZ6ZYfPHhwOuWUU6a7vXfv3mlu+uD2IXN1fdDYVlttVesmALPJdwnt7bvk448/Tt26dSs3mM6qqKxGf9SqadOmpffffz8tvfTSqUOHDjVtGy3/1RQ/JMaPH5+6du1a6+YAMI/xPTLviUpphNLll19+psvWNJh27949dezYMb399tsNbo/ryy233HTLd+7cOV/qW2KJJeZ4O2l78cfEHxQAWsv3yLxlZpXSIgY/derUKW244YbprrvualAFjeubbbZZLZsGAMBcVvND+XFofuDAgWmjjTZKm2yySTr//PPTJ598kkfpAwDQftQ8mO65557pnXfeSSeeeGJ666230vrrr5+GDx8+3YAo5g/RFSPmrG3cJQMAWsL3yPytQ6UlY/cBAGB+n2AfAACCYAoAQBEEUwAAiiCYAgBQBMGUNnfBBRekPn36pC5duqRNN900PfzwwzNcPqYIW2ONNdLCCy+cz+ZxzDHHpM8/d/5ogPbovvvuSzvvvHM+S1Cc1fGmm26a6WNGjhyZNthggzxSf9VVV01Dhw6dK22l7QmmtKlrrrkmz00bU3mMGTMmrbfeemnAgAFpwoQJTS4/bNiw9POf/zwv/8wzz6RLL700P8cvfvGLud52AGov5jKP744ocrTEyy+/nHbcccfUv3//9Nhjj6Wjjz46HXzwwen222+f422l7ZkuijYVFdKNN944DRkypO5MXlEFPfLII3MAbeyII47IgbT+2b+OO+649NBDD6V//etfc7XtAJQlKqY33nhj2m233Zpd5mc/+1n6xz/+kZ588sm62/baa6/04Ycf5nnRmbeomNJmpkyZkkaPHp223XbbutsWWGCBfP2BBx5o8jFf/epX82Oqh/tfeumldOutt6YddthhrrUbgHlXfL/U/94JcaSuue8dylbzMz8x/3j33XfT1KlTpztrV1x/9tlnm3zMd7/73fy4LbbYIkXx/r///W869NBDHcoHoEXirJFNfe9MnDgxffbZZ3n8AvMOFVNqKjqsn3HGGekPf/hD7pN6ww035EMyp512Wq2bBgDMZSqmtJnu3bunjh07prfffrvB7XF9ueWWa/IxJ5xwQtpvv/1yR/Ww7rrr5o7v3//+99Pxxx+fuwIAQHPi+6Wp752uXbuqls6DfOvTZjp16pQ23HDDBgOZYvBTXN9ss82afMynn346XfiMcBuMywNgZuL7pf73Trjzzjub/d6hbCqmtKmYKmrgwIFpo402SptsskmeozQqoAcccEC+f//990+9evVKgwcPztdjrrpzzz03ffnLX84j+seOHZurqHF7NaAC0H5MmjQpfxfUnw4qpoFaaqml0oorrpgGDRqUXn/99XTFFVfk+2NcQswE89Of/jQdeOCB6e67707XXntt7hbGvEcwpU3tueee6Z133kknnnhi7pC+/vrr5+k6qh3Tx40b16BC+stf/jJPBxL/jz80PXr0yKH09NNPr+GrAKBWHnnkkTwnaf2CR4iiR0yc/+abb+bvkqqVV145h9A4Octvf/vbtMIKK6Q//vGPeWQ+8x7zmAIAUAR9TAEAKIJgCgBAEQRTAACKIJgCAFAEwRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBGomzyyyxxBK1bkZ65ZVX8pnR4nSMs2PrrbdORx99dN31Pn365NMFz67vfe97abfddpvt5wGoEkyBeU6c7vbII49Mq6yySurcuXPq3bt3PpXtXXfd1Wan1n3++efTnBbnAP/ud7+bll9++dSlS5d8KsVdd901Pfvss/n+eF1x+sW+ffvO1npuuOGGdNppp6W2Fqd/jBDfXAAGmFULzvIjAGpcRdx8881zRfPss89O6667bvriiy/S7bffng4//PC6UDc7Fl544XyZk6LN2223XVpjjTVycOzZs2d67bXX0m233ZY+/PDDvEzHjh3TcsstN9vrWmqppVJbmjp1aq7kduvWrU2fFyBVAOYh22+/faVXr16VSZMmTXffBx98UPfvV199tbLLLrtUFl100criiy9e2WOPPSpvvfVW3f2PPfZYZeutt64stthi+f4NNtigMmrUqHzf5ZdfXunWrVvdsieddFJlvfXWq1xxxRWVlVZaqdK1a9fKnnvuWZk4cWLdMlOnTq2cccYZlT59+lS6dOlS6devX+W6665r9nU8+uijlfgT/MorrzS7zMsvv5yXiWXDPffck68PHz68sv766+f19O/fv/L2229Xbr311sqaa66ZX8vee+9d+eSTT+qeZ6uttqocddRRddfjNZx33nl1188555xK3759K4ssskhlhRVWqBx22GGVjz/+uO7+6va4+eabK2uttValY8eOuW0DBw6s7LrrrnmZ+He0rf7lpZdeqnzpS1+qnH322U2+9hdeeKHZ1w60Tw7lA/OM999/Pw0fPjxXRhdddNHp7q/2C502bVo+JB7L33vvvenOO+9ML730Uj5EX7XPPvvkQ+ejRo1Ko0ePTj//+c/TQgst1Oy6X3zxxXTTTTelW265JV/iec8888y6+wcPHpyuuOKKdNFFF6WnnnoqHXPMMWnffffNyzWlR48eaYEFFkh//etfcwVyVpx88slpyJAh6f7770/jx49P3/nOd3Kf0WHDhqV//OMf6Y477ki///3vW/x80Y7f/e53ud1/+tOf0t13351++tOfNljm008/TWeddVb64x//mJdbZpllpjusv9lmm6VDDjkkdz+Iy4orrpgOPPDAdPnllzdYNq5/7WtfS6uuuuosvW6gHah1MgZoqYceeihX2m644YYZLnfHHXfkqt64cePqbnvqqafyYx9++OF8PSqLQ4cObfLxTVVMo5pYv0L6k5/8pLLpppvmf3/++ef5/vvvv7/B8xx00EG5etmcIUOG5MdFW6Lyeeqpp1ZefPHFmVZMR4wYUbfM4MGD8231H/eDH/ygMmDAgBZXTBuLSu/SSy/dYHvEOqLKXF/9imlT6wmvv/56fi/ivQtTpkypdO/evdltD7RvKqbAPKNSiXw0c88880weOBSXqrXXXjtXVOO+cOyxx6aDDz44bbvttrnyGRXRGYmR7Isvvnjd9egTOmHChPzvsWPH5opi9BldbLHF6i5RQZ3R80blNwZyXXnllbnaeN1116V11lknV3hnpF+/fnX/XnbZZdMiiyySB4LVv63atpYYMWJE2mabbVKvXr3ya9xvv/3Se++9l19TVadOnRqst6ViYNeOO+6YLrvssnz973//e5o8eXLaY489Zvm5gPmfYArMM1ZbbbU86KYtBjjF4fA4JB2hKQ5dR3C98cYbm12+8WH+aEd0GQiTJk3K/4/D6DG1U/Xy9NNP50P1MxJBMGYUOP3009Pjjz+ettxyy/SrX/1qho+p35Zox4za1pLBZDvttFMOnddff33u1nDBBRfk+6ZMmVK3XAwGi+dtjfgBcPXVV6fPPvssH8aPLhURpgEaE0yBeUaMLh8wYEAOTp988sl091dHs6+11lq572VcqiIkxv0RQKtWX3313Bc0+mTuvvvu0/WFbKl4zpi2aty4cbnfZP1L/artzETwW3PNNZt8bXNKBNEIseecc076yle+krfJG2+80arniqpqU/1ld9hhh9wn+MILL8x9hKPfKUBTBFNgnhKhNMLPJptskit8L7zwQj48H4N34nB4iMPzMY1UDHAaM2ZMevjhh9P++++fttpqq7TRRhvlyt0RRxyRRo4cmV599dX073//Ow+CikDbGlH1/PGPf5xDbgweisP3sd4YgBTXmxIV1RigFRXVCM3RHeDSSy/Nh7zj9rklwnNMXRVtjQFif/7zn/MArtaI7g4PPfRQrsK+++67dVXbmPYqJuMfNGhQrnpX3yeAxgRTYJ4SfSkj9PXv3z8dd9xxefL56NsZk+tHRa5aebz55pvTkksumUd/R1CNx11zzTV1QSn6UEZYjQphjGrffvvt0ymnnNLqdsUE9ieccEIenR8B95vf/GY+tL/yyis3uXzMCBBBLta56aabpg022CCPbI/rxx9/fJpb1ltvvXTuuefmEfexLaO/a7yG1ohwHts2Ksgx60BUkKsOOuig3DXggAMOaMPWA/ObDjECqtaNAGD+9s9//jMPsIruFTE4C6ApgikAc0yMwH/nnXfSwIED81msoiIL0ByH8gGYY6666qq00kor5YFnv/71r2vdHKBwKqYAABRBxRQAgCIIpgAAFEEwBQCgCIIpAABFEEwBACiCYAoAQBEEUwAAiiCYAgCQSvB/AN2wvLmy8IhrAAAAAElFTkSuQmCC",
            "text/plain": [
              "<Figure size 800x500 with 1 Axes>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "import matplotlib.pyplot as plt\n",
        "from collections import Counter\n",
        "\n",
        "# Suppose matched_terms is your dictionary of gene predictions (from auto_match_go_terms)\n",
        "# where each value is a list of dictionaries with a key 'similarity'\n",
        "similarity_scores = []\n",
        "for gene, matches in matched_terms.items():\n",
        "    for match in matches:\n",
        "        similarity_scores.append(match['similarity'])\n",
        "\n",
        "# Bin the similarity scores, e.g., rounded to one decimal place.\n",
        "binned_scores = [round(score, 1) for score in similarity_scores]\n",
        "score_counts = Counter(binned_scores)\n",
        "\n",
        "# Prepare data for the bar plot.\n",
        "bins = sorted(score_counts.keys())\n",
        "counts = [score_counts[b] for b in bins]\n",
        "\n",
        "plt.figure(figsize=(8, 5))\n",
        "plt.bar(bins, counts, width=0.08, edgecolor='black')\n",
        "plt.xlabel('Cosine Similarity')\n",
        "plt.ylabel('Count')\n",
        "plt.title('Distribution of GO Term Similarity Scores')\n",
        "plt.xticks(bins)\n",
        "plt.show()\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ONKEXAOtZtjO"
      },
      "source": [
        "## 6.2. Scoring against NCBI Gene database\n",
        "\n",
        "Need to make a new scoring system: possibly True Positive (what we found), Nobel Positive (NP), Not Detected\n",
        "\n",
        "- for cases where we have Go terms that are alike like \"Wnt signaling pathway\" and \"canonical Wnt signaling pathway\", we can embed the terms and check if they are similar. If they are, then we can consider them as the same term. If not, then we can consider them as different terms."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# term = \"Cul4A-RING E3 ubiquitin ligase complex\"\n",
        "# example_llama_term = \"ubiquitin ligase complex\"\n",
        "# def compute_semantic_similarity(term1, term2, model):\n",
        "#     \"\"\"\n",
        "#     Returns a cosine similarity score between two text strings, using\n",
        "#     a SentenceTransformer model.\n",
        "#     \"\"\"\n",
        "#     emb1 = model.encode(term1, convert_to_tensor=True)\n",
        "#     emb2 = model.encode(term2, convert_to_tensor=True)\n",
        "#     cos_sim = util.cos_sim(emb1, emb2)\n",
        "#     return float(cos_sim[0][0])  # scalar\n",
        "\n",
        "# # Compute the semantic similarity between the Llama term and the official GO term\n",
        "# res = compute_semantic_similarity(example_llama_term, term, sapbert_model)\n",
        "# print(f\"Semantic similarity between '{example_llama_term}' and '{term}': {res:.2f}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The code below takes a bit of time due to:\n",
        "- api calls to retrieve the gene ontology terms from NCBI Gene database for each gene"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 567,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "=== Checking Gene: vitamin D binding protein (ID=2638) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='neutrophil chemotaxis' | Score=0.2954874336719513| PMIDs=['35238975']\n",
            "LocalTerm='fatty acid binding' | Score=0.5969445705413818| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='vitamin D binding' | BestNCBI='vitamin D binding' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='vitamin transport' | BestNCBI='vitamin transport' (diff. references) | Score=1.0000001192092896 | PMIDs=['35238975']\n",
            "LocalTerm='blood microparticle' | BestNCBI='blood microparticle' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='actin binding' | BestNCBI='actin binding' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for vitamin D binding protein (2638) ==\n",
            "TP = 0, NP = 2, DCU = 4, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CYP2R1 (ID=120227) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='liver development' | Score=0.48067033290863037| PMIDs=['35238975']\n",
            "LocalTerm='placenta development' | Score=0.44035327434539795| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='vitamin D biosynthetic process' | BestNCBI='vitamin D biosynthetic process' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "LocalTerm='oxidoreductase activity' | BestNCBI='oxidoreductase activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "LocalTerm='endoplasmic reticulum' | BestNCBI='endoplasmic reticulum' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CYP2R1 (120227) ==\n",
            "TP = 0, NP = 2, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CYP27A1 (ID=1593) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='liver development' | Score=0.6081870794296265| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='vitamin D metabolic process' | BestNCBI='sterol metabolic process' (diff. references) | Score=0.6901999711990356 | PMIDs=['35238975']\n",
            "LocalTerm='mitochondrion' | BestNCBI='mitochondrion' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='monooxygenase activity' | BestNCBI='monooxygenase activity' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CYP27A1 (1593) ==\n",
            "TP = 0, NP = 1, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CYP27B1 (ID=1594) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='renal system process' | Score=0.40947026014328003| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='vitamin D 24-hydroxylase activity' | BestNCBI='calcidiol 1-monooxygenase activity' (diff. references) | Score=0.6804506778717041 | PMIDs=['35238975']\n",
            "LocalTerm='heme binding' | BestNCBI='heme binding' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "LocalTerm='iron ion binding' | BestNCBI='iron ion binding' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='monooxygenase activity' | BestNCBI='monooxygenase activity' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CYP27B1 (1594) ==\n",
            "TP = 0, NP = 1, DCU = 4, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CYP24A1 (ID=1591) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='endoplasmic reticulum' | Score=0.5018866658210754| PMIDs=['35238975']\n",
            "LocalTerm='calcium ion homeostasis' | Score=0.506497859954834| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='vitamin D metabolic process' | BestNCBI='vitamin D metabolic process' (diff. references) | Score=1.0000001192092896 | PMIDs=['35238975']\n",
            "LocalTerm='oxidoreductase activity' | BestNCBI='oxidoreductase activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CYP24A1 (1591) ==\n",
            "TP = 0, NP = 2, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: DBP (ID=1628) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='vitamin D binding' | Score=0.6221728324890137| PMIDs=['35238975']\n",
            "LocalTerm='blood circulation' | Score=0.44067031145095825| PMIDs=['35238975']\n",
            "LocalTerm='response to vitamin D' | Score=0.40129929780960083| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for DBP (1628) ==\n",
            "TP = 0, NP = 3, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: albumin (ID=213) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='vitamin D binding' | Score=0.6221728324890137| PMIDs=['35238975']\n",
            "LocalTerm='blood circulation' | Score=0.3988875150680542| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='extracellular region' | BestNCBI='extracellular region' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='protein binding' | BestNCBI='protein binding' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for albumin (213) ==\n",
            "TP = 0, NP = 2, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: VDR (ID=7421) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='nuclear steroid receptor activity' | BestNCBI='nuclear steroid receptor activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='regulation of gene expression' | BestNCBI='positive regulation of gene expression' (diff. references) | Score=0.8302247524261475 | PMIDs=['35238975']\n",
            "LocalTerm='vitamin D receptor signaling pathway' | BestNCBI='vitamin D receptor signaling pathway' (diff. references) | Score=0.9999998807907104 | PMIDs=['35238975']\n",
            "LocalTerm='cellular response to vitamin D' | BestNCBI='vitamin D receptor signaling pathway' (diff. references) | Score=0.7474405169487 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for VDR (7421) ==\n",
            "TP = 0, NP = 0, DCU = 5, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: PTH (ID=5741) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='parathyroid gland development' | Score=0.6240720748901367| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='hormone activity' | BestNCBI='hormone activity' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='calcium ion homeostasis' | BestNCBI='calcium ion homeostasis' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for PTH (5741) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: pregnane X receptor (ID=8856) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of vitamin D biosynthetic process' | Score=0.4524172246456146| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='nuclear steroid receptor activity' | BestNCBI='nuclear receptor activity' (diff. references) | Score=0.9625775218009949 | PMIDs=['35238975']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='binding' | BestNCBI='protein binding' (diff. references) | Score=0.9121087193489075 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for pregnane X receptor (8856) ==\n",
            "TP = 0, NP = 1, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CYP3A4 (ID=1576) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='oxidoreductase activity' | BestNCBI='oxidoreductase activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "LocalTerm='vitamin D catabolic process' | BestNCBI='vitamin D catabolic process' (diff. references) | Score=0.9999998807907104 | PMIDs=['35238975']\n",
            "LocalTerm='endoplasmic reticulum' | BestNCBI='endoplasmic reticulum' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CYP3A4 (1576) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: FGF-23 (ID=8074) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='osteoblast differentiation' | Score=0.6658450961112976| PMIDs=['35238975']\n",
            "LocalTerm='cell surface' | Score=0.6175786256790161| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='regulation of phosphate metabolic process' | BestNCBI='regulation of phosphate transport' (diff. references) | Score=0.8419133424758911 | PMIDs=['35238975']\n",
            "LocalTerm='negative regulation of vitamin D biosynthetic process' | BestNCBI='vitamin D catabolic process' (diff. references) | Score=0.6944677829742432 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for FGF-23 (8074) ==\n",
            "TP = 0, NP = 2, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TRPV6 (ID=55503) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='calcium ion transmembrane transport' | BestNCBI='calcium ion transmembrane transport' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='plasma membrane' | BestNCBI='plasma membrane' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TRPV6 (55503) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: S100G (ID=795) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='intestinal absorption' | Score=-1.0| PMIDs=['35238975']\n",
            "LocalTerm='calcium ion homeostasis' | Score=-1.0| PMIDs=['35238975']\n",
            "LocalTerm='enterocyte differentiation' | Score=-1.0| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for S100G (795) ==\n",
            "TP = 0, NP = 3, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: ATP2B1 (ID=490) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='calcium ion transport' | BestNCBI='calcium ion transport' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='plasma membrane' | BestNCBI='plasma membrane' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for ATP2B1 (490) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CLDN2 (ID=9075) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='intestinal absorption' | Score=0.6041404604911804| PMIDs=['35238975']\n",
            "LocalTerm='calcium ion transport' | Score=0.60468989610672| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='tight junction' | BestNCBI='tight junction' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CLDN2 (9075) ==\n",
            "TP = 0, NP = 2, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Klotho (ID=9365) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='fibroblast growth factor receptor signaling pathway' | BestNCBI='fibroblast growth factor receptor signaling pathway' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='cell surface' | BestNCBI='plasma membrane' (diff. references) | Score=0.786027193069458 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Klotho (9365) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: osteocalcin (ID=12097) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='bone development' | BestNCBI='bone development' (diff. references) | Score=1.0000001192092896 | PMIDs=['35238975']\n",
            "LocalTerm='osteoblast differentiation' | BestNCBI='osteoblast differentiation' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for osteocalcin (12097) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: p16 (ID=13088) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='negative regulation of cell population proliferation' | Score=0.2524229884147644| PMIDs=['35238975']\n",
            "LocalTerm='cellular senescence' | Score=0.4080943465232849| PMIDs=['35238975']\n",
            "LocalTerm='regulation of cellular senescence' | Score=0.35133153200149536| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for p16 (13088) ==\n",
            "TP = 0, NP = 3, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: p19 (ID=83430) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='negative regulation of cell population proliferation' | Score=0.4022631347179413| PMIDs=['35238975']\n",
            "LocalTerm='cellular senescence' | Score=0.4753301739692688| PMIDs=['35238975']\n",
            "LocalTerm='regulation of apoptotic process' | Score=0.3539924621582031| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for p19 (83430) ==\n",
            "TP = 0, NP = 3, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: IL-10 (ID=3586) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='negative regulation of inflammatory response' | BestNCBI='negative regulation of inflammatory response' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='cytokine production' | BestNCBI='positive regulation of cytokine production' (diff. references) | Score=0.8044841289520264 | PMIDs=['35238975']\n",
            "LocalTerm='immune system process' | BestNCBI='immune response' (diff. references) | Score=0.860161304473877 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for IL-10 (3586) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: IL-1b (ID=3553) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='inflammatory response' | BestNCBI='inflammatory response' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "LocalTerm='cytokine activity' | BestNCBI='cytokine activity' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for IL-1b (3553) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: IL-6 (ID=3569) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='inflammatory response' | BestNCBI='inflammatory response' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "LocalTerm='cytokine activity' | BestNCBI='cytokine activity' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for IL-6 (3569) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TNFalpha (ID=7124) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='inflammatory response' | BestNCBI='inflammatory response' (diff. references) | Score=0.9999999403953552 | PMIDs=['35238975']\n",
            "LocalTerm='cytokine activity' | BestNCBI='cytokine activity' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TNFalpha (7124) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: RANKL (ID=8600) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cytokine activity' | BestNCBI='cytokine activity' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='inflammatory response' | BestNCBI='immune response' (diff. references) | Score=0.7686572670936584 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for RANKL (8600) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: COX2 (ID=4513) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='inflammatory response' | Score=0.4913233518600464| PMIDs=['35238975']\n",
            "LocalTerm='prostaglandin biosynthetic process' | Score=0.3362801671028137| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for COX2 (4513) ==\n",
            "TP = 0, NP = 2, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: insulin (ID=3630) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='insulin secretion' | Score=0.5927863121032715| PMIDs=['35238975']\n",
            "LocalTerm='cellular response to insulin stimulus' | Score=0.6700571775436401| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='glucose homeostasis' | BestNCBI='glucose homeostasis' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for insulin (3630) ==\n",
            "TP = 0, NP = 2, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: renin (ID=5972) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='blood vessel morphogenesis' | Score=0.5310389995574951| PMIDs=['35238975']\n",
            "LocalTerm='circulatory system process' | Score=0.47136417031288147| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='peptidase activity' | BestNCBI='peptidase activity' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for renin (5972) ==\n",
            "TP = 0, NP = 2, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: PHEX (ID=5251) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of phosphate transport' | Score=0.6726651191711426| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='metalloendopeptidase activity' | BestNCBI='metalloendopeptidase activity' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "LocalTerm='cell surface' | BestNCBI='plasma membrane' (diff. references) | Score=0.786027193069458 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for PHEX (5251) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: SLC34A1 (ID=6569) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='calcium ion homeostasis' | BestNCBI='intracellular phosphate ion homeostasis' (diff. references) | Score=0.6963862180709839 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for SLC34A1 (6569) ==\n",
            "TP = 0, NP = 0, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: alpha-fetoprotein (ID=174) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='liver development' | Score=0.45417964458465576| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for alpha-fetoprotein (174) ==\n",
            "TP = 0, NP = 1, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: megalin (ID=4036) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='peptide receptor activity' | Score=0.6407327651977539| PMIDs=['35238975']\n",
            "LocalTerm='pyrenoid tubule' | Score=0.5318969488143921| PMIDs=['35238975']\n",
            "LocalTerm='placenta development' | Score=0.6264646649360657| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='endocytosis' | BestNCBI='endocytosis' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for megalin (4036) ==\n",
            "TP = 0, NP = 3, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: cubilin (ID=8029) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='peptide receptor activity' | Score=0.6740656495094299| PMIDs=['35238975']\n",
            "LocalTerm='pyrenoid tubule' | Score=0.5318969488143921| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='vitamin transport' | BestNCBI='cobalamin transport' (diff. references) | Score=0.7361850738525391 | PMIDs=['35238975']\n",
            "LocalTerm='endocytosis' | BestNCBI='endocytosis' (diff. references) | Score=1.0 | PMIDs=['35238975']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for cubilin (8029) ==\n",
            "TP = 0, NP = 2, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Gc2 (ID=83733) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='vitamin D binding' | Score=0.6221728324890137| PMIDs=['35238975']\n",
            "LocalTerm='blood microparticle' | Score=0.554679811000824| PMIDs=['35238975']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Gc2 (83733) ==\n",
            "TP = 0, NP = 2, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Global Summary Across All Genes ===\n",
            "Total entries: 104\n",
            "TP = 0, NP = 43, DCU = 61, ND = 0\n"
          ]
        }
      ],
      "source": [
        "import json\n",
        "from sentence_transformers import SentenceTransformer, util\n",
        "from Bio import Entrez\n",
        "import xml.etree.ElementTree as ET\n",
        "import time\n",
        "import requests\n",
        "\n",
        "###############################################################################\n",
        "# CONFIGURATION\n",
        "###############################################################################\n",
        "Entrez.email = \"your_email@example.com\"  # Use a valid email address\n",
        "\n",
        "SIMILARITY_THRESHOLD = 0.68\n",
        "MARGIN = 0.1  # The margin by which we'll allow a same-paper match to be \"close enough\"\n",
        "\n",
        "###############################################################################\n",
        "# 1) Utility: map from local \"namespace\" to NCBI's GO Category strings\n",
        "###############################################################################\n",
        "def map_namespace_to_ncbi_category(ns):\n",
        "    if ns == \"biological_process\":\n",
        "        return \"Process\"\n",
        "    elif ns == \"molecular_function\":\n",
        "        return \"Function\"\n",
        "    elif ns == \"cellular_component\":\n",
        "        return \"Component\"\n",
        "    return None\n",
        "\n",
        "###############################################################################\n",
        "# 2) Fetch the GO annotations from NCBI Gene for a given gene_id\n",
        "###############################################################################\n",
        "def fetch_gene_go_terms(gene_id):\n",
        "    \"\"\"\n",
        "    Fetch the NCBI Gene record (XML) for `gene_id` and parse out GO annotations.\n",
        "    Returns a dict of lists, keyed by: \"Function\", \"Process\", \"Component\".\n",
        "    \"\"\"\n",
        "    # Sleep to respect NCBI usage guidelines\n",
        "    time.sleep(0.3)\n",
        "\n",
        "    handle = Entrez.efetch(db=\"gene\", id=gene_id, retmode=\"xml\")\n",
        "    xml_data = handle.read()\n",
        "    handle.close()\n",
        "\n",
        "    root = ET.fromstring(xml_data)\n",
        "    go_annotations = {}\n",
        "\n",
        "    for gene_commentary in root.findall(\".//Gene-commentary\"):\n",
        "        heading_elem = gene_commentary.find(\"Gene-commentary_heading\")\n",
        "        if heading_elem is not None and heading_elem.text == \"GeneOntology\":\n",
        "            for category_block in gene_commentary.findall(\".//Gene-commentary_comment/Gene-commentary\"):\n",
        "                label_elem = category_block.find(\"Gene-commentary_label\")\n",
        "                if label_elem is None:\n",
        "                    continue\n",
        "                category_label = label_elem.text  # \"Function\", \"Process\", or \"Component\"\n",
        "\n",
        "                if category_label not in go_annotations:\n",
        "                    go_annotations[category_label] = []\n",
        "\n",
        "                inner_comment = category_block.find(\"Gene-commentary_comment\")\n",
        "                if inner_comment is None:\n",
        "                    continue\n",
        "\n",
        "                for go_term_block in inner_comment.findall(\"Gene-commentary\"):\n",
        "                    dbtag_db = go_term_block.find(\".//Dbtag_db\")\n",
        "                    if dbtag_db is not None and dbtag_db.text == \"GO\":\n",
        "                        # Parse the GO ID\n",
        "                        object_id_elem = go_term_block.find(\".//Object-id_id\")\n",
        "                        if object_id_elem is not None:\n",
        "                            numeric_id = object_id_elem.text\n",
        "                            padded_id = numeric_id.zfill(7)\n",
        "                            go_id = f\"GO:{padded_id}\"\n",
        "                        else:\n",
        "                            go_id = \"GO:unknown\"\n",
        "\n",
        "                        # Parse the GO term name\n",
        "                        anchor_elem = go_term_block.find(\".//Other-source_anchor\")\n",
        "                        go_term_name = anchor_elem.text if anchor_elem is not None else \"Unknown Term\"\n",
        "\n",
        "                        # Evidence code\n",
        "                        evidence_code = None\n",
        "                        post_text_elem = go_term_block.find(\".//Other-source_post-text\")\n",
        "                        if post_text_elem is not None:\n",
        "                            post_text_str = post_text_elem.text\n",
        "                            if post_text_str and \"evidence: \" in post_text_str:\n",
        "                                evidence_code = post_text_str.split(\"evidence:\")[-1].strip()\n",
        "                            else:\n",
        "                                evidence_code = post_text_str\n",
        "\n",
        "                        # PubMed IDs\n",
        "                        pubmed_ids = []\n",
        "                        refs_block = go_term_block.find(\"Gene-commentary_refs\")\n",
        "                        if refs_block is not None:\n",
        "                            for pub in refs_block.findall(\"Pub\"):\n",
        "                                pmid_elem = pub.find(\".//PubMedId\")\n",
        "                                if pmid_elem is not None:\n",
        "                                    pubmed_ids.append(pmid_elem.text)\n",
        "\n",
        "                        go_annotations[category_label].append({\n",
        "                            \"go_id\": go_id,\n",
        "                            \"term\": go_term_name,\n",
        "                            \"evidence\": evidence_code,\n",
        "                            \"pubmed_ids\": pubmed_ids\n",
        "                        })\n",
        "\n",
        "    return go_annotations\n",
        "\n",
        "###############################################################################\n",
        "# 3) Utility: compute cosine similarity using Sentence-BERT\n",
        "###############################################################################\n",
        "def compute_semantic_similarity(term1, term2, model):\n",
        "    emb1 = model.encode(term1, convert_to_tensor=True)\n",
        "    emb2 = model.encode(term2, convert_to_tensor=True)\n",
        "    cos_sim = util.cos_sim(emb1, emb2)\n",
        "    return float(cos_sim[0][0])\n",
        "\n",
        "###############################################################################\n",
        "# 4) Utility: convert a PMCID to PMID using NCBI E-utilities\n",
        "###############################################################################\n",
        "pmc_cache = {} # Cache for PMCID to PMID conversion\n",
        "\n",
        "def pmcid_to_pmid(pmcid):\n",
        "    \"\"\"\n",
        "    Convert a PMCID (e.g. 'PMC123456') to a numeric PMID string using NCBI E-utilities.\n",
        "    Returns None if not found or conversion fails.\n",
        "    \"\"\"\n",
        "    # cache since we dont want to hit the same PMCID multiple times\n",
        "    if pmcid in pmc_cache:\n",
        "        return pmc_cache[pmcid]\n",
        "    \n",
        "    base = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/\"\n",
        "    url = base + f\"esummary.fcgi?db=pmc&id={pmcid[3:]}&retmode=xml\"\n",
        "    # Note pmcid[3:] strips \"PMC\" prefix if we pass \"PMC123456\" to the 'id=' param\n",
        "\n",
        "    resp = requests.get(url)\n",
        "    if resp.status_code != 200:\n",
        "        return None\n",
        "\n",
        "    root = ET.fromstring(resp.text)\n",
        "\n",
        "    for item in root.findall(\".//Item\"):\n",
        "        if item.get(\"Name\") == \"pmid\":\n",
        "            pmid = item.text\n",
        "            pmc_cache[pmcid] = pmid\n",
        "            return pmid\n",
        "\n",
        "    return None\n",
        "\n",
        "def convert_local_pmcids_to_pmids(local_data):\n",
        "    for gene_entry in local_data:\n",
        "        for local_ann in gene_entry.get(\"go_data\", []):\n",
        "            refs = local_ann.get(\"reference(s)\", [])\n",
        "            new_refs = []\n",
        "            for r in refs:\n",
        "                # If r starts with 'PMC', attempt to convert\n",
        "                if r.startswith(\"PMC\"):\n",
        "                    time.sleep(0.3)  # Respect NCBI usage guidelines\n",
        "                    pmid = pmcid_to_pmid(r)\n",
        "                    if pmid:\n",
        "                        new_refs.append(pmid)\n",
        "                    else:\n",
        "                        # Conversion failed or none found, fallback\n",
        "                        new_refs.append(r)\n",
        "                else:\n",
        "                    new_refs.append(r)\n",
        "            local_ann[\"reference(s)\"] = new_refs\n",
        "\n",
        "###############################################################################\n",
        "# 5) Main comparison & classification logic with tie-break margin\n",
        "###############################################################################\n",
        "def compare_json_vs_ncbi(json_path, model):\n",
        "    \"\"\"\n",
        "    For each local annotation (term + references), we:\n",
        "      1) Gather all possible NCBI annotations in the same GO category\n",
        "      2) Find best match overall & best match among same-paper (i.e., shared PMIDs)\n",
        "      3) If best_same_score >= best_overall_score - MARGIN, prefer the same-paper match\n",
        "      4) Classify that chosen match as TP if it shares references, else DCU (if >= threshold),\n",
        "         or NP if below threshold.\n",
        "    \"\"\"\n",
        "    with open(json_path, \"r\") as f:\n",
        "        local_data = json.load(f)\n",
        "\n",
        "    # Convert PMCIDs in local_data to numeric PMIDs since NCBI uses PMIDs even thorugh GO annotations are curated from PMC\n",
        "    convert_local_pmcids_to_pmids(local_data)\n",
        "\n",
        "    all_comparisons = []\n",
        "\n",
        "    for gene_entry in local_data:\n",
        "        gene_name = gene_entry[\"gene\"]\n",
        "        gene_id = gene_entry[\"geneID\"]\n",
        "        local_go_data = gene_entry.get(\"go_data\", [])\n",
        "\n",
        "        print(f\"\\n=== Checking Gene: {gene_name} (ID={gene_id}) ===\")\n",
        "\n",
        "        ncbi_annotations = fetch_gene_go_terms(gene_id)\n",
        "\n",
        "        # We'll keep track of official GO annotations (for ND detection)\n",
        "        official_by_pmid = set()\n",
        "        ncbi_lookup = {}  # key: (category, term, pmid) -> annotation dict\n",
        "        for category_label, ann_list in ncbi_annotations.items():\n",
        "            for ann in ann_list:\n",
        "                for pmid in ann[\"pubmed_ids\"]:\n",
        "                    key = (category_label, ann[\"term\"], pmid)\n",
        "                    official_by_pmid.add(key)\n",
        "                    ncbi_lookup[key] = ann\n",
        "\n",
        "        matched_officials = set()\n",
        "        comparisons_this_gene = []\n",
        "\n",
        "        # Per-gene counters\n",
        "        count_tp = 0\n",
        "        count_np = 0\n",
        "        count_dcu = 0\n",
        "\n",
        "        #\n",
        "        # 1) CLASSIFY EACH LOCAL ANNOTATION (TP, NP, DCU)\n",
        "        #\n",
        "        for local_ann in local_go_data:\n",
        "            local_term = local_ann[\"term\"]\n",
        "            local_refs = set(local_ann.get(\"reference(s)\", []))\n",
        "            local_ns = local_ann.get(\"namespace\", None)\n",
        "            target_category = map_namespace_to_ncbi_category(local_ns)\n",
        "\n",
        "            classification = None\n",
        "            chosen_match_info = None\n",
        "            chosen_score = -1.0\n",
        "\n",
        "            best_score_all = -1.0\n",
        "            best_info_all = None\n",
        "            best_score_same = -1.0\n",
        "            best_info_same = None\n",
        "\n",
        "            if target_category and target_category in ncbi_annotations:\n",
        "                possible_anns = ncbi_annotations[target_category]\n",
        "\n",
        "                # (A) Find best match overall\n",
        "                for ann in possible_anns:\n",
        "                    sc = compute_semantic_similarity(local_term, ann[\"term\"], model)\n",
        "                    if sc > best_score_all:\n",
        "                        best_score_all = sc\n",
        "                        best_info_all = ann\n",
        "\n",
        "                # (B) Find best match among same-paper\n",
        "                same_paper_anns = [\n",
        "                    ann for ann in possible_anns\n",
        "                    if set(ann[\"pubmed_ids\"]).intersection(local_refs)\n",
        "                ]\n",
        "                for ann in same_paper_anns:\n",
        "                    sc = compute_semantic_similarity(local_term, ann[\"term\"], model)\n",
        "                    if sc > best_score_same:\n",
        "                        best_score_same = sc\n",
        "                        best_info_same = ann\n",
        "\n",
        "                # (C) Tie-break\n",
        "                if best_score_same >= 0:\n",
        "                    # There's at least one same-paper annotation\n",
        "                    if best_score_same >= best_score_all - MARGIN:\n",
        "                        chosen_score = best_score_same\n",
        "                        chosen_match_info = best_info_same\n",
        "                    else:\n",
        "                        chosen_score = best_score_all\n",
        "                        chosen_match_info = best_info_all\n",
        "                else:\n",
        "                    chosen_score = best_score_all\n",
        "                    chosen_match_info = best_info_all\n",
        "\n",
        "                # (D) Classify\n",
        "                if chosen_score >= SIMILARITY_THRESHOLD and chosen_match_info is not None:\n",
        "                    overlap = set(chosen_match_info[\"pubmed_ids\"]).intersection(local_refs)\n",
        "                    if overlap:\n",
        "                        classification = \"TP\"\n",
        "                        count_tp += 1\n",
        "                        # Mark the official annotation as matched\n",
        "                        for pmid in chosen_match_info[\"pubmed_ids\"]:\n",
        "                            if pmid in local_refs:\n",
        "                                matched_officials.add((target_category, chosen_match_info[\"term\"], pmid))\n",
        "                    else:\n",
        "                        classification = \"DCU\"\n",
        "                        count_dcu += 1\n",
        "                else:\n",
        "                    classification = \"NP\"\n",
        "                    count_np += 1\n",
        "            else:\n",
        "                # No recognized category => NP\n",
        "                classification = \"NP\"\n",
        "                count_np += 1\n",
        "\n",
        "            rec = {\n",
        "                \"gene\": gene_name,\n",
        "                \"geneID\": gene_id,\n",
        "                \"local_term\": local_term,\n",
        "                \"namespace\": local_ns,\n",
        "                \"references_used\": list(local_refs),\n",
        "                \"classification\": classification,\n",
        "                \"best_score\": chosen_score,\n",
        "            }\n",
        "            if chosen_match_info:\n",
        "                rec[\"best_ncbi_match\"] = chosen_match_info[\"term\"]\n",
        "                rec[\"best_ncbi_go_id\"] = chosen_match_info[\"go_id\"]\n",
        "                rec[\"best_ncbi_pubmed_ids\"] = chosen_match_info[\"pubmed_ids\"]\n",
        "            else:\n",
        "                rec[\"best_ncbi_match\"] = None\n",
        "                rec[\"best_ncbi_go_id\"] = None\n",
        "                rec[\"best_ncbi_pubmed_ids\"] = []\n",
        "\n",
        "            comparisons_this_gene.append(rec)\n",
        "\n",
        "        #\n",
        "        # 2) DETECT ND (NOT DETECTED) AND STORE AS ADDITIONAL RECORDS\n",
        "        #\n",
        "        # ND means: official NCBI annotation referencing the local paper(s),\n",
        "        # but never matched as a TP above.\n",
        "        #\n",
        "        undetected = official_by_pmid - matched_officials\n",
        "\n",
        "        # The local paper set is all refs mentioned in local_go_data\n",
        "        all_local_pmids_for_gene = set()\n",
        "        for ann in local_go_data:\n",
        "            all_local_pmids_for_gene |= set(ann.get(\"reference(s)\", []))\n",
        "\n",
        "        count_nd = 0\n",
        "        for key in undetected:\n",
        "            cat_label, ncbi_term, pmid = key\n",
        "            if pmid in all_local_pmids_for_gene:\n",
        "                ann = ncbi_lookup.get(key, {})\n",
        "                # Skip if the evidence code is IEA\n",
        "                if ann.get(\"evidence\") == \"IEA\":\n",
        "                    continue\n",
        "                rec = {\n",
        "                    \"gene\": gene_name,\n",
        "                    \"geneID\": gene_id,\n",
        "                    \"local_term\": None,\n",
        "                    \"namespace\": cat_label,\n",
        "                    \"references_used\": [pmid],\n",
        "                    \"classification\": \"ND\",  # now includes evidence info\n",
        "                    \"best_score\": None,\n",
        "                    \"best_ncbi_match\": ncbi_term,\n",
        "                    \"best_ncbi_go_id\": ann.get(\"go_id\"),\n",
        "                    \"best_ncbi_pubmed_ids\": [pmid],\n",
        "                    \"evidence\": ann.get(\"evidence\")\n",
        "                }\n",
        "                comparisons_this_gene.append(rec)\n",
        "                count_nd += 1\n",
        "        \n",
        "        # 3) PRINT RESULTS\n",
        "       \n",
        "        # Summaries\n",
        "        tps_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"TP\"]\n",
        "        nps_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"NP\"]\n",
        "        dcus_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"DCU\"]\n",
        "        nds_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"ND\"]\n",
        "\n",
        "        print(\"\\n-- TPs --\")\n",
        "        for item in tps_for_gene:\n",
        "            print(f\"LocalTerm='{item['local_term']}' | BestNCBI='{item['best_ncbi_match']}' \"\n",
        "                  f\"| Score={item['best_score']:.2f} | PMIDs={item['references_used']}\")\n",
        "\n",
        "        print(\"\\n-- NPs --\")\n",
        "        for item in nps_for_gene:\n",
        "            print(f\"LocalTerm='{item['local_term']}' | Score={item['best_score']}\"\n",
        "                  f\"| PMIDs={item['references_used']}\")\n",
        "\n",
        "        print(\"\\n-- DCUs --\")\n",
        "        for item in dcus_for_gene:\n",
        "            print(f\"LocalTerm='{item['local_term']}' | BestNCBI='{item['best_ncbi_match']}' \"\n",
        "                  f\"(diff. references) | Score={item['best_score']} | PMIDs={item['references_used']}\")\n",
        "\n",
        "        print(\"\\n-- ND --\")\n",
        "        for item in nds_for_gene:\n",
        "            print(f\"NCBITerm='{item['best_ncbi_match']}' | Evidence={item.get('evidence')} | PMID={item['best_ncbi_pubmed_ids']}\")\n",
        "        \n",
        "        print(f\"\\n== Per-gene Summary for {gene_name} ({gene_id}) ==\")\n",
        "        print(f\"TP = {count_tp}, NP = {count_np}, DCU = {count_dcu}, ND = {count_nd}\")\n",
        "        print(\"--------------------------------------------------------\")\n",
        "\n",
        "        all_comparisons.extend(comparisons_this_gene)\n",
        "\n",
        "    return all_comparisons\n",
        "\n",
        "###############################################################################\n",
        "# 6) Example usage\n",
        "###############################################################################\n",
        "json_file_path = f\"./processed_inferences/processed_output_{doc_id}.json\"\n",
        "\n",
        "try:\n",
        "    comparison_results = compare_json_vs_ncbi(json_file_path, sapbert_model)\n",
        "except requests.HTTPError as e:\n",
        "    print(f\"HTTPError: {e.code} - {e.reason}\")\n",
        "    print(\"Please check the gene ID or ensure that the NCBI API is accessible.\")\n",
        "\n",
        "# Now each item in comparison_results has classification in {TP, NP, DCU, ND}\n",
        "\n",
        "# You can also sum ND from the final list if you want a global ND count\n",
        "tp_global = sum(1 for r in comparison_results if r[\"classification\"] == \"TP\")\n",
        "np_global = sum(1 for r in comparison_results if r[\"classification\"] == \"NP\")\n",
        "dcu_global = sum(1 for r in comparison_results if r[\"classification\"] == \"DCU\")\n",
        "nd_global = sum(1 for r in comparison_results if r[\"classification\"] == \"ND\")\n",
        "\n",
        "print(\"\\n=== Global Summary Across All Genes ===\")\n",
        "print(f\"Total entries: {len(comparison_results)}\")\n",
        "print(f\"TP = {tp_global}, NP = {np_global}, DCU = {dcu_global}, ND = {nd_global}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Will put TP, ND, DCU and NP in a excel file in the output folder\n",
        "change the sheet_name to the name of the file you want to save it as in the excel. After that, open excel and copy the sheet to the main sheet. Then delete the other sheet."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 568,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Finished! Data written to sheet 'PMC8892115' in file: /Users/crystalcho/Downloads/Evaluation.xlsx\n"
          ]
        }
      ],
      "source": [
        "import pandas as pd\n",
        "import os\n",
        "from openpyxl import load_workbook\n",
        "from collections import defaultdict\n",
        "\n",
        "###############################################################################\n",
        "# 1) Aggregate classification_counts from comparison_results\n",
        "###############################################################################\n",
        "classification_counts = defaultdict(lambda: {\"TP\": 0, \"NP\": 0, \"DCU\": 0, \"ND\": 0})\n",
        "\n",
        "for row in comparison_results:\n",
        "    gene_id = row[\"geneID\"]\n",
        "    gene_name = row[\"gene\"]\n",
        "    classification = row[\"classification\"]\n",
        "    \n",
        "    # For each PMID in references_used, increment the classification count\n",
        "    for pmid in row[\"references_used\"]:\n",
        "        key = (gene_id, gene_name, pmid)\n",
        "        classification_counts[key][classification] += 1\n",
        "\n",
        "records_for_df = []\n",
        "for (gid, gname, pmid), counts in classification_counts.items():\n",
        "    records_for_df.append({\n",
        "        \"geneID\": gid,\n",
        "        \"gene\": gname,\n",
        "        \"pmid\": pmid,\n",
        "        \"TP\": counts[\"TP\"],\n",
        "        \"ND\": counts[\"ND\"],\n",
        "        \"NP\": counts[\"NP\"],\n",
        "        \"DCU\": counts[\"DCU\"]\n",
        "    })\n",
        "\n",
        "df = pd.DataFrame(\n",
        "    records_for_df,\n",
        "    columns=[\"geneID\", \"gene\", \"pmid\", \"TP\", \"ND\", \"NP\", \"DCU\"]\n",
        ")\n",
        "\n",
        "###############################################################################\n",
        "# 2) Write to Excel, appending a NEW SHEET for each paper\n",
        "###############################################################################\n",
        "output_path = r\"/Users/crystalcho/Downloads/Evaluation.xlsx\"\n",
        "\n",
        "# You can change this to something unique for each paper you analyze.\n",
        "sheet_name = doc_id\n",
        "\n",
        "# If the file already exists, open in append mode. Otherwise, create a new file.\n",
        "mode = \"a\" if os.path.exists(output_path) else \"w\"\n",
        "\n",
        "if mode == \"w\":\n",
        "    # Create a new file and write data\n",
        "    df.to_excel(output_path, sheet_name=sheet_name, index=False)\n",
        "else:\n",
        "    # If the file already exists, append a new sheet\n",
        "    with pd.ExcelWriter(output_path, engine=\"openpyxl\", mode=\"a\", if_sheet_exists=\"new\") as writer:\n",
        "        df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
        "\n",
        "print(f\"Finished! Data written to sheet '{sheet_name}' in file: {output_path}\")\n"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "collapsed_sections": [
        "6oaKSo6iXGOc"
      ],
      "gpuType": "T4",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "3.10.6",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.6"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "097046dc7c4f46f5866068c82b706aca": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "1675d70c8a5e426691d0af5e6ba333f3": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_784d7bf8c3be48b8ad5a3b6e6954690f",
              "IPY_MODEL_730fe49f91204e2cbd3b58d24442be0a",
              "IPY_MODEL_792eb1f8daf34e2590fda2f60e3ff684"
            ],
            "layout": "IPY_MODEL_3e0f9cdf58ef4af1a0b16f9795747dfc"
          }
        },
        "1845ddb7ff634d969fa2538209fc4e0a": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "1b8c25318a2b4b64bab3496083679c85": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "2443ed122977428c98248537123b674b": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_f7dcd748d0944da09948f4372366e8a8",
            "placeholder": "​",
            "style": "IPY_MODEL_466f8ef6ba644d31a8586abaf9830448",
            "value": " 438M/438M [00:01&lt;00:00, 251MB/s]"
          }
        },
        "3117723c389b4e47a97d6e582078e5c9": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "32ef3e232304493d8b7e129c679f4f6f": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_b1c01cfdd57e40e7bc4de9baf81d9f45",
            "placeholder": "​",
            "style": "IPY_MODEL_82a3774cc5cd4acfb1822d9cb74d6f45",
            "value": " 112/112 [00:00&lt;00:00, 9.27kB/s]"
          }
        },
        "331dbfdb6d7849c6a7b6d43bd33a7325": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_8eb65674477e477091f2501704df7eb8",
            "placeholder": "​",
            "style": "IPY_MODEL_6d29b58c64cc4f869d6944699149a60e",
            "value": "model.safetensors: 100%"
          }
        },
        "39794838a2164151a5edf298a63f8147": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "3e0f9cdf58ef4af1a0b16f9795747dfc": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "466f8ef6ba644d31a8586abaf9830448": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "47571c7570f14ab58f96e5d30129da1b": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "4a743c94312c4b5db25087a747f6b879": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_ea731973ab9c44caa479fb7934fd3b4f",
            "max": 198,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_097046dc7c4f46f5866068c82b706aca",
            "value": 198
          }
        },
        "4af857d8e5a94c278ae0fb3d7e092ac5": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "53bb7f0fbad34a818e266d0ea44ffaf1": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "54a0cadcc9ef406db05d36b690d7891c": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c939f017049643b18d5142514937af73",
            "max": 437955508,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_6c463c9a4bef4b5194113acf7ced731e",
            "value": 437955508
          }
        },
        "59dc6c4216cb4c5aad8c3764cf3d3f6d": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_cc7573c531c84e61999e2161710173f3",
            "placeholder": "​",
            "style": "IPY_MODEL_9775c0cd37bc43dcb7953f8b15343332",
            "value": " 198/198 [00:00&lt;00:00, 14.3kB/s]"
          }
        },
        "618fd6a5462e4c5296b560f888b5d638": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_331dbfdb6d7849c6a7b6d43bd33a7325",
              "IPY_MODEL_54a0cadcc9ef406db05d36b690d7891c",
              "IPY_MODEL_2443ed122977428c98248537123b674b"
            ],
            "layout": "IPY_MODEL_c3911d8aca764dd38fad4c1dded8198f"
          }
        },
        "61de99be2c2e4afe90ff06e3bb30a2b7": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "6ac927f5e2a2453d9aae4b5c474cf96d": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_7593152124e04332b709dd30ec87c5a5",
              "IPY_MODEL_4a743c94312c4b5db25087a747f6b879",
              "IPY_MODEL_59dc6c4216cb4c5aad8c3764cf3d3f6d"
            ],
            "layout": "IPY_MODEL_3117723c389b4e47a97d6e582078e5c9"
          }
        },
        "6c463c9a4bef4b5194113acf7ced731e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "6d29b58c64cc4f869d6944699149a60e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "730fe49f91204e2cbd3b58d24442be0a": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_d598fdb3d94a4095ba5a950cef321f1b",
            "max": 226150,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_39794838a2164151a5edf298a63f8147",
            "value": 226150
          }
        },
        "7593152124e04332b709dd30ec87c5a5": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c6b5cf46fbeb41adb8da935949b808c1",
            "placeholder": "​",
            "style": "IPY_MODEL_d4b8ff5330d14128ba304020f7fe8412",
            "value": "tokenizer_config.json: 100%"
          }
        },
        "769b37c671334690b6a7e5ba9059343e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_bdf351e0ccbc4aa18e168f593d1659cf",
              "IPY_MODEL_8901a8884437441c934c53a98a1dac06",
              "IPY_MODEL_32ef3e232304493d8b7e129c679f4f6f"
            ],
            "layout": "IPY_MODEL_77472246b6b0497fbe67b8a3529040e5"
          }
        },
        "77472246b6b0497fbe67b8a3529040e5": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "784d7bf8c3be48b8ad5a3b6e6954690f": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_1845ddb7ff634d969fa2538209fc4e0a",
            "placeholder": "​",
            "style": "IPY_MODEL_9b03af45365c42419bb39779b82be47c",
            "value": "vocab.txt: 100%"
          }
        },
        "792eb1f8daf34e2590fda2f60e3ff684": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_e7eaa149a903445db37dfdf719e14f9e",
            "placeholder": "​",
            "style": "IPY_MODEL_4af857d8e5a94c278ae0fb3d7e092ac5",
            "value": " 226k/226k [00:00&lt;00:00, 1.03MB/s]"
          }
        },
        "7d97b347b7cb40df81041035931d7f27": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_53bb7f0fbad34a818e266d0ea44ffaf1",
            "placeholder": "​",
            "style": "IPY_MODEL_bc435a026437452bb820c14d1ce5d3ee",
            "value": "config.json: 100%"
          }
        },
        "8039b35f8f654a5e93ab69a0d1d8f3ea": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "82a3774cc5cd4acfb1822d9cb74d6f45": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "873ea6832a0447a19938fe764cb7ac58": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "8901a8884437441c934c53a98a1dac06": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_b483173c5fa74a72bb0a88206d3206a3",
            "max": 112,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_8039b35f8f654a5e93ab69a0d1d8f3ea",
            "value": 112
          }
        },
        "8eb65674477e477091f2501704df7eb8": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "9775c0cd37bc43dcb7953f8b15343332": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "9b03af45365c42419bb39779b82be47c": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "b1c01cfdd57e40e7bc4de9baf81d9f45": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b483173c5fa74a72bb0a88206d3206a3": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b85f6bb59d024282b5552ff665da4548": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_47571c7570f14ab58f96e5d30129da1b",
            "max": 462,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_ec42f776af1b48a982636f0c122ab7e6",
            "value": 462
          }
        },
        "bc435a026437452bb820c14d1ce5d3ee": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "bdf351e0ccbc4aa18e168f593d1659cf": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_61de99be2c2e4afe90ff06e3bb30a2b7",
            "placeholder": "​",
            "style": "IPY_MODEL_873ea6832a0447a19938fe764cb7ac58",
            "value": "special_tokens_map.json: 100%"
          }
        },
        "c3911d8aca764dd38fad4c1dded8198f": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "c6b5cf46fbeb41adb8da935949b808c1": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "c939f017049643b18d5142514937af73": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "cc7573c531c84e61999e2161710173f3": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "d4b8ff5330d14128ba304020f7fe8412": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "d598fdb3d94a4095ba5a950cef321f1b": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "de6e3ef0670042378384c41ecd695f1a": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "e4718c03421b41a2a665c26293acd697": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_7d97b347b7cb40df81041035931d7f27",
              "IPY_MODEL_b85f6bb59d024282b5552ff665da4548",
              "IPY_MODEL_f1219528a46a4a8e94bd1c8de9e6e16a"
            ],
            "layout": "IPY_MODEL_1b8c25318a2b4b64bab3496083679c85"
          }
        },
        "e7eaa149a903445db37dfdf719e14f9e": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ea731973ab9c44caa479fb7934fd3b4f": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ec42f776af1b48a982636f0c122ab7e6": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "ec7200455fa34863bec2ae60f9f4aad5": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "f1219528a46a4a8e94bd1c8de9e6e16a": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_de6e3ef0670042378384c41ecd695f1a",
            "placeholder": "​",
            "style": "IPY_MODEL_ec7200455fa34863bec2ae60f9f4aad5",
            "value": " 462/462 [00:00&lt;00:00, 40.9kB/s]"
          }
        },
        "f7dcd748d0944da09948f4372366e8a8": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        }
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
