{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1AfSrrXBI8oi",
        "outputId": "3b20254b-e106-40b9-8f89-a0aa03fba087"
      },
      "outputs": [],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 33,
      "metadata": {
        "id": "UWSHSonXlTeG"
      },
      "outputs": [],
      "source": [
        "# packages\n",
        "import torch\n",
        "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
        "import time"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BuXWNP8jcVkz"
      },
      "source": [
        "# 1. Pubtator API (TODO-Partially done)\n",
        "- using the input folder, check if the file(s) have been pre-annotated by checking if the following APIs return something\n",
        "```\n",
        "https://www.ncbi.nlm.nih.gov/research/pubtator3-api/publications/export/biocxml?pmids={id}\n",
        "or\n",
        "https://www.ncbi.nlm.nih.gov/research/pubtator3-api/publications/pmc_export/biocxml?pmcids=PMC{id}\n",
        "```\n",
        "  - if annotated, retrieve it and put it in the output folder\n",
        "  - if not, use the [example code](https://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/tmTools/download/ExampleCode.Python.zip) they provide to send it to the server to be annotated (Done)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0sl2nmCPFsQq"
      },
      "source": [
        "Using colab:\n",
        "\n",
        "```\n",
        "mkdir input\n",
        "```\n",
        "\n",
        "to make a input folder"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "edkmlyhFFifa",
        "outputId": "304ffdaa-8b04-4d0b-de90-f3caa4e5f27f"
      },
      "outputs": [],
      "source": [
        "!bash"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EQIruerqbU4Q",
        "outputId": "da090b16-ca20-4798-a26d-7bce4bc4090a"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
            "To disable this warning, you can either:\n",
            "\t- Avoid using `tokenizers` before the fork if possible\n",
            "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n"
          ]
        },
        {
          "ename": "UnboundLocalError",
          "evalue": "local variable 'child' referenced before assignment",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py:151\u001b[0m, in \u001b[0;36mProcessHandler.system\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    150\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 151\u001b[0m     child \u001b[38;5;241m=\u001b[39m \u001b[43mpexpect\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspawn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msh\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-c\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcmd\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# Vanilla Pexpect\u001b[39;00m\n\u001b[1;32m    152\u001b[0m flush \u001b[38;5;241m=\u001b[39m sys\u001b[38;5;241m.\u001b[39mstdout\u001b[38;5;241m.\u001b[39mflush\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:205\u001b[0m, in \u001b[0;36mspawn.__init__\u001b[0;34m(self, command, args, timeout, maxread, searchwindowsize, logfile, cwd, env, ignore_sighup, echo, preexec_fn, encoding, codec_errors, dimensions, use_poll)\u001b[0m\n\u001b[1;32m    204\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 205\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_spawn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcommand\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpreexec_fn\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdimensions\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    206\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39muse_poll \u001b[38;5;241m=\u001b[39m use_poll\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:303\u001b[0m, in \u001b[0;36mspawn._spawn\u001b[0;34m(self, command, args, preexec_fn, dimensions)\u001b[0m\n\u001b[1;32m    300\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39margs \u001b[38;5;241m=\u001b[39m [a \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(a, \u001b[38;5;28mbytes\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m a\u001b[38;5;241m.\u001b[39mencode(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mencoding)\n\u001b[1;32m    301\u001b[0m                  \u001b[38;5;28;01mfor\u001b[39;00m a \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39margs]\n\u001b[0;32m--> 303\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mptyproc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_spawnpty\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43menv\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menv\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    304\u001b[0m \u001b[43m                             \u001b[49m\u001b[43mcwd\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcwd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    306\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mptyproc\u001b[38;5;241m.\u001b[39mpid\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:315\u001b[0m, in \u001b[0;36mspawn._spawnpty\u001b[0;34m(self, args, **kwargs)\u001b[0m\n\u001b[1;32m    314\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m'''Spawn a pty and return an instance of PtyProcess.'''\u001b[39;00m\n\u001b[0;32m--> 315\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mptyprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mPtyProcess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspawn\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ptyprocess/ptyprocess.py:315\u001b[0m, in \u001b[0;36mPtyProcess.spawn\u001b[0;34m(cls, argv, cwd, env, echo, preexec_fn, dimensions, pass_fds)\u001b[0m\n\u001b[1;32m    314\u001b[0m os\u001b[38;5;241m.\u001b[39mclose(exec_err_pipe_write)\n\u001b[0;32m--> 315\u001b[0m exec_err_data \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexec_err_pipe_read\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m4096\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    316\u001b[0m os\u001b[38;5;241m.\u001b[39mclose(exec_err_pipe_read)\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
            "\nDuring handling of the above exception, another exception occurred:\n",
            "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[104], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# SubmitText_request.py\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msystem\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpip install unidecode\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mrequests\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mos\u001b[39;00m\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py:657\u001b[0m, in \u001b[0;36mZMQInteractiveShell.system_piped\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    655\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39muser_ns[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_exit_code\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m system(cmd)\n\u001b[1;32m    656\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 657\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39muser_ns[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_exit_code\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43msystem\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar_expand\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcmd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdepth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
            "File \u001b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py:167\u001b[0m, in \u001b[0;36mProcessHandler.system\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    162\u001b[0m         out_size \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlen\u001b[39m(child\u001b[38;5;241m.\u001b[39mbefore)\n\u001b[1;32m    163\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m    164\u001b[0m     \u001b[38;5;66;03m# We need to send ^C to the process.  The ascii code for '^C' is 3\u001b[39;00m\n\u001b[1;32m    165\u001b[0m     \u001b[38;5;66;03m# (the character is known as ETX for 'End of Text', see\u001b[39;00m\n\u001b[1;32m    166\u001b[0m     \u001b[38;5;66;03m# curses.ascii.ETX).\u001b[39;00m\n\u001b[0;32m--> 167\u001b[0m     \u001b[43mchild\u001b[49m\u001b[38;5;241m.\u001b[39msendline(\u001b[38;5;28mchr\u001b[39m(\u001b[38;5;241m3\u001b[39m))\n\u001b[1;32m    168\u001b[0m     \u001b[38;5;66;03m# Read and print any more output the program might produce on its\u001b[39;00m\n\u001b[1;32m    169\u001b[0m     \u001b[38;5;66;03m# way out.\u001b[39;00m\n\u001b[1;32m    170\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
            "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'child' referenced before assignment"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "--- Logging error ---\n",
            "Traceback (most recent call last):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py\", line 151, in system\n",
            "    child = pexpect.spawn(self.sh, args=['-c', cmd])  # Vanilla Pexpect\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py\", line 205, in __init__\n",
            "    self._spawn(command, args, preexec_fn, dimensions)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py\", line 303, in _spawn\n",
            "    self.ptyproc = self._spawnpty(self.args, env=self.env,\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py\", line 315, in _spawnpty\n",
            "    return ptyprocess.PtyProcess.spawn(args, **kwargs)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ptyprocess/ptyprocess.py\", line 289, in spawn\n",
            "    os.execv(command, argv)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py\", line 743, in new_execv\n",
            "    send_process_about_to_be_replaced()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py\", line 985, in send_process_about_to_be_replaced\n",
            "    py_db.send_process_about_to_be_replaced()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/pydevd.py\", line 2043, in send_process_about_to_be_replaced\n",
            "    time.sleep(0.05)\n",
            "KeyboardInterrupt\n",
            "\n",
            "During handling of the above exception, another exception occurred:\n",
            "\n",
            "Traceback (most recent call last):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n",
            "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
            "  File \"/var/folders/s7/c_3jrk0j61l1nkjwz_167y9r0000gn/T/ipykernel_41327/1190397474.py\", line 2, in <module>\n",
            "    get_ipython().system('pip install unidecode')\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 657, in system_piped\n",
            "    self.user_ns[\"_exit_code\"] = system(self.var_expand(cmd, depth=1))\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py\", line 167, in system\n",
            "    child.sendline(chr(3))\n",
            "UnboundLocalError: local variable 'child' referenced before assignment\n",
            "\n",
            "During handling of the above exception, another exception occurred:\n",
            "\n",
            "Traceback (most recent call last):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/logging/__init__.py\", line 1103, in emit\n",
            "    stream.write(msg + self.terminator)\n",
            "OSError: [Errno 9] Bad file descriptor\n",
            "Call stack:\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n",
            "    return _run_code(code, main_globals, None,\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/runpy.py\", line 86, in _run_code\n",
            "    exec(code, run_globals)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
            "    app.launch_new_instance()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
            "    app.start()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
            "    self.io_loop.start()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
            "    self.asyncio_loop.run_forever()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py\", line 600, in run_forever\n",
            "    self._run_once()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py\", line 1896, in _run_once\n",
            "    handle._run()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/asyncio/events.py\", line 80, in _run\n",
            "    self._context.run(self._callback, *self._args)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
            "    await self.process_one()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
            "    await dispatch(*args)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
            "    await result\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
            "    await super().execute_request(stream, ident, parent)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
            "    reply_content = await reply_content\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
            "    res = shell.run_cell(\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
            "    return super().run_cell(*args, **kwargs)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n",
            "    result = self._run_cell(\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n",
            "    result = runner(coro)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
            "    coro.send(None)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n",
            "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n",
            "    if await self.run_code(code, result, async_=asy):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3599, in run_code\n",
            "    self.showtraceback(running_compiled_code=True)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 2179, in showtraceback\n",
            "    self._showtraceback(etype, value, stb)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 569, in _showtraceback\n",
            "    dh.session.send(  # type:ignore[attr-defined]\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/jupyter_client/session.py\", line 831, in send\n",
            "    get_logger().warning(\"WARNING: attempted to send message from fork\\n%s\", msg)\n",
            "Message: 'WARNING: attempted to send message from fork\\n%s'\n",
            "Arguments: {'header': {'msg_id': '9135c61b-d20a3a8c8e9f0be36432a3fe_56627_1043', 'msg_type': 'error', 'username': 'crystalcho', 'session': '9135c61b-d20a3a8c8e9f0be36432a3fe', 'date': datetime.datetime(2025, 3, 27, 9, 1, 19, 312784, tzinfo=datetime.timezone.utc), 'version': '5.3'}, 'msg_id': '9135c61b-d20a3a8c8e9f0be36432a3fe_56627_1043', 'msg_type': 'error', 'parent_header': {'date': datetime.datetime(2025, 3, 27, 9, 1, 12, 778000, tzinfo=tzutc()), 'msg_id': '9db4b385-bdf7-45c8-91a7-8f43e4b90954', 'msg_type': 'execute_request', 'session': 'e3a89fb4-0c8b-4ecc-a37c-6312a14df9e6', 'username': '60edfaf0-327c-463c-a35d-0c9f29c77a14', 'version': '5.2'}, 'content': {'traceback': ['\\x1b[0;31m---------------------------------------------------------------------------\\x1b[0m', '\\x1b[0;31mKeyboardInterrupt\\x1b[0m                         Traceback (most recent call last)', \"File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py:151\\x1b[0m, in \\x1b[0;36mProcessHandler.system\\x1b[0;34m(self, cmd)\\x1b[0m\\n\\x1b[1;32m    150\\x1b[0m \\x1b[38;5;28;01melse\\x1b[39;00m:\\n\\x1b[0;32m--> 151\\x1b[0m     child \\x1b[38;5;241m=\\x1b[39m \\x1b[43mpexpect\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mspawn\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43msh\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43margs\\x1b[49m\\x1b[38;5;241;43m=\\x1b[39;49m\\x1b[43m[\\x1b[49m\\x1b[38;5;124;43m'\\x1b[39;49m\\x1b[38;5;124;43m-c\\x1b[39;49m\\x1b[38;5;124;43m'\\x1b[39;49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43mcmd\\x1b[49m\\x1b[43m]\\x1b[49m\\x1b[43m)\\x1b[49m  \\x1b[38;5;66;03m# Vanilla Pexpect\\x1b[39;00m\\n\\x1b[1;32m    152\\x1b[0m flush \\x1b[38;5;241m=\\x1b[39m sys\\x1b[38;5;241m.\\x1b[39mstdout\\x1b[38;5;241m.\\x1b[39mflush\\n\", 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:205\\x1b[0m, in \\x1b[0;36mspawn.__init__\\x1b[0;34m(self, command, args, timeout, maxread, searchwindowsize, logfile, cwd, env, ignore_sighup, echo, preexec_fn, encoding, codec_errors, dimensions, use_poll)\\x1b[0m\\n\\x1b[1;32m    204\\x1b[0m \\x1b[38;5;28;01melse\\x1b[39;00m:\\n\\x1b[0;32m--> 205\\x1b[0m     \\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43m_spawn\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43mcommand\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43margs\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43mpreexec_fn\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43mdimensions\\x1b[49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m    206\\x1b[0m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39muse_poll \\x1b[38;5;241m=\\x1b[39m use_poll\\n', 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:303\\x1b[0m, in \\x1b[0;36mspawn._spawn\\x1b[0;34m(self, command, args, preexec_fn, dimensions)\\x1b[0m\\n\\x1b[1;32m    300\\x1b[0m     \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39margs \\x1b[38;5;241m=\\x1b[39m [a \\x1b[38;5;28;01mif\\x1b[39;00m \\x1b[38;5;28misinstance\\x1b[39m(a, \\x1b[38;5;28mbytes\\x1b[39m) \\x1b[38;5;28;01melse\\x1b[39;00m a\\x1b[38;5;241m.\\x1b[39mencode(\\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39mencoding)\\n\\x1b[1;32m    301\\x1b[0m                  \\x1b[38;5;28;01mfor\\x1b[39;00m a \\x1b[38;5;129;01min\\x1b[39;00m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39margs]\\n\\x1b[0;32m--> 303\\x1b[0m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39mptyproc \\x1b[38;5;241m=\\x1b[39m \\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43m_spawnpty\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43margs\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43menv\\x1b[49m\\x1b[38;5;241;43m=\\x1b[39;49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43menv\\x1b[49m\\x1b[43m,\\x1b[49m\\n\\x1b[1;32m    304\\x1b[0m \\x1b[43m                             \\x1b[49m\\x1b[43mcwd\\x1b[49m\\x1b[38;5;241;43m=\\x1b[39;49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mcwd\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[38;5;241;43m*\\x1b[39;49m\\x1b[38;5;241;43m*\\x1b[39;49m\\x1b[43mkwargs\\x1b[49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m    306\\x1b[0m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39mpid \\x1b[38;5;241m=\\x1b[39m \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39mptyproc\\x1b[38;5;241m.\\x1b[39mpid\\n', \"File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pexpect/pty_spawn.py:315\\x1b[0m, in \\x1b[0;36mspawn._spawnpty\\x1b[0;34m(self, args, **kwargs)\\x1b[0m\\n\\x1b[1;32m    314\\x1b[0m \\x1b[38;5;250m\\x1b[39m\\x1b[38;5;124;03m'''Spawn a pty and return an instance of PtyProcess.'''\\x1b[39;00m\\n\\x1b[0;32m--> 315\\x1b[0m \\x1b[38;5;28;01mreturn\\x1b[39;00m \\x1b[43mptyprocess\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mPtyProcess\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mspawn\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43margs\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[38;5;241;43m*\\x1b[39;49m\\x1b[38;5;241;43m*\\x1b[39;49m\\x1b[43mkwargs\\x1b[49m\\x1b[43m)\\x1b[49m\\n\", 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ptyprocess/ptyprocess.py:289\\x1b[0m, in \\x1b[0;36mPtyProcess.spawn\\x1b[0;34m(cls, argv, cwd, env, echo, preexec_fn, dimensions, pass_fds)\\x1b[0m\\n\\x1b[1;32m    288\\x1b[0m \\x1b[38;5;28;01mif\\x1b[39;00m env \\x1b[38;5;129;01mis\\x1b[39;00m \\x1b[38;5;28;01mNone\\x1b[39;00m:\\n\\x1b[0;32m--> 289\\x1b[0m     \\x1b[43mos\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mexecv\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43mcommand\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43margv\\x1b[49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m    290\\x1b[0m \\x1b[38;5;28;01melse\\x1b[39;00m:\\n', 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py:743\\x1b[0m, in \\x1b[0;36mcreate_execv.<locals>.new_execv\\x1b[0;34m(path, args)\\x1b[0m\\n\\x1b[1;32m    742\\x1b[0m     send_process_created_message()\\n\\x1b[0;32m--> 743\\x1b[0m     \\x1b[43msend_process_about_to_be_replaced\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m    745\\x1b[0m \\x1b[38;5;28;01mreturn\\x1b[39;00m \\x1b[38;5;28mgetattr\\x1b[39m(os, original_name)(path, args)\\n', 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py:985\\x1b[0m, in \\x1b[0;36msend_process_about_to_be_replaced\\x1b[0;34m()\\x1b[0m\\n\\x1b[1;32m    984\\x1b[0m \\x1b[38;5;28;01mif\\x1b[39;00m py_db \\x1b[38;5;129;01mis\\x1b[39;00m \\x1b[38;5;129;01mnot\\x1b[39;00m \\x1b[38;5;28;01mNone\\x1b[39;00m:\\n\\x1b[0;32m--> 985\\x1b[0m     \\x1b[43mpy_db\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43msend_process_about_to_be_replaced\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43m)\\x1b[49m\\n', 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/pydevd.py:2043\\x1b[0m, in \\x1b[0;36mPyDB.send_process_about_to_be_replaced\\x1b[0;34m(self)\\x1b[0m\\n\\x1b[1;32m   2042\\x1b[0m \\x1b[38;5;28;01mwhile\\x1b[39;00m \\x1b[38;5;129;01mnot\\x1b[39;00m sent[\\x1b[38;5;241m0\\x1b[39m]:\\n\\x1b[0;32m-> 2043\\x1b[0m     \\x1b[43mtime\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43msleep\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;241;43m0.05\\x1b[39;49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m   2045\\x1b[0m     \\x1b[38;5;28;01mif\\x1b[39;00m (time\\x1b[38;5;241m.\\x1b[39mtime() \\x1b[38;5;241m-\\x1b[39m initial_time) \\x1b[38;5;241m>\\x1b[39m timeout:\\n', '\\x1b[0;31mKeyboardInterrupt\\x1b[0m: ', '\\nDuring handling of the above exception, another exception occurred:\\n', '\\x1b[0;31mUnboundLocalError\\x1b[0m                         Traceback (most recent call last)', \"Cell \\x1b[0;32mIn[104], line 2\\x1b[0m\\n\\x1b[1;32m      1\\x1b[0m \\x1b[38;5;66;03m# SubmitText_request.py\\x1b[39;00m\\n\\x1b[0;32m----> 2\\x1b[0m \\x1b[43mget_ipython\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43m)\\x1b[49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43msystem\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;124;43m'\\x1b[39;49m\\x1b[38;5;124;43mpip install unidecode\\x1b[39;49m\\x1b[38;5;124;43m'\\x1b[39;49m\\x1b[43m)\\x1b[49m\\n\\x1b[1;32m      4\\x1b[0m \\x1b[38;5;28;01mimport\\x1b[39;00m\\x1b[38;5;250m \\x1b[39m\\x1b[38;5;21;01mrequests\\x1b[39;00m\\n\\x1b[1;32m      5\\x1b[0m \\x1b[38;5;28;01mimport\\x1b[39;00m\\x1b[38;5;250m \\x1b[39m\\x1b[38;5;21;01mos\\x1b[39;00m\\n\", 'File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/zmqshell.py:657\\x1b[0m, in \\x1b[0;36mZMQInteractiveShell.system_piped\\x1b[0;34m(self, cmd)\\x1b[0m\\n\\x1b[1;32m    655\\x1b[0m         \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39muser_ns[\\x1b[38;5;124m\"\\x1b[39m\\x1b[38;5;124m_exit_code\\x1b[39m\\x1b[38;5;124m\"\\x1b[39m] \\x1b[38;5;241m=\\x1b[39m system(cmd)\\n\\x1b[1;32m    656\\x1b[0m \\x1b[38;5;28;01melse\\x1b[39;00m:\\n\\x1b[0;32m--> 657\\x1b[0m     \\x1b[38;5;28mself\\x1b[39m\\x1b[38;5;241m.\\x1b[39muser_ns[\\x1b[38;5;124m\"\\x1b[39m\\x1b[38;5;124m_exit_code\\x1b[39m\\x1b[38;5;124m\"\\x1b[39m] \\x1b[38;5;241m=\\x1b[39m \\x1b[43msystem\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[38;5;28;43mself\\x1b[39;49m\\x1b[38;5;241;43m.\\x1b[39;49m\\x1b[43mvar_expand\\x1b[49m\\x1b[43m(\\x1b[49m\\x1b[43mcmd\\x1b[49m\\x1b[43m,\\x1b[49m\\x1b[43m \\x1b[49m\\x1b[43mdepth\\x1b[49m\\x1b[38;5;241;43m=\\x1b[39;49m\\x1b[38;5;241;43m1\\x1b[39;49m\\x1b[43m)\\x1b[49m\\x1b[43m)\\x1b[49m\\n', \"File \\x1b[0;32m~/.pyenv/versions/3.10.6/lib/python3.10/site-packages/IPython/utils/_process_posix.py:167\\x1b[0m, in \\x1b[0;36mProcessHandler.system\\x1b[0;34m(self, cmd)\\x1b[0m\\n\\x1b[1;32m    162\\x1b[0m         out_size \\x1b[38;5;241m=\\x1b[39m \\x1b[38;5;28mlen\\x1b[39m(child\\x1b[38;5;241m.\\x1b[39mbefore)\\n\\x1b[1;32m    163\\x1b[0m \\x1b[38;5;28;01mexcept\\x1b[39;00m \\x1b[38;5;167;01mKeyboardInterrupt\\x1b[39;00m:\\n\\x1b[1;32m    164\\x1b[0m     \\x1b[38;5;66;03m# We need to send ^C to the process.  The ascii code for '^C' is 3\\x1b[39;00m\\n\\x1b[1;32m    165\\x1b[0m     \\x1b[38;5;66;03m# (the character is known as ETX for 'End of Text', see\\x1b[39;00m\\n\\x1b[1;32m    166\\x1b[0m     \\x1b[38;5;66;03m# curses.ascii.ETX).\\x1b[39;00m\\n\\x1b[0;32m--> 167\\x1b[0m     \\x1b[43mchild\\x1b[49m\\x1b[38;5;241m.\\x1b[39msendline(\\x1b[38;5;28mchr\\x1b[39m(\\x1b[38;5;241m3\\x1b[39m))\\n\\x1b[1;32m    168\\x1b[0m     \\x1b[38;5;66;03m# Read and print any more output the program might produce on its\\x1b[39;00m\\n\\x1b[1;32m    169\\x1b[0m     \\x1b[38;5;66;03m# way out.\\x1b[39;00m\\n\\x1b[1;32m    170\\x1b[0m     \\x1b[38;5;28;01mtry\\x1b[39;00m:\\n\", \"\\x1b[0;31mUnboundLocalError\\x1b[0m: local variable 'child' referenced before assignment\"], 'ename': 'UnboundLocalError', 'evalue': \"local variable 'child' referenced before assignment\"}, 'metadata': {}}\n",
            "--- Logging error ---\n",
            "Message: 'KeyboardInterrupt caught in kernel.'\n",
            "Arguments: ()\n",
            "Traceback (most recent call last):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
            "    await result\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
            "    await super().execute_request(stream, ident, parent)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 793, in execute_request\n",
            "    reply_msg: dict[str, t.Any] = self.session.send(  # type:ignore[assignment]\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/jupyter_client/session.py\", line 863, in send\n",
            "    stream.send_multipart(to_send, copy=copy)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/zmq/eventloop/zmqstream.py\", line 376, in send_multipart\n",
            "    self._add_io_state(zmq.POLLOUT)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/zmq/eventloop/zmqstream.py\", line 679, in _add_io_state\n",
            "    self._update_handler(self._state)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/zmq/eventloop/zmqstream.py\", line 691, in _update_handler\n",
            "    if state & self.socket.events:\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/zmq/sugar/attrsettr.py\", line 56, in __getattr__\n",
            "    return self._get_attr_opt(upper_key, opt)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/zmq/sugar/attrsettr.py\", line 68, in _get_attr_opt\n",
            "    return self.get(opt)\n",
            "  File \"_zmq.py\", line 864, in zmq.backend.cython._zmq.Socket.get\n",
            "  File \"_zmq.py\", line 1362, in zmq.backend.cython._zmq._getsockopt\n",
            "  File \"_zmq.py\", line 167, in zmq.backend.cython._zmq._check_rc\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/zmq/error.py\", line 121, in __init__\n",
            "    super().__init__(EINTR)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/zmq/error.py\", line 28, in __init__\n",
            "    def __init__(self, errno: int | None = None, msg: str | None = None):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_trace_dispatch_regular.py\", line 354, in __call__\n",
            "    def __call__(self, frame, event, arg):\n",
            "KeyboardInterrupt\n",
            "\n",
            "During handling of the above exception, another exception occurred:\n",
            "\n",
            "Traceback (most recent call last):\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/logging/__init__.py\", line 1104, in emit\n",
            "    self.flush()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/logging/__init__.py\", line 1084, in flush\n",
            "    self.stream.flush()\n",
            "OSError: [Errno 9] Bad file descriptor\n",
            "Call stack:\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n",
            "    return _run_code(code, main_globals, None,\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/runpy.py\", line 86, in _run_code\n",
            "    exec(code, run_globals)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
            "    app.launch_new_instance()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
            "    app.start()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
            "    self.io_loop.start()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
            "    self.asyncio_loop.run_forever()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py\", line 600, in run_forever\n",
            "    self._run_once()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/asyncio/base_events.py\", line 1896, in _run_once\n",
            "    handle._run()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/asyncio/events.py\", line 80, in _run\n",
            "    self._context.run(self._callback, *self._args)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
            "    await self.process_one()\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
            "    await dispatch(*args)\n",
            "  File \"/Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 442, in dispatch_shell\n",
            "    self.log.error(\"KeyboardInterrupt caught in kernel.\")\n"
          ]
        }
      ],
      "source": [
        "# SubmitText_request.py\n",
        "!pip install unidecode\n",
        "\n",
        "import requests\n",
        "import os\n",
        "import sys\n",
        "from unidecode import unidecode\n",
        "import unicodedata\n",
        "import re\n",
        "\n",
        "def submit_text_request(input_folder, bioconcept, output_file_session_number):\n",
        "    pattern = r'[^0-9a-zA-Z\\!\\@\\#\\$\\%\\^\\&\\*\\(\\)\\_\\+\\{\\}\\|\\:\\\"\\<\\>\\?\\-\\=\\[\\]\\\\;\\'\\,\\.\\/ \\t\\n\\r]'\n",
        "    unicode_to_regular = {}\n",
        "    with open('lib/unicode.txt', 'r', encoding='utf-8') as input_file:\n",
        "        for line in input_file:\n",
        "            line = line.strip()\n",
        "            parts = line.split(\"\\t\")\n",
        "            if len(parts) == 2:\n",
        "                uni, reg = parts\n",
        "                if reg == '-whitespace-':\n",
        "                    unicode_to_regular[uni] = ' '\n",
        "                else:\n",
        "                    unicode_to_regular[uni] = reg\n",
        "\n",
        "    with open(output_file_session_number, 'w', encoding='utf-8') as output_file:\n",
        "        for filename in os.listdir(input_folder):\n",
        "            if filename.startswith('.'):\n",
        "                continue  # Skip hidden files\n",
        "            text_str = ''\n",
        "            with open(os.path.join(input_folder, filename), 'r', encoding='utf-8') as file_input:\n",
        "                for line in file_input:\n",
        "                    normalized_line = unicodedata.normalize('NFC', line)\n",
        "                    for uni, reg in unicode_to_regular.items():\n",
        "                        normalized_line = normalized_line.replace(uni, reg)\n",
        "                    normalized_line = unidecode(normalized_line)  # Assuming unidecode does similar work to Perl's version\n",
        "                    text_str += normalized_line\n",
        "            text_str = re.sub(pattern, ' ', text_str)\n",
        "            url = \"https://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/RESTful/request.cgi\"\n",
        "            response = requests.post(url, data={'text': text_str, 'bioconcept': bioconcept})\n",
        "\n",
        "            if response.status_code == 200:\n",
        "                session_number = response.json().get('id', '')\n",
        "                print(f\"Thanks for your submission. The session number is: {session_number}\")\n",
        "                output_file.write(f\"{session_number}\\t{filename}\\n\")\n",
        "            else:\n",
        "                print(f\"Error: HTTP {response.status_code} for {filename}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 38,
      "metadata": {
        "id": "KccYUHYJcEfG"
      },
      "outputs": [],
      "source": [
        "# SubmitText_retreive.py\n",
        "import requests\n",
        "import os\n",
        "import sys\n",
        "\n",
        "def submit_text_retrieve(input_folder, inputfile_session_number, output_folder):\n",
        "    # Ensure output directory exists\n",
        "    if not os.path.exists(output_folder):\n",
        "        os.makedirs(output_folder)\n",
        "\n",
        "    sn_hash = {}\n",
        "    with open(inputfile_session_number, 'r') as input_file:\n",
        "        for line in input_file:\n",
        "            line = line.strip()\n",
        "            if line:\n",
        "                session_number, input_filename = line.split(\"\\t\")\n",
        "                sn_hash[session_number] = input_filename\n",
        "\n",
        "    base_url = \"https://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/RESTful/retrieve.cgi\"\n",
        "\n",
        "    for session_number, input_filename in sn_hash.items():\n",
        "        output_file_path = os.path.join(output_folder, input_filename)\n",
        "        if os.path.exists(output_file_path):\n",
        "            print(f\"{output_file_path} - finished\")\n",
        "            continue\n",
        "\n",
        "        retrieve_url = f\"{base_url}?id={session_number}\"\n",
        "        response = requests.get(retrieve_url)\n",
        "        if response.status_code == 200:\n",
        "            with open(output_file_path, 'w', encoding='utf-8') as output_file:\n",
        "                output_file.write(response.text)\n",
        "            print(f\"{session_number} : Result is retrieved.\")\n",
        "        else:\n",
        "            print(f\"{session_number} : Error retrieving results - {response.status_code}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 61,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8_lpKfmLcaQ-",
        "outputId": "756447fc-8c07-43f6-f6a6-4db2fcb68e62"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Thanks for your submission. The session number is: 0C75BFC8CBEEEC29AFF0\n",
            "Thanks for your submission. The session number is: A95E896E53DA4E110CBF\n"
          ]
        }
      ],
      "source": [
        "input_folder = \"input\"       # pull xml,json, or pubtator formatted files here to be annotated\n",
        "bioconcept = \"Gene\"                                         # we only want the genes\n",
        "output_file_session_number = \"Session_Number.txt\"  # output file for session number which will be used to retrieve the results\n",
        "submit_text_request(input_folder, bioconcept, output_file_session_number)    # submit the text for annotation"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "02JFZ15QfHqt"
      },
      "source": [
        "## Running the code below takes time, so it might say error 400. Just rerun it a bit later."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 75,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "mmQYNjmJdzwO",
        "outputId": "0284e0d6-cf18-4d40-85f9-b49dd3e7890a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "0C75BFC8CBEEEC29AFF0 : Error retrieving results - 400\n",
            "A95E896E53DA4E110CBF : Error retrieving results - 400\n"
          ]
        }
      ],
      "source": [
        "input_folder = \"input\"\n",
        "inputfile_session_number = \"Session_Number.txt\"\n",
        "output_folder = \"output\"\n",
        "submit_text_retrieve(input_folder, inputfile_session_number, output_folder)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "duMZk-ubdK0K"
      },
      "source": [
        "# 2. Relevance (Done)\n",
        "- This can be integrated with Gene Chunking. We can just check if the paper has genes (i.e check if there is an item in the dictionary @ line 171)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S-Gak1R_RF8Y"
      },
      "source": [
        "*3. Gene Chunking*\n",
        "=============\n",
        "Given a **preannotated** article (check input folder) analyze the **XML** file via `parse_xml_file(xml_file_path)` which returns a dictionary of \"Gene Class\" from that file  \n",
        "\n",
        "The Gene Class has the following information:\n",
        "- self.gene_id: the gene id (from the xml file)\n",
        "- self.occurences: A list of occurences of the gene in the text. Each occurence is **3-sentences long**.\n",
        "    - The second sentence contains the explicit mention of the gene.\n",
        "    - The first and third sentences are the context of the gene mention.\n",
        "- self.symbol: the gene symbol (from NCBI. Initialized to None)\n",
        "- self.organism: the organism of the gene (from NCBI. Initialized to None)\n",
        "- self.full_name: the full name of the gene (from NCBI. Initialized to None. Set to name_from_article if not found in NCBI)\n",
        "- self.also_known_as: a list of other names of the gene (from NCBI. Initialized to None. Set to name_from_article if not found in NCBI)\n",
        "- self.name_from_article: the name of the gene as found in the article\n",
        "\n",
        "Steps:\n",
        "1. define the class Gene\n",
        "2. Parse the xml file\n",
        "3. update the Gene class with the information from NCBI\n",
        "Now you have a dictionary of Gene classes.~\n",
        "\n",
        "TIPS:\n",
        "- control + F \"sentence_buffer\" to find where you can adjust the buffer"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 141,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "M9sSOV2PXqOJ",
        "outputId": "6722699e-967f-4dee-c51c-fce74a88168e"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
            "To disable this warning, you can either:\n",
            "\t- Avoid using `tokenizers` before the fork if possible\n",
            "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
            "120119.96s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: biopython in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (1.85)\n",
            "Requirement already satisfied: numpy in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from biopython) (2.2.2)\n",
            "\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip available: \u001b[0m\u001b[31;49m22.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
            "There are 31 genes in the paper.\n",
            "Gene(7049)\n",
            "  Symbol          : TGFBR3\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : transforming growth factor beta receptor 3\n",
            "  Also Known As   : BGCAN, betaglycan\n",
            "  In-text Name    : TGFBR3\n",
            "  Occurrences     : ['Novel role of the clustered miR-23b-3p and miR-27b-3p in enhanced expression of fibrosis-associated genes by targeting TGFBR3 in atrial fibroblasts', \"Overexpression of miR-23b-3p and miR-27b-3p enhanced the expression of collagen, type I, alpha 1 (COL1A1), COL3A1 and ACTA2 in HAFs without significant effects on their proliferation and migration. Luciferase assay showed that miR-23b-3p and miR-27b-3p targeted two different sites in 3'-UTR of transforming growth factor (TGF)-beta1 receptor 3 (TGFBR3) respectively. Consistently, TGFBR3 siRNA could increase fibrosis-related genes expression, along with the Smad1 inactivation and Smad3 activation in HAFs.\", 'Additionally, overexpression of TGFBR3 could alleviate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively. Moreover, Smad3 was activated in HAFs in response to Ang-II treatment and inactivation of Smad3 attenuated up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs. Taken together, these results suggest that the clustered miR-23b-3p and miR-27b-3p consistently promote atrial fibrosis by targeting TGFBR3 to activate Smad3 signalling in HAFs, suggesting that miR-23b-3p and miR-27b-3p are potential therapeutic targets for atrial fibrosis.', 'In light of its essential role in the pathogenesis of fibrosis, TGF-beta has emerged as an attractive therapeutic target.22 TGF-beta1 receptor 3(TGFBR3) is negatively regulated by TGF-beta1 in multiple cell types and mediates TGF-beta superfamily ligand-dependent as well as ligand-independent signalling to both Smad and non-Smad signalling pathways.23 A recent study showed that TGFBR3 inhibited TGF-beta1 expression, Smad2/3 activation and TGFBR1-TGFBR2 complex formation and collagen production in cardiac fibroblasts.24 However, it still not known whether TGFBR3 can be modulated by miR-23b-3p and miR-27b-3p in atrial fibroblasts.', 'In this study, we investigated the expression of miR-23b-3p and miR-27b-3p in atrial appendages of patients with rheumatic AF and in HAFs treated with Ang-II. Based on bio-informatic analysis, we speculated that TGFBR3 is a target of miR-23b-3p and miR-27b-3p and further validated their interactions in this study. By illustrating the underlying mechanism that miR-23b-3p and miR-27b-3p facilitate atrial fibrosis progression, this study will provide new insight into atrial fibrosis target therapy.', \"MiR-23b-3p and miR-27b-3p interact with 3'-UTR of TGF-beta receptor type 3 (TGFBR3)\", \"Analysis of the databases Mirdb (www.mirdb.org) and TargetScan-Vert (www.targetscan.org) showed that TGFBR3 was a potential target gene of miR-23b-3p and miR-27b-3p respectively. The matching positions for miR-23b-3p and miR-27b-3p within 3'-UTR of the targeted TGFBR3 mRNA are shown in Figure 5A. The dual luciferase assay demonstrated that miR-23b-3p reduced the luciferase activity by binding to the site of 285-292 of TGFBR3 3'-UTR (P < 0.01) and miR-27b-3p significantly reduced the luciferase activity by binding to the site of 1525-1531 of TGFBR3 3'-UTR (P < 0.05) (Figure 5B).\", 'Moreover, mRNA and protein expression of TGFBR3 was decreased in HAFs with overexpression of miR-23b-3p, miR-27b-3p and miR-23b-3p combined with miR-27b-3p respectively (P < 0.05 respectively) (Figure 5C,D). However, no significant differences in mRNA and protein expression of TGFBR3 were observed among HAFs with transfection of miR-23b-3p, miR-27b-3p and miR-23b-3p combined with miR-27b-3p respectively (Figure 5C,D). TGFBR3 expression at mRNA and protein level was significantly up-regulated in Ang-II-treated HAFs (P < 0.01 respectively) (Figure S2A), but not in atrial appendages of AF patients (Figure S2B).', 'MiR-23b-3p, miR-27b-3p and TGFBR3 siRNA enhance the expression of fibrosis-associated gene in HAFs', 'MiR-23b-3p, miR-27b-3p mimic and TGFBR3 siRNA were transfected into HAFs followed by determination of the expression of fibrosis-related gene. Western blot results showed that miR-23b-3p, miR-27b-3p and TGFBR3 siRNA efficiently suppressed TGFBR3 expression and increased COL1A1, COL3A1 and ACTA2 expression in HAFs (P < 0.05, P < 0.01 respectively) (Figure 6A). Consistently, phosphorylated Smad1 (p-Smad1) was decreased, but p-Smad3 was increased in HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6A).', 'Meanwhile, Smad3 was significantly decreased in the cytoplasm and accumulated in the nuclei of HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6B,C). Western blot results showed that adenovirous-mediated overexpression of TGFBR3 could markedly attenuate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively (P < 0.01, P < 0.001 respectively) (Figure 6D). Moreover, TGFBR3 overexpression could significantly alleviate Smad3 activation in HAFs with transfection of miR-23b-3p and miR-27b-3p respectively (P < 0.05 respectively) (Figure 6D).', 'As an inhibitory factor of TGF-beta1/Smad3 signalling, TGFBR3 was demonstrated to blunt Tgfbr1/Smad2/3 and potentiate the Acvrl1/Smad1 axis, which was active in lung fibroblasts and antagonized Tgfbr1/Smad2/3 signalling.34 A synthetic peptide from TGFBR3, P144, was reported to efficiently inhibit the TGF-beta1-dependent signalling pathway and collagen type I synthesis in cardiac fibroblasts.35 Further studies showed that overexpression of TGFBR3 deactivated the TGF-beta1 pathway by decreasing the expression of TGF-beta1, Smad3 activation, TGFBR1-TGFBR2 complex formation and inhibiting the pro-fibrotic microRNA-21 expression.24, 36 Therefore, TGFBR3 may act as a protective factor representing a potential therapeutic target for cardiac fibrosis.', \"Our current study has provided several lines of evidence to support the notion that TGFBR3 is a common target of miR-23b-3p and miR-27b-3p. First, the in silico prediction indicated that TGFBR3 was a potential target of miR-23b-3p and miR-27b-3p and the dual luciferase assay revealed that miR-23b-3p and miR-27b-3p specifically bound to the 285-292, 1525-1531 site in the 3'-UTR of TGFBR3 respectively. We confirmed that miR-23b-3p and miR-27b-3p mimic inhibited TGFBR3 expression at both mRNA and protein levels in HAFs.\", 'In parallel with miR-23b-3p and miR-27b-3p, TGFBR3 knockdown enhanced the expressions of COL1A1, COL3A1 and ACTA2 in HAFs. Importantly, Smad1 was inactivated, whereas Smad3 was activated and accumulated in the nuclei of HAFs with transfection of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively. These data indicated that Smad1 inactivation and Smad3 activation mediate the pro-fibrotic effects of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA in atrial fibrosis.', 'Meanwhile, the activation of Smad1 through TGFBR3 and the negative modulation between p-Smad1 and p-Smad3 were in agreement with the previous report.37 Furthermore, TGFBR3 overexpression could significantly block the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p or miR-27b-3p. Therefore, our study demonstrated that the clustered miR-23b-3p and miR-27b-3p increase Smad3 activation to promote atrial fibrosis by targeting TGFBR3 in HAFs.', \"The present study revealed that TGFBR3 is a common target of miR-23b-3p and miR-27b-3p with two different binding sites in 3'-UTR of TGFBR3 by miR-23b-3p and miR-27b-3p respectively. However, no synergetic effect on fibrosis-related gene expression was observed in HAFs with transfection of miR-23b-3p combined with miR-27b-3p. Moreover, no synergetic effect on mRNA and protein expression of TGFBR3 was found in HAFs with transfection of miR-23b-3p combined with miR-27b-3p.\", \"The potential reason may be associated with that binding of miR-23b-3p combined with miR-27b-3p in 3'-UTR of TGFBR3 didn't markedly aggravate TGFBR3 mRNA degradation compare with miR-23b-3p only. These results provide evidence that the inhibitory effect of two or more miRNAs on one gene expression was different with that of two or more small interfering RNAs (siRNAs) targeting one gene.\", 'In this study, our data showed that TGFBR3 was significantly increased in Ang-II-induced HAFs, but no obvious increase of TGFBR3 was observed in the atrial appendages of AF patients. These results indicated that expression of TGFBR3 in atrial appendages was modulated by more factors, including miRNAs, cytokines and other inflammatory factors.', 'We used the Smad3 inhibitors, SIS3 and Naringenin and Smad3 siRNA to verify the participation of Smad3 activation in Ang-II-promoted up-regulation of miR-23b-3p and miR-27b-3p in HAFs. Therefore, we proposed a positive feedback loop that Smad3 activation modulates miR-23b-3p and miR-27b-3p expression, which then activate the Smad3 signalling by targeting TGFBR3 in HAFs.', 'MiR-23b-3p and miR-27b-3p consistently enhanced fibrosis-related gene expression with no significant effects on proliferation and migration of HAFs. TGFBR3 was confirmed as a target of miR-23b-3p and miR-27b-3p and the TGFBR3/Smad1-Smad3 axis mediated the pro-fibrotic effect of miR-23b-3p and miR-27b-3p in HAFs. We also demonstrated that the activation of the Smad3 signalling mediates the up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-induced HAFs.', 'TGFBR3, a potential negative regulator of TGF-beta signaling, protects cardiac fibroblasts from hypoxia-induced apoptosis', 'Glucocorticoids recruit Tgfbr3 and Smad1 to shift transforming growth factor-beta signaling from the Tgfbr1/Smad2/3 axis to the Acvrl1/Smad1 axis in lung fibroblasts']\n",
            "  Document ID     : PMC6484421\n",
            "There are 22 occurrence(s) (each occurrence is at most 3 sentences long) for gene 7049 : TGFBR3.\n",
            "\n",
            "Gene(11606)\n",
            "  Symbol          : Agt\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : angiotensinogen\n",
            "  Also Known As   : AngI, AngII, Aogen, Serpina8\n",
            "  In-text Name    : Ang-II\n",
            "  Occurrences     : ['Human atrial fibroblasts (HAFs) were isolated from atrial appendage tissue of patients with sinus rhythm. A cell model of atrial fibrosis was achieved in Ang-II-induced HAFs. Cell proliferation and migration were detected.', 'We found that miR-23b-3p and miR-27b-3p were markedly increased in atrial appendage tissues of AF patients and in Ang-II-treated HAFs. Overexpression of miR-23b-3p and miR-27b-3p enhanced the expression of collagen, type I, alpha 1 (COL1A1), COL3A1 and ACTA2 in HAFs without significant effects on their proliferation and migration.', 'Additionally, overexpression of TGFBR3 could alleviate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively. Moreover, Smad3 was activated in HAFs in response to Ang-II treatment and inactivation of Smad3 attenuated up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs. Taken together, these results suggest that the clustered miR-23b-3p and miR-27b-3p consistently promote atrial fibrosis by targeting TGFBR3 to activate Smad3 signalling in HAFs, suggesting that miR-23b-3p and miR-27b-3p are potential therapeutic targets for atrial fibrosis.', 'In this study, we investigated the expression of miR-23b-3p and miR-27b-3p in atrial appendages of patients with rheumatic AF and in HAFs treated with Ang-II. Based on bio-informatic analysis, we speculated that TGFBR3 is a target of miR-23b-3p and miR-27b-3p and further validated their interactions in this study.', 'miR-23b-3p and miR-27b-3p are up-regulated in Ang-II-treated human atrial fibroblasts (HAFs)', 'The immunofluorescent staining confirmed that HAFs were successfully isolated and cultured, with positive ACTA2 expression (Figure 2A). Fibrosis-associated genes including COL1A1, COL3A1 and ACTA2, were significantly increased in Ang-II-treated HAFs (Figure 2B). Meanwhile, RT-qPCR results showed that miR-23b/27b precursor, miR-23b-3p and miR-27b-3p were markedly up-regulated in HAFs exposed to Ang-II (P < 0.05 respectively) (Figure 2C,D).', 'However, no significant differences in mRNA and protein expression of TGFBR3 were observed among HAFs with transfection of miR-23b-3p, miR-27b-3p and miR-23b-3p combined with miR-27b-3p respectively (Figure 5C,D). TGFBR3 expression at mRNA and protein level was significantly up-regulated in Ang-II-treated HAFs (P < 0.01 respectively) (Figure S2A), but not in atrial appendages of AF patients (Figure S2B).', 'MiR-23b-3p and miR-27b-3p are up-regulated by Ang-II through the Smad3 pathway', 'We first examined the time-dependent activation of the Smad3 pathway in Ang-II-treated HAFs. Our western blot results showed that the phosphorylation level of Smad3 was significantly increased in HAFs at 10 minutes in response to Ang-II treatment (Figure 7A). Smad3 inhibitors SIS3, Narigenin and Smad3 siRNA were used to suppress Smad3 activation in Ang-II-treated HAFs.', 'RT-qPCR results revealed that treatment with SIS, Narigenin and Smad3 siRNA could prevent Ang-II-induced miR-23b/27b precursor, miR-23b-3p and miR-27b-3p expression respectively (Figure 7B,C). Moreover, SIS3, Narigenin and Smad3 siRNA could significantly attenuate the increase of COL1A1 and COL3A1 in Ang-II-treated HAFs (Figure 7D). Collectively, our results suggest that up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs results from activation of the Smad3 pathway (Figure 7E).', 'Meanwhile, the obvious increase of inflammatory factors, such as IL-1beta and CRP, was also found in atrial appendage tissue of AF patients in the current study, which was supported by the previous reports.28 Some miRNAs, such as miR-2112, 13, 14, 15 and miR-26,16, 17 are involved in the modulation of atrial fibrosis. In this study, we demonstrated that miR-23b-3p and miR-27b-3p were increased in the human rheumatic atrial appendages and in Ang-II-induced HAFs. Functionally, miR-23b-3p and miR-27b-3p could enhance the expression of fibrosis-related gene including COL1A1, COL3A1 and ACTA2, in HAFs.', 'The angiotensin system and TGF-beta1 play an important role in atrial fibrosis-associated with the heart failure models of AF,29, 30 a mouse model of AF30 and in human AF associated with heart failure.31 Ang-II can stimulate NADPH oxidases 2 (Nox2) activity through the TGF-beta1 pathway and inhibition of TGF-beta1 blunts reactive oxygen species (ROS) formation.32, 33 Therefore, renin-angiotensin-aldosterone system (RAAS)-induced TGF-beta1 pathway and TGF-beta1-induced ROS may contribute to fibrotic remodelling.', 'In this study, our data showed that TGFBR3 was significantly increased in Ang-II-induced HAFs, but no obvious increase of TGFBR3 was observed in the atrial appendages of AF patients. These results indicated that expression of TGFBR3 in atrial appendages was modulated by more factors, including miRNAs, cytokines and other inflammatory factors.', 'Smad3 signalling has been shown to participate in cardiac fibrosis.38, 39 Our present study has confirmed that the Smad3 signalling pathway was activated in Ang-II-treated HAFs. We used the Smad3 inhibitors, SIS3 and Naringenin and Smad3 siRNA to verify the participation of Smad3 activation in Ang-II-promoted up-regulation of miR-23b-3p and miR-27b-3p in HAFs. Therefore, we proposed a positive feedback loop that Smad3 activation modulates miR-23b-3p and miR-27b-3p expression, which then activate the Smad3 signalling by targeting TGFBR3 in HAFs.', 'Taken together, our results have demonstrated that miR-23b-3p and miR-27b-3p were up-regulated in the atrial appendages of patients with rheumatic AF and in Ang-II-treated HAFs. MiR-23b-3p and miR-27b-3p consistently enhanced fibrosis-related gene expression with no significant effects on proliferation and migration of HAFs. TGFBR3 was confirmed as a target of miR-23b-3p and miR-27b-3p and the TGFBR3/Smad1-Smad3 axis mediated the pro-fibrotic effect of miR-23b-3p and miR-27b-3p in HAFs.', 'We also demonstrated that the activation of the Smad3 signalling mediates the up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-induced HAFs. Therefore, the present study suggests that miR-23b-3p and miR-27b-3p might be potential targets for AF therapy.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 16 occurrence(s) (each occurrence is at most 3 sentences long) for gene 11606 : Ang-II.\n",
            "\n",
            "Gene(1277)\n",
            "  Symbol          : COL1A1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : collagen type I alpha 1 chain\n",
            "  Also Known As   : CAFYD, EDSARTH1, EDSC, OI1, OI2, OI3, OI4\n",
            "  In-text Name    : collagen, type I, alpha 1\n",
            "  Occurrences     : [\"We found that miR-23b-3p and miR-27b-3p were markedly increased in atrial appendage tissues of AF patients and in Ang-II-treated HAFs. Overexpression of miR-23b-3p and miR-27b-3p enhanced the expression of collagen, type I, alpha 1 (COL1A1), COL3A1 and ACTA2 in HAFs without significant effects on their proliferation and migration. Luciferase assay showed that miR-23b-3p and miR-27b-3p targeted two different sites in 3'-UTR of transforming growth factor (TGF)-beta1 receptor 3 (TGFBR3) respectively.\", 'Consistently, TGFBR3 siRNA could increase fibrosis-related genes expression, along with the Smad1 inactivation and Smad3 activation in HAFs. Additionally, overexpression of TGFBR3 could alleviate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively. Moreover, Smad3 was activated in HAFs in response to Ang-II treatment and inactivation of Smad3 attenuated up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs.', 'Results of Masson trichrome staining revealed that internal fibrosis was markedly increased in the atrial appendages of AF patients (Figure 1A). We detected mRNA expression of Col1a1, Col3a1 and FN1 in atrial appendages of patients with SR or AF. The RT-qPCR results showed that mRNA expression of the above fibrosis-related genes was significantly up-regulated in atrial appendages of AF patients (P < 0.05 respectively) (Figure 1B).', 'The immunofluorescent staining confirmed that HAFs were successfully isolated and cultured, with positive ACTA2 expression (Figure 2A). Fibrosis-associated genes including COL1A1, COL3A1 and ACTA2, were significantly increased in Ang-II-treated HAFs (Figure 2B). Meanwhile, RT-qPCR results showed that miR-23b/27b precursor, miR-23b-3p and miR-27b-3p were markedly up-regulated in HAFs exposed to Ang-II (P < 0.05 respectively) (Figure 2C,D).', 'To demonstrate the potential roles of miR-23b-3p and miR-27b-3p in atrial fibrosis, we determined fibrosis-related gene expression in HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively. Results of RT-qPCR and western blot assay showed that miR-23b-3p efficiently enhanced COL1A1, COL3A1 and ACTA2 expression in HAFs (Figure 4A,B). Consistently, miR-27b-3p also increased the above fibrosis-related gene expression at mRNA and protein level in HAFs (Figure 4C,D).', 'However, no significant difference in COL1A1, COL3A1 and ACTA2 expression was observed in HAFs with transfection of miR-23b-3p together with miR-27b-3p compared with HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively (Figure 4E).', 'MiR-23b-3p, miR-27b-3p mimic and TGFBR3 siRNA were transfected into HAFs followed by determination of the expression of fibrosis-related gene. Western blot results showed that miR-23b-3p, miR-27b-3p and TGFBR3 siRNA efficiently suppressed TGFBR3 expression and increased COL1A1, COL3A1 and ACTA2 expression in HAFs (P < 0.05, P < 0.01 respectively) (Figure 6A). Consistently, phosphorylated Smad1 (p-Smad1) was decreased, but p-Smad3 was increased in HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6A).', 'Meanwhile, Smad3 was significantly decreased in the cytoplasm and accumulated in the nuclei of HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6B,C). Western blot results showed that adenovirous-mediated overexpression of TGFBR3 could markedly attenuate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively (P < 0.01, P < 0.001 respectively) (Figure 6D). Moreover, TGFBR3 overexpression could significantly alleviate Smad3 activation in HAFs with transfection of miR-23b-3p and miR-27b-3p respectively (P < 0.05 respectively) (Figure 6D).', 'RT-qPCR results revealed that treatment with SIS, Narigenin and Smad3 siRNA could prevent Ang-II-induced miR-23b/27b precursor, miR-23b-3p and miR-27b-3p expression respectively (Figure 7B,C). Moreover, SIS3, Narigenin and Smad3 siRNA could significantly attenuate the increase of COL1A1 and COL3A1 in Ang-II-treated HAFs (Figure 7D). Collectively, our results suggest that up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs results from activation of the Smad3 pathway (Figure 7E).', 'Increasing evidence shows that interstitial atrial fibrosis is increased in some forms of AF in both humans and animal models.4, 25, 26, 27 In this study, we observed the obvious increase of fibrosis in atrial appendage tissue of patients with sinus rhythm, along with the significant elevation of COL1A1, COL3A1 and ACTA2. Meanwhile, the obvious increase of inflammatory factors, such as IL-1beta and CRP, was also found in atrial appendage tissue of AF patients in the current study, which was supported by the previous reports.28 Some miRNAs, such as miR-2112, 13, 14, 15 and miR-26,16, 17 are involved in the modulation of atrial fibrosis. In this study, we demonstrated that miR-23b-3p and miR-27b-3p were increased in the human rheumatic atrial appendages and in Ang-II-induced HAFs.', 'Functionally, miR-23b-3p and miR-27b-3p could enhance the expression of fibrosis-related gene including COL1A1, COL3A1 and ACTA2, in HAFs. These results were supported by previous reports that miR-23b and miR-27b enhance myocardial fibrosis.18, 19, 20 Additionally, no significant effects on cell proliferation and migration were observed in HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively.', 'We confirmed that miR-23b-3p and miR-27b-3p mimic inhibited TGFBR3 expression at both mRNA and protein levels in HAFs. In parallel with miR-23b-3p and miR-27b-3p, TGFBR3 knockdown enhanced the expressions of COL1A1, COL3A1 and ACTA2 in HAFs. Importantly, Smad1 was inactivated, whereas Smad3 was activated and accumulated in the nuclei of HAFs with transfection of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively.', 'These data indicated that Smad1 inactivation and Smad3 activation mediate the pro-fibrotic effects of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA in atrial fibrosis. Meanwhile, the activation of Smad1 through TGFBR3 and the negative modulation between p-Smad1 and p-Smad3 were in agreement with the previous report.37 Furthermore, TGFBR3 overexpression could significantly block the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p or miR-27b-3p. Therefore, our study demonstrated that the clustered miR-23b-3p and miR-27b-3p increase Smad3 activation to promote atrial fibrosis by targeting TGFBR3 in HAFs.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 13 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1277 : collagen, type I, alpha 1.\n",
            "\n",
            "Gene(1281)\n",
            "  Symbol          : COL3A1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : collagen type III alpha 1 chain\n",
            "  Also Known As   : EDS4A, EDSVASC, PMGEDSV\n",
            "  In-text Name    : COL3A1\n",
            "  Occurrences     : [\"We found that miR-23b-3p and miR-27b-3p were markedly increased in atrial appendage tissues of AF patients and in Ang-II-treated HAFs. Overexpression of miR-23b-3p and miR-27b-3p enhanced the expression of collagen, type I, alpha 1 (COL1A1), COL3A1 and ACTA2 in HAFs without significant effects on their proliferation and migration. Luciferase assay showed that miR-23b-3p and miR-27b-3p targeted two different sites in 3'-UTR of transforming growth factor (TGF)-beta1 receptor 3 (TGFBR3) respectively.\", 'Consistently, TGFBR3 siRNA could increase fibrosis-related genes expression, along with the Smad1 inactivation and Smad3 activation in HAFs. Additionally, overexpression of TGFBR3 could alleviate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively. Moreover, Smad3 was activated in HAFs in response to Ang-II treatment and inactivation of Smad3 attenuated up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs.', 'Results of Masson trichrome staining revealed that internal fibrosis was markedly increased in the atrial appendages of AF patients (Figure 1A). We detected mRNA expression of Col1a1, Col3a1 and FN1 in atrial appendages of patients with SR or AF. The RT-qPCR results showed that mRNA expression of the above fibrosis-related genes was significantly up-regulated in atrial appendages of AF patients (P < 0.05 respectively) (Figure 1B).', 'The immunofluorescent staining confirmed that HAFs were successfully isolated and cultured, with positive ACTA2 expression (Figure 2A). Fibrosis-associated genes including COL1A1, COL3A1 and ACTA2, were significantly increased in Ang-II-treated HAFs (Figure 2B). Meanwhile, RT-qPCR results showed that miR-23b/27b precursor, miR-23b-3p and miR-27b-3p were markedly up-regulated in HAFs exposed to Ang-II (P < 0.05 respectively) (Figure 2C,D).', 'To demonstrate the potential roles of miR-23b-3p and miR-27b-3p in atrial fibrosis, we determined fibrosis-related gene expression in HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively. Results of RT-qPCR and western blot assay showed that miR-23b-3p efficiently enhanced COL1A1, COL3A1 and ACTA2 expression in HAFs (Figure 4A,B). Consistently, miR-27b-3p also increased the above fibrosis-related gene expression at mRNA and protein level in HAFs (Figure 4C,D).', 'However, no significant difference in COL1A1, COL3A1 and ACTA2 expression was observed in HAFs with transfection of miR-23b-3p together with miR-27b-3p compared with HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively (Figure 4E).', 'MiR-23b-3p, miR-27b-3p mimic and TGFBR3 siRNA were transfected into HAFs followed by determination of the expression of fibrosis-related gene. Western blot results showed that miR-23b-3p, miR-27b-3p and TGFBR3 siRNA efficiently suppressed TGFBR3 expression and increased COL1A1, COL3A1 and ACTA2 expression in HAFs (P < 0.05, P < 0.01 respectively) (Figure 6A). Consistently, phosphorylated Smad1 (p-Smad1) was decreased, but p-Smad3 was increased in HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6A).', 'Meanwhile, Smad3 was significantly decreased in the cytoplasm and accumulated in the nuclei of HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6B,C). Western blot results showed that adenovirous-mediated overexpression of TGFBR3 could markedly attenuate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively (P < 0.01, P < 0.001 respectively) (Figure 6D). Moreover, TGFBR3 overexpression could significantly alleviate Smad3 activation in HAFs with transfection of miR-23b-3p and miR-27b-3p respectively (P < 0.05 respectively) (Figure 6D).', 'RT-qPCR results revealed that treatment with SIS, Narigenin and Smad3 siRNA could prevent Ang-II-induced miR-23b/27b precursor, miR-23b-3p and miR-27b-3p expression respectively (Figure 7B,C). Moreover, SIS3, Narigenin and Smad3 siRNA could significantly attenuate the increase of COL1A1 and COL3A1 in Ang-II-treated HAFs (Figure 7D). Collectively, our results suggest that up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs results from activation of the Smad3 pathway (Figure 7E).', 'Increasing evidence shows that interstitial atrial fibrosis is increased in some forms of AF in both humans and animal models.4, 25, 26, 27 In this study, we observed the obvious increase of fibrosis in atrial appendage tissue of patients with sinus rhythm, along with the significant elevation of COL1A1, COL3A1 and ACTA2. Meanwhile, the obvious increase of inflammatory factors, such as IL-1beta and CRP, was also found in atrial appendage tissue of AF patients in the current study, which was supported by the previous reports.28 Some miRNAs, such as miR-2112, 13, 14, 15 and miR-26,16, 17 are involved in the modulation of atrial fibrosis. In this study, we demonstrated that miR-23b-3p and miR-27b-3p were increased in the human rheumatic atrial appendages and in Ang-II-induced HAFs.', 'Functionally, miR-23b-3p and miR-27b-3p could enhance the expression of fibrosis-related gene including COL1A1, COL3A1 and ACTA2, in HAFs. These results were supported by previous reports that miR-23b and miR-27b enhance myocardial fibrosis.18, 19, 20 Additionally, no significant effects on cell proliferation and migration were observed in HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively.', 'We confirmed that miR-23b-3p and miR-27b-3p mimic inhibited TGFBR3 expression at both mRNA and protein levels in HAFs. In parallel with miR-23b-3p and miR-27b-3p, TGFBR3 knockdown enhanced the expressions of COL1A1, COL3A1 and ACTA2 in HAFs. Importantly, Smad1 was inactivated, whereas Smad3 was activated and accumulated in the nuclei of HAFs with transfection of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively.', 'These data indicated that Smad1 inactivation and Smad3 activation mediate the pro-fibrotic effects of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA in atrial fibrosis. Meanwhile, the activation of Smad1 through TGFBR3 and the negative modulation between p-Smad1 and p-Smad3 were in agreement with the previous report.37 Furthermore, TGFBR3 overexpression could significantly block the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p or miR-27b-3p. Therefore, our study demonstrated that the clustered miR-23b-3p and miR-27b-3p increase Smad3 activation to promote atrial fibrosis by targeting TGFBR3 in HAFs.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 13 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1281 : COL3A1.\n",
            "\n",
            "Gene(59)\n",
            "  Symbol          : ACTA2\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : actin alpha 2, smooth muscle\n",
            "  Also Known As   : ACTSA, SMDYS\n",
            "  In-text Name    : ACTA2\n",
            "  Occurrences     : [\"We found that miR-23b-3p and miR-27b-3p were markedly increased in atrial appendage tissues of AF patients and in Ang-II-treated HAFs. Overexpression of miR-23b-3p and miR-27b-3p enhanced the expression of collagen, type I, alpha 1 (COL1A1), COL3A1 and ACTA2 in HAFs without significant effects on their proliferation and migration. Luciferase assay showed that miR-23b-3p and miR-27b-3p targeted two different sites in 3'-UTR of transforming growth factor (TGF)-beta1 receptor 3 (TGFBR3) respectively.\", 'Consistently, TGFBR3 siRNA could increase fibrosis-related genes expression, along with the Smad1 inactivation and Smad3 activation in HAFs. Additionally, overexpression of TGFBR3 could alleviate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively. Moreover, Smad3 was activated in HAFs in response to Ang-II treatment and inactivation of Smad3 attenuated up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs.', 'The immunofluorescent staining confirmed that HAFs were successfully isolated and cultured, with positive ACTA2 expression (Figure 2A). Fibrosis-associated genes including COL1A1, COL3A1 and ACTA2, were significantly increased in Ang-II-treated HAFs (Figure 2B). Meanwhile, RT-qPCR results showed that miR-23b/27b precursor, miR-23b-3p and miR-27b-3p were markedly up-regulated in HAFs exposed to Ang-II (P < 0.05 respectively) (Figure 2C,D).', 'To demonstrate the potential roles of miR-23b-3p and miR-27b-3p in atrial fibrosis, we determined fibrosis-related gene expression in HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively. Results of RT-qPCR and western blot assay showed that miR-23b-3p efficiently enhanced COL1A1, COL3A1 and ACTA2 expression in HAFs (Figure 4A,B). Consistently, miR-27b-3p also increased the above fibrosis-related gene expression at mRNA and protein level in HAFs (Figure 4C,D).', 'However, no significant difference in COL1A1, COL3A1 and ACTA2 expression was observed in HAFs with transfection of miR-23b-3p together with miR-27b-3p compared with HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively (Figure 4E).', 'MiR-23b-3p, miR-27b-3p mimic and TGFBR3 siRNA were transfected into HAFs followed by determination of the expression of fibrosis-related gene. Western blot results showed that miR-23b-3p, miR-27b-3p and TGFBR3 siRNA efficiently suppressed TGFBR3 expression and increased COL1A1, COL3A1 and ACTA2 expression in HAFs (P < 0.05, P < 0.01 respectively) (Figure 6A). Consistently, phosphorylated Smad1 (p-Smad1) was decreased, but p-Smad3 was increased in HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6A).', 'Meanwhile, Smad3 was significantly decreased in the cytoplasm and accumulated in the nuclei of HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6B,C). Western blot results showed that adenovirous-mediated overexpression of TGFBR3 could markedly attenuate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively (P < 0.01, P < 0.001 respectively) (Figure 6D). Moreover, TGFBR3 overexpression could significantly alleviate Smad3 activation in HAFs with transfection of miR-23b-3p and miR-27b-3p respectively (P < 0.05 respectively) (Figure 6D).', 'Increasing evidence shows that interstitial atrial fibrosis is increased in some forms of AF in both humans and animal models.4, 25, 26, 27 In this study, we observed the obvious increase of fibrosis in atrial appendage tissue of patients with sinus rhythm, along with the significant elevation of COL1A1, COL3A1 and ACTA2. Meanwhile, the obvious increase of inflammatory factors, such as IL-1beta and CRP, was also found in atrial appendage tissue of AF patients in the current study, which was supported by the previous reports.28 Some miRNAs, such as miR-2112, 13, 14, 15 and miR-26,16, 17 are involved in the modulation of atrial fibrosis. In this study, we demonstrated that miR-23b-3p and miR-27b-3p were increased in the human rheumatic atrial appendages and in Ang-II-induced HAFs.', 'Functionally, miR-23b-3p and miR-27b-3p could enhance the expression of fibrosis-related gene including COL1A1, COL3A1 and ACTA2, in HAFs. These results were supported by previous reports that miR-23b and miR-27b enhance myocardial fibrosis.18, 19, 20 Additionally, no significant effects on cell proliferation and migration were observed in HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively.', 'We confirmed that miR-23b-3p and miR-27b-3p mimic inhibited TGFBR3 expression at both mRNA and protein levels in HAFs. In parallel with miR-23b-3p and miR-27b-3p, TGFBR3 knockdown enhanced the expressions of COL1A1, COL3A1 and ACTA2 in HAFs. Importantly, Smad1 was inactivated, whereas Smad3 was activated and accumulated in the nuclei of HAFs with transfection of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively.', 'These data indicated that Smad1 inactivation and Smad3 activation mediate the pro-fibrotic effects of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA in atrial fibrosis. Meanwhile, the activation of Smad1 through TGFBR3 and the negative modulation between p-Smad1 and p-Smad3 were in agreement with the previous report.37 Furthermore, TGFBR3 overexpression could significantly block the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p or miR-27b-3p. Therefore, our study demonstrated that the clustered miR-23b-3p and miR-27b-3p increase Smad3 activation to promote atrial fibrosis by targeting TGFBR3 in HAFs.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 11 occurrence(s) (each occurrence is at most 3 sentences long) for gene 59 : ACTA2.\n",
            "\n",
            "Gene(4086)\n",
            "  Symbol          : SMAD1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : SMAD family member 1\n",
            "  Also Known As   : BSP-1, BSP1, JV4-1, JV41, MADH1, MADR1\n",
            "  In-text Name    : Smad1\n",
            "  Occurrences     : [\"Luciferase assay showed that miR-23b-3p and miR-27b-3p targeted two different sites in 3'-UTR of transforming growth factor (TGF)-beta1 receptor 3 (TGFBR3) respectively. Consistently, TGFBR3 siRNA could increase fibrosis-related genes expression, along with the Smad1 inactivation and Smad3 activation in HAFs. Additionally, overexpression of TGFBR3 could alleviate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively.\", 'Western blot results showed that miR-23b-3p, miR-27b-3p and TGFBR3 siRNA efficiently suppressed TGFBR3 expression and increased COL1A1, COL3A1 and ACTA2 expression in HAFs (P < 0.05, P < 0.01 respectively) (Figure 6A). Consistently, phosphorylated Smad1 (p-Smad1) was decreased, but p-Smad3 was increased in HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6A). Meanwhile, Smad3 was significantly decreased in the cytoplasm and accumulated in the nuclei of HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6B,C).', 'As an inhibitory factor of TGF-beta1/Smad3 signalling, TGFBR3 was demonstrated to blunt Tgfbr1/Smad2/3 and potentiate the Acvrl1/Smad1 axis, which was active in lung fibroblasts and antagonized Tgfbr1/Smad2/3 signalling.34 A synthetic peptide from TGFBR3, P144, was reported to efficiently inhibit the TGF-beta1-dependent signalling pathway and collagen type I synthesis in cardiac fibroblasts.35 Further studies showed that overexpression of TGFBR3 deactivated the TGF-beta1 pathway by decreasing the expression of TGF-beta1, Smad3 activation, TGFBR1-TGFBR2 complex formation and inhibiting the pro-fibrotic microRNA-21 expression.24, 36 Therefore, TGFBR3 may act as a protective factor representing a potential therapeutic target for cardiac fibrosis.', 'In parallel with miR-23b-3p and miR-27b-3p, TGFBR3 knockdown enhanced the expressions of COL1A1, COL3A1 and ACTA2 in HAFs. Importantly, Smad1 was inactivated, whereas Smad3 was activated and accumulated in the nuclei of HAFs with transfection of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively. These data indicated that Smad1 inactivation and Smad3 activation mediate the pro-fibrotic effects of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA in atrial fibrosis.', 'Meanwhile, the activation of Smad1 through TGFBR3 and the negative modulation between p-Smad1 and p-Smad3 were in agreement with the previous report.37 Furthermore, TGFBR3 overexpression could significantly block the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p or miR-27b-3p. Therefore, our study demonstrated that the clustered miR-23b-3p and miR-27b-3p increase Smad3 activation to promote atrial fibrosis by targeting TGFBR3 in HAFs.', 'MiR-23b-3p and miR-27b-3p consistently enhanced fibrosis-related gene expression with no significant effects on proliferation and migration of HAFs. TGFBR3 was confirmed as a target of miR-23b-3p and miR-27b-3p and the TGFBR3/Smad1-Smad3 axis mediated the pro-fibrotic effect of miR-23b-3p and miR-27b-3p in HAFs. We also demonstrated that the activation of the Smad3 signalling mediates the up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-induced HAFs.', 'Glucocorticoids recruit Tgfbr3 and Smad1 to shift transforming growth factor-beta signaling from the Tgfbr1/Smad2/3 axis to the Acvrl1/Smad1 axis in lung fibroblasts']\n",
            "  Document ID     : PMC6484421\n",
            "There are 7 occurrence(s) (each occurrence is at most 3 sentences long) for gene 4086 : Smad1.\n",
            "\n",
            "Gene(4088)\n",
            "  Symbol          : SMAD3\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : SMAD family member 3\n",
            "  Also Known As   : HSPC193, HsT17436, JV15-2, LDS1C, LDS3, MADH3, hMAD-3, hSMAD3, mad3\n",
            "  In-text Name    : Smad3\n",
            "  Occurrences     : [\"Luciferase assay showed that miR-23b-3p and miR-27b-3p targeted two different sites in 3'-UTR of transforming growth factor (TGF)-beta1 receptor 3 (TGFBR3) respectively. Consistently, TGFBR3 siRNA could increase fibrosis-related genes expression, along with the Smad1 inactivation and Smad3 activation in HAFs. Additionally, overexpression of TGFBR3 could alleviate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively.\", 'Moreover, Smad3 was activated in HAFs in response to Ang-II treatment and inactivation of Smad3 attenuated up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs. Taken together, these results suggest that the clustered miR-23b-3p and miR-27b-3p consistently promote atrial fibrosis by targeting TGFBR3 to activate Smad3 signalling in HAFs, suggesting that miR-23b-3p and miR-27b-3p are potential therapeutic targets for atrial fibrosis.', 'Western blot results showed that miR-23b-3p, miR-27b-3p and TGFBR3 siRNA efficiently suppressed TGFBR3 expression and increased COL1A1, COL3A1 and ACTA2 expression in HAFs (P < 0.05, P < 0.01 respectively) (Figure 6A). Consistently, phosphorylated Smad1 (p-Smad1) was decreased, but p-Smad3 was increased in HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6A). Meanwhile, Smad3 was significantly decreased in the cytoplasm and accumulated in the nuclei of HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively (P < 0.05, P < 0.01 respectively) (Figure 6B,C).', 'Western blot results showed that adenovirous-mediated overexpression of TGFBR3 could markedly attenuate the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p and miR-27b-3p respectively (P < 0.01, P < 0.001 respectively) (Figure 6D). Moreover, TGFBR3 overexpression could significantly alleviate Smad3 activation in HAFs with transfection of miR-23b-3p and miR-27b-3p respectively (P < 0.05 respectively) (Figure 6D).', 'MiR-23b-3p and miR-27b-3p are up-regulated by Ang-II through the Smad3 pathway', 'We first examined the time-dependent activation of the Smad3 pathway in Ang-II-treated HAFs. Our western blot results showed that the phosphorylation level of Smad3 was significantly increased in HAFs at 10 minutes in response to Ang-II treatment (Figure 7A). Smad3 inhibitors SIS3, Narigenin and Smad3 siRNA were used to suppress Smad3 activation in Ang-II-treated HAFs.', 'RT-qPCR results revealed that treatment with SIS, Narigenin and Smad3 siRNA could prevent Ang-II-induced miR-23b/27b precursor, miR-23b-3p and miR-27b-3p expression respectively (Figure 7B,C). Moreover, SIS3, Narigenin and Smad3 siRNA could significantly attenuate the increase of COL1A1 and COL3A1 in Ang-II-treated HAFs (Figure 7D). Collectively, our results suggest that up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-treated HAFs results from activation of the Smad3 pathway (Figure 7E).', 'As an inhibitory factor of TGF-beta1/Smad3 signalling, TGFBR3 was demonstrated to blunt Tgfbr1/Smad2/3 and potentiate the Acvrl1/Smad1 axis, which was active in lung fibroblasts and antagonized Tgfbr1/Smad2/3 signalling.34 A synthetic peptide from TGFBR3, P144, was reported to efficiently inhibit the TGF-beta1-dependent signalling pathway and collagen type I synthesis in cardiac fibroblasts.35 Further studies showed that overexpression of TGFBR3 deactivated the TGF-beta1 pathway by decreasing the expression of TGF-beta1, Smad3 activation, TGFBR1-TGFBR2 complex formation and inhibiting the pro-fibrotic microRNA-21 expression.24, 36 Therefore, TGFBR3 may act as a protective factor representing a potential therapeutic target for cardiac fibrosis.', 'In parallel with miR-23b-3p and miR-27b-3p, TGFBR3 knockdown enhanced the expressions of COL1A1, COL3A1 and ACTA2 in HAFs. Importantly, Smad1 was inactivated, whereas Smad3 was activated and accumulated in the nuclei of HAFs with transfection of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA respectively. These data indicated that Smad1 inactivation and Smad3 activation mediate the pro-fibrotic effects of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA in atrial fibrosis.', 'Meanwhile, the activation of Smad1 through TGFBR3 and the negative modulation between p-Smad1 and p-Smad3 were in agreement with the previous report.37 Furthermore, TGFBR3 overexpression could significantly block the increase of COL1A1, COL3A1 and ACTA2 in HAFs after transfection with miR-23b-3p or miR-27b-3p. Therefore, our study demonstrated that the clustered miR-23b-3p and miR-27b-3p increase Smad3 activation to promote atrial fibrosis by targeting TGFBR3 in HAFs.', 'Smad3 signalling has been shown to participate in cardiac fibrosis.38, 39 Our present study has confirmed that the Smad3 signalling pathway was activated in Ang-II-treated HAFs. We used the Smad3 inhibitors, SIS3 and Naringenin and Smad3 siRNA to verify the participation of Smad3 activation in Ang-II-promoted up-regulation of miR-23b-3p and miR-27b-3p in HAFs. Therefore, we proposed a positive feedback loop that Smad3 activation modulates miR-23b-3p and miR-27b-3p expression, which then activate the Smad3 signalling by targeting TGFBR3 in HAFs.', 'MiR-23b-3p and miR-27b-3p consistently enhanced fibrosis-related gene expression with no significant effects on proliferation and migration of HAFs. TGFBR3 was confirmed as a target of miR-23b-3p and miR-27b-3p and the TGFBR3/Smad1-Smad3 axis mediated the pro-fibrotic effect of miR-23b-3p and miR-27b-3p in HAFs. We also demonstrated that the activation of the Smad3 signalling mediates the up-regulation of miR-23b-3p and miR-27b-3p in Ang-II-induced HAFs.', 'Therefore, the present study suggests that miR-23b-3p and miR-27b-3p might be potential targets for AF therapy.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 13 occurrence(s) (each occurrence is at most 3 sentences long) for gene 4088 : Smad3.\n",
            "\n",
            "Gene(406991)\n",
            "  Symbol          : MIR21\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : microRNA 21\n",
            "  Also Known As   : MIRN21, hsa-mir-21, miR-21, miRNA21\n",
            "  In-text Name    : miR-21\n",
            "  Occurrences     : ['Non-coding RNAs (ncRNAs) comprise over 95% of the total RNA transcripts in eukaryotic cells.7 Recently, more than 5000 miRNAs have been discovered in the human genome.8 MicroRNA (miRNA) is an endogenous non-coding RNA with about 22 nucleotides and is involved in a variety of biological processes and diseases by inhibiting the protein translation process of the target gene.9, 10, 11 It has been found that miRNA can also participate in the development of atrial fibrillation through fibrosis-related target genes. For example, miR-21 is markedly up-regulated in AF models and involved in multiple pro-fibrotic pathways by targeting Spry1,12 PIAS3,13 Smad714 and Jagged115 respectively. However, miR-26 down-regulation in AF enhances atrial fibrosis by targeting TRPC3 to increase Ca2+ entry and activate proliferation and differentiation of fibroblasts16 and also by targeting KCNJ2 to enhance I K1 and atrial fibroblast proliferation.17 ', 'As an inhibitory factor of TGF-beta1/Smad3 signalling, TGFBR3 was demonstrated to blunt Tgfbr1/Smad2/3 and potentiate the Acvrl1/Smad1 axis, which was active in lung fibroblasts and antagonized Tgfbr1/Smad2/3 signalling.34 A synthetic peptide from TGFBR3, P144, was reported to efficiently inhibit the TGF-beta1-dependent signalling pathway and collagen type I synthesis in cardiac fibroblasts.35 Further studies showed that overexpression of TGFBR3 deactivated the TGF-beta1 pathway by decreasing the expression of TGF-beta1, Smad3 activation, TGFBR1-TGFBR2 complex formation and inhibiting the pro-fibrotic microRNA-21 expression.24, 36 Therefore, TGFBR3 may act as a protective factor representing a potential therapeutic target for cardiac fibrosis.', 'Role for MicroRNA-21 in atrial profibrillatory fibrotic remodeling associated with experimental postinfarction heart failure', 'Signal transducer and activator of transcription 3/microRNA-21 feedback loop contributes to atrial fibrillation by promoting atrial fibrosis in a rat sterile pericarditis model', 'Rapid atrial pacing induces myocardial fibrosis by down-regulating Smad7 via microRNA-21 in rabbit', 'miR-21 promotes cardiac fibroblast-to-myofibroblast transformation and myocardial fibrosis by targeting Jagged1', 'A novel reciprocal loop between microRNA-21 and TGFbetaRIII is involved in cardiac fibrosis']\n",
            "  Document ID     : PMC6484421\n",
            "There are 7 occurrence(s) (each occurrence is at most 3 sentences long) for gene 406991 : miR-21.\n",
            "\n",
            "Gene(7222)\n",
            "  Symbol          : TRPC3\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : transient receptor potential cation channel subfamily C member 3\n",
            "  Also Known As   : SCA41, TRP3\n",
            "  In-text Name    : TRPC3\n",
            "  Occurrences     : ['For example, miR-21 is markedly up-regulated in AF models and involved in multiple pro-fibrotic pathways by targeting Spry1,12 PIAS3,13 Smad714 and Jagged115 respectively. However, miR-26 down-regulation in AF enhances atrial fibrosis by targeting TRPC3 to increase Ca2+ entry and activate proliferation and differentiation of fibroblasts16 and also by targeting KCNJ2 to enhance I K1 and atrial fibroblast proliferation.17 ']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 7222 : TRPC3.\n",
            "\n",
            "Gene(3759)\n",
            "  Symbol          : KCNJ2\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : potassium inwardly rectifying channel subfamily J member 2\n",
            "  Also Known As   : ATFB9, HHBIRK1, HHIRK1, IRK1, KIR2.1, LQT7, SQT3\n",
            "  In-text Name    : KCNJ2\n",
            "  Occurrences     : ['For example, miR-21 is markedly up-regulated in AF models and involved in multiple pro-fibrotic pathways by targeting Spry1,12 PIAS3,13 Smad714 and Jagged115 respectively. However, miR-26 down-regulation in AF enhances atrial fibrosis by targeting TRPC3 to increase Ca2+ entry and activate proliferation and differentiation of fibroblasts16 and also by targeting KCNJ2 to enhance I K1 and atrial fibroblast proliferation.17 ']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 3759 : KCNJ2.\n",
            "\n",
            "Gene(10320)\n",
            "  Symbol          : IKZF1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : IKAROS family zinc finger 1\n",
            "  Also Known As   : CVID13, Hs.54452, IK1, IKAROS, LYF1, LyF-1, PPP1R92, PRO0758, ZNFN1A1\n",
            "  In-text Name    : I K1\n",
            "  Occurrences     : ['For example, miR-21 is markedly up-regulated in AF models and involved in multiple pro-fibrotic pathways by targeting Spry1,12 PIAS3,13 Smad714 and Jagged115 respectively. However, miR-26 down-regulation in AF enhances atrial fibrosis by targeting TRPC3 to increase Ca2+ entry and activate proliferation and differentiation of fibroblasts16 and also by targeting KCNJ2 to enhance I K1 and atrial fibroblast proliferation.17 ']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 10320 : I K1.\n",
            "\n",
            "Gene(407011)\n",
            "  Symbol          : MIR23B\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : microRNA 23b\n",
            "  Also Known As   : MIRN23B, hsa-mir-23b, miRNA23B, mir-23b\n",
            "  In-text Name    : miR-23b\n",
            "  Occurrences     : ['MiR-23b-3p and miR-27b-3p are located within a single transcription unit from a microRNA cluster on human chromosome 9 (www.mirdb.org). Inhibition of miR-23b could alleviate cardiac fibrosis in caecal ligation puncture (CLP)-induced late sepsis by targeting TGIF1 and PTEN to activate TGF-beta1/Smad2/3 signalling and AKT/N-Cadherin signalling in cardiac fibroblasts respectively.18 The effects of miR-27b on cardiac fibrosis are controversial. Overexpression of miR-27b in cardiomyocytes was reported to contribute to cardiac dysfunction with enhanced myocardial fibrosis by targeting PPAR-gamma19 and MMP1320 respectively.', 'Functionally, miR-23b-3p and miR-27b-3p could enhance the expression of fibrosis-related gene including COL1A1, COL3A1 and ACTA2, in HAFs. These results were supported by previous reports that miR-23b and miR-27b enhance myocardial fibrosis.18, 19, 20 Additionally, no significant effects on cell proliferation and migration were observed in HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 407011 : miR-23b.\n",
            "\n",
            "Gene(7050)\n",
            "  Symbol          : TGIF1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : TGFB induced factor homeobox 1\n",
            "  Also Known As   : HPE4, TGIF\n",
            "  In-text Name    : TGIF1\n",
            "  Occurrences     : ['MiR-23b-3p and miR-27b-3p are located within a single transcription unit from a microRNA cluster on human chromosome 9 (www.mirdb.org). Inhibition of miR-23b could alleviate cardiac fibrosis in caecal ligation puncture (CLP)-induced late sepsis by targeting TGIF1 and PTEN to activate TGF-beta1/Smad2/3 signalling and AKT/N-Cadherin signalling in cardiac fibroblasts respectively.18 The effects of miR-27b on cardiac fibrosis are controversial. Overexpression of miR-27b in cardiomyocytes was reported to contribute to cardiac dysfunction with enhanced myocardial fibrosis by targeting PPAR-gamma19 and MMP1320 respectively.', 'Inhibition of microRNA-23b prevents polymicrobial sepsis-induced cardiac dysfunction by modulating TGIF1 and PTEN']\n",
            "  Document ID     : PMC6484421\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 7050 : TGIF1.\n",
            "\n",
            "Gene(5728)\n",
            "  Symbol          : PTEN\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : phosphatase and tensin homolog\n",
            "  Also Known As   : 10q23del, BZS, CWS1, DEC, GLM2, MHAM, MMAC1, PTEN1, PTENbeta, PTENgama, TEP1\n",
            "  In-text Name    : PTEN\n",
            "  Occurrences     : ['MiR-23b-3p and miR-27b-3p are located within a single transcription unit from a microRNA cluster on human chromosome 9 (www.mirdb.org). Inhibition of miR-23b could alleviate cardiac fibrosis in caecal ligation puncture (CLP)-induced late sepsis by targeting TGIF1 and PTEN to activate TGF-beta1/Smad2/3 signalling and AKT/N-Cadherin signalling in cardiac fibroblasts respectively.18 The effects of miR-27b on cardiac fibrosis are controversial. Overexpression of miR-27b in cardiomyocytes was reported to contribute to cardiac dysfunction with enhanced myocardial fibrosis by targeting PPAR-gamma19 and MMP1320 respectively.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 5728 : PTEN.\n",
            "\n",
            "Gene(7040)\n",
            "  Symbol          : TGFB1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : transforming growth factor beta 1\n",
            "  Also Known As   : CED, DPD1, IBDIMDE, LAP, TGF-beta1, TGFB, TGFbeta\n",
            "  In-text Name    : TGF-beta1\n",
            "  Occurrences     : ['MiR-23b-3p and miR-27b-3p are located within a single transcription unit from a microRNA cluster on human chromosome 9 (www.mirdb.org). Inhibition of miR-23b could alleviate cardiac fibrosis in caecal ligation puncture (CLP)-induced late sepsis by targeting TGIF1 and PTEN to activate TGF-beta1/Smad2/3 signalling and AKT/N-Cadherin signalling in cardiac fibroblasts respectively.18 The effects of miR-27b on cardiac fibrosis are controversial. Overexpression of miR-27b in cardiomyocytes was reported to contribute to cardiac dysfunction with enhanced myocardial fibrosis by targeting PPAR-gamma19 and MMP1320 respectively.', 'In light of its essential role in the pathogenesis of fibrosis, TGF-beta has emerged as an attractive therapeutic target.22 TGF-beta1 receptor 3(TGFBR3) is negatively regulated by TGF-beta1 in multiple cell types and mediates TGF-beta superfamily ligand-dependent as well as ligand-independent signalling to both Smad and non-Smad signalling pathways.23 A recent study showed that TGFBR3 inhibited TGF-beta1 expression, Smad2/3 activation and TGFBR1-TGFBR2 complex formation and collagen production in cardiac fibroblasts.24 However, it still not known whether TGFBR3 can be modulated by miR-23b-3p and miR-27b-3p in atrial fibroblasts.', 'The angiotensin system and TGF-beta1 play an important role in atrial fibrosis-associated with the heart failure models of AF,29, 30 a mouse model of AF30 and in human AF associated with heart failure.31 Ang-II can stimulate NADPH oxidases 2 (Nox2) activity through the TGF-beta1 pathway and inhibition of TGF-beta1 blunts reactive oxygen species (ROS) formation.32, 33 Therefore, renin-angiotensin-aldosterone system (RAAS)-induced TGF-beta1 pathway and TGF-beta1-induced ROS may contribute to fibrotic remodelling.', 'As an inhibitory factor of TGF-beta1/Smad3 signalling, TGFBR3 was demonstrated to blunt Tgfbr1/Smad2/3 and potentiate the Acvrl1/Smad1 axis, which was active in lung fibroblasts and antagonized Tgfbr1/Smad2/3 signalling.34 A synthetic peptide from TGFBR3, P144, was reported to efficiently inhibit the TGF-beta1-dependent signalling pathway and collagen type I synthesis in cardiac fibroblasts.35 Further studies showed that overexpression of TGFBR3 deactivated the TGF-beta1 pathway by decreasing the expression of TGF-beta1, Smad3 activation, TGFBR1-TGFBR2 complex formation and inhibiting the pro-fibrotic microRNA-21 expression.24, 36 Therefore, TGFBR3 may act as a protective factor representing a potential therapeutic target for cardiac fibrosis.', 'TGF-beta signaling in fibrosis', 'Roles for the type III TGF-beta receptor in human cancer', 'TGFBR3, a potential negative regulator of TGF-beta signaling, protects cardiac fibroblasts from hypoxia-induced apoptosis', 'Bone morphogenetic protein-7 antagonizes myocardial fibrosis induced by atrial fibrillation by restraining transforming growth factor-beta (TGF-beta)/Smads signaling', 'New insights into TGF-beta/Smad signaling in tissue fibrosis']\n",
            "  Document ID     : PMC6484421\n",
            "There are 9 occurrence(s) (each occurrence is at most 3 sentences long) for gene 7040 : TGF-beta1.\n",
            "\n",
            "Gene(207)\n",
            "  Symbol          : AKT1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : AKT serine/threonine kinase 1\n",
            "  Also Known As   : AKT, PKB, PKB-ALPHA, PRKBA, RAC, RAC-ALPHA\n",
            "  In-text Name    : AKT\n",
            "  Occurrences     : ['MiR-23b-3p and miR-27b-3p are located within a single transcription unit from a microRNA cluster on human chromosome 9 (www.mirdb.org). Inhibition of miR-23b could alleviate cardiac fibrosis in caecal ligation puncture (CLP)-induced late sepsis by targeting TGIF1 and PTEN to activate TGF-beta1/Smad2/3 signalling and AKT/N-Cadherin signalling in cardiac fibroblasts respectively.18 The effects of miR-27b on cardiac fibrosis are controversial. Overexpression of miR-27b in cardiomyocytes was reported to contribute to cardiac dysfunction with enhanced myocardial fibrosis by targeting PPAR-gamma19 and MMP1320 respectively.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 207 : AKT.\n",
            "\n",
            "Gene(1000)\n",
            "  Symbol          : CDH2\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : cadherin 2\n",
            "  Also Known As   : ACOGS, ADHD8, ARVD14, CD325, CDHN, CDw325, NCAD\n",
            "  In-text Name    : N-Cadherin\n",
            "  Occurrences     : ['MiR-23b-3p and miR-27b-3p are located within a single transcription unit from a microRNA cluster on human chromosome 9 (www.mirdb.org). Inhibition of miR-23b could alleviate cardiac fibrosis in caecal ligation puncture (CLP)-induced late sepsis by targeting TGIF1 and PTEN to activate TGF-beta1/Smad2/3 signalling and AKT/N-Cadherin signalling in cardiac fibroblasts respectively.18 The effects of miR-27b on cardiac fibrosis are controversial. Overexpression of miR-27b in cardiomyocytes was reported to contribute to cardiac dysfunction with enhanced myocardial fibrosis by targeting PPAR-gamma19 and MMP1320 respectively.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1000 : N-Cadherin.\n",
            "\n",
            "Gene(407019)\n",
            "  Symbol          : MIR27B\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : microRNA 27b\n",
            "  Also Known As   : MIR-27b, MIRN27B, miRNA27B\n",
            "  In-text Name    : miR-27b\n",
            "  Occurrences     : ['MiR-23b-3p and miR-27b-3p are located within a single transcription unit from a microRNA cluster on human chromosome 9 (www.mirdb.org). Inhibition of miR-23b could alleviate cardiac fibrosis in caecal ligation puncture (CLP)-induced late sepsis by targeting TGIF1 and PTEN to activate TGF-beta1/Smad2/3 signalling and AKT/N-Cadherin signalling in cardiac fibroblasts respectively.18 The effects of miR-27b on cardiac fibrosis are controversial. Overexpression of miR-27b in cardiomyocytes was reported to contribute to cardiac dysfunction with enhanced myocardial fibrosis by targeting PPAR-gamma19 and MMP1320 respectively.', 'However, a recent study showed that atrial overexpression of miR-27b attenuated angiotensin II-induced atrial fibrosis by targeting ALK5.21 To date, the role and mechanism of miR-23b-3p and miR-27b-3p in atrial fibrosis are not well understood.', 'In light of its essential role in the pathogenesis of fibrosis, TGF-beta has emerged as an attractive therapeutic target.22 TGF-beta1 receptor 3(TGFBR3) is negatively regulated by TGF-beta1 in multiple cell types and mediates TGF-beta superfamily ligand-dependent as well as ligand-independent signalling to both Smad and non-Smad signalling pathways.23 A recent study showed that TGFBR3 inhibited TGF-beta1 expression, Smad2/3 activation and TGFBR1-TGFBR2 complex formation and collagen production in cardiac fibroblasts.24 However, it still not known whether TGFBR3 can be modulated by miR-23b-3p and miR-27b-3p in atrial fibroblasts.', 'Fibrosis-associated genes including COL1A1, COL3A1 and ACTA2, were significantly increased in Ang-II-treated HAFs (Figure 2B). Meanwhile, RT-qPCR results showed that miR-23b/27b precursor, miR-23b-3p and miR-27b-3p were markedly up-regulated in HAFs exposed to Ang-II (P < 0.05 respectively) (Figure 2C,D).', 'Functionally, miR-23b-3p and miR-27b-3p could enhance the expression of fibrosis-related gene including COL1A1, COL3A1 and ACTA2, in HAFs. These results were supported by previous reports that miR-23b and miR-27b enhance myocardial fibrosis.18, 19, 20 Additionally, no significant effects on cell proliferation and migration were observed in HAFs with overexpression of miR-23b-3p and miR-27b-3p respectively.', 'Cardiomyocyte overexpression of miR-27b induces cardiac hypertrophy and dysfunction in mice', 'Cardiomycyte overexpression of miR-27b resulted in cardiac fibrosis and mitochondria injury in mice']\n",
            "  Document ID     : PMC6484421\n",
            "There are 7 occurrence(s) (each occurrence is at most 3 sentences long) for gene 407019 : miR-27b.\n",
            "\n",
            "Gene(183)\n",
            "  Symbol          : AGT\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : angiotensinogen\n",
            "  Also Known As   : ANHU, SERPINA8, hFLT1\n",
            "  In-text Name    : angiotensin II\n",
            "  Occurrences     : ['Overexpression of miR-27b in cardiomyocytes was reported to contribute to cardiac dysfunction with enhanced myocardial fibrosis by targeting PPAR-gamma19 and MMP1320 respectively. However, a recent study showed that atrial overexpression of miR-27b attenuated angiotensin II-induced atrial fibrosis by targeting ALK5.21 To date, the role and mechanism of miR-23b-3p and miR-27b-3p in atrial fibrosis are not well understood.', 'Atrial overexpression of microRNA-27b attenuates angiotensin II-induced atrial fibrosis and fibrillation by targeting ALK5']\n",
            "  Document ID     : PMC6484421\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 183 : angiotensin II.\n",
            "\n",
            "Gene(7046)\n",
            "  Symbol          : TGFBR1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : transforming growth factor beta receptor 1\n",
            "  Also Known As   : AAT5, ACVRLK4, ALK-5, ALK5, ESS1, LDS1, LDS1A, LDS2A, MSSE, SKR4, TBR-i, TBRI, TGFR-1, tbetaR-I\n",
            "  In-text Name    : TGFBR1\n",
            "  Occurrences     : ['In light of its essential role in the pathogenesis of fibrosis, TGF-beta has emerged as an attractive therapeutic target.22 TGF-beta1 receptor 3(TGFBR3) is negatively regulated by TGF-beta1 in multiple cell types and mediates TGF-beta superfamily ligand-dependent as well as ligand-independent signalling to both Smad and non-Smad signalling pathways.23 A recent study showed that TGFBR3 inhibited TGF-beta1 expression, Smad2/3 activation and TGFBR1-TGFBR2 complex formation and collagen production in cardiac fibroblasts.24 However, it still not known whether TGFBR3 can be modulated by miR-23b-3p and miR-27b-3p in atrial fibroblasts.', 'As an inhibitory factor of TGF-beta1/Smad3 signalling, TGFBR3 was demonstrated to blunt Tgfbr1/Smad2/3 and potentiate the Acvrl1/Smad1 axis, which was active in lung fibroblasts and antagonized Tgfbr1/Smad2/3 signalling.34 A synthetic peptide from TGFBR3, P144, was reported to efficiently inhibit the TGF-beta1-dependent signalling pathway and collagen type I synthesis in cardiac fibroblasts.35 Further studies showed that overexpression of TGFBR3 deactivated the TGF-beta1 pathway by decreasing the expression of TGF-beta1, Smad3 activation, TGFBR1-TGFBR2 complex formation and inhibiting the pro-fibrotic microRNA-21 expression.24, 36 Therefore, TGFBR3 may act as a protective factor representing a potential therapeutic target for cardiac fibrosis.', 'Glucocorticoids recruit Tgfbr3 and Smad1 to shift transforming growth factor-beta signaling from the Tgfbr1/Smad2/3 axis to the Acvrl1/Smad1 axis in lung fibroblasts']\n",
            "  Document ID     : PMC6484421\n",
            "There are 3 occurrence(s) (each occurrence is at most 3 sentences long) for gene 7046 : TGFBR1.\n",
            "\n",
            "Gene(7048)\n",
            "  Symbol          : TGFBR2\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : transforming growth factor beta receptor 2\n",
            "  Also Known As   : AAT3, FAA3, LDS1B, LDS2, LDS2B, MFS2, RIIC, TAAD2, TBR-ii, TBRII, TGFR-2, TGFbeta-RII, tbetaR-II\n",
            "  In-text Name    : TGFBR2\n",
            "  Occurrences     : ['In light of its essential role in the pathogenesis of fibrosis, TGF-beta has emerged as an attractive therapeutic target.22 TGF-beta1 receptor 3(TGFBR3) is negatively regulated by TGF-beta1 in multiple cell types and mediates TGF-beta superfamily ligand-dependent as well as ligand-independent signalling to both Smad and non-Smad signalling pathways.23 A recent study showed that TGFBR3 inhibited TGF-beta1 expression, Smad2/3 activation and TGFBR1-TGFBR2 complex formation and collagen production in cardiac fibroblasts.24 However, it still not known whether TGFBR3 can be modulated by miR-23b-3p and miR-27b-3p in atrial fibroblasts.', 'As an inhibitory factor of TGF-beta1/Smad3 signalling, TGFBR3 was demonstrated to blunt Tgfbr1/Smad2/3 and potentiate the Acvrl1/Smad1 axis, which was active in lung fibroblasts and antagonized Tgfbr1/Smad2/3 signalling.34 A synthetic peptide from TGFBR3, P144, was reported to efficiently inhibit the TGF-beta1-dependent signalling pathway and collagen type I synthesis in cardiac fibroblasts.35 Further studies showed that overexpression of TGFBR3 deactivated the TGF-beta1 pathway by decreasing the expression of TGF-beta1, Smad3 activation, TGFBR1-TGFBR2 complex formation and inhibiting the pro-fibrotic microRNA-21 expression.24, 36 Therefore, TGFBR3 may act as a protective factor representing a potential therapeutic target for cardiac fibrosis.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 7048 : TGFBR2.\n",
            "\n",
            "Gene(2335)\n",
            "  Symbol          : FN1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : fibronectin 1\n",
            "  Also Known As   : CIG, ED-B, FINC, FN, FNZ, GFND, GFND2, LETS, MSF, SMDCF\n",
            "  In-text Name    : FN1\n",
            "  Occurrences     : ['Results of Masson trichrome staining revealed that internal fibrosis was markedly increased in the atrial appendages of AF patients (Figure 1A). We detected mRNA expression of Col1a1, Col3a1 and FN1 in atrial appendages of patients with SR or AF. The RT-qPCR results showed that mRNA expression of the above fibrosis-related genes was significantly up-regulated in atrial appendages of AF patients (P < 0.05 respectively) (Figure 1B).']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 2335 : FN1.\n",
            "\n",
            "Gene(3553)\n",
            "  Symbol          : IL1B\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : interleukin 1 beta\n",
            "  Also Known As   : IL-1, IL1-BETA, IL1F2, IL1beta\n",
            "  In-text Name    : IL-1beta\n",
            "  Occurrences     : ['The RT-qPCR results showed that mRNA expression of the above fibrosis-related genes was significantly up-regulated in atrial appendages of AF patients (P < 0.05 respectively) (Figure 1B). Moreover, the inflammation-related genes including IL-1beta and CRP, but not TNF-alpha were significantly up-regulated in atrial appendages of AF patients (Figure S1). However, no significant changes of RAAS-associated genes, such as ACE1, ACE2 and Apelin were observed in the atrial appendages of AF patients (Figure S2).', 'Increasing evidence shows that interstitial atrial fibrosis is increased in some forms of AF in both humans and animal models.4, 25, 26, 27 In this study, we observed the obvious increase of fibrosis in atrial appendage tissue of patients with sinus rhythm, along with the significant elevation of COL1A1, COL3A1 and ACTA2. Meanwhile, the obvious increase of inflammatory factors, such as IL-1beta and CRP, was also found in atrial appendage tissue of AF patients in the current study, which was supported by the previous reports.28 Some miRNAs, such as miR-2112, 13, 14, 15 and miR-26,16, 17 are involved in the modulation of atrial fibrosis. In this study, we demonstrated that miR-23b-3p and miR-27b-3p were increased in the human rheumatic atrial appendages and in Ang-II-induced HAFs.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 3553 : IL-1beta.\n",
            "\n",
            "Gene(1401)\n",
            "  Symbol          : CRP\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : C-reactive protein\n",
            "  Also Known As   : PTX1\n",
            "  In-text Name    : CRP\n",
            "  Occurrences     : ['The RT-qPCR results showed that mRNA expression of the above fibrosis-related genes was significantly up-regulated in atrial appendages of AF patients (P < 0.05 respectively) (Figure 1B). Moreover, the inflammation-related genes including IL-1beta and CRP, but not TNF-alpha were significantly up-regulated in atrial appendages of AF patients (Figure S1). However, no significant changes of RAAS-associated genes, such as ACE1, ACE2 and Apelin were observed in the atrial appendages of AF patients (Figure S2).', 'Increasing evidence shows that interstitial atrial fibrosis is increased in some forms of AF in both humans and animal models.4, 25, 26, 27 In this study, we observed the obvious increase of fibrosis in atrial appendage tissue of patients with sinus rhythm, along with the significant elevation of COL1A1, COL3A1 and ACTA2. Meanwhile, the obvious increase of inflammatory factors, such as IL-1beta and CRP, was also found in atrial appendage tissue of AF patients in the current study, which was supported by the previous reports.28 Some miRNAs, such as miR-2112, 13, 14, 15 and miR-26,16, 17 are involved in the modulation of atrial fibrosis. In this study, we demonstrated that miR-23b-3p and miR-27b-3p were increased in the human rheumatic atrial appendages and in Ang-II-induced HAFs.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1401 : CRP.\n",
            "\n",
            "Gene(7124)\n",
            "  Symbol          : TNF\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : tumor necrosis factor\n",
            "  Also Known As   : DIF, IMD127, TNF-alpha, TNFA, TNFSF2, TNLG1F\n",
            "  In-text Name    : TNF-alpha\n",
            "  Occurrences     : ['The RT-qPCR results showed that mRNA expression of the above fibrosis-related genes was significantly up-regulated in atrial appendages of AF patients (P < 0.05 respectively) (Figure 1B). Moreover, the inflammation-related genes including IL-1beta and CRP, but not TNF-alpha were significantly up-regulated in atrial appendages of AF patients (Figure S1). However, no significant changes of RAAS-associated genes, such as ACE1, ACE2 and Apelin were observed in the atrial appendages of AF patients (Figure S2).']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 7124 : TNF-alpha.\n",
            "\n",
            "Gene(1636)\n",
            "  Symbol          : ACE\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : angiotensin I converting enzyme\n",
            "  Also Known As   : ACE1, CD143, DCP, DCP1\n",
            "  In-text Name    : ACE1\n",
            "  Occurrences     : ['Moreover, the inflammation-related genes including IL-1beta and CRP, but not TNF-alpha were significantly up-regulated in atrial appendages of AF patients (Figure S1). However, no significant changes of RAAS-associated genes, such as ACE1, ACE2 and Apelin were observed in the atrial appendages of AF patients (Figure S2). The expression levels of miR-23b/27b precursor, miR-23b-3p and miR-27b-3p were up-regulated in the atrial appendages of AF patients (Figure 1C), consistent with miRNA profiling data of atrial appendages from SR and AF patients (data not shown).']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 1636 : ACE1.\n",
            "\n",
            "Gene(59272)\n",
            "  Symbol          : ACE2\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : angiotensin converting enzyme 2\n",
            "  Also Known As   : ACEH\n",
            "  In-text Name    : ACE2\n",
            "  Occurrences     : ['Moreover, the inflammation-related genes including IL-1beta and CRP, but not TNF-alpha were significantly up-regulated in atrial appendages of AF patients (Figure S1). However, no significant changes of RAAS-associated genes, such as ACE1, ACE2 and Apelin were observed in the atrial appendages of AF patients (Figure S2). The expression levels of miR-23b/27b precursor, miR-23b-3p and miR-27b-3p were up-regulated in the atrial appendages of AF patients (Figure 1C), consistent with miRNA profiling data of atrial appendages from SR and AF patients (data not shown).']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 59272 : ACE2.\n",
            "\n",
            "Gene(8862)\n",
            "  Symbol          : APLN\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : apelin\n",
            "  Also Known As   : APEL, XNPEP2\n",
            "  In-text Name    : Apelin\n",
            "  Occurrences     : ['Moreover, the inflammation-related genes including IL-1beta and CRP, but not TNF-alpha were significantly up-regulated in atrial appendages of AF patients (Figure S1). However, no significant changes of RAAS-associated genes, such as ACE1, ACE2 and Apelin were observed in the atrial appendages of AF patients (Figure S2). The expression levels of miR-23b/27b precursor, miR-23b-3p and miR-27b-3p were up-regulated in the atrial appendages of AF patients (Figure 1C), consistent with miRNA profiling data of atrial appendages from SR and AF patients (data not shown).']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 8862 : Apelin.\n",
            "\n",
            "Gene(21803)\n",
            "  Symbol          : Tgfb1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : transforming growth factor, beta 1\n",
            "  Also Known As   : TGF-beta1, TGFbeta1, Tgfb, Tgfb-1\n",
            "  In-text Name    : TGF-beta1\n",
            "  Occurrences     : ['The angiotensin system and TGF-beta1 play an important role in atrial fibrosis-associated with the heart failure models of AF,29, 30 a mouse model of AF30 and in human AF associated with heart failure.31 Ang-II can stimulate NADPH oxidases 2 (Nox2) activity through the TGF-beta1 pathway and inhibition of TGF-beta1 blunts reactive oxygen species (ROS) formation.32, 33 Therefore, renin-angiotensin-aldosterone system (RAAS)-induced TGF-beta1 pathway and TGF-beta1-induced ROS may contribute to fibrotic remodelling.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 21803 : TGF-beta1.\n",
            "\n",
            "Gene(13058)\n",
            "  Symbol          : Cybb\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : cytochrome b-245, beta polypeptide\n",
            "  Also Known As   : CGD91-phox, Cgd, Cyd, Nox2, gp91-1, gp91phox\n",
            "  In-text Name    : NADPH oxidases 2\n",
            "  Occurrences     : ['The angiotensin system and TGF-beta1 play an important role in atrial fibrosis-associated with the heart failure models of AF,29, 30 a mouse model of AF30 and in human AF associated with heart failure.31 Ang-II can stimulate NADPH oxidases 2 (Nox2) activity through the TGF-beta1 pathway and inhibition of TGF-beta1 blunts reactive oxygen species (ROS) formation.32, 33 Therefore, renin-angiotensin-aldosterone system (RAAS)-induced TGF-beta1 pathway and TGF-beta1-induced ROS may contribute to fibrotic remodelling.']\n",
            "  Document ID     : PMC6484421\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 13058 : NADPH oxidases 2.\n",
            "\n",
            "Gene(94)\n",
            "  Symbol          : ACVRL1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : activin A receptor like type 1\n",
            "  Also Known As   : ACVRLK1, ALK-1, ALK1, HHT, HHT2, ORW2, SKR3, TSR-I\n",
            "  In-text Name    : Acvrl1\n",
            "  Occurrences     : ['As an inhibitory factor of TGF-beta1/Smad3 signalling, TGFBR3 was demonstrated to blunt Tgfbr1/Smad2/3 and potentiate the Acvrl1/Smad1 axis, which was active in lung fibroblasts and antagonized Tgfbr1/Smad2/3 signalling.34 A synthetic peptide from TGFBR3, P144, was reported to efficiently inhibit the TGF-beta1-dependent signalling pathway and collagen type I synthesis in cardiac fibroblasts.35 Further studies showed that overexpression of TGFBR3 deactivated the TGF-beta1 pathway by decreasing the expression of TGF-beta1, Smad3 activation, TGFBR1-TGFBR2 complex formation and inhibiting the pro-fibrotic microRNA-21 expression.24, 36 Therefore, TGFBR3 may act as a protective factor representing a potential therapeutic target for cardiac fibrosis.', 'Glucocorticoids recruit Tgfbr3 and Smad1 to shift transforming growth factor-beta signaling from the Tgfbr1/Smad2/3 axis to the Acvrl1/Smad1 axis in lung fibroblasts']\n",
            "  Document ID     : PMC6484421\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 94 : Acvrl1.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!pip install biopython\n",
        "import xml.etree.ElementTree as ET\n",
        "from Bio import Entrez\n",
        "import re\n",
        "import time\n",
        "\n",
        "class Gene:\n",
        "    def __init__(self, gene_id):\n",
        "        self.gene_id = gene_id\n",
        "        self.occurrences = []  # list to store snippet(s) where the gene was mentioned\n",
        "        self.symbol = None\n",
        "        self.organism = None\n",
        "        self.full_name = None\n",
        "        self.also_known_as = None\n",
        "        self.name_from_article = None\n",
        "        self.doc_id = None\n",
        "\n",
        "    def add_occurrence(self, snippet):\n",
        "        if snippet not in self.occurrences:  # Avoid duplicates\n",
        "            self.occurrences.append(snippet)\n",
        "\n",
        "    def set_name_from_article(self, name_from_article):\n",
        "        \"\"\"Sets the temporary name of the gene. This is the name according to the article\"\"\"\n",
        "        self.name_from_article = name_from_article\n",
        "\n",
        "    def get_name_from_article(self):\n",
        "        return self.name_from_article\n",
        "\n",
        "    def get_occurrences(self):\n",
        "        return self.occurrences\n",
        "\n",
        "    def get_also_known_as(self):\n",
        "        return self.also_known_as\n",
        "\n",
        "    def get_gene_id(self):\n",
        "        return self.gene_id\n",
        "    \n",
        "    def get_organsim(self):\n",
        "        return self.organism\n",
        "    \n",
        "    def get_doc_id(self):\n",
        "        return self.doc_id\n",
        "    \n",
        "    def set_doc_id(self, doc_id):\n",
        "        self.doc_id = doc_id\n",
        "\n",
        "    def update_info(self, symbol, organism, full_name, also_known_as):\n",
        "        self.symbol = symbol\n",
        "        self.organism = organism\n",
        "        self.full_name = full_name\n",
        "        self.also_known_as = also_known_as\n",
        "\n",
        "    def __repr__(self):\n",
        "        return (f\"Gene({self.gene_id})\\n\"\n",
        "                f\"  Symbol          : {self.symbol}\\n\"\n",
        "                f\"  Organism        : {self.organism}\\n\"\n",
        "                f\"  Full Name       : {self.full_name}\\n\"\n",
        "                f\"  Also Known As   : {self.also_known_as}\\n\"\n",
        "                f\"  In-text Name    : {self.name_from_article}\\n\"\n",
        "                f\"  Occurrences     : {self.occurrences}\\n\"\n",
        "                f\"  Document ID     : {self.doc_id}\")\n",
        "\n",
        "def parse_xml_file(xml_path):\n",
        "    \"\"\"Parses the XML file and returns a gene dictionary keyed by gene ID.\"\"\"\n",
        "    tree = ET.parse(xml_path)\n",
        "    root = tree.getroot()\n",
        "    gene_dict = {}\n",
        "\n",
        "    for document in root.findall('document'):\n",
        "        \n",
        "        # get the document ID (pmid or pmcid)\n",
        "        doc_id_elem = document.find('id')\n",
        "        if doc_id_elem is not None:\n",
        "            doc_id = doc_id_elem.text\n",
        "        else:\n",
        "            doc_id = None\n",
        "\n",
        "\n",
        "        # get the gene annotations in the document from the passages\n",
        "        for passage in document.findall('passage'):       \n",
        "            # Also check for 'article-id_pmc'\n",
        "            article_id_pmc_elem = passage.find(\"infon[@key='article-id_pmc']\")\n",
        "            if article_id_pmc_elem is not None:\n",
        "                pmc_value = article_id_pmc_elem.text.strip()\n",
        "                if not pmc_value.startswith(\"PMC\"):\n",
        "                    pmc_value = \"PMC\" + pmc_value\n",
        "                doc_id = pmc_value     \n",
        "            section_type_elem = passage.find(\"infon[@key='section_type']\")\n",
        "            if section_type_elem is not None and section_type_elem.text and section_type_elem.text.upper() in (\"METHODS\", \"FIG\", \"TABLE\"):\n",
        "                continue  # Skip passages under METHODS\n",
        "            # Get the full passage text.\n",
        "            passage_text_elem = passage.find(\"text\")\n",
        "            passage_text = passage_text_elem.text if passage_text_elem is not None else \"\"\n",
        "            # Determine the starting offset for this passage.\n",
        "            passage_offset_elem = passage.find(\"offset\")\n",
        "            passage_offset = int(passage_offset_elem.text) if passage_offset_elem is not None else 0\n",
        "            # Split the passage into sentences using regex.\n",
        "            sentences = re.split(r'(?<=[.!?])\\s+', passage_text)\n",
        "            # Compute start indices for each sentence within the passage text.\n",
        "            start_indices = []\n",
        "            current_index = passage_offset\n",
        "            for sentence in sentences:\n",
        "                start_indices.append(current_index)\n",
        "                current_index += len(sentence) + 1  # account for the delimiter space\n",
        "\n",
        "            # Dictionary to store annotations per gene in this passage.\n",
        "            # For each gene_id, we collect the sentence indices and a temporary in-text name.\n",
        "            gene_annotations = {}\n",
        "            for annotation in passage.findall('annotation'):\n",
        "                ann_type = annotation.find(\"infon[@key='type']\")\n",
        "                if ann_type is not None and ann_type.text == \"Gene\":\n",
        "                    # Check for both 'identifier' and 'NCBI Gene' keys.\n",
        "                    gene_id_elem = annotation.find(\"infon[@key='identifier']\")\n",
        "                    if gene_id_elem is None:\n",
        "                        gene_id_elem = annotation.find(\"infon[@key='NCBI Gene']\")\n",
        "                    gene_id = gene_id_elem.text if gene_id_elem is not None else None\n",
        "\n",
        "                    # Handle multiple IDs by keeping only the first if needed.\n",
        "                    if gene_id and \";\" in gene_id:\n",
        "                        gene_id = gene_id.split(\";\")[0].strip()\n",
        "\n",
        "                    # Extract the gene name from the annotation text.\n",
        "                    in_text_gene_name_elem = annotation.find(\"text\")\n",
        "                    in_text_gene_name = in_text_gene_name_elem.text if in_text_gene_name_elem is not None else None\n",
        "\n",
        "                    # Extract the annotation location (offset) to find the sentence.\n",
        "                    location_elem = annotation.find(\"location\")\n",
        "                    ann_offset = int(location_elem.attrib.get('offset', 0)) if location_elem is not None else 0\n",
        "\n",
        "                    # Determine which sentence contains the annotation based on its offset.\n",
        "                    sentence_index = None\n",
        "                    for i, start in enumerate(start_indices):\n",
        "                        if start <= ann_offset < start + len(sentences[i]):\n",
        "                            sentence_index = i\n",
        "                            break\n",
        "\n",
        "                    if sentence_index is not None and gene_id:\n",
        "                        if gene_id not in gene_annotations:\n",
        "                            gene_annotations[gene_id] = {'indices': [], 'name': in_text_gene_name}\n",
        "                        gene_annotations[gene_id]['indices'].append(sentence_index)\n",
        "\n",
        "            # For each gene in this passage, process the annotation indices.\n",
        "            for gene_id, data in gene_annotations.items():\n",
        "                # Remove duplicate indices and sort.\n",
        "                indices = sorted(set(data['indices']))\n",
        "                candidate_windows = []\n",
        "                # For each annotation occurrence, form a candidate window:\n",
        "                # one sentence before and one sentence after the annotated sentence.\n",
        "                for idx in indices:\n",
        "                    start_sentence = max(0, idx - 1)\n",
        "                    end_sentence = min(len(sentences), idx + 2)  # end index is non-inclusive\n",
        "                    candidate_windows.append((start_sentence, end_sentence))\n",
        "                # Merge overlapping candidate windows.\n",
        "                merged_windows = []\n",
        "                for window in sorted(candidate_windows):\n",
        "                    if not merged_windows:\n",
        "                        merged_windows.append(window)\n",
        "                    else:\n",
        "                        last = merged_windows[-1]\n",
        "                        # If windows overlap or touch, merge them.\n",
        "                        if window[0] <= last[1]:\n",
        "                            merged_windows[-1] = (last[0], max(last[1], window[1]))\n",
        "                        else:\n",
        "                            merged_windows.append(window)\n",
        "                # For each merged window, if it spans more than 3 sentences, split it into chunks of 3 sentences maximum.\n",
        "                for window in merged_windows:\n",
        "                    start, end = window\n",
        "                    window_size = end - start\n",
        "                    if window_size <= 3:\n",
        "                        snippet = \" \".join(sentences[start:end])\n",
        "                        if gene_id in gene_dict:\n",
        "                            gene_dict[gene_id].add_occurrence(snippet)\n",
        "                        else:\n",
        "                            gene_obj = Gene(gene_id)\n",
        "                            gene_obj.add_occurrence(snippet)\n",
        "                            gene_obj.set_doc_id(doc_id)\n",
        "                            gene_obj.set_name_from_article(data['name'])\n",
        "                            gene_dict[gene_id] = gene_obj\n",
        "                    else:\n",
        "                        # Split the merged window into non-overlapping chunks of at most 3 sentences.\n",
        "                        for i in range(start, end, 3):\n",
        "                            chunk_end = min(i + 3, end)\n",
        "                            snippet = \" \".join(sentences[i:chunk_end])\n",
        "                            if snippet.strip():\n",
        "                                if gene_id in gene_dict:\n",
        "                                    gene_dict[gene_id].add_occurrence(snippet)\n",
        "                                else:\n",
        "                                    gene_obj = Gene(gene_id)\n",
        "                                    gene_obj.set_doc_id(doc_id)\n",
        "                                    gene_obj.add_occurrence(snippet)\n",
        "                                    gene_obj.set_name_from_article(data['name'])\n",
        "                                    gene_dict[gene_id] = gene_obj\n",
        "    return gene_dict\n",
        "\n",
        "def fetch_and_update_gene_info(gene_dict):\n",
        "    \"\"\"Retrieves gene information from NCBI and updates the genes in gene_dict.\"\"\"\n",
        "    gene_ids = list(gene_dict.keys())\n",
        "    if gene_ids:\n",
        "        # Post the gene IDs to NCBI.\n",
        "        handle = Entrez.epost(db=\"gene\", id=\",\".join(gene_ids))\n",
        "        result = Entrez.read(handle)\n",
        "        handle.close()\n",
        "\n",
        "        webenv = result[\"WebEnv\"]\n",
        "        query_key = result[\"QueryKey\"]\n",
        "\n",
        "        handle = Entrez.esummary(db=\"gene\", webenv=webenv, query_key=query_key)\n",
        "        record = Entrez.read(handle)\n",
        "        handle.close()\n",
        "\n",
        "        for docsum in record[\"DocumentSummarySet\"][\"DocumentSummary\"]:\n",
        "            gene_id = docsum.attributes[\"uid\"]\n",
        "            symbol = docsum.get('NomenclatureSymbol', 'No symbol')\n",
        "            organism = docsum.get('Organism', {}).get('ScientificName', 'No organism')\n",
        "            full_name = docsum.get('NomenclatureName', gene_dict[gene_id].get_name_from_article())\n",
        "            also_known_as = docsum.get('OtherAliases', gene_dict[gene_id].get_name_from_article())\n",
        "\n",
        "            if full_name == '':\n",
        "                full_name = gene_dict[gene_id].get_name_from_article()\n",
        "            if also_known_as == '':\n",
        "                also_known_as = gene_dict[gene_id].get_name_from_article()\n",
        "\n",
        "            if gene_id in gene_dict:\n",
        "                gene_dict[gene_id].update_info(symbol, organism, full_name, also_known_as)\n",
        "            time.sleep(0.5)  # Pause briefly to avoid overwhelming NCBI servers.\n",
        "\n",
        "# Set your email (and API key if available)\n",
        "Entrez.email = \"email here\"\n",
        "# Entrez.api_key = \"your_api_key\"\n",
        "\n",
        "xml_path = \"./output/PMC6484421.xml\"\n",
        "gene_dict = parse_xml_file(xml_path)\n",
        "fetch_and_update_gene_info(gene_dict)  # Optionally fetch gene info from NCBI\n",
        "\n",
        "print(\"There are {} genes in the paper.\".format(len(gene_dict)))\n",
        "if len(gene_dict) == 0:\n",
        "    print(\"No genes found in the paper.\")\n",
        "else:\n",
        "    for gene in gene_dict.values():\n",
        "        print(gene)\n",
        "        print(f\"There are {len(gene.get_occurrences())} occurrence(s) (each occurrence is at most 3 sentences long) for gene {gene.get_gene_id()} : {gene.get_name_from_article()}.\")\n",
        "        print(\"\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "33A7yYBLXZes"
      },
      "source": [
        "# 3.5 Prompt Engineering\n",
        "- Creates the prompt using the Gene dictionary (from Gene chunking)\n",
        "- batch_size: allows you to have more than 1 Gene in the prompt such that you have something like the following. Ideally keep it at batch_size=1 to ensure that the GO terms generated are exclusive to the Gene being analyzed.\n",
        "\n",
        "```\n",
        "# batch_size > 1 has the basic format:\n",
        "Basic Prompt\n",
        "Gene 1\n",
        "Gene 2\n",
        "...\n",
        "# batch_size = 1 has the basic format:\n",
        "Basic Prompt\n",
        "Gene 1\n",
        "```\n",
        "- batch_index: allows you to iterate through the gene dictionary so that if you have a paper with x amount of genes, you can just increment  the index and run the inference x amount of times.\n",
        "- direct: ignore this. This is for the future.\n",
        "- customized_prompt: ignore this as well. This is for the future."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 142,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UsRz3Wb5YTQl",
        "outputId": "dc245a37-b770-4af2-f071-f0b973bda774"
      },
      "outputs": [],
      "source": [
        "def make_go_term_prompt(gene_dict, batch_size=1, batch_index=0, direct=False, customized_prompt=None):\n",
        "    \"\"\"\n",
        "    Create a prompt for Llama to output Gene Ontology (GO) terms (without IDs) for a batch of genes\n",
        "    from the gene dictionary. The output is expected to be categorized into three sections for each gene:\n",
        "      1. Biological Processes\n",
        "      2. Cellular Components\n",
        "      3. Molecular Functions\n",
        "\n",
        "    For each gene, the output should follow the format below without mixing results across different genes.\n",
        "\n",
        "    :param gene_dict: A dictionary where each key is a gene ID and each value is a Gene object that\n",
        "                      contains attributes such as symbol, organism, full_name, and occurrences.\n",
        "    :param batch_size: The number of genes to include in this prompt batch.\n",
        "    :param batch_index: The index (0-based) of the batch to process. For instance, if batch_size=1,\n",
        "                        batch_index=2 will include the 3rd gene in the dictionary.\n",
        "    :param direct: If True, use a more direct instruction for GO term extraction.\n",
        "    :param customized_prompt: If provided, use this custom prompt text in place of the default.\n",
        "    :return: A string containing the complete prompt.\n",
        "    \"\"\"\n",
        "\n",
        "    # Instruction blocks\n",
        "\n",
        "    # context = \"\"\"You are an efficient and insightful assistant to a geneticist.\"\"\"\n",
        "\n",
        "    general_instructions = \"\"\"Analyze the following gene information and text snippets under \"Occurrences\".\n",
        "Identify all potentially relevant Gene Ontology (GO) terms using only evidence explicitly stated in the text.\n",
        "Do not include any GO terms based on inferences or assumptions not explicitly supported by the text.\n",
        "\n",
        "After completing your analysis, for each Gene Ontology (GO) term you list, provide a short one-line justification referencing the text as well as a relationship between the gene and the term.\n",
        "If the text does not explicitly support a GO term (even if it might be inferred), do not include it. If no GO term is supported for a particular gene, say None found.\n",
        "\n",
        "Be concise, do not use unnecessary words.\n",
        "Be factual, do not editorialize.\n",
        "Be specific, avoid overly general statements like \"it is involved in many cellular processes\".\n",
        "\n",
        "\"\"\"\n",
        "\n",
        "    example = \"\"\"To help you in your work, I am providing an example of gene information and the corresponding example analysis output structured in JSON.\n",
        "\n",
        "Example gene information is:\n",
        "Gene: myogenin\n",
        "Organism: Mus musculus\n",
        "Occurrences:\n",
        "- Activation of NF-kappaB increases the expression of the inducible nitric oxide synthase (iNOS) in muscle cells that sequester HuR (RNA-binding protein) by preventing transcription of MyoD. Denervation-induced atrophy is related to the upregulation of specific histone deacetylases (HDAC), able to repress a negative regulator of myogenin, with an increase in MuRF1 expression and muscle wasting. AMPK (protein kinase activated by 5'adenosine monophosphate) is also involved in muscle atrophy.\n",
        "\n",
        "Example anaylysis output is:\n",
        "[\n",
        "  {\n",
        "    \"gene\": \"myogenin\",\n",
        "    \"go_data\": [\n",
        "      {\n",
        "        \"term\": \"DNA-binding transcription activator activity\",\n",
        "        \"relationship\": \"enables\",\n",
        "        \"namespace\": \"molecular_function\",\n",
        "        \"justification\": \"Myogenin is a key transcription factor in muscle differentiation that binds to E-box sequences in muscle-specific genes and activates their transcription through RNA polymerase II.\"\n",
        "      },\n",
        "      {\n",
        "        \"term\": \"skeletal muscle cell differentiation\",\n",
        "        \"relationship\": \"involved_in\",\n",
        "        \"namespace\": \"biological_process\",\n",
        "        \"justification\": \"Myogenin is one of the four myogenic regulatory factors (MRFs) that drive skeletal muscle differentiation by promoting the transition from myoblasts to myotubes.\"\n",
        "      },\n",
        "      {\n",
        "        \"term\": \"nucleus\",\n",
        "        \"relationship\": \"located_in\",\n",
        "        \"namespace\": \"cellular_component\",\n",
        "        \"justification\": \"As a transcription factor, myogenin localizes to the nucleus to regulate the expression of muscle-specific genes.\"\n",
        "      }\n",
        "    ]\n",
        "  }\n",
        "]\n",
        "\"\"\"\n",
        "\n",
        "\n",
        "    # Assemble the prompt text based on the chosen mode\n",
        "    if direct:\n",
        "        # prompt_text = context\n",
        "        # prompt_text += direct_instructions\n",
        "        # prompt_text += format_instructions\n",
        "        # prompt_text += example_output\n",
        "        pass\n",
        "    elif customized_prompt:\n",
        "        # prompt_text = context\n",
        "        # prompt_text += customized_prompt\n",
        "        # prompt_text += format_instructions\n",
        "        # prompt_text += example_output\n",
        "        pass\n",
        "    else:\n",
        "        prompt_text = general_instructions\n",
        "        prompt_text += example\n",
        "        prompt_text += \"\\nHere is the gene information:\\n\"\n",
        "\n",
        "    # Convert the gene dictionary to a list and determine the batch slice.\n",
        "    gene_items = list(gene_dict.items())\n",
        "    start_index = batch_index * batch_size\n",
        "    end_index = start_index + batch_size\n",
        "    batch_genes = gene_items[start_index:end_index]\n",
        "\n",
        "    # Add details and occurrence context for each gene in the selected batch.\n",
        "    for gene_id, gene_obj in batch_genes:\n",
        "        # prompt_text += f\"\\nGene ID: {gene_id}\\n\"\n",
        "        # prompt_text += f\"Symbol: {gene_obj.symbol}\\n\"\n",
        "        # prompt_text += f\"Organism: {gene_obj.organism}\\n\"\n",
        "        # prompt_text += f\"Full Name: {gene_obj.full_name}\\n\"\n",
        "        prompt_text += f\"\\nGene: {gene_obj.get_name_from_article()}\\n\"\n",
        "        if gene_obj.get_organsim() != None:\n",
        "          # add organism if available\n",
        "          prompt_text += f\"Organism: {gene_obj.get_organsim()}\\n\"\n",
        "        prompt_text += \"Occurrences:\\n\"\n",
        "        for occ in gene_obj.occurrences:\n",
        "            prompt_text += f\"- {occ}\\n\"\n",
        "\n",
        "    return prompt_text\n",
        "\n",
        "# Example usage:\n",
        "# Assuming gene_dict is defined and contains Gene objects with attributes: symbol, organism, full_name, and occurrences.\n",
        "# To process only one gene at a time:\n",
        "# prompt = make_go_term_prompt(gene_dict, batch_size=1, batch_index=0)\n",
        "# print(prompt)\n",
        "\n",
        "# prompt = make_go_term_prompt(gene_dict, batch_size=116, batch_index=0)\n",
        "# print(prompt)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6oaKSo6iXGOc"
      },
      "source": [
        "# 4. Inference (Local)\n",
        "The two cells below rely on local inference meaning that you must have the model locally. Skip them if youre using the API like GROQ"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hzFESXHRXBHG"
      },
      "outputs": [],
      "source": [
        "print(\"Loading LLaMA model.\")\n",
        "# model_dir = \"/content/drive/MyDrive/Llama 3.2-3B-Instruct-model\"\n",
        "model_dir = \"/content/drive/MyDrive/GOLLM/Llama 3.2-3B-Instruct-model\"\n",
        "device = torch.device(\"cuda\")\n",
        "print(torch.cuda.is_available())\n",
        "\n",
        "tokenizer = AutoTokenizer.from_pretrained(model_dir, local_files_only=True)\n",
        "model = AutoModelForCausalLM.from_pretrained(model_dir, local_files_only=True).to(device)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "IzBYJajjYYly"
      },
      "outputs": [],
      "source": [
        "def run_inference(prompt):\n",
        "    \"\"\"\n",
        "    Run inference on the provided prompt using the specified model and tokenizer.\n",
        "\n",
        "    :param prompt: The prompt string to send to the model.\n",
        "    :return: The generated text.\n",
        "    \"\"\"\n",
        "\n",
        "    # Encode the prompt and move to the appropriate device\n",
        "    inputs = tokenizer(prompt, return_tensors=\"pt\").to(\"cuda\")\n",
        "    attention_mask = inputs[\"attention_mask\"]\n",
        "\n",
        "    # Greedy or minimal variation == When: You want consistent, short answers, and you dont mind if its a bit blunt.\n",
        "    # outputs = model.generate(\n",
        "    #     input_ids=inputs[\"input_ids\"],\n",
        "    #     attention_mask=attention_mask,\n",
        "    #     max_new_tokens=150,\n",
        "    #     do_sample=False,     # no sampling\n",
        "    #     num_beams=1,         # purely greedy\n",
        "    #     no_repeat_ngram_size=3,\n",
        "    #     pad_token_id=tokenizer.eos_token_id\n",
        "    # )\n",
        "\n",
        "    # Beam Search (Higher Quality / Less Randomness) == You want a more global optimum text.\n",
        "    # outputs = model.generate(\n",
        "    #     input_ids=inputs[\"input_ids\"],\n",
        "    #     attention_mask=attention_mask,\n",
        "    #     max_new_tokens=200,\n",
        "    #     do_sample=False,\n",
        "    #     num_beams=4,      # search multiple beams\n",
        "    #     length_penalty=1.0,  # see if you want to encourage or discourage long outputs\n",
        "    #     no_repeat_ngram_size=3,\n",
        "    #     pad_token_id=tokenizer.eos_token_id\n",
        "    # )\n",
        "\n",
        "    # Beam SearchV2\n",
        "    outputs = model.generate(\n",
        "        input_ids=inputs[\"input_ids\"],\n",
        "        attention_mask=attention_mask,\n",
        "        max_new_tokens=400,       # Enough tokens for a concise but thorough answer\n",
        "        do_sample=False,          # Turn off sampling; we want a more deterministic, stable answer\n",
        "        num_beams=3,              # Explore multiple beams for higher-quality completions\n",
        "        length_penalty=0.1,       # 1.0 means \"neutral\" length preference (>=1.0 encourages longer outputs)\n",
        "        no_repeat_ngram_size=4,   # Helps avoid repeating the same phrase\n",
        "        early_stopping=True,      # Stops as soon as the best beam is complete\n",
        "        pad_token_id=tokenizer.eos_token_id,\n",
        "        temperature=None,\n",
        "        top_p=None,\n",
        "    )\n",
        "    # Sampling (More Creative / Less Deterministic)\n",
        "    # outputs = model.generate(\n",
        "    #     input_ids=inputs[\"input_ids\"],\n",
        "    #     attention_mask=attention_mask,\n",
        "    #     max_new_tokens=200,\n",
        "    #     do_sample=True,       # sampling\n",
        "    #     temperature=0.7,      # moderate creativity\n",
        "    #     top_p=0.9,            # nucleus sampling\n",
        "    #     no_repeat_ngram_size=3,\n",
        "    #     pad_token_id=tokenizer.eos_token_id\n",
        "    # )\n",
        "\n",
        "    # Decode the output and return the generated text\n",
        "    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)\n",
        "    completion = generated_text[len(prompt):].strip()\n",
        "    return completion\n",
        "\n",
        "for i in range(3):\n",
        "    start_time = time.time()\n",
        "    output = run_inference(test_prompt)\n",
        "    print(output)\n",
        "\n",
        "    # with open(f'/content/drive/MyDrive/GOLLM/Inference Outputs/output_{i+1}.txt', 'w') as file:\n",
        "    #     file.write(output)\n",
        "\n",
        "    print(f\"-->Model loaded on {device} named {torch.cuda.get_device_name(0)} and it took {time.time()-start_time:.2f} seconds.\")\n",
        "    print(\"-\"*50)\n",
        "\n",
        "# prompt = \"what is 2+2\"\n",
        "# print(\"Running inference...\")\n",
        "# start_time = time.time()\n",
        "# output = run_inference(prompt)\n",
        "# print(f\"Model loaded on {device} in {time.time()-start_time:.2f} seconds.\")\n",
        "# print(output)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gSz3gKoaY7yu"
      },
      "source": [
        "# 4v2. GROQ API Inference\n",
        "- need API Key\n",
        "- uses llama 3.3-70b-versatile"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 119,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "J_k_yNbOwpkb",
        "outputId": "16a5c4d1-e0bd-4c8b-a736-77a7dd2e2130"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
            "To disable this warning, you can either:\n",
            "\t- Avoid using `tokenizers` before the fork if possible\n",
            "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
            "111120.36s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: Groq in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (0.18.0)\n",
            "Requirement already satisfied: pydantic<3,>=1.9.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (2.10.6)\n",
            "Requirement already satisfied: anyio<5,>=3.5.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (4.8.0)\n",
            "Requirement already satisfied: httpx<1,>=0.23.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (0.28.1)\n",
            "Requirement already satisfied: typing-extensions<5,>=4.10 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (4.12.2)\n",
            "Requirement already satisfied: distro<2,>=1.7.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (1.9.0)\n",
            "Requirement already satisfied: sniffio in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from Groq) (1.3.1)\n",
            "Requirement already satisfied: idna>=2.8 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from anyio<5,>=3.5.0->Groq) (3.10)\n",
            "Requirement already satisfied: exceptiongroup>=1.0.2 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from anyio<5,>=3.5.0->Groq) (1.2.2)\n",
            "Requirement already satisfied: certifi in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from httpx<1,>=0.23.0->Groq) (2024.12.14)\n",
            "Requirement already satisfied: httpcore==1.* in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from httpx<1,>=0.23.0->Groq) (1.0.7)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from httpcore==1.*->httpx<1,>=0.23.0->Groq) (0.14.0)\n",
            "Requirement already satisfied: pydantic-core==2.27.2 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from pydantic<3,>=1.9.0->Groq) (2.27.2)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from pydantic<3,>=1.9.0->Groq) (0.7.0)\n",
            "\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip available: \u001b[0m\u001b[31;49m22.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
          ]
        }
      ],
      "source": [
        "# using GROQ API\n",
        "!pip install Groq\n",
        "from groq import Groq\n",
        "\n",
        "client = Groq(\n",
        "    api_key=\"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\"\n",
        ")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pmOgXgUc_DAC"
      },
      "source": [
        "## Run it again if it stops. It will skip documents that already exists."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 143,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6eOnOkZsztbK",
        "outputId": "f9dbfdd0-0f1e-4238-cfb7-7cca62af4790"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Finished processing 31 genes. Final merged JSON saved to /Users/crystalcho/GOLLM/Llama3.2-3B-Instruct-Inference/llama_inferences/llama_output_PMC6484421.json\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "import json\n",
        "import tempfile\n",
        "import shutil \n",
        "\n",
        "# Create a temporary directory to store individual JSON outputs.\n",
        "temp_dir = os.path.join(os.getcwd(), \"temp_inferences\")\n",
        "if not os.path.exists(temp_dir):\n",
        "    os.makedirs(temp_dir)\n",
        "# Create a  directory to store individual llama JSON outputs.\n",
        "llama_dir = os.path.join(os.getcwd(), \"llama_inferences\")\n",
        "if not os.path.exists(llama_dir):\n",
        "    os.makedirs(llama_dir)\n",
        "\n",
        "# retrieve the doc_id from the gene_dict of any gene (since all genes in the gene_dict are from the same document)\n",
        "doc_id = None\n",
        "for gene in gene_dict.values():\n",
        "    if gene.get_doc_id() is not None:\n",
        "        doc_id = gene.get_doc_id()\n",
        "        break\n",
        "\n",
        "# Create a subdirectory under temp_inferences named after doc_id.\n",
        "# If doc_id is None, fall back to a placeholder name, e.g. \"unknown_doc_id\".\n",
        "temp_doc_dir = os.path.join(temp_dir, doc_id if doc_id else \"unknown_doc_id\")\n",
        "if not os.path.exists(temp_doc_dir):\n",
        "    os.makedirs(temp_doc_dir)\n",
        "\n",
        "# List of gene items from gene_dict\n",
        "gene_items = list(gene_dict.items())\n",
        "final_results = []\n",
        "\n",
        "# Process each gene one-by-one (batch_size=1)\n",
        "for idx, (gene_id, gene_obj) in enumerate(gene_items):\n",
        "    output_filename = os.path.join(temp_doc_dir, f\"gene_{gene_id}.json\")\n",
        "\n",
        "    # Check if the gene's output file already exists\n",
        "    if os.path.exists(output_filename):\n",
        "        print(f\"Skipping gene {gene_id}: output file already exists.\")\n",
        "        try:\n",
        "            with open(output_filename, \"r\", encoding=\"utf-8\") as f:\n",
        "                gene_output = f.read()\n",
        "            final_results.append(json.loads(gene_output))\n",
        "        except json.JSONDecodeError:\n",
        "            print(f\"Warning: Existing output for gene {gene_id} is not valid JSON.\")\n",
        "        continue\n",
        "\n",
        "    # Generate a prompt for a single gene.\n",
        "    prompt = make_go_term_prompt(gene_dict, batch_size=1, batch_index=idx)\n",
        "    # Use your GROQ API client to request the inference.\n",
        "    chat_completion = client.chat.completions.create(\n",
        "        messages=[\n",
        "            {\"role\": \"user\", \"content\": prompt},\n",
        "        ],\n",
        "        model=\"llama-3.3-70b-versatile\",\n",
        "        temperature=0,\n",
        "        max_completion_tokens=4096,\n",
        "        top_p=0.5,\n",
        "        stream=False,\n",
        "        response_format={\"type\": \"json_object\"},\n",
        "        stop=None,\n",
        "    )\n",
        "\n",
        "    # Extract the JSON output from the response.\n",
        "    gene_output = chat_completion.choices[0].message.content\n",
        "\n",
        "    # Add the output to the final results list.\n",
        "    try:\n",
        "        model_json = json.loads(gene_output)\n",
        "\n",
        "        # Construct a new dict to ensure key order: gene, geneID, go_data\n",
        "        ordered_model_json = {\n",
        "            \"gene\": model_json.get(\"gene\", \"\"),\n",
        "            \"geneID\": gene_id,\n",
        "            \"go_data\": model_json.get(\"go_data\", [])\n",
        "        }\n",
        "\n",
        "        # If go_data is not a list, set it to an empty list\n",
        "        if not isinstance(ordered_model_json[\"go_data\"], list):\n",
        "            ordered_model_json[\"go_data\"] = []\n",
        "\n",
        "        # Add reference(s) to each GO entry\n",
        "        for entry in ordered_model_json[\"go_data\"]:\n",
        "            entry[\"reference(s)\"] = [gene_obj.get_doc_id()]\n",
        "\n",
        "        # Save the updated JSON in the temp folder\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            json.dump(ordered_model_json, f, indent=2)\n",
        "\n",
        "        final_results.append(ordered_model_json)\n",
        "    except json.JSONDecodeError:\n",
        "        print(f\"Warning: Output for gene {gene_id} is not valid JSON.\")\n",
        "\n",
        "# Merge all outputs into one finalized JSON file, using doc_id in the filename\n",
        "llama_output_path = os.path.join(llama_dir, f\"llama_output_{doc_id}.json\")\n",
        "with open(llama_output_path, \"w\", encoding=\"utf-8\") as f:\n",
        "    json.dump(final_results, f, indent=2)\n",
        "\n",
        "print(f\"Finished processing {len(gene_items)} genes. Final merged JSON saved to {llama_output_path}\")\n",
        "\n",
        "# Remove the temporary directory\n",
        "# shutil.rmtree(temp_dir)\n",
        "# print(f\"Removed temporary folder: {temp_dir}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "P58Ns1lwdbbg"
      },
      "source": [
        "# 5. Normalize\n",
        "- use an embedding model (SAPBERT) to ensure that the terms generated by Llama are actual GO terms. You will need the Go.obo file from the [Gene Ontology Consortium](https://geneontology.org/docs/download-ontology/)\n",
        "  - Essentially, for each go term created by Llama, Sapbert will do a similarity search against the Go.obo and retrieve the actual term + GO:ID"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 80,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "n5lKdeAZ1Ssh",
        "outputId": "31f1d594-43c0-4f83-e46c-4873128724d0"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "74050.60s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: sentence-transformers in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (3.4.1)\n",
            "Requirement already satisfied: transformers<5.0.0,>=4.41.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (4.48.1)\n",
            "Requirement already satisfied: scikit-learn in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (1.6.1)\n",
            "Requirement already satisfied: scipy in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (1.15.1)\n",
            "Requirement already satisfied: tqdm in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (4.67.1)\n",
            "Requirement already satisfied: Pillow in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (11.1.0)\n",
            "Requirement already satisfied: huggingface-hub>=0.20.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (0.27.1)\n",
            "Requirement already satisfied: torch>=1.11.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sentence-transformers) (2.5.1)\n",
            "Requirement already satisfied: fsspec>=2023.5.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (2024.12.0)\n",
            "Requirement already satisfied: typing-extensions>=3.7.4.3 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (4.12.2)\n",
            "Requirement already satisfied: packaging>=20.9 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (24.2)\n",
            "Requirement already satisfied: requests in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (2.32.3)\n",
            "Requirement already satisfied: filelock in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (3.17.0)\n",
            "Requirement already satisfied: pyyaml>=5.1 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (6.0.2)\n",
            "Requirement already satisfied: networkx in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from torch>=1.11.0->sentence-transformers) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from torch>=1.11.0->sentence-transformers) (3.1.5)\n",
            "Requirement already satisfied: sympy==1.13.1 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from torch>=1.11.0->sentence-transformers) (1.13.1)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from sympy==1.13.1->torch>=1.11.0->sentence-transformers) (1.3.0)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (2024.11.6)\n",
            "Requirement already satisfied: numpy>=1.17 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (2.2.2)\n",
            "Requirement already satisfied: safetensors>=0.4.1 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (0.5.2)\n",
            "Requirement already satisfied: tokenizers<0.22,>=0.21 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (0.21.0)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from scikit-learn->sentence-transformers) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from scikit-learn->sentence-transformers) (3.5.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from jinja2->torch>=1.11.0->sentence-transformers) (3.0.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (3.10)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (2024.12.14)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (3.4.1)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (2.3.0)\n",
            "\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip available: \u001b[0m\u001b[31;49m22.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "74056.90s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: obonet in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (1.1.0)\n",
            "Requirement already satisfied: networkx in /Users/crystalcho/.pyenv/versions/3.10.6/lib/python3.10/site-packages (from obonet) (3.4.2)\n",
            "\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip available: \u001b[0m\u001b[31;49m22.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0.1\u001b[0m\n",
            "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
          ]
        }
      ],
      "source": [
        "# Install required packages (run these in your environment if not already installed)\n",
        "!pip install sentence-transformers\n",
        "!pip install obonet"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 81,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8TPMPZKF1P1X",
        "outputId": "80689670-c815-40d7-8bb8-d202d1fc88c4"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Loaded 40267 GO terms from /Users/crystalcho/Downloads/go.obo\n"
          ]
        }
      ],
      "source": [
        "import obonet\n",
        "import json\n",
        "from sentence_transformers import SentenceTransformer, util\n",
        "import numpy as np\n",
        "\n",
        "##########################################\n",
        "# 1. Load and parse the GO.obo file\n",
        "##########################################\n",
        "\n",
        "def load_go_ontology(obo_filepath):\n",
        "    \"\"\"\n",
        "    Parse the GO.obo file using obonet.\n",
        "    Returns a dictionary mapping GO IDs to their term names.\n",
        "    \"\"\"\n",
        "    graph = obonet.read_obo(obo_filepath)\n",
        "    go_terms = {}\n",
        "    for go_id, data in graph.nodes(data=True):\n",
        "        # Only include nodes that have a 'name' field\n",
        "        if 'name' in data:\n",
        "            go_terms[go_id] = data\n",
        "    return go_terms\n",
        "\n",
        "# set the path to your local go.obo (or something else later) file\n",
        "obo_file = \"/Users/crystalcho/Downloads/go.obo\"\n",
        "go_terms_dict = load_go_ontology(obo_file)\n",
        "print(f\"Loaded {len(go_terms_dict)} GO terms from {obo_file}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 340,
          "referenced_widgets": [
            "e4718c03421b41a2a665c26293acd697",
            "7d97b347b7cb40df81041035931d7f27",
            "b85f6bb59d024282b5552ff665da4548",
            "f1219528a46a4a8e94bd1c8de9e6e16a",
            "1b8c25318a2b4b64bab3496083679c85",
            "53bb7f0fbad34a818e266d0ea44ffaf1",
            "bc435a026437452bb820c14d1ce5d3ee",
            "47571c7570f14ab58f96e5d30129da1b",
            "ec42f776af1b48a982636f0c122ab7e6",
            "de6e3ef0670042378384c41ecd695f1a",
            "ec7200455fa34863bec2ae60f9f4aad5",
            "618fd6a5462e4c5296b560f888b5d638",
            "331dbfdb6d7849c6a7b6d43bd33a7325",
            "54a0cadcc9ef406db05d36b690d7891c",
            "2443ed122977428c98248537123b674b",
            "c3911d8aca764dd38fad4c1dded8198f",
            "8eb65674477e477091f2501704df7eb8",
            "6d29b58c64cc4f869d6944699149a60e",
            "c939f017049643b18d5142514937af73",
            "6c463c9a4bef4b5194113acf7ced731e",
            "f7dcd748d0944da09948f4372366e8a8",
            "466f8ef6ba644d31a8586abaf9830448",
            "6ac927f5e2a2453d9aae4b5c474cf96d",
            "7593152124e04332b709dd30ec87c5a5",
            "4a743c94312c4b5db25087a747f6b879",
            "59dc6c4216cb4c5aad8c3764cf3d3f6d",
            "3117723c389b4e47a97d6e582078e5c9",
            "c6b5cf46fbeb41adb8da935949b808c1",
            "d4b8ff5330d14128ba304020f7fe8412",
            "ea731973ab9c44caa479fb7934fd3b4f",
            "097046dc7c4f46f5866068c82b706aca",
            "cc7573c531c84e61999e2161710173f3",
            "9775c0cd37bc43dcb7953f8b15343332",
            "1675d70c8a5e426691d0af5e6ba333f3",
            "784d7bf8c3be48b8ad5a3b6e6954690f",
            "730fe49f91204e2cbd3b58d24442be0a",
            "792eb1f8daf34e2590fda2f60e3ff684",
            "3e0f9cdf58ef4af1a0b16f9795747dfc",
            "1845ddb7ff634d969fa2538209fc4e0a",
            "9b03af45365c42419bb39779b82be47c",
            "d598fdb3d94a4095ba5a950cef321f1b",
            "39794838a2164151a5edf298a63f8147",
            "e7eaa149a903445db37dfdf719e14f9e",
            "4af857d8e5a94c278ae0fb3d7e092ac5",
            "769b37c671334690b6a7e5ba9059343e",
            "bdf351e0ccbc4aa18e168f593d1659cf",
            "8901a8884437441c934c53a98a1dac06",
            "32ef3e232304493d8b7e129c679f4f6f",
            "77472246b6b0497fbe67b8a3529040e5",
            "61de99be2c2e4afe90ff06e3bb30a2b7",
            "873ea6832a0447a19938fe764cb7ac58",
            "b483173c5fa74a72bb0a88206d3206a3",
            "8039b35f8f654a5e93ab69a0d1d8f3ea",
            "b1c01cfdd57e40e7bc4de9baf81d9f45",
            "82a3774cc5cd4acfb1822d9cb74d6f45"
          ]
        },
        "id": "zfuUI3PF1boL",
        "outputId": "e257d393-bdbb-4695-d2ac-d8f2ee4ecbea"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "No sentence-transformers model found with name cambridgeltl/SapBERT-from-PubMedBERT-fulltext. Creating a new one with mean pooling.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Loaded embedding model: cambridgeltl/SapBERT-from-PubMedBERT-fulltext\n"
          ]
        }
      ],
      "source": [
        "##########################################\n",
        "# 2. Load the SAPBERT embedding model (or an alternative)\n",
        "##########################################\n",
        "\n",
        "# Here we use a SAPBERT model available from Hugging Face.\n",
        "# may choose another model (e.g., \"sentence-transformers/all-MiniLM-L6-v2\")\n",
        "model_name = \"cambridgeltl/SapBERT-from-PubMedBERT-fulltext\"  # SAPBERT model fine-tuned on UMLS\n",
        "sapbert_model = SentenceTransformer(model_name)\n",
        "print(f\"Loaded embedding model: {model_name}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 144,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OM1So4sX2JLT",
        "outputId": "8477c5ed-2546-409c-8284-f1d957453bde"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Embedded GO terms with additional metadata.\n"
          ]
        }
      ],
      "source": [
        "def embed_go_terms(go_terms, model):\n",
        "    \"\"\"\n",
        "    Given a dictionary of GO terms (GO ID -> term data) and an embedding model,\n",
        "    returns:\n",
        "      - a list of GO IDs,\n",
        "      - a list of GO term names (in the same order),\n",
        "      - a tensor of embeddings for the GO term names,\n",
        "      - a list of GO namespaces,\n",
        "      - a list of GO relationships (if available, otherwise default to 'unknown')\n",
        "    \"\"\"\n",
        "    go_ids = []\n",
        "    go_names = []\n",
        "    go_namespaces = []\n",
        "    go_relationships = []\n",
        "\n",
        "    for go_id, data in go_terms.items():\n",
        "        if 'name' in data:\n",
        "            go_ids.append(go_id)\n",
        "            go_names.append(data['name'])\n",
        "            go_namespaces.append(data.get('namespace', 'unknown'))\n",
        "            # Some GO entries might not have a \"relationship\" field; default to 'unknown'\n",
        "            go_relationships.append(data.get('relationship', 'unknown'))\n",
        "\n",
        "    go_embeddings = model.encode(go_names, convert_to_tensor=True)\n",
        "    return go_ids, go_names, go_embeddings, go_namespaces, go_relationships\n",
        "\n",
        "# Reload GO ontology using the updated load_go_ontology:\n",
        "# go_terms_dict = load_go_ontology(obo_file) # already loaded in the previous cell\n",
        "go_ids, go_names, go_embeddings, go_namespaces, go_relationships = embed_go_terms(go_terms_dict, sapbert_model)\n",
        "print(\"Embedded GO terms with additional metadata.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 145,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "G0yR2eF6228J",
        "outputId": "3bfeca3b-8998-4598-f8f4-56505af53917"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Example: Llama term 'muscle cell differentiation' matched with 'muscle cell differentiation' (GO:0042692) with score 1.00\n"
          ]
        }
      ],
      "source": [
        "import torch\n",
        "\n",
        "def get_best_matching_go_term(llama_term, go_ids, go_names, go_embeddings, model, threshold=0.6):\n",
        "    \"\"\"\n",
        "    Given a GO term from Llama, compute its embedding and find the best matching GO term\n",
        "    from the provided GO ontology embeddings.\n",
        "\n",
        "    Parameters:\n",
        "      - llama_term: The GO term string generated by Llama.\n",
        "      - go_ids, go_names, go_embeddings: Outputs from embed_go_terms().\n",
        "      - model: The embedding model.\n",
        "      - threshold: Minimum cosine similarity required to accept the match.\n",
        "\n",
        "    Returns:\n",
        "      - best_go_id: The GO:ID of the best match (or None if score is below threshold)\n",
        "      - best_go_name: The GO term name of the best match.\n",
        "      - best_score: The cosine similarity score.\n",
        "    \"\"\"\n",
        "    # Encode the query using the model\n",
        "    query_embedding = model.encode(llama_term, convert_to_tensor=True)\n",
        "    # Compute cosine similarities between the query and all GO term embeddings\n",
        "    cos_scores = torch.squeeze(util.cos_sim(query_embedding, go_embeddings))\n",
        "    # Use torch.argmax to find the best matching index\n",
        "    best_idx = torch.argmax(cos_scores)\n",
        "    best_score = cos_scores[best_idx].item()\n",
        "\n",
        "    if best_score >= threshold:\n",
        "        return go_ids[best_idx], go_names[best_idx], best_score, best_idx\n",
        "    else:\n",
        "        return None, None, best_score, None\n",
        "\n",
        "# Test the matching function with an example Llama-generated term:\n",
        "example_llama_term = \"muscle cell differentiation\"\n",
        "best_go_id, best_go_name, score, best_idx = get_best_matching_go_term(example_llama_term, go_ids, go_names, go_embeddings, sapbert_model)\n",
        "print(f\"Example: Llama term '{example_llama_term}' matched with '{best_go_name}' ({best_go_id}) with score {score:.2f}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 146,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zp0nhP0_3pHz",
        "outputId": "331efcf4-a0ca-4146-ceab-abf18abd3981"
      },
      "outputs": [],
      "source": [
        "import json\n",
        "\n",
        "def parse_llama_output(llama_output_path):\n",
        "    \"\"\"\n",
        "    Parses Llama's output JSON file.\n",
        "    Returns the parsed JSON data (a list of gene entries).\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(llama_output_path, 'r') as f:\n",
        "            data = json.load(f)\n",
        "        # If the JSON is a single dictionary, convert it to a list\n",
        "        if isinstance(data, dict):\n",
        "            data = [data]\n",
        "        return data\n",
        "    except Exception as e:\n",
        "        print(f\"Error parsing JSON: {e}\")\n",
        "        return []\n",
        "\n",
        "# print(parse_llama_output(\"./llama_inferences/llama_output_22118460))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 147,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rpyVuCmM_EtU",
        "outputId": "c8d869f2-162d-4661-cf94-91f28390fa6e"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "there are 31 genes with matched GO terms.\n",
            "Gene: TGFBR3 (7049)\n",
            "  Llama Term: transforming growth factor beta-activated receptor activity\n",
            "  Matched GO Term: transforming growth factor beta receptor activity (GO:0005024)\n",
            "  Similarity: 0.93\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TGFBR3 mediates TGF-beta superfamily ligand-dependent as well as ligand-independent signalling to both Smad and non-Smad signalling pathways.\n",
            "\n",
            "  Llama Term: negative regulation of transforming growth factor beta receptor signaling pathway\n",
            "  Matched GO Term: negative regulation of transforming growth factor beta receptor signaling pathway (GO:0030512)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['negatively_regulates GO:0007179']\n",
            "  Namespace: biological_process\n",
            "  Justification: TGFBR3 is a potential negative regulator of TGF-beta signaling.\n",
            "\n",
            "  Llama Term: cell surface\n",
            "  Matched GO Term: cell surface (GO:0009986)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: TGFBR3 is a receptor and thus located on the cell surface.\n",
            "\n",
            "  Llama Term: regulation of fibroblast proliferation\n",
            "  Matched GO Term: regulation of fibroblast proliferation (GO:0048145)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0048144']\n",
            "  Namespace: biological_process\n",
            "  Justification: TGFBR3 knockdown enhanced the expressions of COL1A1, COL3A1 and ACTA2 in HAFs, indicating a role in fibrosis.\n",
            "\n",
            "  Llama Term: apoptotic process\n",
            "  Matched GO Term: apoptotic process (GO:0006915)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['has_part GO:0097190', 'has_part GO:0097194']\n",
            "  Namespace: biological_process\n",
            "  Justification: TGFBR3 protects cardiac fibroblasts from hypoxia-induced apoptosis.\n",
            "\n",
            "Gene: Ang-II (11606)\n",
            "  Llama Term: peptide hormone receptor binding\n",
            "  Matched GO Term: peptide hormone receptor binding (GO:0051428)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Ang-II can stimulate NADPH oxidases 2 (Nox2) activity through the TGF-beta1 pathway.\n",
            "\n",
            "  Llama Term: atrial fibrosis\n",
            "  Matched GO Term: atrial cardiac myofibril assembly (GO:0055004)\n",
            "  Similarity: 0.66\n",
            "  Relationship: ['part_of GO:0055014']\n",
            "  Namespace: biological_process\n",
            "  Justification: Ang-II-induced HAFs were used as a cell model of atrial fibrosis.\n",
            "\n",
            "  Llama Term: Smad3 signaling pathway\n",
            "  Matched GO Term: SMAD protein signal transduction (GO:0060395)\n",
            "  Similarity: 0.77\n",
            "  Relationship: ['part_of GO:0007178']\n",
            "  Namespace: biological_process\n",
            "  Justification: Smad3 signalling has been shown to participate in cardiac fibrosis and was activated in Ang-II-treated HAFs.\n",
            "\n",
            "  Llama Term: cellular response to angiotensin\n",
            "  Matched GO Term: cellular response to angiotensin (GO:1904385)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Ang-II-treated HAFs were used to study the effects of Ang-II on cell proliferation and migration.\n",
            "\n",
            "Gene: collagen, type I, alpha 1 (1277)\n",
            "  Llama Term: fibril organization\n",
            "  Matched GO Term: supramolecular fiber organization (GO:0097435)\n",
            "  Similarity: 0.88\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: COL1A1 is a key component of collagen fibrils, which are essential for tissue structure and function.\n",
            "\n",
            "  Llama Term: extracellular matrix\n",
            "  Matched GO Term: extracellular matrix (GO:0031012)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: COL1A1 is a major component of the extracellular matrix in connective tissue.\n",
            "\n",
            "  Llama Term: fibrosis\n",
            "  Matched GO Term: connective tissue replacement (GO:0097709)\n",
            "  Similarity: 0.63\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: COL1A1 expression is increased in atrial fibrosis, suggesting its involvement in the fibrotic process.\n",
            "\n",
            "Gene: COL3A1 (1281)\n",
            "  Llama Term: fibril organization\n",
            "  Matched GO Term: supramolecular fiber organization (GO:0097435)\n",
            "  Similarity: 0.88\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: COL3A1 is a component of collagen fibrils, which are involved in fibril organization.\n",
            "\n",
            "  Llama Term: extracellular matrix\n",
            "  Matched GO Term: extracellular matrix (GO:0031012)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: COL3A1 is a type of collagen that is secreted into the extracellular matrix.\n",
            "\n",
            "  Llama Term: fibrosis\n",
            "  Matched GO Term: connective tissue replacement (GO:0097709)\n",
            "  Similarity: 0.63\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: COL3A1 expression is increased in atrial fibrosis, suggesting its involvement in the fibrotic process.\n",
            "\n",
            "Gene: ACTA2 (59)\n",
            "  Llama Term: muscle contraction\n",
            "  Matched GO Term: muscle contraction (GO:0006936)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: ACTA2 is expressed in HAFs and its expression is increased in response to Ang-II treatment, which is related to atrial fibrosis.\n",
            "\n",
            "  Llama Term: cytoskeleton\n",
            "  Matched GO Term: cytoskeleton (GO:0005856)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: ACTA2 is an actin protein that is a component of the cytoskeleton.\n",
            "\n",
            "  Llama Term: fibrosis\n",
            "  Matched GO Term: connective tissue replacement (GO:0097709)\n",
            "  Similarity: 0.63\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: ACTA2 expression is increased in HAFs in response to miR-23b-3p and miR-27b-3p, which promote atrial fibrosis.\n",
            "\n",
            "Gene: Smad1 (4086)\n",
            "  Llama Term: transforming growth factor beta receptor signaling pathway\n",
            "  Matched GO Term: transforming growth factor beta receptor signaling pathway (GO:0007179)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0071560']\n",
            "  Namespace: biological_process\n",
            "  Justification: Smad1 inactivation and Smad3 activation mediate the pro-fibrotic effects of miR-23b-3p, miR-27b-3p and TGFBR3 siRNA in atrial fibrosis.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: Phosphorylated Smad1 (p-Smad1) was decreased, but no direct information about Smad1 location is provided, however Smad3 accumulated in the nuclei of HAFs, and given the context, Smad1 is likely to be in the nucleus as well.\n",
            "\n",
            "  Llama Term: intracellular signal transduction\n",
            "  Matched GO Term: intracellular signal transduction (GO:0035556)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['occurs_in GO:0005622']\n",
            "  Namespace: biological_process\n",
            "  Justification: The activation of Smad1 through TGFBR3 and the negative modulation between p-Smad1 and p-Smad3 were in agreement with the previous report.\n",
            "\n",
            "  Llama Term: protein phosphorylation\n",
            "  Matched GO Term: protein phosphorylation (GO:0006468)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Phosphorylated Smad1 (p-Smad1) was decreased.\n",
            "\n",
            "Gene: Smad3 (4088)\n",
            "  Llama Term: transcription coregulator activity\n",
            "  Matched GO Term: transcription coregulator activity (GO:0003712)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Smad3 activation modulates miR-23b-3p and miR-27b-3p expression.\n",
            "\n",
            "  Llama Term: atrial fibrosis\n",
            "  Matched GO Term: atrial cardiac myofibril assembly (GO:0055004)\n",
            "  Similarity: 0.66\n",
            "  Relationship: ['part_of GO:0055014']\n",
            "  Namespace: biological_process\n",
            "  Justification: Smad3 signalling pathway was activated in Ang-II-treated HAFs to promote atrial fibrosis.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: Smad3 accumulated in the nuclei of HAFs after transfection with miR-23b-3p, miR-27b-3p and TGFBR3 siRNA.\n",
            "\n",
            "  Llama Term: signal transduction\n",
            "  Matched GO Term: signal transduction (GO:0007165)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0007154', 'part_of GO:0023052', 'part_of GO:0051716']\n",
            "  Namespace: biological_process\n",
            "  Justification: Smad3 activation modulates miR-23b-3p and miR-27b-3p expression through the Smad3 signalling pathway.\n",
            "\n",
            "  Llama Term: phosphorylation\n",
            "  Matched GO Term: phosphorylation (GO:0016310)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: The phosphorylation level of Smad3 was significantly increased in HAFs at 10 minutes in response to Ang-II treatment.\n",
            "\n",
            "Gene: miR-21 (406991)\n",
            "  Llama Term: negative regulation of gene expression\n",
            "  Matched GO Term: negative regulation of gene expression (GO:0010629)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['negatively_regulates GO:0010467']\n",
            "  Namespace: biological_process\n",
            "  Justification: miR-21 is involved in a variety of biological processes and diseases by inhibiting the protein translation process of the target gene.\n",
            "\n",
            "  Llama Term: fibrosis\n",
            "  Matched GO Term: connective tissue replacement (GO:0097709)\n",
            "  Similarity: 0.63\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: miR-21 is involved in multiple pro-fibrotic pathways by targeting Spry1, PIAS3, Smad7, and Jagged1.\n",
            "\n",
            "  Llama Term: regulation of atrial fibrillation\n",
            "  Matched GO Term: regulation of atrial cardiac muscle cell membrane depolarization (GO:0060371)\n",
            "  Similarity: 0.69\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: miR-21 participates in the development of atrial fibrillation through fibrosis-related target genes.\n",
            "\n",
            "  Llama Term: microRNA-mediated gene silencing\n",
            "  Matched GO Term: miRNA-mediated post-transcriptional gene silencing (GO:0035195)\n",
            "  Similarity: 0.92\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: miR-21 inhibits the protein translation process of the target gene.\n",
            "\n",
            "Gene: TRPC3 (7222)\n",
            "  Llama Term: calcium channel activity\n",
            "  Matched GO Term: calcium channel activity (GO:0005262)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TRPC3 is targeted by miR-26 to increase Ca2+ entry.\n",
            "\n",
            "  Llama Term: fibroblast proliferation\n",
            "  Matched GO Term: fibroblast proliferation (GO:0048144)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: TRPC3 targeting by miR-26 leads to activation of proliferation of fibroblasts.\n",
            "\n",
            "Gene: KCNJ2 (3759)\n",
            "  Llama Term: atrial cardiac muscle cell differentiation\n",
            "  Matched GO Term: atrial cardiac muscle cell differentiation (GO:0055011)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: KCNJ2 is targeted by miR-26 to enhance atrial fibroblast proliferation, suggesting a role in atrial muscle cell differentiation.\n",
            "\n",
            "  Llama Term: voltage-gated potassium channel activity\n",
            "  Matched GO Term: voltage-gated potassium channel activity (GO:0005249)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: KCNJ2 is involved in enhancing I K1, indicating a role in potassium channel activity.\n",
            "\n",
            "Gene: I K1 (10320)\n",
            "  Llama Term: atrial cardiac muscle cell differentiation\n",
            "  Matched GO Term: atrial cardiac muscle cell differentiation (GO:0055011)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: I K1 enhancement is associated with proliferation and differentiation of atrial fibroblasts.\n",
            "\n",
            "  Llama Term: potassium ion transmembrane transport\n",
            "  Matched GO Term: potassium ion transmembrane transport (GO:0071805)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: I K1 is associated with potassium ion transport.\n",
            "\n",
            "Gene: miR-23b (407011)\n",
            "  Llama Term: negative regulation of gene expression\n",
            "  Matched GO Term: negative regulation of gene expression (GO:0010629)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['negatively_regulates GO:0010467']\n",
            "  Namespace: biological_process\n",
            "  Justification: Inhibition of miR-23b could alleviate cardiac fibrosis by targeting TGIF1 and PTEN.\n",
            "\n",
            "  Llama Term: regulation of fibrosis\n",
            "  Matched GO Term: regulation of connective tissue replacement (GO:1905203)\n",
            "  Similarity: 0.73\n",
            "  Relationship: ['regulates GO:0097709']\n",
            "  Namespace: biological_process\n",
            "  Justification: MiR-23b-3p could enhance the expression of fibrosis-related genes including COL1A1, COL3A1 and ACTA2.\n",
            "\n",
            "  Llama Term: microRNA-mediated gene silencing\n",
            "  Matched GO Term: miRNA-mediated post-transcriptional gene silencing (GO:0035195)\n",
            "  Similarity: 0.92\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MiR-23b targets TGIF1 and PTEN to activate TGF-beta1/Smad2/3 signalling and AKT/N-Cadherin signalling.\n",
            "\n",
            "Gene: TGIF1 (7050)\n",
            "  Llama Term: negative regulation of TGF-beta signaling pathway\n",
            "  Matched GO Term: negative regulation of transforming growth factor beta receptor signaling pathway (GO:0030512)\n",
            "  Similarity: 0.93\n",
            "  Relationship: ['negatively_regulates GO:0007179']\n",
            "  Namespace: biological_process\n",
            "  Justification: TGIF1 is targeted by miR-23b to activate TGF-beta1/Smad2/3 signalling.\n",
            "\n",
            "  Llama Term: regulation of cardiac fibroblast activation\n",
            "  Matched GO Term: regulation of myofibroblast differentiation (GO:1904760)\n",
            "  Similarity: 0.71\n",
            "  Relationship: ['regulates GO:0036446']\n",
            "  Namespace: biological_process\n",
            "  Justification: Inhibition of miR-23b could alleviate cardiac fibrosis by targeting TGIF1 to activate TGF-beta1/Smad2/3 signalling in cardiac fibroblasts.\n",
            "\n",
            "Gene: PTEN (5728)\n",
            "  Llama Term: negative regulation of protein kinase B signaling\n",
            "  Matched GO Term: negative regulation of protein kinase C signaling (GO:0090038)\n",
            "  Similarity: 0.82\n",
            "  Relationship: ['negatively_regulates GO:0070528']\n",
            "  Namespace: biological_process\n",
            "  Justification: PTEN is targeted by miR-23b to activate AKT/N-Cadherin signalling in cardiac fibroblasts.\n",
            "\n",
            "Gene: TGF-beta1 (7040)\n",
            "  Llama Term: transforming growth factor beta receptor signaling pathway\n",
            "  Matched GO Term: transforming growth factor beta receptor signaling pathway (GO:0007179)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0071560']\n",
            "  Namespace: biological_process\n",
            "  Justification: TGF-beta1 is involved in the TGF-beta1/Smad2/3 signalling pathway.\n",
            "\n",
            "  Llama Term: cellular response to angiotensin\n",
            "  Matched GO Term: cellular response to angiotensin (GO:1904385)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Ang-II can stimulate NADPH oxidases 2 (Nox2) activity through the TGF-beta1 pathway.\n",
            "\n",
            "  Llama Term: fibrosis\n",
            "  Matched GO Term: connective tissue replacement (GO:0097709)\n",
            "  Similarity: 0.63\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: TGF-beta1 plays an important role in atrial fibrosis-associated with the heart failure models of AF.\n",
            "\n",
            "  Llama Term: extracellular space\n",
            "  Matched GO Term: extracellular space (GO:0005615)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005576']\n",
            "  Namespace: cellular_component\n",
            "  Justification: TGF-beta1 is a secreted protein that signals through cell surface receptors.\n",
            "\n",
            "  Llama Term: growth factor activity\n",
            "  Matched GO Term: growth factor activity (GO:0008083)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TGF-beta1 is a growth factor that regulates cell growth and differentiation.\n",
            "\n",
            "Gene: AKT (207)\n",
            "  Llama Term: cell signaling\n",
            "  Matched GO Term: cell-cell signaling (GO:0007267)\n",
            "  Similarity: 0.97\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: AKT is involved in AKT/N-Cadherin signalling in cardiac fibroblasts.\n",
            "\n",
            "  Llama Term: protein kinase activity\n",
            "  Matched GO Term: protein kinase activity (GO:0004672)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: AKT is a protein kinase.\n",
            "\n",
            "Gene: N-Cadherin (1000)\n",
            "  Llama Term: cell adhesion molecule binding\n",
            "  Matched GO Term: cell adhesion molecule binding (GO:0050839)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: N-Cadherin is involved in AKT/N-Cadherin signalling.\n",
            "\n",
            "  Llama Term: cell-cell adhesion\n",
            "  Matched GO Term: cell-cell adhesion (GO:0098609)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: N-Cadherin is involved in cell adhesion processes.\n",
            "\n",
            "  Llama Term: cell junction\n",
            "  Matched GO Term: cell junction (GO:0030054)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: N-Cadherin is involved in cell-cell adhesion, which occurs at cell junctions.\n",
            "\n",
            "Gene: miR-27b (407019)\n",
            "  Llama Term: negative regulation of gene expression\n",
            "  Matched GO Term: negative regulation of gene expression (GO:0010629)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['negatively_regulates GO:0010467']\n",
            "  Namespace: biological_process\n",
            "  Justification: miR-27b targets PPAR-gamma, ALK5, and other genes to regulate their expression.\n",
            "\n",
            "  Llama Term: regulation of fibrosis\n",
            "  Matched GO Term: regulation of connective tissue replacement (GO:1905203)\n",
            "  Similarity: 0.73\n",
            "  Relationship: ['regulates GO:0097709']\n",
            "  Namespace: biological_process\n",
            "  Justification: miR-27b contributes to cardiac dysfunction with enhanced myocardial fibrosis and attenuated angiotensin II-induced atrial fibrosis.\n",
            "\n",
            "  Llama Term: microRNA-mediated gene silencing\n",
            "  Matched GO Term: miRNA-mediated post-transcriptional gene silencing (GO:0035195)\n",
            "  Similarity: 0.92\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: miR-27b targets specific genes, such as TGIF1, PTEN, and PPAR-gamma, to regulate their expression.\n",
            "\n",
            "  Llama Term: cytoplasm\n",
            "  Matched GO Term: cytoplasm (GO:0005737)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005622']\n",
            "  Namespace: cellular_component\n",
            "  Justification: miR-27b is a microRNA, which is typically located in the cytoplasm.\n",
            "\n",
            "Gene: angiotensin II (183)\n",
            "  Llama Term: atrial fibrosis\n",
            "  Matched GO Term: atrial cardiac myofibril assembly (GO:0055004)\n",
            "  Similarity: 0.66\n",
            "  Relationship: ['part_of GO:0055014']\n",
            "  Namespace: biological_process\n",
            "  Justification: Atrial overexpression of microRNA-27b attenuates angiotensin II-induced atrial fibrosis\n",
            "\n",
            "  Llama Term: blood vessel\n",
            "  Matched GO Term: vascular process in circulatory system (GO:0003018)\n",
            "  Similarity: 0.77\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Angiotensin II is known to have effects on blood vessels, although this is not explicitly stated in the text, it is implied through its involvement in fibrosis\n",
            "\n",
            "Gene: TGFBR1 (7046)\n",
            "  Llama Term: transforming growth factor beta-activated receptor activity\n",
            "  Matched GO Term: transforming growth factor beta receptor activity (GO:0005024)\n",
            "  Similarity: 0.93\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TGFBR1 forms a complex with TGFBR2 to mediate TGF-beta superfamily ligand-dependent signalling.\n",
            "\n",
            "  Llama Term: cellular response to transforming growth factor beta stimulus\n",
            "  Matched GO Term: cellular response to transforming growth factor beta stimulus (GO:0071560)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: TGFBR1 is involved in TGF-beta1-dependent signalling pathway.\n",
            "\n",
            "  Llama Term: TGF-beta receptor complex\n",
            "  Matched GO Term: transforming growth factor beta complex (GO:0099126)\n",
            "  Similarity: 0.86\n",
            "  Relationship: ['part_of GO:0062023']\n",
            "  Namespace: cellular_component\n",
            "  Justification: TGFBR1 forms a complex with TGFBR2.\n",
            "\n",
            "Gene: TGFBR2 (7048)\n",
            "  Llama Term: transforming growth factor beta-activated receptor activity\n",
            "  Matched GO Term: transforming growth factor beta receptor activity (GO:0005024)\n",
            "  Similarity: 0.93\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TGFBR2 is mentioned in the context of TGF-beta1 receptor complex formation.\n",
            "\n",
            "  Llama Term: cell surface receptor signaling pathway\n",
            "  Matched GO Term: cell surface receptor signaling pathway (GO:0007166)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: TGFBR2 is part of the TGF-beta signaling pathway, which is a cell surface receptor signaling pathway.\n",
            "\n",
            "  Llama Term: plasma membrane\n",
            "  Matched GO Term: plasma membrane (GO:0005886)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0071944']\n",
            "  Namespace: cellular_component\n",
            "  Justification: As a receptor, TGFBR2 is expected to be located in the plasma membrane.\n",
            "\n",
            "Gene: FN1 (2335)\n",
            "  Llama Term: extracellular matrix organization\n",
            "  Matched GO Term: extracellular matrix organization (GO:0030198)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: FN1 mRNA expression is up-regulated in atrial appendages of AF patients, suggesting a role in fibrosis.\n",
            "\n",
            "  Llama Term: fibrinogen binding\n",
            "  Matched GO Term: fibrinogen binding (GO:0070051)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Not explicitly stated in the text, however FN1 is known to bind fibrinogen.\n",
            "\n",
            "  Llama Term: extracellular matrix\n",
            "  Matched GO Term: extracellular matrix (GO:0031012)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: FN1 is a component of the extracellular matrix.\n",
            "\n",
            "Gene: IL-1beta (3553)\n",
            "  Llama Term: inflammatory response\n",
            "  Matched GO Term: inflammatory response (GO:0006954)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: IL-1beta is described as an inflammation-related gene that is significantly up-regulated in atrial appendages of AF patients.\n",
            "\n",
            "  Llama Term: cytokine activity\n",
            "  Matched GO Term: cytokine activity (GO:0005125)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: IL-1beta is implied to have cytokine activity as it is mentioned alongside other inflammatory factors.\n",
            "\n",
            "Gene: CRP (1401)\n",
            "  Llama Term: acute-phase response\n",
            "  Matched GO Term: acute-phase response (GO:0006953)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: CRP is mentioned alongside other inflammation-related genes, such as IL-1beta, which are significantly up-regulated in atrial appendages of AF patients.\n",
            "\n",
            "  Llama Term: inflammatory response\n",
            "  Matched GO Term: inflammatory response (GO:0006954)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: CRP is described as an inflammatory factor, with its obvious increase found in atrial appendage tissue of AF patients.\n",
            "\n",
            "Gene: TNF-alpha (7124)\n",
            "  Llama Term: inflammatory response\n",
            "  Matched GO Term: inflammatory response (GO:0006954)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: TNF-alpha is mentioned alongside other inflammation-related genes, such as IL-1beta and CRP.\n",
            "\n",
            "Gene: ACE1 (1636)\n",
            "  Llama Term: peptidyl-dipeptidase activity\n",
            "  Matched GO Term: peptidyl-dipeptidase activity (GO:0008241)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: ACE1 is known for its role in converting angiotensin I to angiotensin II, a function that involves peptidyl-dipeptidase activity.\n",
            "\n",
            "  Llama Term: blood vessel\n",
            "  Matched GO Term: vascular process in circulatory system (GO:0003018)\n",
            "  Similarity: 0.77\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: ACE1 is associated with the renin-angiotensin-aldosterone system (RAAS), which plays a crucial role in regulating blood pressure and is primarily located in blood vessels.\n",
            "\n",
            "Gene: ACE2 (59272)\n",
            "  Llama Term: regulation of blood pressure\n",
            "  Matched GO Term: regulation of blood pressure (GO:0008217)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0008015']\n",
            "  Namespace: biological_process\n",
            "  Justification: ACE2 is part of the RAAS system, which is involved in blood pressure regulation.\n",
            "\n",
            "  Llama Term: peptidyl-dipeptidase activity\n",
            "  Matched GO Term: peptidyl-dipeptidase activity (GO:0008241)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: ACE2 is known to have peptidyl-dipeptidase activity, although not explicitly stated in the given text, it is a well-known function of the ACE2 gene.\n",
            "\n",
            "Gene: Apelin (8862)\n",
            "  Llama Term: regulation of blood pressure\n",
            "  Matched GO Term: regulation of blood pressure (GO:0008217)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0008015']\n",
            "  Namespace: biological_process\n",
            "  Justification: Apelin is associated with RAAS (Renin-Angiotensin-Aldosterone System), which plays a crucial role in regulating blood pressure.\n",
            "\n",
            "  Llama Term: atrial appendage\n",
            "  Matched GO Term: cardiac atrium formation (GO:0003210)\n",
            "  Similarity: 0.69\n",
            "  Relationship: ['part_of GO:0003209']\n",
            "  Namespace: biological_process\n",
            "  Justification: No significant changes of RAAS-associated genes, such as ACE1, ACE2 and Apelin were observed in the atrial appendages of AF patients.\n",
            "\n",
            "Gene: TGF-beta1 (21803)\n",
            "  Llama Term: transforming growth factor beta receptor signaling pathway\n",
            "  Matched GO Term: transforming growth factor beta receptor signaling pathway (GO:0007179)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0071560']\n",
            "  Namespace: biological_process\n",
            "  Justification: TGF-beta1 plays a role in the TGF-beta1 pathway.\n",
            "\n",
            "  Llama Term: cellular response to angiotensin\n",
            "  Matched GO Term: cellular response to angiotensin (GO:1904385)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Ang-II can stimulate NADPH oxidases 2 (Nox2) activity through the TGF-beta1 pathway.\n",
            "\n",
            "  Llama Term: regulation of reactive oxygen species metabolic process\n",
            "  Matched GO Term: regulation of reactive oxygen species metabolic process (GO:2000377)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0072593']\n",
            "  Namespace: biological_process\n",
            "  Justification: Inhibition of TGF-beta1 blunts reactive oxygen species (ROS) formation.\n",
            "\n",
            "  Llama Term: extracellular space\n",
            "  Matched GO Term: extracellular space (GO:0005615)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005576']\n",
            "  Namespace: cellular_component\n",
            "  Justification: TGF-beta1 is secreted and acts through a signaling pathway.\n",
            "\n",
            "Gene: NADPH oxidases 2 (13058)\n",
            "  Llama Term: cellular response to angiotensin\n",
            "  Matched GO Term: cellular response to angiotensin (GO:1904385)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Ang-II can stimulate NADPH oxidases 2 (Nox2) activity through the TGF-beta1 pathway.\n",
            "\n",
            "  Llama Term: transforming growth factor beta signaling pathway\n",
            "  Matched GO Term: transforming growth factor beta receptor signaling pathway (GO:0007179)\n",
            "  Similarity: 0.98\n",
            "  Relationship: ['part_of GO:0071560']\n",
            "  Namespace: biological_process\n",
            "  Justification: Ang-II can stimulate NADPH oxidases 2 (Nox2) activity through the TGF-beta1 pathway.\n",
            "\n",
            "  Llama Term: generation of reactive oxygen species\n",
            "  Matched GO Term: reactive oxygen species biosynthetic process (GO:1903409)\n",
            "  Similarity: 0.88\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: inhibition of TGF-beta1 blunts reactive oxygen species (ROS) formation.\n",
            "\n",
            "Gene: Acvrl1 (94)\n",
            "  Llama Term: transforming growth factor beta receptor activity\n",
            "  Matched GO Term: transforming growth factor beta receptor activity (GO:0005024)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Acvrl1 is part of the Acvrl1/Smad1 axis, which is involved in TGF-beta signaling.\n",
            "\n",
            "  Llama Term: TGF-beta signaling pathway\n",
            "  Matched GO Term: transforming growth factor beta receptor signaling pathway (GO:0007179)\n",
            "  Similarity: 0.94\n",
            "  Relationship: ['part_of GO:0071560']\n",
            "  Namespace: biological_process\n",
            "  Justification: Acvrl1 is part of the Acvrl1/Smad1 axis, which is active in lung fibroblasts and antagonizes Tgfbr1/Smad2/3 signalling.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "def auto_match_go_terms(llama_data, go_ids, go_names, go_embeddings, go_namespaces, go_relationships, model, threshold=0.6):\n",
        "    \"\"\"\n",
        "    Given parsed Llama JSON data, match each Llama-generated GO term to the best official GO term.\n",
        "    \"\"\"\n",
        "    matched_go_terms = {}\n",
        "\n",
        "    # If the JSON is nested under \"genes\", extract that list.\n",
        "    if isinstance(llama_data, list) and len(llama_data) > 0 and \"genes\" in llama_data[0]:\n",
        "        gene_entries = llama_data[0][\"genes\"]\n",
        "    else:\n",
        "        gene_entries = llama_data\n",
        "\n",
        "    for gene_entry in gene_entries:\n",
        "        gene_name = gene_entry.get(\"gene\", \"UnknownGene\")\n",
        "        gene_id = gene_entry.get(\"geneID\", \"UnknownID\")     # using id instead of name for the dictionary since some genes have the same name\n",
        "        go_data = gene_entry.get(\"go_data\", [])\n",
        "        unique_matches = {}\n",
        "\n",
        "        for entry in go_data:\n",
        "            llama_term = entry.get(\"term\")\n",
        "            if not llama_term:\n",
        "                continue\n",
        "\n",
        "            best_go_id, best_go_name, best_score, best_idx = get_best_matching_go_term(\n",
        "                llama_term, go_ids, go_names, go_embeddings, model, threshold=threshold\n",
        "            )\n",
        "\n",
        "            if best_go_id and best_idx is not None:\n",
        "                official_relationship = go_relationships[best_idx]\n",
        "                official_namespace = go_namespaces[best_idx]\n",
        "\n",
        "                if best_go_id in unique_matches:\n",
        "                    if best_score > unique_matches[best_go_id][\"similarity\"]:\n",
        "                        unique_matches[best_go_id] = {\n",
        "                            \"llama_term\": llama_term,\n",
        "                            \"matched_go_id\": best_go_id,\n",
        "                            \"matched_go_name\": best_go_name,\n",
        "                            \"similarity\": best_score,\n",
        "                            \"relationship\": official_relationship,\n",
        "                            \"namespace\": official_namespace,\n",
        "                            \"justification\": entry.get(\"justification\")\n",
        "                        }\n",
        "                else:\n",
        "                    unique_matches[best_go_id] = {\n",
        "                        \"llama_term\": llama_term,\n",
        "                        \"matched_go_id\": best_go_id,\n",
        "                        \"matched_go_name\": best_go_name,\n",
        "                        \"similarity\": best_score,\n",
        "                        \"relationship\": official_relationship,\n",
        "                        \"namespace\": official_namespace,\n",
        "                        \"justification\": entry.get(\"justification\")\n",
        "                    }\n",
        "\n",
        "        matched_go_terms[gene_id] = list(unique_matches.values())\n",
        "\n",
        "    return matched_go_terms\n",
        "\n",
        "def replace_terms_with_go(llama_data, matched_go_terms, similarity_threshold=0.65):\n",
        "    \"\"\"\n",
        "    Replaces each llama_data entry's GO term with the matched GO term info from matched_go_terms.\n",
        "    \"\"\"\n",
        "    # If llama_data is nested under \"genes\", re-point gene_entries\n",
        "    if isinstance(llama_data, list) and len(llama_data) > 0 and \"genes\" in llama_data[0]:\n",
        "        gene_entries = llama_data[0][\"genes\"]\n",
        "    else:\n",
        "        gene_entries = llama_data\n",
        "\n",
        "    for gene_entry in gene_entries:\n",
        "        gene_id = gene_entry.get(\"geneID\")\n",
        "        if gene_id in matched_go_terms:\n",
        "            new_go_data = []\n",
        "            for match in matched_go_terms[gene_id]:\n",
        "                # Skip if the match similarity is below threshold\n",
        "                if match[\"similarity\"] < similarity_threshold:\n",
        "                    continue\n",
        "                new_entry = {\n",
        "                    \"term\": match[\"matched_go_name\"],      # replace Llama term\n",
        "                    \"go_id\": match[\"matched_go_id\"],       # new field\n",
        "                    \"relationship\": match[\"relationship\"], # official relationship\n",
        "                    \"namespace\": match[\"namespace\"],\n",
        "                    \"justification\": match[\"justification\"],\n",
        "                    # carry over references if you want:\n",
        "                    \"reference(s)\": gene_entry[\"go_data\"][0].get(\"reference(s)\", [])\n",
        "                }\n",
        "                new_go_data.append(new_entry)\n",
        "            gene_entry[\"go_data\"] = new_go_data\n",
        "    return llama_data\n",
        "\n",
        "# Create a  directory to store PROCESSED llama JSON outputs.\n",
        "processed_dir = os.path.join(os.getcwd(), \"processed_inferences\")\n",
        "if not os.path.exists(processed_dir):\n",
        "    os.makedirs(processed_dir)\n",
        "\n",
        "# Usage example:\n",
        "# 1. Load the Llama output JSON file and parse it\n",
        "llama_output_path = \"./llama_inferences/llama_output_PMC6484421.json\"\n",
        "llama_data = parse_llama_output(llama_output_path)\n",
        "\n",
        "# 2. Match the Llama terms to official GO terms\n",
        "matched_terms = auto_match_go_terms(\n",
        "    llama_data, go_ids, go_names, go_embeddings, go_namespaces, go_relationships, sapbert_model, threshold=0.6\n",
        ")\n",
        "\n",
        "# 3. Replace llama terms with the official GO terms\n",
        "updated_data = replace_terms_with_go(llama_data, matched_terms)\n",
        "\n",
        "# 4. save the processed JSON output to a new file\n",
        "procesed_output_path = os.path.join(processed_dir, f\"processed_output_{doc_id}.json\")\n",
        "with open(procesed_output_path, \"w\", encoding=\"utf-8\") as f:\n",
        "    json.dump(updated_data, f, indent=2)\n",
        "\n",
        "print(\"there are {} genes with matched GO terms.\".format(len(matched_terms)))\n",
        "for gene_id, matches in matched_terms.items():\n",
        "    print(f\"Gene: {gene_dict[gene_id].get_name_from_article()} ({gene_id})\")\n",
        "    for match in matches:\n",
        "        print(f\"  Llama Term: {match['llama_term']}\")\n",
        "        print(f\"  Matched GO Term: {match['matched_go_name']} ({match['matched_go_id']})\")\n",
        "        print(f\"  Similarity: {match['similarity']:.2f}\")\n",
        "        print(f\"  Relationship: {match['relationship']}\")\n",
        "        print(f\"  Namespace: {match['namespace']}\")\n",
        "        print(f\"  Justification: {match['justification']}\\n\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "52hJV2NbJMzp"
      },
      "source": [
        "# 6. Evaluate"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6DBtiTOtJgFO"
      },
      "source": [
        "6.1. Counting and Plotting Similarity Distributions"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 148,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 487
        },
        "id": "sPcTb_-UJLvq",
        "outputId": "4754b2f1-b823-432a-f485-1a874be690ee"
      },
      "outputs": [
        {
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAq4AAAHWCAYAAAC2Zgs3AAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAAPEBJREFUeJzt3Qd4FNXex/F/gJDQO4SSUKQFERCQIqiAaERFePEqKCoioiKggOXK5VJENF59KSJN76XYEOUKKoqghKJIB7HQpBoEEwTpkgRh3ud/3mf32V02lU02J/v9PM9AdnZ39uzMbPLb/5w5E+Y4jiMAAABAPlco2A0AAAAAsoLgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKBMiYMWMkLCwsT16rQ4cOZnJZuXKlee3//ve/efL6Dz74oNSqVUvyszNnzsjDDz8sUVFRZt0MGTIk2E0KSbnxufC3/+lr6GsF0pw5c8xyDxw4ENDlAsg5giuQwR8s1xQZGSnVqlWTuLg4mTx5spw+fTogr3P48GHzx3br1q2S3+TntmXFSy+9ZLbjgAED5J133pH7778/w8dfvHhR3n77bbnpppukYsWKEh4eLpUrV5abb75Z3nzzTUlNTb3kOWfPnpUXXnhBmjRpIsWLF5cyZcrIddddZ5aT2dW0ffex9Kb8+gVBvxiMHj1aGjduLCVKlJAKFSpIs2bN5MknnzT7TkE1bdo0s+0C7ffffzfrrmHDhlKsWDGz77Vq1Ur+/ve/m3UN4P+FOZn9dgVCkP5h6tu3r4wdO1Zq164t58+fl6SkJFPZ/OqrryQmJkY+/fRTE1hc/vrrLzNpyM2qTZs2yTXXXCOzZ882VaSsSktLM/8XLVrU/K/t6tixo8yfP1/+9re/Zeu95qRtuj406EVEREh+1aZNGylSpIisXr0608eeO3dO/ud//keWLl0q1157rXTt2lWqVKkif/zxh6xatUoWL14sffr0kZkzZ7qfk5ycLDfeeKPs2LFDevXqJTfccIOkpKTIRx99JF9//bX07NlT3nvvPSlcuLDf19y3b5+sWbPGa55WiDWsPPLII+55JUuWlO7du0t+otu/devWsnPnTrNeNLBquNq2bZssWrTI7IeuIwI5+Vxk5fV99z8N+RqkA1l1vXDhgnktfR1X1ViDun6x0c9coOh+dvXVV8upU6fkoYceMuH12LFj8sMPP8hnn31m/s+vX2CAvFYkz18RsEiXLl2kZcuW7tvDhw+X5cuXy+233y533HGHCS1aHVEaknTKTX/++aep7LkCa7BoNTK/O3LkiDRq1ChLjx06dKgJrZMmTTJVL09PPfWU7N6923xh8aSBTbf/woULzb7g8sQTT8gzzzwj//u//2vCiFbM/KlTp46ZPD322GNm3n333SeXS0O07ieFCgX+wNrHH38s3333nQnm99577yWv6/pilVufi9ze/7SSrlVk/dKR3hePQNIvRImJifLtt9+aL06eNMzm5efd9d6BfEsrrgC8zZ49W49EOBs3bvR7/0svvWTuf/PNN93zRo8ebeZ5+vLLL5127do5ZcqUcUqUKOHUr1/fGT58uLlvxYoV5vG+k762uuGGG5wrr7zS2bRpk3Pdddc5xYoVc5588kn3fTq5uJY1b948s/wqVao4xYsXd7p27eokJiZ6talmzZpOnz59LnlPnsvMrG36fF2OpzNnzjjDhg1zatSo4RQtWtS811dffdW5ePGi1+N0OQMHDnQWLlxo3p8+tlGjRs4XX3yRpW2TnJzsPPTQQ07lypWdiIgIp0mTJs6cOXMuWRe+0/79+/0uT9dP4cKFnVtuucXJqrVr15plajv8OX/+vFOvXj2nXLlyzp9//pnl5eo+4rttfv31V6dv377m/brW1cyZM70e43rP77//vjNixAinWrVqTlhYmHP8+HGzPF3uL7/84tx2223mZ71/ypQp5rk//PCD07FjR7O/xMTEOO+9916m7YyPjzevd+DAgUwf6+9z4doHPvzwQyc2NtaJjIx02rRpY9qiZsyY4VxxxRVm++o+6bvt/O1/ukx9LRdt24ABA8x+qMsvX76887e//e2SZbk+6ytXrjSPr1SpklO2bFmv+1zP0df03a+0fXv37jU/T5gw4ZL3/+2335r75s6dm+46evTRR80+eOHCBScr1q1b53Tp0sW0U7fbVVdd5UyaNMnrMQkJCU779u3N/fr754477nC2b9/ud9ts27bNueeee8zymjVr5r7/nXfecZo3b27Wn+7LPXv2vOT3yc8//+z06NHD/M7R7VW9enXzuBMnTmTpvQDZRcUVyAHtL/mPf/xDvvzyS+nfv7/fx+hhU63MancC7XKghxv37NljqioqNjbWzB81apQ5NKx9I5VnxUUPF2rVVw9FaxVOD19n5MUXXzSHNLXKpxVHrSB27tzZ9FN1VYazIitt86S5QauOK1askH79+plDx1rB1MrjoUOHZOLEiV6P18P3CxYskMcff1xKlSpl+g3feeedpuqkfSUzOqSvh6B1PQ4aNMh049DD0tqV4cSJE6Zaqm3XPq1aRa1Ro4apmKpKlSr5XeYXX3xhDglnp8qph8PVAw884Pd+rTBqJfL5558321u3QU5odwTt8qDbVN+vvgdtr65jrcT5nnCm/W21Ovf000+bPrmuSp2+P92Prr/+ennllVdMpVSXp5W1ESNGSO/evaVHjx4yY8YM857atm1r1m16atasaf7Xvrz//Oc/c3Ty1TfffGO62wwcONDcjo+PN5+XZ5991vQj1X3j+PHjpr16+FyPdGTHxo0bTVcM/ezofqAnWE2fPt3sP9u3bzdHLjzp6+n61X1eq47+6Odp8ODBpvuGrjeln0mtkrdr186sV93vPOk83ce7deuW4frUbaT7rVbyM6KVf11PVatWNfu7nnyolX/tUuA6WrBs2TKzvbVd2nVCPzevv/66aeOWLVsu6XZw1113Sb169Uy/cFfvQf1dMnLkSLn77rtNFxbtg6vL0H1Iq+1ly5Y1lXXt96/7mq4XbYt+3rUt+nnUPt9AwGU76gIhILOKq9IqxtVXX51uZWnixInm9u+//57uMnT5npVMT1rJ0fu0+uTvPn8VV612nDp1yj1fK1o6/7XXXstWxTWztvlWvD7++GPz2HHjxnk9TitcWvnbs2ePe54+TiuHnvO+//57M//11193MqJVJX3cu+++656XlpbmtG3b1ilZsqTXe9f2aYUxM0OHDjXL3Lp1q9f81NRUs+1c09GjR933de/e3TxHK5rpWbBggXnM5MmTnZxWXPv16+dUrVrV67VVr169zP7nqua6tn+dOnUuqfDq8vQ+PUrgou3WCr5uG63Su+zcufOSyqU/+hoNGjQwj9X1/OCDD5oqsFbDs1px1eqcZ/XzjTfeMPOjoqK8tqMeQfCtmGel4uqv0u2qlL/99tuXfNa1OvnXX395Pd634qr0KIHn58S3/Tt27PDaNytWrOj38+YpKSnJVHr1+Q0bNnQee+wxU6H1rVpq+2rXrm3eu+++53lkQ6umWqE/duyY12esUKFCzgMPPHDJttFqqyetVmsF+MUXX/Sa/+OPPzpFihRxz//uu+/M8+fPn5/h+wMCiVEFgBzSqktGowtoRUJ98skn5kSSnNAqrZ4kllVaLdPqjoueqKWVGT25KDfp8rUvoPbv9KTVTs0UWiX0pBXIK664wn1bq9KlS5c2Jyxl9jpa1bnnnnu8+jvq6+rJQXoiVXZp5dK1PX1fSytwrslVZVSu7e65rn257nMtP7t0vemJXnqimP589OhR96RVrpMnT5rqmSet1qVXWdeqmee+2aBBA1Nx1Yqai87T+zLbDvoa69evNxV118mMWgXWfU0rb/5GYPClJ7Z5Vv70ZC+llXfP9eqan1mb/LXRRU+w0qMXdevWNe/Pd70pPXJyOf1ZdT3qCWhaYXXRow66vTKr5mvV9vvvvzd9nLXKrJVvrdjryAJaRXdVQbXSuX//flNpd/1+cXFVvX/77TdzhEWPQpQvX97rM6YjZvj7XaCv60mPhujvLH1Pnvudfva0MqtHVpSroqrvU/vfA3mB4ArkkAaljIKLnlWuh+Y0MOgfJj1k+eGHH2YrxFavXj1bJ2boHxXfP2b6xzq3x6H85ZdfzHBhvutDD9u77vekozL4KleunPmjndnr6Hv0PeEovdfJClebfYcc0m2nh2V10iGx/D0noy8uWQm3GdFDs3q4VYfi8gzQOrm+zGh3EE/pHd7XQOXbVUJDhx5C9z3Mr/Mz2w6ux+lhfN23dNITjDT4TpkyxYStzPjuA64QFB0d7Xd+VtrkSQ+P62F/XZ5+AdSRAHQd6DrV0O8ro64RWaFBUr9kzJ071z1PQ6x+hjt16pTp8zX0a1cGDZ67du0y3WdcXRdco1ns3bvXPbJBelyfAd0WvvRzogHUtyuE73vXkxE1LOtnzXff024Jrv1Onzds2DD5z3/+Y9avfqGaOnWq3/ULBAp9XIEc+PXXX80vZw2FGVV8dFgkrU58/vnnsmTJEvnggw/MHzHtG5uV6k52+qVmVXr9EbWPXV6cQa3Se51gjM6nQw+pn376SZo2beqer3+kXX1T33333UsCgJ5Zr8MUaZ8/f/Q+ldWRDXy5vuBotS69fo+ew7FltL+kt74DtR20Gq39UHVIMe1XqYFt3LhxGT4nt9uklV8dyk2rk9pnVwOw7vv6BdLfl8dAfNb0iIf2uda+tVdddZXpw6t9Z7MzsoO2sX79+ma67bbbTHjU9elZMQ803/eu60fboUdK/G0Pz6MT48ePN9VdPbKkv9f06If2V163bp35YgQEGsEVyAE9iUJphSEj+gdLD4nqNGHCBHPyg57UoWFWQ1GgryiklRLfP/Z6IpNnwNHKplad/FVqPIdnyk7bNLjoCSFaZfSsMOo4n677A0GXo4FQ/7B6hoHLeR09iUX/OGs40JOUskJPjtE/znpykr/gql8CtPKm61ortzmhwVnXpS4rpyd35TV9v9oFRL8EBJteRU4DvwYrz6G6/O372ZHR5+KWW24x2033Je3ioIfPM7vwRUb086jrVKuwytW9RtdvevuE6zOgVVtf+jnRymhmw13p6+jvDq2oaoDOjIZ0nfREPQ3tus9rd4fMvrwAOUFXASCb9OxmPRSqv9QzCjo6qLgvPdteufoAuv6AXO4fUxcNUp6Hr/WPt/7R03Dm+UdJqyGeY23qWcAHDx70WlZ22nbrrbeagKWHiT3paAL6h97z9S+Hvo5eCEIr1y46uL2e7axVIL0IQHbpIWutFmp1ybf96VX7dHQFDQ5a0dN150u/nPz888/mDPmcVvI0TGt/T+3n6i8IaleCYNH+mHrI2d+XHz1j399h6rym6893u+l+ovvp5dDPRXqfCR1NQvtfa5cg7ferYc63Ku6P9hf2N5LBhg0bTN9c1/ps3ry5+b2joxv4tsH1XrXLgf6eeeutt7weo/uQVkT1M5QZHWFC15+OiuG7DvW2tsnVf1s/f570PeuXyqz0cwZygoorkAENM1ql0F/OOjSRhlbt86hVDT0MmNHVgHQ4Ke0qoIf79PHaL0yH+dHDZ+3bt3eHSO0bp9UJra7pH0Wt1OS0v52ejKHL1j6Q2l79A6fdGTyH7NJDjhpotTqkJ19ovzk9FO55slR226Z9+/TKXRrYtL+jHnLXP5J6+FAP1fouO6d0aK433njDHJrcvHmzOblH34sOOaXvNaf9SfW5etKLHl6eN2+eeT96YoyGM122Dn/lG8b0S4JW0nWYIz2RRocM0z/WemKLXlVJ+zi7Tl7KqZdfftlU53W96zbUbgf6hUhPLtIKt78vR3lBPwN6lSodAk2H69IvDXry1KxZs8w6COTVq3JKq+J6ZES7COh6W7t2rVlnGQ23lhUtWrQwfVG1mqifLd1PPPuwancB7Z+q2+1f//pXlpap7dQqrXa10OVrv3btS6rrU3/H6NB7SgOhvrbunxpO9XOuQVV/R+nwe3qSlHr11VfNl0XtIqEnzbmGw9J1kZVto59XfX96wRX9POuV2/SzpZ8RveCGfg51yDX9fajDqulwWlqZ1d+T+l5cX7qAXBHQMQqAAsI1DI5r0uGbdJiem266yQwt5TlcT3rD/ugA4N26dTODvevz9X8ddkYH7Pb0ySefmEHldZgZfxcg8Ce94bB0AHodPkiHwtHhjnQ4KB143tf48ePN0Fk6JJFeIEEvcuC7zIza5m84otOnT5uhpfR9hoeHmwH4M7oAga/0hunypUMu6YD8OsyQrlcdfN3fkF1ZHQ7Lc6ghXU6nTp3MYPX6nvU1brzxRjMk2blz5y55jr7nMWPGmO2k67tUqVJmfeoFEXzfd04vQKDvV9dXdHS0Wa+6H2qbPC9+4dr+/oYlcl2AwFd6+1dW1tu+ffucUaNGmYsG6L6m60qHc9LnLV++PMsXIPCkQ07pfN1nPPl7b1kZDkuHi3LtJzpUWlxcnBnuy3c/y2joO3/DYenQVfo+dVu7LkDgS9erDj2lF4/ICr3wwjPPPGMG+3ftezoM2l133eVs2bLlksevXr3a/C7SNui21Ytw+A4lt2zZMrMv6n5ZunRpczGS9C5AkN6QfR999JEZJkxfQycdqku3265du9z7gV6EQy8W4brIg17MQl8byC1h+k/uRGIAAEKPXupXj34kJCQEuylAgUMfVwAAAmTTpk1mHNX0rqoG4PJQcQUA4DLpyU/a71pHMdC+0drnN6M+8AByhoorAACXSU8S1JOl9Cpd77//PqEVyCVUXAEAAGAFKq4AAACwAsEVAAAAVijwFyDQS0MePnzYDJ4c6MtrAgAA4PJpz1W98mO1atW8LukdcsFVQ2t0dHSwmwEAAIBM6OXH9QqTIRtcXZeA1BVRunTpYDcHAAAAPk6dOmUKjZldurvAB1dX9wANrQRXAACA/Cuzbp2cnAUAAAArEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALBC0IProUOH5L777pMKFSpIsWLF5KqrrpJNmzZ5DUg7atQoqVq1qrm/c+fOsnv37qC2GQAAACEWXI8fPy7t2rWT8PBw+eKLL2T79u0yfvx4KVeunPsxr7zyikyePFlmzJgh69evlxIlSkhcXJykpKQEs+kAAADIY2GOljSD5LnnnpNvv/1WvvnmG7/3a9P00l9PPfWUPP3002beyZMnpUqVKjJnzhzp1atXlga0LVOmjHke47gCAADkP1nNa0GtuH766afSsmVLueuuu6Ry5cpy9dVXy7///W/3/fv375ekpCTTPcBF31Tr1q1l7dq1fpeZmppq3rznBAAAAPsFNbju27dPpk+fLvXq1ZOlS5fKgAED5IknnpC33nrL3K+hVWmF1ZPedt3nKz4+3oRb16SXDwMAAID9ghpcL168KM2bN5eXXnrJVFsfeeQR6d+/v+nPmlPDhw83ZWbXdPDgwYC2GQAAACEYXHWkgEaNGnnNi42NlcTERPNzVFSU+T85OdnrMXrbdZ+viIgI0zfCcwIAAID9ghpcdUSBXbt2ec37+eefpWbNmubn2rVrm4CakJDgvl/7rOroAm3bts3z9gIAACB4igTxtWXo0KFy7bXXmq4Cd999t2zYsEHefPNNM6mwsDAZMmSIjBs3zvSD1SA7cuRIM9JA9+7dg9l0AAAAhFJwveaaa2ThwoWmX+rYsWNNMJ00aZL07t3b/Zhnn31Wzp49a/q/njhxQtq3by9LliyRyMjIYDYdAADkEe1CePTo0WA3I+RUrFhRYmJiJD8J6jiueYFxXAEAsDu0NmgYKynn/gx2U0JOZLHismvnjjwJr1nNa0GtuAIAAGREK60aWivc/pSEV2CIy7xy/thBOfbZeLP+81PVleAKAADyPQ2tEVF1g90MhPKoAgAAAEBWEVwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKAAAAKxBcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAViC4AgAAwAoEVwAAAFiB4AoAAAArEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKAAAAKxBcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAViC4AgAAwAoEVwAAAFiB4AoAAAArBDW4jhkzRsLCwrymhg0buu9PSUmRgQMHSoUKFaRkyZJy5513SnJycjCbDAAAgFCtuF555ZXy22+/uafVq1e77xs6dKgsWrRI5s+fL6tWrZLDhw9Ljx49gtpeAAAABEeRoDegSBGJioq6ZP7Jkydl5syZMnfuXOnUqZOZN3v2bImNjZV169ZJmzZtgtBaAAAAhGzFdffu3VKtWjWpU6eO9O7dWxITE838zZs3y/nz56Vz587ux2o3gpiYGFm7dm26y0tNTZVTp055TQAAALBfUINr69atZc6cObJkyRKZPn267N+/X6677jo5ffq0JCUlSdGiRaVs2bJez6lSpYq5Lz3x8fFSpkwZ9xQdHZ0H7wQAAAAFuqtAly5d3D83adLEBNmaNWvKhx9+KMWKFcvRMocPHy7Dhg1z39aKK+EVAADAfkHvKuBJq6v169eXPXv2mH6vaWlpcuLECa/H6KgC/vrEukREREjp0qW9JgAAANgvXwXXM2fOyN69e6Vq1arSokULCQ8Pl4SEBPf9u3btMn1g27ZtG9R2AgAAIMS6Cjz99NPStWtX0z1Ah7oaPXq0FC5cWO655x7TP7Vfv37msH/58uVN5XTw4MEmtDKiAAAAQOgJanD99ddfTUg9duyYVKpUSdq3b2+GutKf1cSJE6VQoULmwgM6WkBcXJxMmzYtmE0GAABAKAbXefPmZXh/ZGSkTJ061UwAAAAIbfmqjysAAACQHoIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKAAAAKxBcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAViC4AgAAwAoEVwAAAFiB4AoAAAArEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKAAAAKxBcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAViC4AgAAwAoEVwAAAFiB4AoAAAArEFwBAABgBYIrAAAArEBwBQAAgBXyTXB9+eWXJSwsTIYMGeKel5KSIgMHDpQKFSpIyZIl5c4775Tk5OSgthMAAAAhHFw3btwob7zxhjRp0sRr/tChQ2XRokUyf/58WbVqlRw+fFh69OgRtHYCAAAghIPrmTNnpHfv3vLvf/9bypUr555/8uRJmTlzpkyYMEE6deokLVq0kNmzZ8uaNWtk3bp1QW0zAAAAQjC4aleA2267TTp37uw1f/PmzXL+/Hmv+Q0bNpSYmBhZu3ZtustLTU2VU6dOeU0AAACwX5Fgvvi8efNky5YtpquAr6SkJClatKiULVvWa36VKlXMfemJj4+X559/PlfaCwAAgBCsuB48eFCefPJJee+99yQyMjJgyx0+fLjpZuCa9HUAAABgv6AFV+0KcOTIEWnevLkUKVLETHoC1uTJk83PWllNS0uTEydOeD1PRxWIiopKd7kRERFSunRprwkAAAD2C1pXgRtvvFF+/PFHr3l9+/Y1/Vj//ve/S3R0tISHh0tCQoIZBkvt2rVLEhMTpW3btkFqNQAAAEIuuJYqVUoaN27sNa9EiRJmzFbX/H79+smwYcOkfPnypnI6ePBgE1rbtGkTpFYDAAAgJE/OyszEiROlUKFCpuKqowXExcXJtGnTgt0sAAAAhHpwXblypddtPWlr6tSpZgIAAEBoC/o4rgAAAEBWEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKAAAAKxBcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAViC4AgAAwAoEVwAAAFiB4AoAAAArEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAACg4AbXOnXqyLFjxy6Zf+LECXMfAAAAkC+C64EDB+TChQuXzE9NTZVDhw4Fol0AAACAlyKSDZ9++qn756VLl0qZMmXctzXIJiQkSK1atbKzSAAAACDwwbV79+7m/7CwMOnTp4/XfeHh4Sa0jh8/PjuLBAAAAAIfXC9evGj+r127tmzcuFEqVqyYnacDAAAAeRNcXfbv35/zVwQAAADyKrgq7c+q05EjR9yVWJdZs2bldLEAAABA4ILr888/L2PHjpWWLVtK1apVTZ9XAAAAIN8F1xkzZsicOXPk/vvvD3yLAAAAgECN45qWlibXXnttTp4KAAAA5F1wffjhh2Xu3Lk5e0UAAAAgr7oKpKSkyJtvvinLli2TJk2amDFcPU2YMCEniwUAAAACG1x/+OEHadasmfn5p59+8rqPE7UAAACQb4LrihUrAt8SAAAAINB9XAEAAAArKq4dO3bMsEvA8uXLL6dNAAAAQGCCq6t/q8v58+dl69atpr9rnz59crJIAAAAIPDBdeLEiX7njxkzRs6cOZOTRQIAAAB518f1vvvuk1mzZgVykQAAAEDgg+vatWslMjIykIsEAAAAct5VoEePHl63HceR3377TTZt2iQjR47M8nKmT59upgMHDpjbV155pYwaNUq6dOnivtDBU089JfPmzZPU1FSJi4uTadOmSZUqVXLSbAAAAIRaxbVMmTJeU/ny5aVDhw6yePFiGT16dJaXU6NGDXn55Zdl8+bNJvR26tRJunXrJtu2bTP3Dx06VBYtWiTz58+XVatWyeHDhy8JzQAAAAgNOaq4zp49OyAv3rVrV6/bL774oqnArlu3zoTamTNnyty5c02gdb1ubGysub9NmzYBaQMAAAAKcHB10Urpjh073If5r7766hwv68KFC6ayevbsWWnbtq1Ztg6z1blzZ/djGjZsKDExMaYvbXrBVbsU6ORy6tSpHLcJAAAAlgfXI0eOSK9evWTlypVStmxZM+/EiRPmwgTaH7VSpUpZXtaPP/5ogqr2Zy1ZsqQsXLhQGjVqZMaFLVq0qHv5Ltq/NSkpKd3lxcfHy/PPP5+TtwUAAICC1sd18ODBcvr0adMX9Y8//jCTXnxAq5tPPPFEtpbVoEEDE1LXr18vAwYMMBcw2L59u+TU8OHD5eTJk+7p4MGDOV4WAAAALK+4LlmyRJYtW2b6m7polXTq1Kly8803Z2tZWlWtW7eu+blFixayceNGee2116Rnz56SlpZmKrmeVdfk5GSJiopKd3kRERFmAgAAQMGSo4rrxYsXJTw8/JL5Ok/vuxz6fO2jqiFWl5eQkOC+b9euXZKYmGi6FgAAACC05Kjiqmf5P/nkk/L+++9LtWrVzLxDhw6Z4atuvPHGbB3W1zFb9YQr7XqgIwhov9mlS5eaYbb69esnw4YNM8NtlS5d2nRR0NDKiAIAAAChJ0fBdcqUKXLHHXdIrVq1JDo62szTvqSNGzeWd999N1sneT3wwAPm4gUaVJs0aWJC60033WTunzhxohQqVEjuvPNOrwsQAAAAIPTkKLhqWN2yZYvp57pz504zT/u7eg5dlRU6TmtG9PKx2m9WJwAAAIS2bPVxXb58uTkJS0cPCAsLM5VRPXyv0zXXXGPGcv3mm29yr7UAAAAIWdkKrpMmTZL+/fub/qa+9FD/o48+KhMmTAhk+wAAAIDsB9fvv/9ebrnllnTv16Gw9IpXAAAAQFCDq46h6m8YLJciRYrI77//Hoh2AQAAADkPrtWrVzdXyErPDz/8IFWrVs3OIgEAAIDAB9dbb71VRo4cKSkpKZfcd+7cORk9erTcfvvt2VkkAAAAEPjhsP75z3/KggULpH79+jJo0CBp0KCBma9DYumQVRcuXJARI0ZkZ5EAAABA4INrlSpVZM2aNTJgwABz1SvHccx8HRpLLw6g4VUfAwAAAAT9AgQ1a9aUxYsXy/Hjx2XPnj0mvNarV0/KlSsX8MYBAAAAl3XlLKVBVS86AAAAAOS7k7MAAACAYCG4AgAAwAoEVwAAAFiB4AoAAAArEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKAAAAKxBcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAViC4AgAAwAoEVwAAAFiB4AoAAAArEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKQQ2u8fHxcs0110ipUqWkcuXK0r17d9m1a5fXY1JSUmTgwIFSoUIFKVmypNx5552SnJwctDYDAAAgBIPrqlWrTChdt26dfPXVV3L+/Hm5+eab5ezZs+7HDB06VBYtWiTz5883jz98+LD06NEjmM0GAABAEBSRIFqyZInX7Tlz5pjK6+bNm+X666+XkydPysyZM2Xu3LnSqVMn85jZs2dLbGysCbtt2rS5ZJmpqalmcjl16lQevBMAAACEVB9XDaqqfPny5n8NsFqF7dy5s/sxDRs2lJiYGFm7dm263Q/KlCnjnqKjo/Oo9QAAAAiJ4Hrx4kUZMmSItGvXTho3bmzmJSUlSdGiRaVs2bJej61SpYq5z5/hw4ebAOyaDh48mCftBwAAQAHuKuBJ+7r+9NNPsnr16staTkREhJkAAABQsOSLiuugQYPks88+kxUrVkiNGjXc86OioiQtLU1OnDjh9XgdVUDvAwAAQOgIanB1HMeE1oULF8ry5culdu3aXve3aNFCwsPDJSEhwT1Ph8tKTEyUtm3bBqHFAAAACMmuAto9QEcM+OSTT8xYrq5+q3pSVbFixcz//fr1k2HDhpkTtkqXLi2DBw82odXfiAIAAAAouIIaXKdPn27+79Chg9d8HfLqwQcfND9PnDhRChUqZC48oMNcxcXFybRp04LSXgAAAIRocNWuApmJjIyUqVOnmgkAAAChK1+cnAUAAABkhuAKAAAAKxBcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAViC4AgAAwAoEVwAAAFiB4AoAAAArEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKAAAAKxBcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAViC4AgAAwAoEVwAAAFiB4AoAAAArEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWKFIsBsAAECgJCYmytGjR4PdjJBTsWJFiYmJCXYzEAIIrgCAAhNaGzSMlZRzfwa7KSEnslhx2bVzB+EVuY7gCgAoELTSqqG1wu1PSXiF6GA3J2ScP3ZQjn023qx/gityG8EVAFCgaGiNiKob7GYAyAWcnAUAAAArEFwBAABgBYIrAAAArEBwBQAAgBUIrgAAALACwRUAAABWCGpw/frrr6Vr165SrVo1CQsLk48//tjrfsdxZNSoUVK1alUpVqyYdO7cWXbv3h209gIAACBEg+vZs2eladOmMnXqVL/3v/LKKzJ58mSZMWOGrF+/XkqUKCFxcXGSkpKS520FAABACF+AoEuXLmbyR6utkyZNkn/+85/SrVs3M+/tt9+WKlWqmMpsr1698ri1AAAACKZ828d1//79kpSUZLoHuJQpU0Zat24ta9euTfd5qampcurUKa8JAAAA9su3wVVDq9IKqye97brPn/j4eBNwXVN0NNerBgAAKAjybXDNqeHDh8vJkyfd08GDB4PdJAAAABTk4BoVFWX+T05O9pqvt133+RMRESGlS5f2mgAAAGC/fBtca9eubQJqQkKCe572V9XRBdq2bRvUtgEAACDERhU4c+aM7Nmzx+uErK1bt0r58uUlJiZGhgwZIuPGjZN69eqZIDty5Egz5mv37t2D2WwAAACEWnDdtGmTdOzY0X172LBh5v8+ffrInDlz5NlnnzVjvT7yyCNy4sQJad++vSxZskQiIyOD2GoAAACEXHDt0KGDGa81PXo1rbFjx5oJAAAAoS3f9nEFAAAAPBFcAQAAYAWCKwAAAKxAcAUAAIAVgnpyFmCTxMREOXr0aLCbEVIqVqxohsYDAEARXIEshtYGDWMl5dyfwW5KSIksVlx27dxBeAUAGARXIAu00qqhtcLtT0l4hehgNycknD92UI59Nt6se4IrAEARXIFs0NAaEVU32M0AACAkcXIWAAAArEBwBQAAgBUIrgAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKAAAAKxQJdgMKosTERDl69GiwmxFyKlasKDExMcFuBgAAyCUE11wIrQ0axkrKuT+D3ZSQE1msuOzauYPwCgBAAUVwDTCttGporXD7UxJeITrYzQkZ548dlGOfjTfrn+AKAEDBRHDNJRpaI6LqBrsZAAAABQYnZwEAAMAKBFcAAABYga4CAEIWI4AEByOAAMgpgiuAkMQIIMHDCCAAcorgCiAkMQJIcDACCIDLQXAFENIYAQQA7MHJWQAAALACwRUAAABWILgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAIAAMAKBFcAAABYgeAKAAAAKxBcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAVrAiuE6dOlVq1aolkZGR0rp1a9mwYUOwmwQAAIA8lu+D6wcffCDDhg2T0aNHy5YtW6Rp06YSFxcnR44cCXbTAAAAkIfyfXCdMGGC9O/fX/r27SuNGjWSGTNmSPHixWXWrFnBbhoAAADyUBHJx9LS0mTz5s0yfPhw97xChQpJ586dZe3atX6fk5qaaiaXkydPmv9PnTqVBy0WOXPmzP+3I2mPXExLyZPXhMj5P351r//c2NZs17zHNi2YcnO7sk2Dg89qwXQ+l7erL9drOI6T8QOdfOzQoUPaemfNmjVe85955hmnVatWfp8zevRo8xwmJiYmJiYmJiaxajp48GCG2TBfV1xzQquz2ifW5eLFi/LHH39IhQoVJCwsLKhty+/02050dLQcPHhQSpcuHezmIADYpgUT27XgYZsWTGzXrNNK6+nTp6VatWoZPi5fB9eKFStK4cKFJTk52Wu+3o6KivL7nIiICDN5Klu2bK62s6DRDxcfsIKFbVowsV0LHrZpwcR2zZoyZcrYfXJW0aJFpUWLFpKQkOBVQdXbbdu2DWrbAAAAkLfydcVV6WH/Pn36SMuWLaVVq1YyadIkOXv2rBllAAAAAKEj3wfXnj17yu+//y6jRo2SpKQkadasmSxZskSqVKkS7KYVONrFQsfL9e1qAXuxTQsmtmvBwzYtmNiugRemZ2jlwnIBAACAgMrXfVwBAAAAF4IrAAAArEBwBQAAgBUIrgAAALACwTXETJ06VWrVqiWRkZHSunVr2bBhQ4aPP3HihAwcOFCqVq1qzoqsX7++LF68OM/ai8Bu0w4dOpgryPlOt912W562GYH/rOpQgQ0aNJBixYqZK/UMHTpUUlK4rrut2/T8+fMyduxYueKKK8zjmzZtakbUQf7x9ddfS9euXc2VnvT36Mcff5zpc1auXCnNmzc3f0/r1q0rc+bMyZO2FigZXhAWBcq8efOcokWLOrNmzXK2bdvm9O/f3ylbtqyTnJzs9/GpqalOy5YtnVtvvdVZvXq1s3//fmflypXO1q1b87ztCMw2PXbsmPPbb7+5p59++skpXLiwM3v27DxvOwK3Xd977z0nIiLC/K+f06VLlzpVq1Z1hg4dmudtR2C26bPPPutUq1bN+fzzz529e/c606ZNcyIjI50tW7bkedvh3+LFi50RI0Y4CxYs0NGZnIULF2b4+H379jnFixd3hg0b5mzfvt15/fXXze/fJUuW5FmbCwKCawhp1aqVM3DgQPftCxcumF+M8fHxfh8/ffp0p06dOk5aWloethK5uU19TZw40SlVqpRz5syZXGwlcnu76mM7derkNU//OLZr1y7X24rc2ab6xWPKlCle83r06OH07t0719uK7MtKcNUvI1deeaXXvJ49ezpxcXG53LqCha4CISItLU02b94snTt3ds8rVKiQub127Vq/z/n000/NpXW1q4Be8KFx48by0ksvyYULF/Kw5QjkNvU1c+ZM6dWrl5QoUSIXW4rc3q7XXnuteY7r0PO+fftMl55bb701z9qNwG7T1NRU00XAk3YDWb16da63F7lDt7XnPqDi4uKy/Psallw5C4Fx9OhREzh9rzimt3fu3On3OfrHb/ny5dK7d2/zR3DPnj3y+OOPm75XeiUQ2LdNPWnI+emnn0x4hd3b9d577zXPa9++vR5Fk7/++ksee+wx+cc//pFHrUagt6kGmgkTJsj1119v+rkmJCTIggULKBxYTK/+6W8fOHXqlJw7d858MUHmqLgiXRcvXpTKlSvLm2++KS1atDCX3x0xYoTMmDEj2E1DAGhgveqqq6RVq1bBbgouk57woUdDpk2bJlu2bDEB5/PPP5cXXngh2E1DDr322mtSr149adiwoRQtWlQGDRokffv2NZVaIJRRcQ0RFStWlMKFC0tycrLXfL0dFRXl9zk6kkB4eLh5nktsbKz51qiHvvSXKezapi5nz56VefPmmbOWYf92HTlypNx///3y8MMPm9v6hUS38SOPPGK+bBJ27NumlSpVMmep68gQx44dM2euP/fcc1KnTp08ajUCTbe1v32gdOnSVFuzgd9mIUJDplZN9XCTZ0VVb2s/Vn/atWtnugfo41x+/vlnE2gJrXZuU5f58+ebPnT33XdfHrQUub1d//zzz0vCqesL5/+fNwJbP6vaz7V69eqm+8dHH30k3bp1y4MWIzfotvbcB9RXX32V6T4AH8E+Owx5OxyLDpkzZ84cMxTHI488YoZjSUpKMvfff//9znPPPed+fGJiojnjfNCgQc6uXbuczz77zKlcubIzbty4IL4LXM42dWnfvr05mxUFY7uOHj3afFbff/99M+TOl19+6VxxxRXO3XffHcR3gcvZpuvWrXM++ugjMxTW119/bUaNqF27tnP8+PEgvgt4On36tPPdd9+ZSePUhAkTzM+//PKLuV+3p25X3+GwnnnmGWfHjh3O1KlTGQ4rBwiuIUbHjYuJiTHjCerwLPrL0eWGG25w+vTp4/X4NWvWOK1btza/cHVorBdffNH566+/gtByBGqb7ty50/yS1XCDgrFdz58/74wZM8aEVR3rMzo62nn88ccJORZvUx0zOzY21vzurVChgglAhw4dClLL4c+KFSvM71LfybUd9X/drr7PadasmdkH9G8qY2hnX5j+41uFBQAAAPIb+rgCAADACgRXAAAAWIHgCgAAACsQXAEAAGAFgisAAACsQHAFAACAFQiuAAAAsALBFQAAAFYguAJADsyZM0fKli0b7GbIgQMHJCwsTLZu3XpZy+nQoYMMGTLEfbtWrVoyadKky27fgw8+KN27d7/s5QCAIrgCKJCSkpJk8ODBUqdOHYmIiJDo6Gjp2rWrJCQkBGT5PXv2lJ9//lly2/79++Xee++VatWqSWRkpNSoUUO6desmO3fuNPfr+/rtt9+kcePGl/U6CxYskBdeeEEC7bXXXjMhP72ADADZUSRbjwYAC2gVsl27dqYi+uqrr8pVV10l58+fl6VLl8rAgQPdoe9yFCtWzEy5Sdt80003SYMGDUywrFq1qvz666/yxRdfyIkTJ8xjChcuLFFRUZf9WuXLl5dAunDhgqkElylTJqDLBRDiHAAoYLp06eJUr17dOXPmzCX3HT9+3P3zL7/84txxxx1OiRIlnFKlSjl33XWXk5SU5L5/69atTocOHZySJUua+5s3b+5s3LjR3Dd79mynTJky7seOHj3aadq0qfP22287NWvWdEqXLu307NnTOXXqlPsxFy5ccF566SWnVq1aTmRkpNOkSRNn/vz56b6P7777ztFf0wcOHEj3Mfv37zeP0ceqFStWmNtLlixxmjVrZl6nY8eOTnJysrN48WKnYcOG5r3cc889ztmzZ93LueGGG5wnn3zSfVvfw8SJE923x48f7zRu3NgpXry4U6NGDWfAgAHO6dOn3fe71scnn3zixMbGOoULFzZt69Onj9OtWzfzGP1Z2+Y57du3z7niiiucV1991e973717d7rvHUDooasAgALljz/+kCVLlpjKaokSJS6539Uv9eLFi+aQuz5+1apV8tVXX8m+fftMFwCX3r17m0PzGzdulM2bN8tzzz0n4eHh6b723r175eOPP5bPPvvMTLrcl19+2X1/fHy8vP322zJjxgzZtm2bDB06VO677z7zOH8qVaokhQoVkv/+97+mgpkdY8aMkSlTpsiaNWvk4MGDcvfdd5s+q3PnzpXPP/9cvvzyS3n99dezvDxtx+TJk02733rrLVm+fLk8++yzXo/5888/5V//+pf85z//MY+rXLnyJd0G2rZtK/379zfdG3SKiYmRhx56SGbPnu31WL19/fXXS926dbP1vgEUcMFOzgAQSOvXrzeVugULFmT4uC+//NJUBRMTE93ztm3bZp67YcMGc1srk3PmzPH7fH8VV61GelZYn3nmGad169bm55SUFHP/mjVrvJbTr18/U/1Mz5QpU8zztC1aOR07dqyzd+/eTCuuy5Ytcz8mPj7ezPN83qOPPurExcVlueLqSyvFFSpU8Fof+hpapfbkWXH19zrq0KFDZlvotlNpaWlOxYoV0133AEIXFVcABYrjaH7K3I4dO8yJTTq5NGrUyFRk9T41bNgwefjhh6Vz586mcqoV1YzomfilSpVy39Y+qUeOHDE/79mzx1Qktc9qyZIl3ZNWYDNarlaO9USz9957z1Qr58+fL1deeaWpEGekSZMm7p+rVKkixYsXNyeqec5ztS0rli1bJjfeeKNUr17dvMf7779fjh07Zt6TS9GiRb1eN6v0xLPbbrtNZs2aZW4vWrRIUlNT5a677sr2sgAUbARXAAVKvXr1zElBgTgBSw+36yFvDVV6aFyD7cKFC9N9vG83Am2HdklQZ86cMf/rYXoduso1bd++3XQFyIgGRR0R4cUXX5Tvv/9errvuOhk3blyGz/Fsi7Yjo7Zl5WS322+/3YTSjz76yHSbmDp1qrkvLS3N/Tg9WU2XmxP6BWHevHly7tw5001Au2xo2AYATwRXAAWKnh0fFxdngtXZs2cvud91Nn5sbKzp+6mTi4ZIvV8Dqkv9+vVNX1TtE9qjR49L+mJmlS5Th+VKTEw0/TY9J8+qb2Y0GDZs2NDve8stGlQ15I4fP17atGlj1snhw4dztCytyvrrr3vrrbeaPsnTp083fZS13ysA+CK4AihwNLRqOGrVqpWpEO7evdsc/teTi/Rwu9LD/zpMlp6AtWXLFtmwYYM88MADcsMNN0jLli1N5W/QoEGycuVK+eWXX+Tbb781J2lp4M0JrZo+/fTTJgTryU3aPUBfV0+Q0tv+aEVWTyDTiqyGau1uMHPmTHNIXefnFQ3XOjSXtlVPYHvnnXfMCWY5od0p1q9fb6q4R48edVd9dVgvvVjB8OHDTdXctZ0AwBPBFUCBo305NRR27NhRnnrqKTM4v/Yt1YsPaEXPVbn85JNPpFy5cubsdQ2y+rwPPvjAHaS0D6eGWa0w6ln5Xbp0keeffz7H7dIB/keOHGlGF9AAfMstt5iuA7Vr1/b7eB3RQIOevmbr1q2lefPm5sx8vT1ixAjJK02bNpUJEyaYEQN0XWp/W30POaHhXdetVqB11AStQLv069fPdD3o27dvAFsPoCAJ0zO0gt0IAAC++eYbcwKYdt/Qk8cAwBfBFQAQVDqCwO+//y59+vQxVwHTii4A+ENXAQBAUL3//vtSs2ZNc2LcK6+8EuzmAMjHqLgCAADAClRcAQAAYAWCKwAAAKxAcAUAAIAVCK4AAACwAsEVAAAAViC4AgAAwAoEVwAAAFiB4AoAAACxwf8BiFfELV/O4IQAAAAASUVORK5CYII=",
            "text/plain": [
              "<Figure size 800x500 with 1 Axes>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "import matplotlib.pyplot as plt\n",
        "from collections import Counter\n",
        "\n",
        "# Suppose matched_terms is your dictionary of gene predictions (from auto_match_go_terms)\n",
        "# where each value is a list of dictionaries with a key 'similarity'\n",
        "similarity_scores = []\n",
        "for gene, matches in matched_terms.items():\n",
        "    for match in matches:\n",
        "        similarity_scores.append(match['similarity'])\n",
        "\n",
        "# Bin the similarity scores, e.g., rounded to one decimal place.\n",
        "binned_scores = [round(score, 1) for score in similarity_scores]\n",
        "score_counts = Counter(binned_scores)\n",
        "\n",
        "# Prepare data for the bar plot.\n",
        "bins = sorted(score_counts.keys())\n",
        "counts = [score_counts[b] for b in bins]\n",
        "\n",
        "plt.figure(figsize=(8, 5))\n",
        "plt.bar(bins, counts, width=0.08, edgecolor='black')\n",
        "plt.xlabel('Cosine Similarity')\n",
        "plt.ylabel('Count')\n",
        "plt.title('Distribution of GO Term Similarity Scores')\n",
        "plt.xticks(bins)\n",
        "plt.show()\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ONKEXAOtZtjO"
      },
      "source": [
        "## 6.2. Scoring against NCBI Gene database\n",
        "\n",
        "Need to make a new scoring system: possibly True Positive (what we found), Nobel Positive (NP), Not Detected\n",
        "\n",
        "- for cases where we have Go terms that are alike like \"Wnt signaling pathway\" and \"canonical Wnt signaling pathway\", we can embed the terms and check if they are similar. If they are, then we can consider them as the same term. If not, then we can consider them as different terms."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# term = \"Cul4A-RING E3 ubiquitin ligase complex\"\n",
        "# example_llama_term = \"ubiquitin ligase complex\"\n",
        "# def compute_semantic_similarity(term1, term2, model):\n",
        "#     \"\"\"\n",
        "#     Returns a cosine similarity score between two text strings, using\n",
        "#     a SentenceTransformer model.\n",
        "#     \"\"\"\n",
        "#     emb1 = model.encode(term1, convert_to_tensor=True)\n",
        "#     emb2 = model.encode(term2, convert_to_tensor=True)\n",
        "#     cos_sim = util.cos_sim(emb1, emb2)\n",
        "#     return float(cos_sim[0][0])  # scalar\n",
        "\n",
        "# # Compute the semantic similarity between the Llama term and the official GO term\n",
        "# res = compute_semantic_similarity(example_llama_term, term, sapbert_model)\n",
        "# print(f\"Semantic similarity between '{example_llama_term}' and '{term}': {res:.2f}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The code below takes a bit of time due to:\n",
        "- api calls to retrieve the gene ontology terms from NCBI Gene database for each gene"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "=== Checking Gene: TGFBR3 (ID=7049) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of fibroblast proliferation' | Score=0.4910426437854767| PMIDs=['30729664']\n",
            "LocalTerm='apoptotic process' | Score=0.4449974000453949| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='transforming growth factor beta receptor activity' | BestNCBI='transforming growth factor beta receptor activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "LocalTerm='negative regulation of transforming growth factor beta receptor signaling pathway' | BestNCBI='negative regulation of transforming growth factor beta receptor signaling pathway' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='cell surface' | BestNCBI='cell surface' (diff. references) | Score=1.0000001192092896 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "NCBITerm='negative regulation of extracellular matrix assembly' | PMID=['30729664']\n",
            "NCBITerm='negative regulation of SMAD protein signal transduction' | PMID=['30729664']\n",
            "NCBITerm='positive regulation of SMAD protein signal transduction' | PMID=['30729664']\n",
            "\n",
            "-- GO with experiment codes (NCBI) --\n",
            "NCBITerm='negative regulation of SMAD protein signal transduction' | Evidence=IMP | PMIDs=['30729664']\n",
            "NCBITerm='negative regulation of extracellular matrix assembly' | Evidence=IMP | PMIDs=['30729664']\n",
            "NCBITerm='positive regulation of SMAD protein signal transduction' | Evidence=IMP | PMIDs=['30729664']\n",
            "\n",
            "== Per-gene Summary for TGFBR3 (7049) ==\n",
            "TP = 0, NP = 2, DCU = 3, ND = 3, GO with experiment codes (NCBI) = 3\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Ang-II (ID=11606) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='peptide hormone receptor binding' | Score=0.6756502389907837| PMIDs=['30729664']\n",
            "LocalTerm='atrial cardiac myofibril assembly' | Score=0.5668154358863831| PMIDs=['30729664']\n",
            "LocalTerm='SMAD protein signal transduction' | Score=0.5593180060386658| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cellular response to angiotensin' | BestNCBI='response to angiotensin' (diff. references) | Score=0.8775870203971863 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Ang-II (11606) ==\n",
            "TP = 0, NP = 3, DCU = 1, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: collagen, type I, alpha 1 (ID=1277) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='supramolecular fiber organization' | BestNCBI='collagen fibril organization' (diff. references) | Score=0.8089128732681274 | PMIDs=['30729664']\n",
            "LocalTerm='extracellular matrix' | BestNCBI='collagen-containing extracellular matrix' (diff. references) | Score=0.7784955501556396 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for collagen, type I, alpha 1 (1277) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: COL3A1 (ID=1281) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='supramolecular fiber organization' | BestNCBI='supramolecular fiber organization' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='extracellular matrix' | BestNCBI='collagen-containing extracellular matrix' (diff. references) | Score=0.7784955501556396 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for COL3A1 (1281) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: ACTA2 (ID=59) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='muscle contraction' | Score=0.578100323677063| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cytoskeleton' | BestNCBI='actin cytoskeleton' (diff. references) | Score=0.8474626541137695 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for ACTA2 (59) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Smad1 (ID=4086) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='protein phosphorylation' | Score=0.5457684993743896| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='transforming growth factor beta receptor signaling pathway' | BestNCBI='transforming growth factor beta receptor signaling pathway' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='intracellular signal transduction' | BestNCBI='signal transduction' (diff. references) | Score=0.9288161993026733 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Smad1 (4086) ==\n",
            "TP = 0, NP = 1, DCU = 3, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Smad3 (ID=4088) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='atrial cardiac myofibril assembly' | Score=0.5146453976631165| PMIDs=['30729664']\n",
            "LocalTerm='phosphorylation' | Score=0.5067050457000732| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='transcription coregulator activity' | BestNCBI='transcription coactivator binding' (diff. references) | Score=0.8656408786773682 | PMIDs=['30729664']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='signal transduction' | BestNCBI='signal transduction involved in regulation of gene expression' (diff. references) | Score=0.7374023199081421 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "NCBITerm='SMAD protein signal transduction' | PMID=['30729664']\n",
            "NCBITerm='positive regulation of extracellular matrix assembly' | PMID=['30729664']\n",
            "NCBITerm='positive regulation of miRNA transcription' | PMID=['30729664']\n",
            "NCBITerm='response to angiotensin' | PMID=['30729664']\n",
            "\n",
            "-- GO with experiment codes (NCBI) --\n",
            "NCBITerm='SMAD protein signal transduction' | Evidence=IMP | PMIDs=['30729664']\n",
            "NCBITerm='positive regulation of extracellular matrix assembly' | Evidence=IMP | PMIDs=['30729664']\n",
            "NCBITerm='positive regulation of miRNA transcription' | Evidence=IMP | PMIDs=['30729664']\n",
            "NCBITerm='response to angiotensin' | Evidence=IMP | PMIDs=['30729664']\n",
            "\n",
            "== Per-gene Summary for Smad3 (4088) ==\n",
            "TP = 0, NP = 2, DCU = 3, ND = 4, GO with experiment codes (NCBI) = 4\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: miR-21 (ID=406991) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of atrial cardiac muscle cell membrane depolarization' | Score=0.49364739656448364| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='negative regulation of gene expression' | BestNCBI='negative regulation of gene expression' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "LocalTerm='miRNA-mediated post-transcriptional gene silencing' | BestNCBI='miRNA-mediated post-transcriptional gene silencing' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for miR-21 (406991) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TRPC3 (ID=7222) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='fibroblast proliferation' | Score=0.37589114904403687| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='calcium channel activity' | BestNCBI='calcium channel activity' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TRPC3 (7222) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: KCNJ2 (ID=3759) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='atrial cardiac muscle cell differentiation' | Score=0.5030057430267334| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='voltage-gated potassium channel activity' | BestNCBI='inward rectifier potassium channel activity' (diff. references) | Score=0.7689232230186462 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for KCNJ2 (3759) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: KCNJ2 (ID=10320) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='atrial cardiac muscle cell differentiation' | Score=0.571267306804657| PMIDs=['30729664']\n",
            "LocalTerm='potassium ion transmembrane transport' | Score=0.3337082266807556| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for KCNJ2 (10320) ==\n",
            "TP = 0, NP = 2, DCU = 0, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: miR-23b (ID=407011) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='negative regulation of gene expression' | Score=0.5483413934707642| PMIDs=['30729664']\n",
            "LocalTerm='regulation of connective tissue replacement' | Score=0.37762781977653503| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='miRNA-mediated post-transcriptional gene silencing' | BestNCBI='miRNA-mediated post-transcriptional gene silencing' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for miR-23b (407011) ==\n",
            "TP = 0, NP = 2, DCU = 1, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TGIF1 (ID=7050) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='negative regulation of transforming growth factor beta receptor signaling pathway' | Score=0.31124287843704224| PMIDs=['30729664']\n",
            "LocalTerm='regulation of myofibroblast differentiation' | Score=0.37340211868286133| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TGIF1 (7050) ==\n",
            "TP = 0, NP = 2, DCU = 0, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: PTEN (ID=5728) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='negative regulation of protein kinase C signaling' | Score=0.5765676498413086| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for PTEN (5728) ==\n",
            "TP = 0, NP = 1, DCU = 0, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TGF-beta1 (ID=7040) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='transforming growth factor beta receptor signaling pathway' | BestNCBI='transforming growth factor beta receptor signaling pathway' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='cellular response to angiotensin' | BestNCBI='cellular response to acetaldehyde' (diff. references) | Score=0.6963939070701599 | PMIDs=['30729664']\n",
            "LocalTerm='extracellular space' | BestNCBI='extracellular space' (diff. references) | Score=1.0000001192092896 | PMIDs=['30729664']\n",
            "LocalTerm='growth factor activity' | BestNCBI='growth factor activity' (diff. references) | Score=0.9999998807907104 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TGF-beta1 (7040) ==\n",
            "TP = 0, NP = 0, DCU = 4, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: AKT (ID=207) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cell-cell signaling' | BestNCBI='intracellular signal transduction' (diff. references) | Score=0.8332735300064087 | PMIDs=['30729664']\n",
            "LocalTerm='protein kinase activity' | BestNCBI='protein kinase activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for AKT (207) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: N-Cadherin (ID=1000) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cell adhesion molecule binding' | BestNCBI='cadherin binding' (diff. references) | Score=0.8014683723449707 | PMIDs=['30729664']\n",
            "LocalTerm='cell-cell adhesion' | BestNCBI='cell-cell adhesion' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='cell junction' | BestNCBI='cell junction' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for N-Cadherin (1000) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: miR-27b (ID=407019) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of connective tissue replacement' | Score=0.4282189607620239| PMIDs=['30729664']\n",
            "LocalTerm='cytoplasm' | Score=0.6482447981834412| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='negative regulation of gene expression' | BestNCBI='negative regulation of gene expression' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "LocalTerm='miRNA-mediated post-transcriptional gene silencing' | BestNCBI='miRNA-mediated post-transcriptional gene silencing' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "NCBITerm='positive regulation of extracellular matrix assembly' | PMID=['30729664']\n",
            "NCBITerm='miRNA-mediated gene silencing by mRNA destabilization' | PMID=['30729664']\n",
            "NCBITerm='response to angiotensin' | PMID=['30729664']\n",
            "NCBITerm='mRNA base-pairing translational repressor activity' | PMID=['30729664']\n",
            "\n",
            "-- GO with experiment codes (NCBI) --\n",
            "NCBITerm='mRNA base-pairing translational repressor activity' | Evidence=IDA | PMIDs=['30729664']\n",
            "NCBITerm='miRNA-mediated gene silencing by mRNA destabilization' | Evidence=IDA | PMIDs=['30729664']\n",
            "NCBITerm='positive regulation of extracellular matrix assembly' | Evidence=IDA | PMIDs=['30729664']\n",
            "NCBITerm='response to angiotensin' | Evidence=IDA | PMIDs=['30729664']\n",
            "\n",
            "== Per-gene Summary for miR-27b (407019) ==\n",
            "TP = 0, NP = 2, DCU = 2, ND = 4, GO with experiment codes (NCBI) = 4\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: angiotensin II (ID=183) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='atrial cardiac myofibril assembly' | Score=0.3798832297325134| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='vascular process in circulatory system' | BestNCBI='blood vessel remodeling' (diff. references) | Score=0.694534420967102 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "NCBITerm='positive regulation of extracellular matrix assembly' | PMID=['30729664']\n",
            "NCBITerm='response to angiotensin' | PMID=['30729664']\n",
            "NCBITerm='positive regulation of miRNA transcription' | PMID=['30729664']\n",
            "\n",
            "-- GO with experiment codes (NCBI) --\n",
            "NCBITerm='positive regulation of extracellular matrix assembly' | Evidence=IDA | PMIDs=['30729664']\n",
            "NCBITerm='positive regulation of miRNA transcription' | Evidence=IMP | PMIDs=['30729664']\n",
            "NCBITerm='response to angiotensin' | Evidence=IDA | PMIDs=['30729664']\n",
            "\n",
            "== Per-gene Summary for angiotensin II (183) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 3, GO with experiment codes (NCBI) = 3\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TGFBR1 (ID=7046) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='transforming growth factor beta receptor activity' | BestNCBI='transforming growth factor beta receptor activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "LocalTerm='cellular response to transforming growth factor beta stimulus' | BestNCBI='cellular response to transforming growth factor beta stimulus' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='transforming growth factor beta complex' | BestNCBI='transforming growth factor beta ligand-receptor complex' (diff. references) | Score=0.8034414649009705 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TGFBR1 (7046) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TGFBR2 (ID=7048) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='cell surface receptor signaling pathway' | Score=0.6054239273071289| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='transforming growth factor beta receptor activity' | BestNCBI='transforming growth factor beta receptor activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "LocalTerm='plasma membrane' | BestNCBI='plasma membrane' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TGFBR2 (7048) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: FN1 (ID=2335) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='extracellular matrix organization' | Score=0.6059841513633728| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='fibrinogen binding' | BestNCBI='heparin binding' (diff. references) | Score=0.7102800607681274 | PMIDs=['30729664']\n",
            "LocalTerm='extracellular matrix' | BestNCBI='extracellular matrix' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for FN1 (2335) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: IL-1beta (ID=3553) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='inflammatory response' | BestNCBI='inflammatory response' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "LocalTerm='cytokine activity' | BestNCBI='cytokine activity' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for IL-1beta (3553) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CRP (ID=1401) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='acute-phase response' | BestNCBI='acute-phase response' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='inflammatory response' | BestNCBI='inflammatory response' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CRP (1401) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TNF-alpha (ID=7124) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='inflammatory response' | BestNCBI='inflammatory response' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TNF-alpha (7124) ==\n",
            "TP = 0, NP = 0, DCU = 1, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: ACE1 (ID=1636) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='peptidyl-dipeptidase activity' | BestNCBI='peptidyl-dipeptidase activity' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='vascular process in circulatory system' | BestNCBI='blood vessel remodeling' (diff. references) | Score=0.694534420967102 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for ACE1 (1636) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: ACE2 (ID=59272) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='regulation of blood pressure' | BestNCBI='regulation of vasoconstriction' (diff. references) | Score=0.7293990850448608 | PMIDs=['30729664']\n",
            "LocalTerm='peptidyl-dipeptidase activity' | BestNCBI='peptidyl-dipeptidase activity' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for ACE2 (59272) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Apelin (ID=8862) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='cardiac atrium formation' | Score=0.6414111256599426| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='regulation of blood pressure' | BestNCBI='negative regulation of blood pressure' (diff. references) | Score=0.7808612585067749 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Apelin (8862) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TGF-beta1 (ID=21803) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of reactive oxygen species metabolic process' | Score=0.5235404968261719| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='transforming growth factor beta receptor signaling pathway' | BestNCBI='transforming growth factor beta receptor signaling pathway' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "LocalTerm='cellular response to angiotensin' | BestNCBI='cellular response to acetaldehyde' (diff. references) | Score=0.6963939070701599 | PMIDs=['30729664']\n",
            "LocalTerm='extracellular space' | BestNCBI='extracellular space' (diff. references) | Score=1.0000001192092896 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TGF-beta1 (21803) ==\n",
            "TP = 0, NP = 1, DCU = 3, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: NADPH oxidases 2 (ID=13058) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='transforming growth factor beta receptor signaling pathway' | Score=0.4137794077396393| PMIDs=['30729664']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cellular response to angiotensin' | BestNCBI='response to angiotensin' (diff. references) | Score=0.8775870203971863 | PMIDs=['30729664']\n",
            "LocalTerm='reactive oxygen species biosynthetic process' | BestNCBI='superoxide anion generation' (diff. references) | Score=0.7540382146835327 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for NADPH oxidases 2 (13058) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Acvrl1 (ID=94) ===\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='transforming growth factor beta receptor activity' | BestNCBI='transforming growth factor beta receptor activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['30729664']\n",
            "LocalTerm='transforming growth factor beta receptor signaling pathway' | BestNCBI='transforming growth factor beta receptor signaling pathway' (diff. references) | Score=1.0 | PMIDs=['30729664']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Acvrl1 (94) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0, GO with experiment codes (NCBI) = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Global Summary Across All Genes ===\n",
            "Total entries: 111\n",
            "TP = 0, NP = 27, DCU = 56, ND = 14, GO with experiment codes (NCBI) = 14\n"
          ]
        }
      ],
      "source": [
        "import json\n",
        "from sentence_transformers import SentenceTransformer, util\n",
        "from Bio import Entrez\n",
        "import xml.etree.ElementTree as ET\n",
        "import time\n",
        "import requests\n",
        "\n",
        "###############################################################################\n",
        "# CONFIGURATION\n",
        "###############################################################################\n",
        "Entrez.email = \"your_email@example.com\"  # Use a valid email address\n",
        "\n",
        "SIMILARITY_THRESHOLD = 0.68\n",
        "MARGIN = 0.1  # The margin by which we'll allow a same-paper match to be \"close enough\"\n",
        "\n",
        "# Define which evidence codes are considered experimental/curated.\n",
        "experimental_codes = {\"IDA\", \"IMP\", \"IPI\", \"IEP\"}\n",
        "\n",
        "###############################################################################\n",
        "# 1) Utility: map from local \"namespace\" to NCBI's GO Category strings\n",
        "###############################################################################\n",
        "def map_namespace_to_ncbi_category(ns):\n",
        "    if ns == \"biological_process\":\n",
        "        return \"Process\"\n",
        "    elif ns == \"molecular_function\":\n",
        "        return \"Function\"\n",
        "    elif ns == \"cellular_component\":\n",
        "        return \"Component\"\n",
        "    return None\n",
        "\n",
        "###############################################################################\n",
        "# 2) Fetch the GO annotations from NCBI Gene for a given gene_id\n",
        "###############################################################################\n",
        "def fetch_gene_go_terms(gene_id):\n",
        "    \"\"\"\n",
        "    Fetch the NCBI Gene record (XML) for `gene_id` and parse out GO annotations.\n",
        "    Returns a dict of lists, keyed by: \"Function\", \"Process\", \"Component\".\n",
        "    \"\"\"\n",
        "    # Sleep to respect NCBI usage guidelines\n",
        "    time.sleep(0.3)\n",
        "\n",
        "    handle = Entrez.efetch(db=\"gene\", id=gene_id, retmode=\"xml\")\n",
        "    xml_data = handle.read()\n",
        "    handle.close()\n",
        "\n",
        "    root = ET.fromstring(xml_data)\n",
        "    go_annotations = {}\n",
        "\n",
        "    for gene_commentary in root.findall(\".//Gene-commentary\"):\n",
        "        heading_elem = gene_commentary.find(\"Gene-commentary_heading\")\n",
        "        if heading_elem is not None and heading_elem.text == \"GeneOntology\":\n",
        "            for category_block in gene_commentary.findall(\".//Gene-commentary_comment/Gene-commentary\"):\n",
        "                label_elem = category_block.find(\"Gene-commentary_label\")\n",
        "                if label_elem is None:\n",
        "                    continue\n",
        "                category_label = label_elem.text  # \"Function\", \"Process\", or \"Component\"\n",
        "\n",
        "                if category_label not in go_annotations:\n",
        "                    go_annotations[category_label] = []\n",
        "\n",
        "                inner_comment = category_block.find(\"Gene-commentary_comment\")\n",
        "                if inner_comment is None:\n",
        "                    continue\n",
        "\n",
        "                for go_term_block in inner_comment.findall(\"Gene-commentary\"):\n",
        "                    dbtag_db = go_term_block.find(\".//Dbtag_db\")\n",
        "                    if dbtag_db is not None and dbtag_db.text == \"GO\":\n",
        "                        # Parse the GO ID\n",
        "                        object_id_elem = go_term_block.find(\".//Object-id_id\")\n",
        "                        if object_id_elem is not None:\n",
        "                            numeric_id = object_id_elem.text\n",
        "                            padded_id = numeric_id.zfill(7)\n",
        "                            go_id = f\"GO:{padded_id}\"\n",
        "                        else:\n",
        "                            go_id = \"GO:unknown\"\n",
        "\n",
        "                        # Parse the GO term name\n",
        "                        anchor_elem = go_term_block.find(\".//Other-source_anchor\")\n",
        "                        go_term_name = anchor_elem.text if anchor_elem is not None else \"Unknown Term\"\n",
        "\n",
        "                        # Evidence code\n",
        "                        evidence_code = None\n",
        "                        post_text_elem = go_term_block.find(\".//Other-source_post-text\")\n",
        "                        if post_text_elem is not None:\n",
        "                            post_text_str = post_text_elem.text\n",
        "                            if post_text_str and \"evidence: \" in post_text_str:\n",
        "                                evidence_code = post_text_str.split(\"evidence:\")[-1].strip()\n",
        "                            else:\n",
        "                                evidence_code = post_text_str\n",
        "\n",
        "                        # PubMed IDs\n",
        "                        pubmed_ids = []\n",
        "                        refs_block = go_term_block.find(\"Gene-commentary_refs\")\n",
        "                        if refs_block is not None:\n",
        "                            for pub in refs_block.findall(\"Pub\"):\n",
        "                                pmid_elem = pub.find(\".//PubMedId\")\n",
        "                                if pmid_elem is not None:\n",
        "                                    pubmed_ids.append(pmid_elem.text)\n",
        "\n",
        "                        go_annotations[category_label].append({\n",
        "                            \"go_id\": go_id,\n",
        "                            \"term\": go_term_name,\n",
        "                            \"evidence\": evidence_code,\n",
        "                            \"pubmed_ids\": pubmed_ids\n",
        "                        })\n",
        "\n",
        "    return go_annotations\n",
        "\n",
        "###############################################################################\n",
        "# 3) Utility: compute cosine similarity using Sentence-BERT\n",
        "###############################################################################\n",
        "def compute_semantic_similarity(term1, term2, model):\n",
        "    emb1 = model.encode(term1, convert_to_tensor=True)\n",
        "    emb2 = model.encode(term2, convert_to_tensor=True)\n",
        "    cos_sim = util.cos_sim(emb1, emb2)\n",
        "    return float(cos_sim[0][0])\n",
        "\n",
        "###############################################################################\n",
        "# 4) Utility: convert a PMCID to PMID using NCBI E-utilities\n",
        "###############################################################################\n",
        "pmc_cache = {} # Cache for PMCID to PMID conversion\n",
        "\n",
        "def pmcid_to_pmid(pmcid):\n",
        "    \"\"\"\n",
        "    Convert a PMCID (e.g. 'PMC123456') to a numeric PMID string using NCBI E-utilities.\n",
        "    Returns None if not found or conversion fails.\n",
        "    \"\"\"\n",
        "    # cache since we dont want to hit the same PMCID multiple times\n",
        "    if pmcid in pmc_cache:\n",
        "        return pmc_cache[pmcid]\n",
        "    \n",
        "    base = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/\"\n",
        "    url = base + f\"esummary.fcgi?db=pmc&id={pmcid[3:]}&retmode=xml\"\n",
        "    # Note pmcid[3:] strips \"PMC\" prefix if we pass \"PMC123456\" to the 'id=' param\n",
        "\n",
        "    resp = requests.get(url)\n",
        "    if resp.status_code != 200:\n",
        "        return None\n",
        "\n",
        "    root = ET.fromstring(resp.text)\n",
        "\n",
        "    for item in root.findall(\".//Item\"):\n",
        "        if item.get(\"Name\") == \"pmid\":\n",
        "            pmid = item.text\n",
        "            pmc_cache[pmcid] = pmid\n",
        "            return pmid\n",
        "\n",
        "    return None\n",
        "\n",
        "def convert_local_pmcids_to_pmids(local_data):\n",
        "    for gene_entry in local_data:\n",
        "        for local_ann in gene_entry.get(\"go_data\", []):\n",
        "            refs = local_ann.get(\"reference(s)\", [])\n",
        "            new_refs = []\n",
        "            for r in refs:\n",
        "                # If r starts with 'PMC', attempt to convert\n",
        "                if r.startswith(\"PMC\"):\n",
        "                    time.sleep(0.3)  # Respect NCBI usage guidelines\n",
        "                    pmid = pmcid_to_pmid(r)\n",
        "                    if pmid:\n",
        "                        new_refs.append(pmid)\n",
        "                    else:\n",
        "                        # Conversion failed or none found, fallback\n",
        "                        new_refs.append(r)\n",
        "                else:\n",
        "                    new_refs.append(r)\n",
        "            local_ann[\"reference(s)\"] = new_refs\n",
        "\n",
        "###############################################################################\n",
        "# 5) Main comparison & classification logic with tie-break margin\n",
        "###############################################################################\n",
        "def compare_json_vs_ncbi(json_path, model):\n",
        "    \"\"\"\n",
        "    For each local annotation (term + references), we:\n",
        "      1) Gather all possible NCBI annotations in the same GO category\n",
        "      2) Find best match overall & best match among same-paper (i.e., shared PMIDs)\n",
        "      3) If best_same_score >= best_overall_score - MARGIN, prefer the same-paper match\n",
        "      4) Classify that chosen match as TP if it shares references, else DCU (if >= threshold),\n",
        "         or NP if below threshold.\n",
        "    \"\"\"\n",
        "    with open(json_path, \"r\") as f:\n",
        "        local_data = json.load(f)\n",
        "\n",
        "    # Convert PMCIDs in local_data to numeric PMIDs since NCBI uses PMIDs even thorugh GO annotations are curated from PMC\n",
        "    convert_local_pmcids_to_pmids(local_data)\n",
        "\n",
        "    all_comparisons = []\n",
        "\n",
        "    for gene_entry in local_data:\n",
        "        gene_name = gene_entry[\"gene\"]\n",
        "        gene_id = gene_entry[\"geneID\"]\n",
        "        local_go_data = gene_entry.get(\"go_data\", [])\n",
        "\n",
        "        print(f\"\\n=== Checking Gene: {gene_name} (ID={gene_id}) ===\")\n",
        "\n",
        "        ncbi_annotations = fetch_gene_go_terms(gene_id)\n",
        "\n",
        "        # We'll keep track of official GO annotations (for ND detection)\n",
        "        official_by_pmid = set()\n",
        "        for category_label, ann_list in ncbi_annotations.items():\n",
        "            for ann in ann_list:\n",
        "                for pmid in ann[\"pubmed_ids\"]:\n",
        "                    official_by_pmid.add((category_label, ann[\"term\"], pmid))\n",
        "\n",
        "        matched_officials = set()\n",
        "        comparisons_this_gene = []\n",
        "\n",
        "        # Per-gene counters\n",
        "        count_tp = 0\n",
        "        count_np = 0\n",
        "        count_dcu = 0\n",
        "\n",
        "        #\n",
        "        # 1) CLASSIFY EACH LOCAL ANNOTATION (TP, NP, DCU)\n",
        "        #\n",
        "        for local_ann in local_go_data:\n",
        "            local_term = local_ann[\"term\"]\n",
        "            local_refs = set(local_ann.get(\"reference(s)\", []))\n",
        "            local_ns = local_ann.get(\"namespace\", None)\n",
        "            target_category = map_namespace_to_ncbi_category(local_ns)\n",
        "\n",
        "            classification = None\n",
        "            chosen_match_info = None\n",
        "            chosen_score = -1.0\n",
        "\n",
        "            best_score_all = -1.0\n",
        "            best_info_all = None\n",
        "            best_score_same = -1.0\n",
        "            best_info_same = None\n",
        "\n",
        "            if target_category and target_category in ncbi_annotations:\n",
        "                possible_anns = ncbi_annotations[target_category]\n",
        "\n",
        "                # (A) Find best match overall\n",
        "                for ann in possible_anns:\n",
        "                    sc = compute_semantic_similarity(local_term, ann[\"term\"], model)\n",
        "                    if sc > best_score_all:\n",
        "                        best_score_all = sc\n",
        "                        best_info_all = ann\n",
        "\n",
        "                # (B) Find best match among same-paper\n",
        "                same_paper_anns = [\n",
        "                    ann for ann in possible_anns\n",
        "                    if set(ann[\"pubmed_ids\"]).intersection(local_refs)\n",
        "                ]\n",
        "                for ann in same_paper_anns:\n",
        "                    sc = compute_semantic_similarity(local_term, ann[\"term\"], model)\n",
        "                    if sc > best_score_same:\n",
        "                        best_score_same = sc\n",
        "                        best_info_same = ann\n",
        "\n",
        "                # (C) Tie-break\n",
        "                if best_score_same >= 0:\n",
        "                    # There's at least one same-paper annotation\n",
        "                    if best_score_same >= best_score_all - MARGIN:\n",
        "                        chosen_score = best_score_same\n",
        "                        chosen_match_info = best_info_same\n",
        "                    else:\n",
        "                        chosen_score = best_score_all\n",
        "                        chosen_match_info = best_info_all\n",
        "                else:\n",
        "                    chosen_score = best_score_all\n",
        "                    chosen_match_info = best_info_all\n",
        "\n",
        "                # (D) Classify\n",
        "                if chosen_score >= SIMILARITY_THRESHOLD and chosen_match_info is not None:\n",
        "                    overlap = set(chosen_match_info[\"pubmed_ids\"]).intersection(local_refs)\n",
        "                    if overlap:\n",
        "                        classification = \"TP\"\n",
        "                        count_tp += 1\n",
        "                        # Mark the official annotation as matched\n",
        "                        for pmid in chosen_match_info[\"pubmed_ids\"]:\n",
        "                            if pmid in local_refs:\n",
        "                                matched_officials.add((target_category, chosen_match_info[\"term\"], pmid))\n",
        "                    else:\n",
        "                        classification = \"DCU\"\n",
        "                        count_dcu += 1\n",
        "                else:\n",
        "                    classification = \"NP\"\n",
        "                    count_np += 1\n",
        "            else:\n",
        "                # No recognized category => NP\n",
        "                classification = \"NP\"\n",
        "                count_np += 1\n",
        "\n",
        "            rec = {\n",
        "                \"gene\": gene_name,\n",
        "                \"geneID\": gene_id,\n",
        "                \"local_term\": local_term,\n",
        "                \"namespace\": local_ns,\n",
        "                \"references_used\": list(local_refs),\n",
        "                \"classification\": classification,\n",
        "                \"best_score\": chosen_score,\n",
        "            }\n",
        "            if chosen_match_info:\n",
        "                rec[\"best_ncbi_match\"] = chosen_match_info[\"term\"]\n",
        "                rec[\"best_ncbi_go_id\"] = chosen_match_info[\"go_id\"]\n",
        "                rec[\"best_ncbi_pubmed_ids\"] = chosen_match_info[\"pubmed_ids\"]\n",
        "            else:\n",
        "                rec[\"best_ncbi_match\"] = None\n",
        "                rec[\"best_ncbi_go_id\"] = None\n",
        "                rec[\"best_ncbi_pubmed_ids\"] = []\n",
        "\n",
        "            comparisons_this_gene.append(rec)\n",
        "\n",
        "        #\n",
        "        # 2) DETECT ND (NOT DETECTED) AND STORE AS ADDITIONAL RECORDS\n",
        "        #\n",
        "        # ND means: official NCBI annotation referencing the local paper(s),\n",
        "        # but never matched as a TP above.\n",
        "        #\n",
        "        undetected = official_by_pmid - matched_officials\n",
        "\n",
        "        # The local paper set is all refs mentioned in local_go_data\n",
        "        all_local_pmids_for_gene = set()\n",
        "        for ann in local_go_data:\n",
        "            all_local_pmids_for_gene |= set(ann.get(\"reference(s)\", []))\n",
        "\n",
        "        actual_nd = []\n",
        "        for (cat_label, term_text, pmid) in undetected:\n",
        "            if pmid in all_local_pmids_for_gene:\n",
        "                actual_nd.append((cat_label, term_text, pmid))\n",
        "\n",
        "        count_nd = len(actual_nd)\n",
        "\n",
        "        # Storing ND in comparisons_this_gene\n",
        "        # For each official annotation that is ND, we add a new record\n",
        "        for (cat_label, ncbi_term, pmid) in actual_nd:\n",
        "            rec = {\n",
        "                \"gene\": gene_name,\n",
        "                \"geneID\": gene_id,\n",
        "                \"local_term\": None,  # No local annotation matched\n",
        "                \"namespace\": cat_label,\n",
        "                \"references_used\": [pmid],  # or empty if you prefer\n",
        "                \"classification\": \"ND\",\n",
        "                \"best_score\": None,\n",
        "                # We might store the official annotation in 'best_ncbi_match'\n",
        "                \"best_ncbi_match\": ncbi_term,\n",
        "                \"best_ncbi_go_id\": None,  # We didn't parse that here, but could\n",
        "                \"best_ncbi_pubmed_ids\": [pmid],\n",
        "            }\n",
        "            comparisons_this_gene.append(rec)\n",
        "        \n",
        "         #\n",
        "        # 3) ADDITIONAL: Look for any NCBI annotation that has experimental evidence\n",
        "        #    and shares a PubMed ID with the local articles, but was not already\n",
        "        #    matched as TP (or reported as ND).\n",
        "        #\n",
        "        extra_added = set()  # to avoid duplicates (using (namespace, go_id) as key)\n",
        "        for category_label, ann_list in ncbi_annotations.items():\n",
        "            for ann in ann_list:\n",
        "                if (ann[\"evidence\"] in experimental_codes and \n",
        "                    set(ann[\"pubmed_ids\"]).intersection(all_local_pmids_for_gene)):\n",
        "                    key = (category_label, ann[\"go_id\"])\n",
        "                    if key not in extra_added:\n",
        "                        # We add this extra record only if it was not already included above\n",
        "                        rec = {\n",
        "                            \"gene\": gene_name,\n",
        "                            \"geneID\": gene_id,\n",
        "                            \"local_term\": None,\n",
        "                            \"namespace\": category_label,\n",
        "                            \"references_used\": list(set(ann[\"pubmed_ids\"]).intersection(all_local_pmids_for_gene)),\n",
        "                            \"classification\": \"GO with experiment codes (NCBI)\",\n",
        "                            \"best_score\": None,\n",
        "                            \"best_ncbi_match\": ann[\"term\"],\n",
        "                            \"best_ncbi_go_id\": ann[\"go_id\"],\n",
        "                            \"best_ncbi_pubmed_ids\": ann[\"pubmed_ids\"],\n",
        "                            \"evidence\": ann[\"evidence\"]\n",
        "                        }\n",
        "                        comparisons_this_gene.append(rec)\n",
        "                        extra_added.add(key)\n",
        "\n",
        "        #\n",
        "        # 3) PRINT RESULTS\n",
        "       \n",
        "        # Summaries\n",
        "        tps_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"TP\"]\n",
        "        nps_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"NP\"]\n",
        "        dcus_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"DCU\"]\n",
        "        nds_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"ND\"]\n",
        "\n",
        "        print(\"\\n-- TPs --\")\n",
        "        for item in tps_for_gene:\n",
        "            print(f\"LocalTerm='{item['local_term']}' | BestNCBI='{item['best_ncbi_match']}' \"\n",
        "                  f\"| Score={item['best_score']:.2f} | PMIDs={item['references_used']}\")\n",
        "\n",
        "        print(\"\\n-- NPs --\")\n",
        "        for item in nps_for_gene:\n",
        "            print(f\"LocalTerm='{item['local_term']}' | Score={item['best_score']}\"\n",
        "                  f\"| PMIDs={item['references_used']}\")\n",
        "\n",
        "        print(\"\\n-- DCUs --\")\n",
        "        for item in dcus_for_gene:\n",
        "            print(f\"LocalTerm='{item['local_term']}' | BestNCBI='{item['best_ncbi_match']}' \"\n",
        "                  f\"(diff. references) | Score={item['best_score']} | PMIDs={item['references_used']}\")\n",
        "\n",
        "        print(\"\\n-- ND --\")\n",
        "        for item in nds_for_gene:\n",
        "            print(f\"NCBITerm='{item['best_ncbi_match']}' | PMID={item['best_ncbi_pubmed_ids']}\")\n",
        "\n",
        "        exp_records = [x for x in comparisons_this_gene if x[\"classification\"] == \"GO with experiment codes (NCBI)\"]\n",
        "        if exp_records:\n",
        "            print(\"\\n-- GO with experiment codes (NCBI) --\")\n",
        "            for item in exp_records:\n",
        "                print(f\"NCBITerm='{item['best_ncbi_match']}' | Evidence={item['evidence']} | PMIDs={item['references_used']}\")\n",
        "\n",
        "        \n",
        "        print(f\"\\n== Per-gene Summary for {gene_name} ({gene_id}) ==\")\n",
        "        print(f\"TP = {count_tp}, NP = {count_np}, DCU = {count_dcu}, ND = {count_nd}, \"\n",
        "              f\"GO with experiment codes (NCBI) = {len(exp_records)}\")\n",
        "        print(\"--------------------------------------------------------\")\n",
        "\n",
        "        all_comparisons.extend(comparisons_this_gene)\n",
        "\n",
        "    return all_comparisons\n",
        "\n",
        "###############################################################################\n",
        "# 6) Example usage\n",
        "###############################################################################\n",
        "json_file_path = \"./processed_inferences/processed_output_PMC6484421.json\"\n",
        "\n",
        "try:\n",
        "    comparison_results = compare_json_vs_ncbi(json_file_path, sapbert_model)\n",
        "except requests.HTTPError as e:\n",
        "    print(f\"HTTPError: {e.code} - {e.reason}\")\n",
        "    print(\"Please check the gene ID or ensure that the NCBI API is accessible.\")\n",
        "\n",
        "# Now each item in comparison_results has classification in {TP, NP, DCU, ND}\n",
        "\n",
        "# You can also sum ND from the final list if you want a global ND count\n",
        "tp_global = sum(1 for r in comparison_results if r[\"classification\"] == \"TP\")\n",
        "np_global = sum(1 for r in comparison_results if r[\"classification\"] == \"NP\")\n",
        "dcu_global = sum(1 for r in comparison_results if r[\"classification\"] == \"DCU\")\n",
        "nd_global = sum(1 for r in comparison_results if r[\"classification\"] == \"ND\")\n",
        "exp_global = sum(1 for r in comparison_results if r[\"classification\"] == \"GO with experiment codes (NCBI)\")\n",
        "\n",
        "\n",
        "print(\"\\n=== Global Summary Across All Genes ===\")\n",
        "print(f\"Total entries: {len(comparison_results)}\")\n",
        "print(f\"TP = {tp_global}, NP = {np_global}, DCU = {dcu_global}, ND = {nd_global}, \"\n",
        "      f\"GO with experiment codes (NCBI) = {exp_global}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Will put TP, ND, DCU and NP in a excel file in the output folder\n",
        "change the sheet_name to the name of the file you want to save it as in the excel. After that, open excel and copy the sheet to the main sheet. Then delete the other sheet."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 139,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Finished! Data written to sheet 'PMC6742646' in file: /Users/crystalcho/Downloads/Evaluation.xlsx\n"
          ]
        }
      ],
      "source": [
        "import pandas as pd\n",
        "import os\n",
        "from openpyxl import load_workbook\n",
        "from collections import defaultdict\n",
        "\n",
        "###############################################################################\n",
        "# 1) Aggregate classification_counts from comparison_results\n",
        "###############################################################################\n",
        "classification_counts = defaultdict(lambda: {\"TP\": 0, \"NP\": 0, \"DCU\": 0, \"ND\": 0})\n",
        "\n",
        "for row in comparison_results:\n",
        "    gene_id = row[\"geneID\"]\n",
        "    gene_name = row[\"gene\"]\n",
        "    classification = row[\"classification\"]\n",
        "    \n",
        "    # For each PMID in references_used, increment the classification count\n",
        "    for pmid in row[\"references_used\"]:\n",
        "        key = (gene_id, gene_name, pmid)\n",
        "        classification_counts[key][classification] += 1\n",
        "\n",
        "records_for_df = []\n",
        "for (gid, gname, pmid), counts in classification_counts.items():\n",
        "    records_for_df.append({\n",
        "        \"geneID\": gid,\n",
        "        \"gene\": gname,\n",
        "        \"pmid\": pmid,\n",
        "        \"TP\": counts[\"TP\"],\n",
        "        \"ND\": counts[\"ND\"],\n",
        "        \"NP\": counts[\"NP\"],\n",
        "        \"DCU\": counts[\"DCU\"]\n",
        "    })\n",
        "\n",
        "df = pd.DataFrame(\n",
        "    records_for_df,\n",
        "    columns=[\"geneID\", \"gene\", \"pmid\", \"TP\", \"ND\", \"NP\", \"DCU\"]\n",
        ")\n",
        "\n",
        "###############################################################################\n",
        "# 2) Write to Excel, appending a NEW SHEET for each paper\n",
        "###############################################################################\n",
        "output_path = r\"/Users/crystalcho/Downloads/Evaluation.xlsx\"\n",
        "\n",
        "# You can change this to something unique for each paper you analyze.\n",
        "sheet_name = \"PMC6742646\"\n",
        "\n",
        "# If the file already exists, open in append mode. Otherwise, create a new file.\n",
        "mode = \"a\" if os.path.exists(output_path) else \"w\"\n",
        "\n",
        "if mode == \"w\":\n",
        "    # Create a new file and write data\n",
        "    df.to_excel(output_path, sheet_name=sheet_name, index=False)\n",
        "else:\n",
        "    # If the file already exists, append a new sheet\n",
        "    with pd.ExcelWriter(output_path, engine=\"openpyxl\", mode=\"a\", if_sheet_exists=\"new\") as writer:\n",
        "        df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
        "\n",
        "print(f\"Finished! Data written to sheet '{sheet_name}' in file: {output_path}\")\n"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "collapsed_sections": [
        "6oaKSo6iXGOc"
      ],
      "gpuType": "T4",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "BIOIN311C1",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.11"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "097046dc7c4f46f5866068c82b706aca": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "1675d70c8a5e426691d0af5e6ba333f3": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_784d7bf8c3be48b8ad5a3b6e6954690f",
              "IPY_MODEL_730fe49f91204e2cbd3b58d24442be0a",
              "IPY_MODEL_792eb1f8daf34e2590fda2f60e3ff684"
            ],
            "layout": "IPY_MODEL_3e0f9cdf58ef4af1a0b16f9795747dfc"
          }
        },
        "1845ddb7ff634d969fa2538209fc4e0a": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "1b8c25318a2b4b64bab3496083679c85": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "2443ed122977428c98248537123b674b": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_f7dcd748d0944da09948f4372366e8a8",
            "placeholder": "",
            "style": "IPY_MODEL_466f8ef6ba644d31a8586abaf9830448",
            "value": "438M/438M[00:01&lt;00:00,251MB/s]"
          }
        },
        "3117723c389b4e47a97d6e582078e5c9": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "32ef3e232304493d8b7e129c679f4f6f": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_b1c01cfdd57e40e7bc4de9baf81d9f45",
            "placeholder": "",
            "style": "IPY_MODEL_82a3774cc5cd4acfb1822d9cb74d6f45",
            "value": "112/112[00:00&lt;00:00,9.27kB/s]"
          }
        },
        "331dbfdb6d7849c6a7b6d43bd33a7325": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_8eb65674477e477091f2501704df7eb8",
            "placeholder": "",
            "style": "IPY_MODEL_6d29b58c64cc4f869d6944699149a60e",
            "value": "model.safetensors:100%"
          }
        },
        "39794838a2164151a5edf298a63f8147": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "3e0f9cdf58ef4af1a0b16f9795747dfc": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "466f8ef6ba644d31a8586abaf9830448": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "47571c7570f14ab58f96e5d30129da1b": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "4a743c94312c4b5db25087a747f6b879": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_ea731973ab9c44caa479fb7934fd3b4f",
            "max": 198,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_097046dc7c4f46f5866068c82b706aca",
            "value": 198
          }
        },
        "4af857d8e5a94c278ae0fb3d7e092ac5": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "53bb7f0fbad34a818e266d0ea44ffaf1": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "54a0cadcc9ef406db05d36b690d7891c": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c939f017049643b18d5142514937af73",
            "max": 437955508,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_6c463c9a4bef4b5194113acf7ced731e",
            "value": 437955508
          }
        },
        "59dc6c4216cb4c5aad8c3764cf3d3f6d": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_cc7573c531c84e61999e2161710173f3",
            "placeholder": "",
            "style": "IPY_MODEL_9775c0cd37bc43dcb7953f8b15343332",
            "value": "198/198[00:00&lt;00:00,14.3kB/s]"
          }
        },
        "618fd6a5462e4c5296b560f888b5d638": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_331dbfdb6d7849c6a7b6d43bd33a7325",
              "IPY_MODEL_54a0cadcc9ef406db05d36b690d7891c",
              "IPY_MODEL_2443ed122977428c98248537123b674b"
            ],
            "layout": "IPY_MODEL_c3911d8aca764dd38fad4c1dded8198f"
          }
        },
        "61de99be2c2e4afe90ff06e3bb30a2b7": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "6ac927f5e2a2453d9aae4b5c474cf96d": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_7593152124e04332b709dd30ec87c5a5",
              "IPY_MODEL_4a743c94312c4b5db25087a747f6b879",
              "IPY_MODEL_59dc6c4216cb4c5aad8c3764cf3d3f6d"
            ],
            "layout": "IPY_MODEL_3117723c389b4e47a97d6e582078e5c9"
          }
        },
        "6c463c9a4bef4b5194113acf7ced731e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "6d29b58c64cc4f869d6944699149a60e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "730fe49f91204e2cbd3b58d24442be0a": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_d598fdb3d94a4095ba5a950cef321f1b",
            "max": 226150,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_39794838a2164151a5edf298a63f8147",
            "value": 226150
          }
        },
        "7593152124e04332b709dd30ec87c5a5": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c6b5cf46fbeb41adb8da935949b808c1",
            "placeholder": "",
            "style": "IPY_MODEL_d4b8ff5330d14128ba304020f7fe8412",
            "value": "tokenizer_config.json:100%"
          }
        },
        "769b37c671334690b6a7e5ba9059343e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_bdf351e0ccbc4aa18e168f593d1659cf",
              "IPY_MODEL_8901a8884437441c934c53a98a1dac06",
              "IPY_MODEL_32ef3e232304493d8b7e129c679f4f6f"
            ],
            "layout": "IPY_MODEL_77472246b6b0497fbe67b8a3529040e5"
          }
        },
        "77472246b6b0497fbe67b8a3529040e5": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "784d7bf8c3be48b8ad5a3b6e6954690f": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_1845ddb7ff634d969fa2538209fc4e0a",
            "placeholder": "",
            "style": "IPY_MODEL_9b03af45365c42419bb39779b82be47c",
            "value": "vocab.txt:100%"
          }
        },
        "792eb1f8daf34e2590fda2f60e3ff684": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_e7eaa149a903445db37dfdf719e14f9e",
            "placeholder": "",
            "style": "IPY_MODEL_4af857d8e5a94c278ae0fb3d7e092ac5",
            "value": "226k/226k[00:00&lt;00:00,1.03MB/s]"
          }
        },
        "7d97b347b7cb40df81041035931d7f27": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_53bb7f0fbad34a818e266d0ea44ffaf1",
            "placeholder": "",
            "style": "IPY_MODEL_bc435a026437452bb820c14d1ce5d3ee",
            "value": "config.json:100%"
          }
        },
        "8039b35f8f654a5e93ab69a0d1d8f3ea": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "82a3774cc5cd4acfb1822d9cb74d6f45": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "873ea6832a0447a19938fe764cb7ac58": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "8901a8884437441c934c53a98a1dac06": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_b483173c5fa74a72bb0a88206d3206a3",
            "max": 112,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_8039b35f8f654a5e93ab69a0d1d8f3ea",
            "value": 112
          }
        },
        "8eb65674477e477091f2501704df7eb8": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "9775c0cd37bc43dcb7953f8b15343332": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "9b03af45365c42419bb39779b82be47c": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "b1c01cfdd57e40e7bc4de9baf81d9f45": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b483173c5fa74a72bb0a88206d3206a3": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b85f6bb59d024282b5552ff665da4548": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_47571c7570f14ab58f96e5d30129da1b",
            "max": 462,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_ec42f776af1b48a982636f0c122ab7e6",
            "value": 462
          }
        },
        "bc435a026437452bb820c14d1ce5d3ee": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "bdf351e0ccbc4aa18e168f593d1659cf": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_61de99be2c2e4afe90ff06e3bb30a2b7",
            "placeholder": "",
            "style": "IPY_MODEL_873ea6832a0447a19938fe764cb7ac58",
            "value": "special_tokens_map.json:100%"
          }
        },
        "c3911d8aca764dd38fad4c1dded8198f": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "c6b5cf46fbeb41adb8da935949b808c1": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "c939f017049643b18d5142514937af73": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "cc7573c531c84e61999e2161710173f3": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "d4b8ff5330d14128ba304020f7fe8412": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "d598fdb3d94a4095ba5a950cef321f1b": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "de6e3ef0670042378384c41ecd695f1a": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "e4718c03421b41a2a665c26293acd697": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_7d97b347b7cb40df81041035931d7f27",
              "IPY_MODEL_b85f6bb59d024282b5552ff665da4548",
              "IPY_MODEL_f1219528a46a4a8e94bd1c8de9e6e16a"
            ],
            "layout": "IPY_MODEL_1b8c25318a2b4b64bab3496083679c85"
          }
        },
        "e7eaa149a903445db37dfdf719e14f9e": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ea731973ab9c44caa479fb7934fd3b4f": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ec42f776af1b48a982636f0c122ab7e6": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "ec7200455fa34863bec2ae60f9f4aad5": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "f1219528a46a4a8e94bd1c8de9e6e16a": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_de6e3ef0670042378384c41ecd695f1a",
            "placeholder": "",
            "style": "IPY_MODEL_ec7200455fa34863bec2ae60f9f4aad5",
            "value": "462/462[00:00&lt;00:00,40.9kB/s]"
          }
        },
        "f7dcd748d0944da09948f4372366e8a8": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        }
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
