{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1AfSrrXBI8oi",
        "outputId": "3b20254b-e106-40b9-8f89-a0aa03fba087"
      },
      "outputs": [],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "UWSHSonXlTeG"
      },
      "outputs": [],
      "source": [
        "# packages\n",
        "import torch\n",
        "from transformers import AutoModelForCausalLM, AutoTokenizer\n",
        "import time"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BuXWNP8jcVkz"
      },
      "source": [
        "# 1. Pubtator API (TODO-Partially done)\n",
        "- using the input folder, check if the file(s) have been pre-annotated by checking if the following APIs return something\n",
        "```\n",
        "https://www.ncbi.nlm.nih.gov/research/pubtator3-api/publications/export/biocxml?pmids={id}\n",
        "or\n",
        "https://www.ncbi.nlm.nih.gov/research/pubtator3-api/publications/pmc_export/biocxml?pmcids=PMC{id}\n",
        "```\n",
        "  - if annotated, retrieve it and put it in the output folder\n",
        "  - if not, use the [example code](https://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/tmTools/download/ExampleCode.Python.zip) they provide to send it to the server to be annotated (Done)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0sl2nmCPFsQq"
      },
      "source": [
        "Using colab:\n",
        "\n",
        "```\n",
        "mkdir input\n",
        "```\n",
        "\n",
        "to make a input folder"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "edkmlyhFFifa",
        "outputId": "304ffdaa-8b04-4d0b-de90-f3caa4e5f27f"
      },
      "outputs": [],
      "source": [
        "!bash"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EQIruerqbU4Q",
        "outputId": "da090b16-ca20-4798-a26d-7bce4bc4090a"
      },
      "outputs": [],
      "source": [
        "# SubmitText_request.py\n",
        "!pip install unidecode\n",
        "\n",
        "import requests\n",
        "import os\n",
        "import sys\n",
        "from unidecode import unidecode\n",
        "import unicodedata\n",
        "import re\n",
        "\n",
        "def submit_text_request(input_folder, bioconcept, output_file_session_number):\n",
        "    pattern = r'[^0-9a-zA-Z\\!\\@\\#\\$\\%\\^\\&\\*\\(\\)\\_\\+\\{\\}\\|\\:\\\"\\<\\>\\?\\-\\=\\[\\]\\\\;\\'\\,\\.\\/ \\t\\n\\r]'\n",
        "    unicode_to_regular = {}\n",
        "    with open('lib/unicode.txt', 'r', encoding='utf-8') as input_file:\n",
        "        for line in input_file:\n",
        "            line = line.strip()\n",
        "            parts = line.split(\"\\t\")\n",
        "            if len(parts) == 2:\n",
        "                uni, reg = parts\n",
        "                if reg == '-whitespace-':\n",
        "                    unicode_to_regular[uni] = ' '\n",
        "                else:\n",
        "                    unicode_to_regular[uni] = reg\n",
        "\n",
        "    with open(output_file_session_number, 'w', encoding='utf-8') as output_file:\n",
        "        for filename in os.listdir(input_folder):\n",
        "            if filename.startswith('.'):\n",
        "                continue  # Skip hidden files\n",
        "            text_str = ''\n",
        "            with open(os.path.join(input_folder, filename), 'r', encoding='utf-8') as file_input:\n",
        "                for line in file_input:\n",
        "                    normalized_line = unicodedata.normalize('NFC', line)\n",
        "                    for uni, reg in unicode_to_regular.items():\n",
        "                        normalized_line = normalized_line.replace(uni, reg)\n",
        "                    normalized_line = unidecode(normalized_line)  # Assuming unidecode does similar work to Perl's version\n",
        "                    text_str += normalized_line\n",
        "            text_str = re.sub(pattern, ' ', text_str)\n",
        "            url = \"https://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/RESTful/request.cgi\"\n",
        "            response = requests.post(url, data={'text': text_str, 'bioconcept': bioconcept})\n",
        "\n",
        "            if response.status_code == 200:\n",
        "                session_number = response.json().get('id', '')\n",
        "                print(f\"Thanks for your submission. The session number is: {session_number}\")\n",
        "                output_file.write(f\"{session_number}\\t{filename}\\n\")\n",
        "            else:\n",
        "                print(f\"Error: HTTP {response.status_code} for {filename}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KccYUHYJcEfG"
      },
      "outputs": [],
      "source": [
        "# SubmitText_retreive.py\n",
        "import requests\n",
        "import os\n",
        "import sys\n",
        "\n",
        "def submit_text_retrieve(input_folder, inputfile_session_number, output_folder):\n",
        "    # Ensure output directory exists\n",
        "    if not os.path.exists(output_folder):\n",
        "        os.makedirs(output_folder)\n",
        "\n",
        "    sn_hash = {}\n",
        "    with open(inputfile_session_number, 'r') as input_file:\n",
        "        for line in input_file:\n",
        "            line = line.strip()\n",
        "            if line:\n",
        "                session_number, input_filename = line.split(\"\\t\")\n",
        "                sn_hash[session_number] = input_filename\n",
        "\n",
        "    base_url = \"https://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/RESTful/retrieve.cgi\"\n",
        "\n",
        "    for session_number, input_filename in sn_hash.items():\n",
        "        output_file_path = os.path.join(output_folder, input_filename)\n",
        "        if os.path.exists(output_file_path):\n",
        "            print(f\"{output_file_path} - finished\")\n",
        "            continue\n",
        "\n",
        "        retrieve_url = f\"{base_url}?id={session_number}\"\n",
        "        response = requests.get(retrieve_url)\n",
        "        if response.status_code == 200:\n",
        "            with open(output_file_path, 'w', encoding='utf-8') as output_file:\n",
        "                output_file.write(response.text)\n",
        "            print(f\"{session_number} : Result is retrieved.\")\n",
        "        else:\n",
        "            print(f\"{session_number} : Error retrieving results - {response.status_code}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8_lpKfmLcaQ-",
        "outputId": "756447fc-8c07-43f6-f6a6-4db2fcb68e62"
      },
      "outputs": [],
      "source": [
        "input_folder = \"input\"       # pull xml,json, or pubtator formatted files here to be annotated\n",
        "bioconcept = \"Gene\"                                         # we only want the genes\n",
        "output_file_session_number = \"Session_Number.txt\"  # output file for session number which will be used to retrieve the results\n",
        "submit_text_request(input_folder, bioconcept, output_file_session_number)    # submit the text for annotation"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "02JFZ15QfHqt"
      },
      "source": [
        "## Running the code below takes time, so it might say error 400. Just rerun it a bit later."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "mmQYNjmJdzwO",
        "outputId": "0284e0d6-cf18-4d40-85f9-b49dd3e7890a"
      },
      "outputs": [],
      "source": [
        "input_folder = \"input\"\n",
        "inputfile_session_number = \"Session_Number.txt\"\n",
        "output_folder = \"output\"\n",
        "submit_text_retrieve(input_folder, inputfile_session_number, output_folder)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "duMZk-ubdK0K"
      },
      "source": [
        "# 2. Relevance (Done)\n",
        "- This can be integrated with Gene Chunking. We can just check if the paper has genes (i.e check if there is an item in the dictionary @ line 171)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S-Gak1R_RF8Y"
      },
      "source": [
        "*3. Gene Chunking*\n",
        "=============\n",
        "Given a **preannotated** article (check input folder) analyze the **XML** file via `parse_xml_file(xml_file_path)` which returns a dictionary of \"Gene Class\" from that file  \n",
        "\n",
        "The Gene Class has the following information:\n",
        "- self.gene_id: the gene id (from the xml file)\n",
        "- self.occurences: A list of occurences of the gene in the text. Each occurence is **3-sentences long**.\n",
        "    - The second sentence contains the explicit mention of the gene.\n",
        "    - The first and third sentences are the context of the gene mention.\n",
        "- self.symbol: the gene symbol (from NCBI. Initialized to None)\n",
        "- self.organism: the organism of the gene (from NCBI. Initialized to None)\n",
        "- self.full_name: the full name of the gene (from NCBI. Initialized to None. Set to name_from_article if not found in NCBI)\n",
        "- self.also_known_as: a list of other names of the gene (from NCBI. Initialized to None. Set to name_from_article if not found in NCBI)\n",
        "- self.name_from_article: the name of the gene as found in the article\n",
        "\n",
        "Steps:\n",
        "1. define the class Gene\n",
        "2. Parse the xml file\n",
        "3. update the Gene class with the information from NCBI\n",
        "Now you have a dictionary of Gene classes.~\n",
        "\n",
        "TIPS:\n",
        "- control + F \"sentence_buffer\" to find where you can adjust the buffer"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "M9sSOV2PXqOJ",
        "outputId": "6722699e-967f-4dee-c51c-fce74a88168e"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com\n",
            "Requirement already satisfied: biopython in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (1.85)\n",
            "Requirement already satisfied: numpy in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from biopython) (1.26.4)\n",
            "There are 61 genes in the paper.\n",
            "Gene(433766)\n",
            "  Symbol          : Trim63\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : tripartite motif-containing 63\n",
            "  Also Known As   : MuRF1, RF1, Rnf28\n",
            "  In-text Name    : MuRF1\n",
            "  Occurrences     : ['MuRF1 mono-ubiquitinates TRalpha to inhibit T3-induced cardiac hypertrophy in vivo', \"However, how the UPS regulates T3-mediated activity is not well understood. In this study, we aim to determine the role of the muscle-specific ubiquitin ligase muscle ring finger-1 (MuRF1) in regulating T3-induced cardiomyocyte growth. An increase in MuRF1 expression inhibits T3-induced physiological cardiac hypertrophy, whereas a decrease in MuRF1 expression enhances T3's activity both in vitro and in cardiomyocytes in vivo.\", 'MuRF1 interacts directly with TRalpha to inhibit its activity by posttranslational ubiquitination in a non-canonical manner. We then demonstrated that a nuclear localization apparatus that regulates/inhibits nuclear receptors by sequestering them within a subcompartment of the nucleus was necessary for MuRF1 to inhibit T3 activity. This work implicates a novel mechanism that enhances the beneficial T3 activity specifically within the heart, thereby offering a potential target to enhance cardiac T3 activity in an organ-specific manner.', 'For example, estrogen, progesterone, androgen, and glucocorticoid nuclear receptors are degraded in a proteasome-dependent manner in some conditions, with their heterodimeric binding partners RARgamma and RXR also degraded by parallel mechanisms. The muscle-specific ubiquitin ligase MuRF1 has recently been shown to specifically inhibit PPARalpha, but not PPARbeta/delta or PPARgamma1, activity in vivo. In contrast to the previous studies on nuclear receptors, MuRF1 was found to inhibit PPARalpha in the cardiomyocyte through mono-ubiquitination that enhanced nuclear export without promoting PPARalpha degradation.', 'By mono-ubiquitinating one to three lysines adjacent to a newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 inhibited PPARgamma1, whereas MuRF3 inhibited PPARbeta activity in vivo by mono-ubiquitination. In this study, we identify the cardiac ubiquitin ligase MuRF1 as an inhibitor of T3-induced physiological cardiac hypertrophy in vivo.', 'We determined that MuRF1 directly interacts with a specific thyroid receptor to mediate its posttranslational modification and inhibit its activity. These studies then determined a previously undescribed mechanism in which MuRF1 inhibits T3 activity, by altering TRalpha localization via posttranslational ubiquitination.', \"We next aim to further define how MuRF1 inhibits T3-induced cardiomyocyte hypertrophy mechanistically. We focused our attention on MuRF1's regulation of thyroid receptor (TR) TRalpha1 activity, as it is the only isoform in cardiomyocytes that stimulates growth (the less common TRalpha2, and TRbeta1 are present but lack this activity) (,). In these studies, we identified that increasing MuRF1 expression significantly inhibited TRalpha-induced TRE-luciferase activity (Fig.\", '2A). As another way to measure TRalpha activity, we next evaluated how increasing MuRF1 would affect TR target genes and identified that MuRF1 significantly inhibited sarco/endoplasmic reticulum Ca2 -ATPase (SERCA expression (Fig. 2B)).', \"2B). As MuRF1's inhibition of TRalpha could result from poly-ubiquitination and proteasome-dependent degradation, we performed immunoblot analysis of TRalpha levels under conditions where MuRF1 inhibited its activity and found that increasing MuRF1 did not change TRalpha protein concentration (Fig. 2C).\", 'As both increasing and decreasing MuRF1 expression in T3-induced cardiomyocyte hypertrophy did not affect the steady-state levels of TRalpha (Fig. 2C), we were able to rule out MuRF1-mediated TRalpha degradation as a mechanism.', \"In recent studies, we demonstrated that MuRF1 inhibited PPARalpha activity similarly without affecting PPARalpha protein concentration. Using confocal microscopy, we found that increasing MuRF1 removed nuclear PPARalpha in a majority of the cells that subsequent studies found was due to MuRF1-targeted increased in nuclear export. Therefore, we next determined whether MuRF1 inhibits T3-induced TRalpha1 activity by altering TRalpha1's nuclear localization.\", 'HL-1 cardiomyocytes were transfected with TRalpha1 and/or MuRF1 ( /-T3) and analyzed by confocal microscopy. Interestingly, T3 altered the nuclear localization of TRalpha1 (Fig.', '3A, column 3). Specifically, it appeared that MuRF1 altered TRalpha1 nuclear localization by moving it to the nuclear periphery, a region closely aligned with the endoplasmic reticulum and microtubule anchoring complexes. To confirm this localization, we first identified the co-localization of the nuclear proteins lamin beta1 and desmin with TRalpha1 using confocal immunofluorescent analysis (Fig.', '4A). Increasing MuRF1 expression significantly increased the co-localization of TRalpha1 with nuclear lamin beta1 (Fig. 4), while similarly increasing the co-localization of TRalpha1 with desmin (Supplementary Fig.', \"S2). Lastly, we assessed MuRF1's effects on TRalpha1 nuclear localization by isolating the cardiomyocyte nuclear fraction by differential centrifugation and measuring the amount of TRalpha1 by immunoblot (Fig. 4B).\", 'Together, these studies demonstrate that increasing MuRF1 sequestered TRalpha1 in the nucleus, to inhibit TRalpha1 nuclear activity.', 'To assess whether TRalpha is a MuRF1 substrate, we next determined whether MuRF1 interacted directly with TRalpha by transfecting cells with MuRF1 and immunoprecipitating the myc-labeled MuRF1 (Fig. 5A). To determine the specific region of TRalpha that MuRF1 interacts, we next performed a pull-down assay using recombinant TRalpha domains to identify their ability to interact with full-length MuRF1 (Fig.', '5B). We found that MuRF1 binds the TRalpha hinge (or D, as denoted on the diagram) domain of the receptor (Fig. 5B, right).', \"With evidence that MuRF1 interacts with TRalpha, we next tested MuRF1's ability to ubiquitinate TRalpha in vitro. We identified that MuRF1 catalyzes TRalpha mono-ubiquitination (Fig. 5C), a posttranslational modification that does not target substrate for degradation, consistent with our initial findings in cardiomyocytes.\", \"With evidence that MuRF1 mono-ubiquitinated TR in a cell-free system, we next aim to identify these in cells. We observed an unexpected complication of identifying MuRF1's mono-ubiquitination in cells where a strong background of TRpoly-ubiquitination occurs (Fig. 5D).\", '5E, as described in TRbeta). As immunoprecipitated TRalpha was polyubiquitinated in control cells (Figure 5D, red box), we hypothesized that the poly-ubiquitination competed for the sample lysines that MuRF1 mono-ubiquitinated. Support for this hypothesis came from studies with increased MuRF1 (AdMuRF1), where there was decreased poly-ubiquitination (Fig.', '5D, blue box).', \"The centrosome-associated protein (CAP) CAP350 has recently been identified as protein involved in regulating nuclear receptor activity, including PPAR. Initially, we analyzed MuRF1's effects on the co-localization of endogenous CAP350 and TRalpha in cardiomyocytes by confocal imaging (Fig. 6A).\", 'We found that increasing MuRF1 significantly increased nuclear CAP350 and the co-localization of CAP350 and TRalpha (Fig. 6A, right two columns). The increased co-localization of CAP350 and TRalpha driven by increasing MuRF1 was most apparent in the perinuclear region following T3 stimulation (Fig.', \"6A, in yellow). To solidify CAP350's mechanistic relationship to MuRF1's inhibition of TRalpha, we next knocked down CAP350 (to  lt; 50 , Fig. 6B) and determined its effect on cardiomyocyte cross-sectional area (Fig.\", '6C). The MuRF1 attenuation of T3-stimulated cardiomyocyte growth (experimental group 6 vs group 3, Fig. 6C) was significantly overcome (i.e increased in cross-sectional area size) with CAP350 knock-down (experimental group 5, Fig.', \"6C). Together, these studies demonstrate a role of CAP350 in MuRF1's regulation of TR in cardiomyocytes.\", 'We next aim to identify the region within TRalpha protein that MuRF1 ubiquitinates to modify its activity. A well-known challenge to studying protein ubiquitination is its heterogeneous nature; when a major ubiquitination site (lysine) is mutated, it is common for nearby lysines to be alternatively targeted.', 'S3): TRalpha A/B (AF1, N-terminus), TRalpha C (DNA-binding), TRalpha D (hinge), and TRalpha E/F (ligand-binding). These TRalpha constructs were then expressed in cardiomyocytes and assessed for their response to MuRF1 alterations in localization by confocal microscopy (). Increasing MuRF1 expression localized the TRA/B, C, and D lysine mutants to the perinuclear region as observed with the native receptor (Fig.', '7A). In contrast, cells expressing the E/F TR domain mutant exhibit a distinctly nuclear localization of the receptor rather than a perinuclear one as observed with the native form implying that MuRF1 ubiquitinates residues in this domain to negatively regulate TR function (Fig. 7A, far right).', '7B, far right). Together, these studies illustrate that MuRF1 inhibits TRalpha activity by mono-ubiquitinating it in the E/F region, resulting in an interaction with CAP350 and nuclear sequestration in cardiomyocytes.', 'To determine the physiological relevance of these mechanistic findings in vivo, we next challenged MuRF1-/- and MHC-MuRF1 cardiomyocyte-specific transgenic mice (MuRF1Tg ) with daily T3 i.p. for 2 weeks. We followed the cardiac phenotype by conscious echocardiography and histological analysis.', 'Echocardiographic and histological assessment identified that T3 induced an exaggerated cardiac hypertrophy in MuRF1-/- compared with sibling-matched wild-type controls, evidenced by increases in left ventricular anterior wall thickness and LV mass (Fig. 8A, Table 1 for full echocardiographic data) and cardiomyocyte area (Fig. 8B).', 'Interestingly, MuRF1-/- had an enhanced systolic function, reflecting increases in contractility and reflective of the T3-induced enhancements driven by increases in SERCA and phospholamban. Overall, the MuRF1-/- mice had a 53.6  increase in cardiomyocyte area compared with the 30.5  increase found in the wild-type controls (Fig. 8B).', '8C), consistent with an increase in pathological hypertrophy, identified by increases in BNP, alpha1-skeletal muscle actin, and ANF (data not shown). The observed effects on TH-induced physiological hypertrophy in our mouse models were most likely attributable to MuRF1 activity as circulating concentrations of T3/T4 were not altered in MuRF1-/- or MuRF1Tg  mice (Supplementary Fig. S5A and B).', 'All studies included parallel PBS-injected controls, with no alterations seen (as detailed in Supplementary Tables S1 and S2). Overall, these studies demonstrate that the MuRF1-/- mice exhibited an exaggerated cardiac hypertrophy, while MuRF1Tg  hearts lacked the development of cardiac hypertrophy in response to T3, consistent with a role for MuRF1 in inhibiting TRalpha activity in vivo. With no evidence for alterations in steady-state TRalpha protein levels in vivo (Supplementary Fig.', \"S4), like those found in vitro (Fig. 2C), cardiomyocyte MuRF1's inhibition of TRalpha is relevant in the intact heart illustrating alternative mechanisms that could be targeted to protect against heart failure in a tissue-specific manner.\", 'The closely related receptors that heterodimerize with PPARs, such as RARgamma and RXR, are also degraded by the ubiquitin proteasome system with fewer details available. We have recently extended these findings to demonstrate that the muscle-specific ubiquitin ligase MuRF1 specifically inhibits PPARalpha, but not PPARbeta/delta or PPARgamma1 activity in vivo. In contrast to previous studies on nuclear receptors, we identified that MuRF1 had an inhibitory effect on PPARalpha in cardiomyocytes by enhancing nuclear export, not degradation.', 'By multi-mono-ubiquitination of lysines adjacent to the newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 was found to inhibit PPARgamma1, while MuRF3 inhibits predominantly PPARbeta activity in vivo through mono-ubiquitination of substrates.', 'While cardiac MuRF1 has multiple roles in the heart, a consistent theme of its activity has been its anti-hypertrophic activity and regulation of metabolism. In pressure overload-induced cardiac hypertrophy, MuRF1 inhibits pathological hypertrophy by inhibiting the serum response transcription factor. In physiological hypertrophy induced by exercise (IGF-1), MuRF1 transcriptionally inhibits cardiomyocyte growth by inhibiting cJun-mediated increases in IGF-1 signaling proteins, by preferentially recognizing phospho-cJun, poly-ubiquitinating, and targeting it for proteasome-dependent degradation in vivo.', 'Metabolically, MuRF1 has been found to ubiquitinate oxidized creatine kinase, mediating its turnover in vivo. Increasing MuRF1 decreases creatine kinase activity significantly in the intact heart, resulting in a reduction of this critical transporter of ATP from the mitochondria to the sarcomere, resulting in heart failure. MuRF1 also inhibits PPARalpha, interacts with a host of energy-related proteins, and alters the cardiac metabolome in vivo.', \"Building upon this larger theme, this study demonstrates MuRF1's anti-hypertrophic role in T3-mediated cardiomyocyte growth, with broader implications in T3-regulated metabolic rates critical to cardiac function in health and disease.\", \"The link between MuRF1 and the centrosome-associated protein 350 (CAP350) stems from the observation that MuRF1 alters nuclear TRalpha distribution in a similar manner as CAP350 does with PPARalpha. Although the role of CAP350 in regulating nuclear hormone receptors (such as TRalpha) is established, the mechanistic details have yet to be elucidated. In previous studies, the recently defined centrosome- and cytoskeleton-associated protein CAP350 was found to inhibit PPARalpha activity by redirecting PPARalpha's nuclear distribution to the perinuclear area with punctate distribution.\", \"In this study, we similarly observed that MuRF1 altered the nuclear distribution of TR (Fig. 3A), which led to subsequent studies to delineate the underlying mechanistic relationship between MuRF1's anti-hypertrophic activity with T3 stimulation and MuRF1's reliance upon CAP350 for preventing hypertrophic growth (Fig. 6C).\", \"The present studies illustrate that MuRF1's attenuation of T3-induced cardiomyocyte growth is dependent on TRalpha mono-ubiquitination of the ligand-binding domain (Fig. 7) and interaction with CAP350 (with significantly smaller cardiomyocytes with MuRF1 and CAP350 present, Fig. 6C, group 6).\", 'Our working hypothesis on the role of MuRF1 in this process is that MuRF1 mono-ubiquitinates TRalpha to enhance its affinity for CAP350, which sequesters TRalpha-Ub in nuclear compartments making it inaccessible to DNA promoters to diminish transcriptional activity.', 'SUMOylation and mono-ubiquitination also inhibit T3-induced TRalpha activity. Interestingly, TRalpha is SUMOylated at lysines 284 and 389, both of which are within the ligand-binding domain (E/F region) of TRalpha, consistent with our findings that MuRF1 interacts with and catalyzes the addition of ubiquitin in the E/F domain of TRalpha are critical to regulation by MuRF1 (Fig. 7, lysines 187 through 389, Supplementary Fig.', 'The exaggerated hypertrophy in the MuRF1-/- hearts is consistent with an enhanced cardiomyocyte TRalpha activity (Fig. 1A, Fig. 8A).', 'Similarly, the lack of cardiac hypertrophy in the MuRF1Tg  hearts was consistent with in vitro findings (Fig. 1B and C, Fig.', '8C). One explanation for this may be the energetic defects seen in the MuRF1Tg  mice, such as decreased creatine kinase activity and altered fatty acid oxidation. Conversely, the enhanced MuRF1-/- function seen with T3 stimulation (Fig.', '8A) may reflect an athletic phenotype previously identified in the MuRF1-/- hearts that may be related, but not fully due to enhanced T3 activity resulting from decreased ubiquitination.', \"The use of T3 in heart failure has been a topic of discussion for many decades, with a growing number of animal and human studies suggesting that TH treatment may contribute to improved cardiovascular outcomes, although off-target effects may limit application, particularly in patients who appear systemically euthyroid. The present studies offer insight into how one might specifically enhance the actions of TH by inhibiting MuRF1's antagonism, thereby targeting muscle-specific activation without necessarily supplementing T3 systemically.\", 'Muscle ring finger protein-1 inhibits PKC{epsilon} activation and prevents cardiomyocyte hypertrophy', 'Non-targeted metabolomics analysis of cardiac muscle ring finger-1 (MuRF1), MuRF2, and MuRF3 in vivo reveals novel and redundant metabolic changes', 'The ubiquitin ligase MuRF1 protects against cardiac ischemia/reperfusion injury by its proteasome-dependent degradation of phospho-c-Jun', 'The ubiquitin ligase MuRF1 regulates PPARalpha activity in the heart by enhancing nuclear export via monoubiquitination', 'MURF-1 and MURF-2 target a specific subset of myofibrillar proteins redundantly: Towards understanding MURF-dependent muscle ubiquitination']\n",
            "  Document ID     : PMC5453669\n",
            "There are 57 occurrence(s) (each occurrence is at most 3 sentences long) for gene 433766 : MuRF1.\n",
            "\n",
            "Gene(14685)\n",
            "  Symbol          : Gnat1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : G protein subunit alpha transducin 1\n",
            "  Also Known As   : Gnat-1, Hg1f, Ird1, Ird2, Tralpha, irdc, irdr, transducin\n",
            "  In-text Name    : TRalpha\n",
            "  Occurrences     : ['MuRF1 mono-ubiquitinates TRalpha to inhibit T3-induced cardiac hypertrophy in vivo', \"An increase in MuRF1 expression inhibits T3-induced physiological cardiac hypertrophy, whereas a decrease in MuRF1 expression enhances T3's activity both in vitro and in cardiomyocytes in vivo. MuRF1 interacts directly with TRalpha to inhibit its activity by posttranslational ubiquitination in a non-canonical manner. We then demonstrated that a nuclear localization apparatus that regulates/inhibits nuclear receptors by sequestering them within a subcompartment of the nucleus was necessary for MuRF1 to inhibit T3 activity.\", 'Its biologically active form (T3) is involved in directing myocyte alignment and geometry, extracellular matrix remodeling, and cardiac function through the effects on contraction. T3 mediates both direct influences on cardiac function by affecting calcium flux and calcium handling proteins, in addition to voltage-gated potassium channels to affect contractility, and through transcriptional regulation as TH is a ligand for the thyroid hormone receptor, predominantly the alpha isoform, TRalpha in the heart. A decrease in TH concentrations has been observed in acute myocardial infarction, and low circulating TH is commonly observed with severe heart failure (HF).', 'We determined that MuRF1 directly interacts with a specific thyroid receptor to mediate its posttranslational modification and inhibit its activity. These studies then determined a previously undescribed mechanism in which MuRF1 inhibits T3 activity, by altering TRalpha localization via posttranslational ubiquitination.', \"We focused our attention on MuRF1's regulation of thyroid receptor (TR) TRalpha1 activity, as it is the only isoform in cardiomyocytes that stimulates growth (the less common TRalpha2, and TRbeta1 are present but lack this activity) (,). In these studies, we identified that increasing MuRF1 expression significantly inhibited TRalpha-induced TRE-luciferase activity (Fig. 2A).\", 'As another way to measure TRalpha activity, we next evaluated how increasing MuRF1 would affect TR target genes and identified that MuRF1 significantly inhibited sarco/endoplasmic reticulum Ca2 -ATPase (SERCA expression (Fig. 2B)). Other well-known TRalpha targets (alphaMHC and ANT) were not increased in these studies, which may be due to the limited 30-min T3 stimulation (Fig.', \"2B). As MuRF1's inhibition of TRalpha could result from poly-ubiquitination and proteasome-dependent degradation, we performed immunoblot analysis of TRalpha levels under conditions where MuRF1 inhibited its activity and found that increasing MuRF1 did not change TRalpha protein concentration (Fig. 2C).\", 'As both increasing and decreasing MuRF1 expression in T3-induced cardiomyocyte hypertrophy did not affect the steady-state levels of TRalpha (Fig. 2C), we were able to rule out MuRF1-mediated TRalpha degradation as a mechanism.', 'To assess whether TRalpha is a MuRF1 substrate, we next determined whether MuRF1 interacted directly with TRalpha by transfecting cells with MuRF1 and immunoprecipitating the myc-labeled MuRF1 (Fig. 5A). To determine the specific region of TRalpha that MuRF1 interacts, we next performed a pull-down assay using recombinant TRalpha domains to identify their ability to interact with full-length MuRF1 (Fig.', '5B). We found that MuRF1 binds the TRalpha hinge (or D, as denoted on the diagram) domain of the receptor (Fig. 5B, right).', \"With evidence that MuRF1 interacts with TRalpha, we next tested MuRF1's ability to ubiquitinate TRalpha in vitro. We identified that MuRF1 catalyzes TRalpha mono-ubiquitination (Fig. 5C), a posttranslational modification that does not target substrate for degradation, consistent with our initial findings in cardiomyocytes.\", '5E, as described in TRbeta). As immunoprecipitated TRalpha was polyubiquitinated in control cells (Figure 5D, red box), we hypothesized that the poly-ubiquitination competed for the sample lysines that MuRF1 mono-ubiquitinated. Support for this hypothesis came from studies with increased MuRF1 (AdMuRF1), where there was decreased poly-ubiquitination (Fig.', '5D, blue box). While the identification of a mono-ubiquitinated TRalpha was not achieved, we did identify the common dichotomy competing mono- and poly-ubiquitin for substrates, known to be dependent on the types of E2 present and ratios of E3 to substrates.', \"The centrosome-associated protein (CAP) CAP350 has recently been identified as protein involved in regulating nuclear receptor activity, including PPAR. Initially, we analyzed MuRF1's effects on the co-localization of endogenous CAP350 and TRalpha in cardiomyocytes by confocal imaging (Fig. 6A).\", 'We found that increasing MuRF1 significantly increased nuclear CAP350 and the co-localization of CAP350 and TRalpha (Fig. 6A, right two columns). The increased co-localization of CAP350 and TRalpha driven by increasing MuRF1 was most apparent in the perinuclear region following T3 stimulation (Fig.', \"6A, in yellow). To solidify CAP350's mechanistic relationship to MuRF1's inhibition of TRalpha, we next knocked down CAP350 (to  lt; 50 , Fig. 6B) and determined its effect on cardiomyocyte cross-sectional area (Fig.\", 'We next aim to identify the region within TRalpha protein that MuRF1 ubiquitinates to modify its activity. A well-known challenge to studying protein ubiquitination is its heterogeneous nature; when a major ubiquitination site (lysine) is mutated, it is common for nearby lysines to be alternatively targeted.', 'As the goal of mutating a lysine would be to prevent specific ubiquitination, we took an alternative approach and mutated lysine residues in specific regions of the TR protein. Four full-length TRalpha constructs were created with lysine residues mutated to arginine (KR) (Supplementary Fig. S3): TRalpha A/B (AF1, N-terminus), TRalpha C (DNA-binding), TRalpha D (hinge), and TRalpha E/F (ligand-binding).', 'These TRalpha constructs were then expressed in cardiomyocytes and assessed for their response to MuRF1 alterations in localization by confocal microscopy (). Increasing MuRF1 expression localized the TRA/B, C, and D lysine mutants to the perinuclear region as observed with the native receptor (Fig. 7A).', '7B, far right). Together, these studies illustrate that MuRF1 inhibits TRalpha activity by mono-ubiquitinating it in the E/F region, resulting in an interaction with CAP350 and nuclear sequestration in cardiomyocytes.', 'All studies included parallel PBS-injected controls, with no alterations seen (as detailed in Supplementary Tables S1 and S2). Overall, these studies demonstrate that the MuRF1-/- mice exhibited an exaggerated cardiac hypertrophy, while MuRF1Tg  hearts lacked the development of cardiac hypertrophy in response to T3, consistent with a role for MuRF1 in inhibiting TRalpha activity in vivo. With no evidence for alterations in steady-state TRalpha protein levels in vivo (Supplementary Fig.', \"S4), like those found in vitro (Fig. 2C), cardiomyocyte MuRF1's inhibition of TRalpha is relevant in the intact heart illustrating alternative mechanisms that could be targeted to protect against heart failure in a tissue-specific manner.\", 'Understanding how the thyroid hormone receptor alpha is regulated is critical to our understanding of heart disease. Inhibiting TRalpha in myocardial infarction experimentally impairs post-ischemic performance due to activation of p38 and an increased phospholamban. Conversely, T3 therapy protects the heart against ischemia:reperfusion injury in a TRalpha-dependent manner.', 'T3 also seems to counteract the maladaptive responses implicated in adrenergic induced cardiomyocyte remodeling. Stimulation of cardiomyocytes with phenylephrine results in increased fetal gene expression (betaMHC); co-stimulating cardiomyocytes with phenylephrine and T3 blocks these maladaptive changes, while increasing TRalpha expression in the nucleus and cytosol. In models of myocardial infarction (acute coronary artery ligation) leading to compensated heart failure, a distinct pattern of TRalpha expression was found during the course of post-ischemic remodeling.', 'TRalpha was upregulated initially, then markedly decreased during the transition to decompensated heart failure, resulting in tissue hypothyroidism. While circulating T3 and T4 levels remained unchanged, a link between mTOR and TRalpha was identified. Together, these studies illustrate the critical role the TRalpha plays in heart failure and how regulating TRalpha expression can critically affect disease progression and severity.', \"The link between MuRF1 and the centrosome-associated protein 350 (CAP350) stems from the observation that MuRF1 alters nuclear TRalpha distribution in a similar manner as CAP350 does with PPARalpha. Although the role of CAP350 in regulating nuclear hormone receptors (such as TRalpha) is established, the mechanistic details have yet to be elucidated. In previous studies, the recently defined centrosome- and cytoskeleton-associated protein CAP350 was found to inhibit PPARalpha activity by redirecting PPARalpha's nuclear distribution to the perinuclear area with punctate distribution.\", \"6C). The present studies illustrate that MuRF1's attenuation of T3-induced cardiomyocyte growth is dependent on TRalpha mono-ubiquitination of the ligand-binding domain (Fig. 7) and interaction with CAP350 (with significantly smaller cardiomyocytes with MuRF1 and CAP350 present, Fig.\", '6C, group 6). Our working hypothesis on the role of MuRF1 in this process is that MuRF1 mono-ubiquitinates TRalpha to enhance its affinity for CAP350, which sequesters TRalpha-Ub in nuclear compartments making it inaccessible to DNA promoters to diminish transcriptional activity.', 'This is an emerging, common theme, further illustrated with the estrogen receptor (ER), when the BRCA1 ubiquitin ligase mono-ubiquitinates it to inhibit its activity and the suppression of DNA polymerase Eta, when mono-ubiquitinated by the Pirh2 ubiquitin ligase. The present studies demonstrate a novel role for CAP350 activity in the heart and an emerging role in regulating TRalpha activity that has not been previously reported.', 'Posttranslational modification of TRalpha by ubiquitin has not previously been described; however, its regulation by the closely related small ubiquitin-related modifier (SUMO) has been in cancer. Paralleling the ubiquitination process, SUMOylation involves the addition of a single SUMO modification, added by the cooperation of the SUMO E1, E2, and E3 (SUMO ligase) enzymes.', 'SUMOylated TRalpha has been reported in both liver (HepG2) and pituitary (GH3) cell lines (Liu et al.2012), mediated by the SUMO ligase PIAS1. In the absence of ligand, PIAS1 SUMOylates TRalpha, whereas the SUMO ligase PIASx SUMOylates TRbeta only in the presence of ligand. Both SUMO1 and SUMO3 (two different types of SUMO modification) have been reported in T3-dependent gene regulation, closely matching TR activity and cofactor recruitment to the TRE in ChIP assays.', 'This study describes a non-canonical ubiquitination (mono-ubiquitination) of TRalpha, which parallels the process of TRalpha SUMOylation. Both SUMOylation and mono-ubiquitination occur at specific lysines and do not result in degradation of TRalpha. SUMOylation and mono-ubiquitination also inhibit T3-induced TRalpha activity.', 'Interestingly, TRalpha is SUMOylated at lysines 284 and 389, both of which are within the ligand-binding domain (E/F region) of TRalpha, consistent with our findings that MuRF1 interacts with and catalyzes the addition of ubiquitin in the E/F domain of TRalpha are critical to regulation by MuRF1 (Fig. 7, lysines 187 through 389, Supplementary Fig. S3 for additional information).', 'The physiological importance of SUMOylation in health and disease is not known; however, the common T3-mediated growth functions of TRalpha make it likely that SUMOylation in hepatocytes and pituitary parallel the critical regulation of mono-ubiquitination in cardiomyocytes reported here.', 'The exaggerated hypertrophy in the MuRF1-/- hearts is consistent with an enhanced cardiomyocyte TRalpha activity (Fig. 1A, Fig.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 35 occurrence(s) (each occurrence is at most 3 sentences long) for gene 14685 : TRalpha.\n",
            "\n",
            "Gene(11651)\n",
            "  Symbol          : Akt1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : thymoma viral proto-oncogene 1\n",
            "  Also Known As   : Akt, LTR-akt, PKB, PKB/Akt, PKBalpha, Rac\n",
            "  In-text Name    : Akt\n",
            "  Occurrences     : ['Thyroid hormone (TH) is recognized for its role in cellular metabolism and growth and participates in homeostasis of the heart. T3 activates pro-survival pathways including Akt and mTOR. Treatment with T3 after myocardial infarction is cardioprotective and promotes elements of physiological hypertrophic response after cardiac injury.', 'Treatment with TH following myocardial infarction is cardioprotective and promotes elements of physiological hypertrophic response after cardiac injury. T3 activates many of the physiological changes that would benefit the heart, activating pro-survival pathways Akt and mTOR and paralleling changes induced by exercise-mediated by IGF-1. Despite our rich knowledge of the cardioprotective effects of T3, little is known about the intracellular regulation of its action.', 'This study may offer several clinically relevant insights into approaches to treating heart failure. TH induces molecular changes that benefit the failing heart, including the stimulation of pro-survival pathways such as Akt and mTOR that is accompanied by improved cardiac function, paralleling changes induced by exercise, mediated by IGF-1, to drive increases in contractile proteins critical to cardiac function and performance. Exercise has also been shown to increase myocardial TR expression and T3 levels in a small cohort of HF patients with ventricular assist devices, underscoring the role for TH in promoting positive adaptive remodeling of the injured heart.', 'Thyroid hormone improves the mechanical performance of the post-infarcted diabetic myocardium: A response associated with up-regulation of Akt/mTOR and AMPK activation', 'Dose-dependent effects of thyroid hormone on post-ischemic cardiac performance: Potential involvement of Akt and ERK signalings']\n",
            "  Document ID     : PMC5453669\n",
            "There are 5 occurrence(s) (each occurrence is at most 3 sentences long) for gene 11651 : Akt.\n",
            "\n",
            "Gene(56717)\n",
            "  Symbol          : Mtor\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : mechanistic target of rapamycin kinase\n",
            "  Also Known As   : 2610315D21Rik, FRAP, FRAP2, Frap1, RAFT1, RAPT1, flat\n",
            "  In-text Name    : mTOR\n",
            "  Occurrences     : ['Thyroid hormone (TH) is recognized for its role in cellular metabolism and growth and participates in homeostasis of the heart. T3 activates pro-survival pathways including Akt and mTOR. Treatment with T3 after myocardial infarction is cardioprotective and promotes elements of physiological hypertrophic response after cardiac injury.', 'Treatment with TH following myocardial infarction is cardioprotective and promotes elements of physiological hypertrophic response after cardiac injury. T3 activates many of the physiological changes that would benefit the heart, activating pro-survival pathways Akt and mTOR and paralleling changes induced by exercise-mediated by IGF-1. Despite our rich knowledge of the cardioprotective effects of T3, little is known about the intracellular regulation of its action.', 'TRalpha was upregulated initially, then markedly decreased during the transition to decompensated heart failure, resulting in tissue hypothyroidism. While circulating T3 and T4 levels remained unchanged, a link between mTOR and TRalpha was identified. Together, these studies illustrate the critical role the TRalpha plays in heart failure and how regulating TRalpha expression can critically affect disease progression and severity.', 'This study may offer several clinically relevant insights into approaches to treating heart failure. TH induces molecular changes that benefit the failing heart, including the stimulation of pro-survival pathways such as Akt and mTOR that is accompanied by improved cardiac function, paralleling changes induced by exercise, mediated by IGF-1, to drive increases in contractile proteins critical to cardiac function and performance. Exercise has also been shown to increase myocardial TR expression and T3 levels in a small cohort of HF patients with ventricular assist devices, underscoring the role for TH in promoting positive adaptive remodeling of the injured heart.', 'Thyroid hormone improves the mechanical performance of the post-infarcted diabetic myocardium: A response associated with up-regulation of Akt/mTOR and AMPK activation']\n",
            "  Document ID     : PMC5453669\n",
            "There are 5 occurrence(s) (each occurrence is at most 3 sentences long) for gene 56717 : mTOR.\n",
            "\n",
            "Gene(21823)\n",
            "  Symbol          : Th\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : tyrosine hydroxylase\n",
            "  Also Known As   : TH\n",
            "  In-text Name    : TH\n",
            "  Occurrences     : ['Thyroid hormone (TH) is recognized for its role in cellular metabolism and growth and participates in homeostasis of the heart. T3 activates pro-survival pathways including Akt and mTOR.', 'Thyroid hormone (TH) is recognized for its role in cellular metabolism and growth and participates in homeostasis of the heart. Its biologically active form (T3) is involved in directing myocyte alignment and geometry, extracellular matrix remodeling, and cardiac function through the effects on contraction. T3 mediates both direct influences on cardiac function by affecting calcium flux and calcium handling proteins, in addition to voltage-gated potassium channels to affect contractility, and through transcriptional regulation as TH is a ligand for the thyroid hormone receptor, predominantly the alpha isoform, TRalpha in the heart.', 'A decrease in TH concentrations has been observed in acute myocardial infarction, and low circulating TH is commonly observed with severe heart failure (HF). Treatment with TH following myocardial infarction is cardioprotective and promotes elements of physiological hypertrophic response after cardiac injury. T3 activates many of the physiological changes that would benefit the heart, activating pro-survival pathways Akt and mTOR and paralleling changes induced by exercise-mediated by IGF-1.', '5D). As described in cancer cells, poly-ubiquitination is a common way in which thyroid hormone receptors have been described natively (Fig. 5E, as described in TRbeta).', '8C), consistent with an increase in pathological hypertrophy, identified by increases in BNP, alpha1-skeletal muscle actin, and ANF (data not shown). The observed effects on TH-induced physiological hypertrophy in our mouse models were most likely attributable to MuRF1 activity as circulating concentrations of T3/T4 were not altered in MuRF1-/- or MuRF1Tg  mice (Supplementary Fig. S5A and B).', 'This study may offer several clinically relevant insights into approaches to treating heart failure. TH induces molecular changes that benefit the failing heart, including the stimulation of pro-survival pathways such as Akt and mTOR that is accompanied by improved cardiac function, paralleling changes induced by exercise, mediated by IGF-1, to drive increases in contractile proteins critical to cardiac function and performance. Exercise has also been shown to increase myocardial TR expression and T3 levels in a small cohort of HF patients with ventricular assist devices, underscoring the role for TH in promoting positive adaptive remodeling of the injured heart.', 'Thyroid hormone signalling is altered in response to physical training in patients with end-stage heart failure and mechanical assist devices: Potential physiological consequences?', 'Hormone binding induces rapid proteasome-mediated degradation of thyroid hormone receptors', 'Cardiac hypertrophy and thyroid hormone signaling', 'Divergent regulation of cardiac KCND3 potassium channel expression by the thyroid hormone receptors alpha1 and beta1', 'Restoration of thyroid hormone balance: A game changer in the treatment of heart failure?', 'Thyroid hormone induces myocardial matrix degradation by activating matrix metalloproteinase-1', 'Thyroid hormone and cardiovascular system: From basic concepts to clinical application', 'Thyroid hormone receptor isoform-specific modification by small ubiquitin-like modifier (SUMO) modulates thyroid hormone-dependent gene regulation', 'Thyroid hormone and cardiac disease: From basic concepts to clinical application', 'Thyroid hormone improves the mechanical performance of the post-infarcted diabetic myocardium: A response associated with up-regulation of Akt/mTOR and AMPK activation', 'Dose-dependent effects of thyroid hormone on post-ischemic cardiac performance: Potential involvement of Akt and ERK signalings', 'Signaling mechanisms in thyroid hormone-induced cardiac hypertrophy', 'Thyroid hormone attenuates cardiac remodeling and improves hemodynamics early after acute myocardial infarction in rats', 'Thyroid hormone changes cardiomyocyte shape and geometry via ERK signaling pathway: Potential therapeutic implications in reversing cardiac remodeling?']\n",
            "  Document ID     : PMC5453669\n",
            "There are 20 occurrence(s) (each occurrence is at most 3 sentences long) for gene 21823 : TH.\n",
            "\n",
            "Gene(16000)\n",
            "  Symbol          : Igf1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : insulin-like growth factor 1\n",
            "  Also Known As   : C730016P09Rik, Igf-1, Igf-I\n",
            "  In-text Name    : IGF-1\n",
            "  Occurrences     : ['Treatment with TH following myocardial infarction is cardioprotective and promotes elements of physiological hypertrophic response after cardiac injury. T3 activates many of the physiological changes that would benefit the heart, activating pro-survival pathways Akt and mTOR and paralleling changes induced by exercise-mediated by IGF-1. Despite our rich knowledge of the cardioprotective effects of T3, little is known about the intracellular regulation of its action.', 'In pressure overload-induced cardiac hypertrophy, MuRF1 inhibits pathological hypertrophy by inhibiting the serum response transcription factor. In physiological hypertrophy induced by exercise (IGF-1), MuRF1 transcriptionally inhibits cardiomyocyte growth by inhibiting cJun-mediated increases in IGF-1 signaling proteins, by preferentially recognizing phospho-cJun, poly-ubiquitinating, and targeting it for proteasome-dependent degradation in vivo. Metabolically, MuRF1 has been found to ubiquitinate oxidized creatine kinase, mediating its turnover in vivo.', 'This study may offer several clinically relevant insights into approaches to treating heart failure. TH induces molecular changes that benefit the failing heart, including the stimulation of pro-survival pathways such as Akt and mTOR that is accompanied by improved cardiac function, paralleling changes induced by exercise, mediated by IGF-1, to drive increases in contractile proteins critical to cardiac function and performance. Exercise has also been shown to increase myocardial TR expression and T3 levels in a small cohort of HF patients with ventricular assist devices, underscoring the role for TH in promoting positive adaptive remodeling of the injured heart.', 'Muscle RING Finger-1 Attenuates IGF-1-dependent cardiomyocyte hypertrophy by inhibiting JNK signaling']\n",
            "  Document ID     : PMC5453669\n",
            "There are 4 occurrence(s) (each occurrence is at most 3 sentences long) for gene 16000 : IGF-1.\n",
            "\n",
            "Gene(19411)\n",
            "  Symbol          : Rarg\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : retinoic acid receptor, gamma\n",
            "  Also Known As   : Nr1b3, RAR-gamma, RARD, RARgamma2\n",
            "  In-text Name    : RARgamma\n",
            "  Occurrences     : ['Intracellular regulation of nuclear receptors by posttranslational modification by ubiquitin is well known and involves two distinct mechanisms: 1) targeted degradation of nuclear receptors by the proteasome and 2) ubiquitination-altering activity in non-degradatory ways. For example, estrogen, progesterone, androgen, and glucocorticoid nuclear receptors are degraded in a proteasome-dependent manner in some conditions, with their heterodimeric binding partners RARgamma and RXR also degraded by parallel mechanisms. The muscle-specific ubiquitin ligase MuRF1 has recently been shown to specifically inhibit PPARalpha, but not PPARbeta/delta or PPARgamma1, activity in vivo.', 'Alternatively, the androgen receptor and the PPARalpha are targeted to the proteasome in the absence of ligand. The closely related receptors that heterodimerize with PPARs, such as RARgamma and RXR, are also degraded by the ubiquitin proteasome system with fewer details available. We have recently extended these findings to demonstrate that the muscle-specific ubiquitin ligase MuRF1 specifically inhibits PPARalpha, but not PPARbeta/delta or PPARgamma1 activity in vivo.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 19411 : RARgamma.\n",
            "\n",
            "Gene(19013)\n",
            "  Symbol          : Ppara\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : peroxisome proliferator activated receptor alpha\n",
            "  Also Known As   : 4933429D07Rik, Nr1c1, PPAR-alpha, PPARalpha, Ppar\n",
            "  In-text Name    : PPARalpha\n",
            "  Occurrences     : ['For example, estrogen, progesterone, androgen, and glucocorticoid nuclear receptors are degraded in a proteasome-dependent manner in some conditions, with their heterodimeric binding partners RARgamma and RXR also degraded by parallel mechanisms. The muscle-specific ubiquitin ligase MuRF1 has recently been shown to specifically inhibit PPARalpha, but not PPARbeta/delta or PPARgamma1, activity in vivo. In contrast to the previous studies on nuclear receptors, MuRF1 was found to inhibit PPARalpha in the cardiomyocyte through mono-ubiquitination that enhanced nuclear export without promoting PPARalpha degradation.', 'By mono-ubiquitinating one to three lysines adjacent to a newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 inhibited PPARgamma1, whereas MuRF3 inhibited PPARbeta activity in vivo by mono-ubiquitination.', \"In recent studies, we demonstrated that MuRF1 inhibited PPARalpha activity similarly without affecting PPARalpha protein concentration. Using confocal microscopy, we found that increasing MuRF1 removed nuclear PPARalpha in a majority of the cells that subsequent studies found was due to MuRF1-targeted increased in nuclear export. Therefore, we next determined whether MuRF1 inhibits T3-induced TRalpha1 activity by altering TRalpha1's nuclear localization.\", \"The centrosome-associated protein (CAP) CAP350 has recently been identified as protein involved in regulating nuclear receptor activity, including PPAR. Initially, we analyzed MuRF1's effects on the co-localization of endogenous CAP350 and TRalpha in cardiomyocytes by confocal imaging (Fig.\", 'Some nuclear hormone receptors are degraded in both the presence and absence of their cognate ligands, such as the estrogen, progesterone, and glucocorticoid receptors ( ). Alternatively, the androgen receptor and the PPARalpha are targeted to the proteasome in the absence of ligand. The closely related receptors that heterodimerize with PPARs, such as RARgamma and RXR, are also degraded by the ubiquitin proteasome system with fewer details available.', 'We have recently extended these findings to demonstrate that the muscle-specific ubiquitin ligase MuRF1 specifically inhibits PPARalpha, but not PPARbeta/delta or PPARgamma1 activity in vivo. In contrast to previous studies on nuclear receptors, we identified that MuRF1 had an inhibitory effect on PPARalpha in cardiomyocytes by enhancing nuclear export, not degradation. By multi-mono-ubiquitination of lysines adjacent to the newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo.', 'Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 was found to inhibit PPARgamma1, while MuRF3 inhibits predominantly PPARbeta activity in vivo through mono-ubiquitination of substrates.', \"Increasing MuRF1 decreases creatine kinase activity significantly in the intact heart, resulting in a reduction of this critical transporter of ATP from the mitochondria to the sarcomere, resulting in heart failure. MuRF1 also inhibits PPARalpha, interacts with a host of energy-related proteins, and alters the cardiac metabolome in vivo. Building upon this larger theme, this study demonstrates MuRF1's anti-hypertrophic role in T3-mediated cardiomyocyte growth, with broader implications in T3-regulated metabolic rates critical to cardiac function in health and disease.\", \"The link between MuRF1 and the centrosome-associated protein 350 (CAP350) stems from the observation that MuRF1 alters nuclear TRalpha distribution in a similar manner as CAP350 does with PPARalpha. Although the role of CAP350 in regulating nuclear hormone receptors (such as TRalpha) is established, the mechanistic details have yet to be elucidated. In previous studies, the recently defined centrosome- and cytoskeleton-associated protein CAP350 was found to inhibit PPARalpha activity by redirecting PPARalpha's nuclear distribution to the perinuclear area with punctate distribution.\", 'In this study, we similarly observed that MuRF1 altered the nuclear distribution of TR (Fig.', 'Stabilization of peroxisome proliferator-activated receptor alpha by the ligand', 'Activity and subcellular compartmentalization of peroxisome proliferator-activated receptor alpha are altered by the centrosome-associated protein CAP350', 'The ubiquitin ligase MuRF1 regulates PPARalpha activity in the heart by enhancing nuclear export via monoubiquitination']\n",
            "  Document ID     : PMC5453669\n",
            "There are 13 occurrence(s) (each occurrence is at most 3 sentences long) for gene 19013 : PPARalpha.\n",
            "\n",
            "Gene(19015)\n",
            "  Symbol          : Ppard\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : peroxisome proliferator activator receptor delta\n",
            "  Also Known As   : NUC-1, NUC1, Nr1c2, PPAR-beta, PPAR-delta, PPAR[b], PPARdelta, Pparb, Pparb/d\n",
            "  In-text Name    : PPARbeta/delta\n",
            "  Occurrences     : ['For example, estrogen, progesterone, androgen, and glucocorticoid nuclear receptors are degraded in a proteasome-dependent manner in some conditions, with their heterodimeric binding partners RARgamma and RXR also degraded by parallel mechanisms. The muscle-specific ubiquitin ligase MuRF1 has recently been shown to specifically inhibit PPARalpha, but not PPARbeta/delta or PPARgamma1, activity in vivo. In contrast to the previous studies on nuclear receptors, MuRF1 was found to inhibit PPARalpha in the cardiomyocyte through mono-ubiquitination that enhanced nuclear export without promoting PPARalpha degradation.', 'By mono-ubiquitinating one to three lysines adjacent to a newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 inhibited PPARgamma1, whereas MuRF3 inhibited PPARbeta activity in vivo by mono-ubiquitination. In this study, we identify the cardiac ubiquitin ligase MuRF1 as an inhibitor of T3-induced physiological cardiac hypertrophy in vivo.', 'The closely related receptors that heterodimerize with PPARs, such as RARgamma and RXR, are also degraded by the ubiquitin proteasome system with fewer details available. We have recently extended these findings to demonstrate that the muscle-specific ubiquitin ligase MuRF1 specifically inhibits PPARalpha, but not PPARbeta/delta or PPARgamma1 activity in vivo. In contrast to previous studies on nuclear receptors, we identified that MuRF1 had an inhibitory effect on PPARalpha in cardiomyocytes by enhancing nuclear export, not degradation.', 'By multi-mono-ubiquitination of lysines adjacent to the newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 was found to inhibit PPARgamma1, while MuRF3 inhibits predominantly PPARbeta activity in vivo through mono-ubiquitination of substrates.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 4 occurrence(s) (each occurrence is at most 3 sentences long) for gene 19015 : PPARbeta/delta.\n",
            "\n",
            "Gene(381485)\n",
            "  Symbol          : Trim55\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : tripartite motif-containing 55\n",
            "  Also Known As   : D830041C10Rik, MuRF-2, Murf2, Rnf29\n",
            "  In-text Name    : MuRF2\n",
            "  Occurrences     : ['By mono-ubiquitinating one to three lysines adjacent to a newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 inhibited PPARgamma1, whereas MuRF3 inhibited PPARbeta activity in vivo by mono-ubiquitination. In this study, we identify the cardiac ubiquitin ligase MuRF1 as an inhibitor of T3-induced physiological cardiac hypertrophy in vivo.', 'By multi-mono-ubiquitination of lysines adjacent to the newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 was found to inhibit PPARgamma1, while MuRF3 inhibits predominantly PPARbeta activity in vivo through mono-ubiquitination of substrates.', 'Non-targeted metabolomics analysis of cardiac muscle ring finger-1 (MuRF1), MuRF2, and MuRF3 in vivo reveals novel and redundant metabolic changes', 'MuRF2 regulates PPARgamma1 activity to protect against diabetic cardiomyopathy and enhance weight gain induced by a high fat diet', 'MURF-1 and MURF-2 target a specific subset of myofibrillar proteins redundantly: Towards understanding MURF-dependent muscle ubiquitination']\n",
            "  Document ID     : PMC5453669\n",
            "There are 5 occurrence(s) (each occurrence is at most 3 sentences long) for gene 381485 : MuRF2.\n",
            "\n",
            "Gene(58522)\n",
            "  Symbol          : Trim54\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : tripartite motif-containing 54\n",
            "  Also Known As   : 4930486E09Rik, 4930566I02Rik, MURF, MURF-3, MuRF3, Rnf30\n",
            "  In-text Name    : MuRF3\n",
            "  Occurrences     : ['By mono-ubiquitinating one to three lysines adjacent to a newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 inhibited PPARgamma1, whereas MuRF3 inhibited PPARbeta activity in vivo by mono-ubiquitination. In this study, we identify the cardiac ubiquitin ligase MuRF1 as an inhibitor of T3-induced physiological cardiac hypertrophy in vivo.', 'By multi-mono-ubiquitination of lysines adjacent to the newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 was found to inhibit PPARgamma1, while MuRF3 inhibits predominantly PPARbeta activity in vivo through mono-ubiquitination of substrates.', 'Non-targeted metabolomics analysis of cardiac muscle ring finger-1 (MuRF1), MuRF2, and MuRF3 in vivo reveals novel and redundant metabolic changes', 'MURF-1 and MURF-2 target a specific subset of myofibrillar proteins redundantly: Towards understanding MURF-dependent muscle ubiquitination']\n",
            "  Document ID     : PMC5453669\n",
            "There are 4 occurrence(s) (each occurrence is at most 3 sentences long) for gene 58522 : MuRF3.\n",
            "\n",
            "Gene(84585)\n",
            "  Symbol          : Rnf123\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : ring finger protein 123\n",
            "  Also Known As   : Kpc1\n",
            "  In-text Name    : ubiquitin ligase\n",
            "  Occurrences     : ['For example, estrogen, progesterone, androgen, and glucocorticoid nuclear receptors are degraded in a proteasome-dependent manner in some conditions, with their heterodimeric binding partners RARgamma and RXR also degraded by parallel mechanisms. The muscle-specific ubiquitin ligase MuRF1 has recently been shown to specifically inhibit PPARalpha, but not PPARbeta/delta or PPARgamma1, activity in vivo. In contrast to the previous studies on nuclear receptors, MuRF1 was found to inhibit PPARalpha in the cardiomyocyte through mono-ubiquitination that enhanced nuclear export without promoting PPARalpha degradation.', 'By mono-ubiquitinating one to three lysines adjacent to a newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 inhibited PPARgamma1, whereas MuRF3 inhibited PPARbeta activity in vivo by mono-ubiquitination. In this study, we identify the cardiac ubiquitin ligase MuRF1 as an inhibitor of T3-induced physiological cardiac hypertrophy in vivo.', 'We determined that MuRF1 directly interacts with a specific thyroid receptor to mediate its posttranslational modification and inhibit its activity.', 'The closely related receptors that heterodimerize with PPARs, such as RARgamma and RXR, are also degraded by the ubiquitin proteasome system with fewer details available. We have recently extended these findings to demonstrate that the muscle-specific ubiquitin ligase MuRF1 specifically inhibits PPARalpha, but not PPARbeta/delta or PPARgamma1 activity in vivo. In contrast to previous studies on nuclear receptors, we identified that MuRF1 had an inhibitory effect on PPARalpha in cardiomyocytes by enhancing nuclear export, not degradation.', 'By multi-mono-ubiquitination of lysines adjacent to the newly identified nuclear export sequence in PPARalpha, MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo. Complementary to these studies, the closely related muscle-specific ubiquitin ligase MuRF2 was found to inhibit PPARgamma1, while MuRF3 inhibits predominantly PPARbeta activity in vivo through mono-ubiquitination of substrates.', 'Protein mono-ubiquitination that results in an enhanced affinity to other proteins is a common molecular process noted in the biology of cancer. An established example is the ubiquitin ligase MDM2, which mediates p53 mono-ubiquitination. In this system, p53 mono-ubiquitination alters its conformation to preferentially bind CRM1 (a subunit of the nuclear export apparatus), which then enhances p53 nuclear export resulting in activity inhibition.', 'This is an emerging, common theme, further illustrated with the estrogen receptor (ER), when the BRCA1 ubiquitin ligase mono-ubiquitinates it to inhibit its activity and the suppression of DNA polymerase Eta, when mono-ubiquitinated by the Pirh2 ubiquitin ligase. The present studies demonstrate a novel role for CAP350 activity in the heart and an emerging role in regulating TRalpha activity that has not been previously reported.', 'Pirh2 E3 ubiquitin ligase monoubiquitinates DNA polymerase eta to suppress translesion DNA synthesis', 'Muscle-specific RING finger 1 is a bona fide ubiquitin ligase that degrades cardiac troponin I', 'The ubiquitin ligase MuRF1 protects against cardiac ischemia/reperfusion injury by its proteasome-dependent degradation of phospho-c-Jun', 'Atrogin-1/muscle atrophy F-box inhibits calcineurin-dependent cardiac hypertrophy by participating in an SCF ubiquitin ligase complex', 'The ubiquitin ligase MuRF1 regulates PPARalpha activity in the heart by enhancing nuclear export via monoubiquitination']\n",
            "  Document ID     : PMC5453669\n",
            "There are 12 occurrence(s) (each occurrence is at most 3 sentences long) for gene 84585 : ubiquitin ligase.\n",
            "\n",
            "Gene(140939)\n",
            "  Symbol          : Trim63\n",
            "  Organism        : Rattus norvegicus\n",
            "  Full Name       : tripartite motif containing 63\n",
            "  Also Known As   : Murf, Murf1, Rnf28\n",
            "  In-text Name    : MuRF1\n",
            "  Occurrences     : ['To identify the role of MuRF1 in regulating TH-induced physiological hypertrophy, HL-1 cardiomyocyte-derived cells and neonatal rat cardiomyocytes (NRVMs) were stimulated with tri-iodothyronine (T3). As expected, T3 induced cardiomyocyte hypertrophy, reflected by increased CSA (Fig.', 'S1A, B and C, uncropped). T3 stimulation of cardiomyocytes with MuRF1-knockdown (using AdshMuRF1,  50  decrease in protein, Fig. 2C) resulted in a significant increase in CSA compared with T3 alone (Fig.', '1A). Conversely, increasing MuRF1 expression (AdMuRF1) significantly inhibited T3-induced increase in CSA in both HL-1 cells and NRVM (B and C). Together, these studies indicated a role of endogenous MuRF1 in attenuating T3-induced cardiac hypertrophy.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 3 occurrence(s) (each occurrence is at most 3 sentences long) for gene 140939 : MuRF1.\n",
            "\n",
            "Gene(71991)\n",
            "  Symbol          : Ercc8\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : excision repaiross-complementing rodent repair deficiency, complementation group 8\n",
            "  Also Known As   : 2410022P04Rik, 2810431L23Rik, 4631412O06Rik, B130065P18Rik, Ckn1, Csa\n",
            "  In-text Name    : CSA\n",
            "  Occurrences     : ['To identify the role of MuRF1 in regulating TH-induced physiological hypertrophy, HL-1 cardiomyocyte-derived cells and neonatal rat cardiomyocytes (NRVMs) were stimulated with tri-iodothyronine (T3). As expected, T3 induced cardiomyocyte hypertrophy, reflected by increased CSA (Fig. 1A, B and C, Supplementary Fig.', 'T3 stimulation of cardiomyocytes with MuRF1-knockdown (using AdshMuRF1,  50  decrease in protein, Fig. 2C) resulted in a significant increase in CSA compared with T3 alone (Fig. 1A).', 'Conversely, increasing MuRF1 expression (AdMuRF1) significantly inhibited T3-induced increase in CSA in both HL-1 cells and NRVM (B and C). Together, these studies indicated a role of endogenous MuRF1 in attenuating T3-induced cardiac hypertrophy.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 3 occurrence(s) (each occurrence is at most 3 sentences long) for gene 71991 : CSA.\n",
            "\n",
            "Gene(25085)\n",
            "  Symbol          : Th\n",
            "  Organism        : Rattus norvegicus\n",
            "  Full Name       : tyrosine hydroxylase\n",
            "  Also Known As   : The\n",
            "  In-text Name    : TH\n",
            "  Occurrences     : ['To identify the role of MuRF1 in regulating TH-induced physiological hypertrophy, HL-1 cardiomyocyte-derived cells and neonatal rat cardiomyocytes (NRVMs) were stimulated with tri-iodothyronine (T3). As expected, T3 induced cardiomyocyte hypertrophy, reflected by increased CSA (Fig.', 'Thyroid hormone replacement therapy attenuates atrial remodeling and reduces atrial fibrillation inducibility in a rat myocardial infarction-heart failure model']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 25085 : TH.\n",
            "\n",
            "Gene(104218)\n",
            "  Symbol          : Dtc1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : detected by T cells 1\n",
            "  Also Known As   : Dtc-1, tra-1\n",
            "  In-text Name    : TRalpha1\n",
            "  Occurrences     : [\"We next aim to further define how MuRF1 inhibits T3-induced cardiomyocyte hypertrophy mechanistically. We focused our attention on MuRF1's regulation of thyroid receptor (TR) TRalpha1 activity, as it is the only isoform in cardiomyocytes that stimulates growth (the less common TRalpha2, and TRbeta1 are present but lack this activity) (,). In these studies, we identified that increasing MuRF1 expression significantly inhibited TRalpha-induced TRE-luciferase activity (Fig.\", \"Using confocal microscopy, we found that increasing MuRF1 removed nuclear PPARalpha in a majority of the cells that subsequent studies found was due to MuRF1-targeted increased in nuclear export. Therefore, we next determined whether MuRF1 inhibits T3-induced TRalpha1 activity by altering TRalpha1's nuclear localization. HL-1 cardiomyocytes were transfected with TRalpha1 and/or MuRF1 ( /-T3) and analyzed by confocal microscopy.\", 'Interestingly, T3 altered the nuclear localization of TRalpha1 (Fig. 3A, column 4) compared with controls (Fig. 3A, column 3).', 'Specifically, it appeared that MuRF1 altered TRalpha1 nuclear localization by moving it to the nuclear periphery, a region closely aligned with the endoplasmic reticulum and microtubule anchoring complexes. To confirm this localization, we first identified the co-localization of the nuclear proteins lamin beta1 and desmin with TRalpha1 using confocal immunofluorescent analysis (Fig. 4A).', 'Increasing MuRF1 expression significantly increased the co-localization of TRalpha1 with nuclear lamin beta1 (Fig. 4), while similarly increasing the co-localization of TRalpha1 with desmin (Supplementary Fig. S2).', \"Lastly, we assessed MuRF1's effects on TRalpha1 nuclear localization by isolating the cardiomyocyte nuclear fraction by differential centrifugation and measuring the amount of TRalpha1 by immunoblot (Fig. 4B). Together, these studies demonstrate that increasing MuRF1 sequestered TRalpha1 in the nucleus, to inhibit TRalpha1 nuclear activity.\", 'Acute T3 treatment protects the heart against ischemia-reperfusion injury via TRalpha1 receptor']\n",
            "  Document ID     : PMC5453669\n",
            "There are 7 occurrence(s) (each occurrence is at most 3 sentences long) for gene 104218 : TRalpha1.\n",
            "\n",
            "Gene(21833)\n",
            "  Symbol          : Thra\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : thyroid hormone receptor alpha\n",
            "  Also Known As   : 6430529J03Rik, Erba, Nr1a1, Rvr, T3R[a], T3Ralpha, Thra1, Thra2, c-erbA-1, c-erbA-alpha, c-erbAalpha\n",
            "  In-text Name    : TRalpha2\n",
            "  Occurrences     : [\"We next aim to further define how MuRF1 inhibits T3-induced cardiomyocyte hypertrophy mechanistically. We focused our attention on MuRF1's regulation of thyroid receptor (TR) TRalpha1 activity, as it is the only isoform in cardiomyocytes that stimulates growth (the less common TRalpha2, and TRbeta1 are present but lack this activity) (,). In these studies, we identified that increasing MuRF1 expression significantly inhibited TRalpha-induced TRE-luciferase activity (Fig.\", 'Understanding how the thyroid hormone receptor alpha is regulated is critical to our understanding of heart disease. Inhibiting TRalpha in myocardial infarction experimentally impairs post-ischemic performance due to activation of p38 and an increased phospholamban.', 'Thyroid hormone receptor alpha can control action potential duration in mouse ventricular myocytes through the KCNE1 ion channel subunit']\n",
            "  Document ID     : PMC5453669\n",
            "There are 3 occurrence(s) (each occurrence is at most 3 sentences long) for gene 21833 : TRalpha2.\n",
            "\n",
            "Gene(211770)\n",
            "  Symbol          : Trib1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : tribbles pseudokinase 1\n",
            "  Also Known As   : A530090O15Rik, TRB-1, Trb1\n",
            "  In-text Name    : TRbeta1\n",
            "  Occurrences     : [\"We next aim to further define how MuRF1 inhibits T3-induced cardiomyocyte hypertrophy mechanistically. We focused our attention on MuRF1's regulation of thyroid receptor (TR) TRalpha1 activity, as it is the only isoform in cardiomyocytes that stimulates growth (the less common TRalpha2, and TRbeta1 are present but lack this activity) (,). In these studies, we identified that increasing MuRF1 expression significantly inhibited TRalpha-induced TRE-luciferase activity (Fig.\"]\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 211770 : TRbeta1.\n",
            "\n",
            "Gene(53313)\n",
            "  Symbol          : Atp2a3\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : ATPase, Ca++ transporting, ubiquitous\n",
            "  Also Known As   : SERCA3b, Serca3\n",
            "  In-text Name    : sarco/endoplasmic reticulum Ca2 -ATPase\n",
            "  Occurrences     : ['2A). As another way to measure TRalpha activity, we next evaluated how increasing MuRF1 would affect TR target genes and identified that MuRF1 significantly inhibited sarco/endoplasmic reticulum Ca2 -ATPase (SERCA expression (Fig. 2B)).', '8B). Interestingly, MuRF1-/- had an enhanced systolic function, reflecting increases in contractility and reflective of the T3-induced enhancements driven by increases in SERCA and phospholamban. Overall, the MuRF1-/- mice had a 53.6  increase in cardiomyocyte area compared with the 30.5  increase found in the wild-type controls (Fig.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 53313 : sarco/endoplasmic reticulum Ca2 -ATPase.\n",
            "\n",
            "Gene(13346)\n",
            "  Symbol          : Des\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : desmin\n",
            "  Also Known As   : desmin\n",
            "  In-text Name    : desmin\n",
            "  Occurrences     : ['Specifically, it appeared that MuRF1 altered TRalpha1 nuclear localization by moving it to the nuclear periphery, a region closely aligned with the endoplasmic reticulum and microtubule anchoring complexes. To confirm this localization, we first identified the co-localization of the nuclear proteins lamin beta1 and desmin with TRalpha1 using confocal immunofluorescent analysis (Fig. 4A).', 'Increasing MuRF1 expression significantly increased the co-localization of TRalpha1 with nuclear lamin beta1 (Fig. 4), while similarly increasing the co-localization of TRalpha1 with desmin (Supplementary Fig. S2).']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 13346 : desmin.\n",
            "\n",
            "Gene(17869)\n",
            "  Symbol          : Myc\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : myelocytomatosis oncogene\n",
            "  Also Known As   : Myc2, Niard, Nird, bHLHe39\n",
            "  In-text Name    : myc\n",
            "  Occurrences     : ['To assess whether TRalpha is a MuRF1 substrate, we next determined whether MuRF1 interacted directly with TRalpha by transfecting cells with MuRF1 and immunoprecipitating the myc-labeled MuRF1 (Fig. 5A).']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 17869 : myc.\n",
            "\n",
            "Gene(56321)\n",
            "  Symbol          : Aatf\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : apoptosis antagonizing transcription factor\n",
            "  Also Known As   : 4933415H02Rik, 5830465M17Rik, Che-1, Trb\n",
            "  In-text Name    : TRbeta\n",
            "  Occurrences     : ['As described in cancer cells, poly-ubiquitination is a common way in which thyroid hormone receptors have been described natively (Fig. 5E, as described in TRbeta). As immunoprecipitated TRalpha was polyubiquitinated in control cells (Figure 5D, red box), we hypothesized that the poly-ubiquitination competed for the sample lysines that MuRF1 mono-ubiquitinated.', 'SUMOylated TRalpha has been reported in both liver (HepG2) and pituitary (GH3) cell lines (Liu et al.2012), mediated by the SUMO ligase PIAS1. In the absence of ligand, PIAS1 SUMOylates TRalpha, whereas the SUMO ligase PIASx SUMOylates TRbeta only in the presence of ligand. Both SUMO1 and SUMO3 (two different types of SUMO modification) have been reported in T3-dependent gene regulation, closely matching TR activity and cofactor recruitment to the TRE in ChIP assays.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 56321 : TRbeta.\n",
            "\n",
            "Gene(74081)\n",
            "  Symbol          : Cep350\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : centrosomal protein 350\n",
            "  Also Known As   : 4933409L06Rik, 6430546F08Rik\n",
            "  In-text Name    : CAP350\n",
            "  Occurrences     : [\"The centrosome-associated protein (CAP) CAP350 has recently been identified as protein involved in regulating nuclear receptor activity, including PPAR. Initially, we analyzed MuRF1's effects on the co-localization of endogenous CAP350 and TRalpha in cardiomyocytes by confocal imaging (Fig. 6A).\", 'We found that increasing MuRF1 significantly increased nuclear CAP350 and the co-localization of CAP350 and TRalpha (Fig. 6A, right two columns). The increased co-localization of CAP350 and TRalpha driven by increasing MuRF1 was most apparent in the perinuclear region following T3 stimulation (Fig.', \"6A, in yellow). To solidify CAP350's mechanistic relationship to MuRF1's inhibition of TRalpha, we next knocked down CAP350 (to  lt; 50 , Fig. 6B) and determined its effect on cardiomyocyte cross-sectional area (Fig.\", '6C). In control cells, CAP350 knock-down attenuated the T3-induced cardiomyocyte hypertrophy (experimental group 3 vs group 2, Fig. 6C).', 'The MuRF1 attenuation of T3-stimulated cardiomyocyte growth (experimental group 6 vs group 3, Fig. 6C) was significantly overcome (i.e increased in cross-sectional area size) with CAP350 knock-down (experimental group 5, Fig. 6C).', \"Together, these studies demonstrate a role of CAP350 in MuRF1's regulation of TR in cardiomyocytes.\", '7A, far right). Interestingly, when the E/F TR domain mutant was expressed, CAP350 localization was affected, moving from a perinuclear to a pan-nuclear distribution (Fig. 7B, far right).', 'Together, these studies illustrate that MuRF1 inhibits TRalpha activity by mono-ubiquitinating it in the E/F region, resulting in an interaction with CAP350 and nuclear sequestration in cardiomyocytes.', \"The link between MuRF1 and the centrosome-associated protein 350 (CAP350) stems from the observation that MuRF1 alters nuclear TRalpha distribution in a similar manner as CAP350 does with PPARalpha. Although the role of CAP350 in regulating nuclear hormone receptors (such as TRalpha) is established, the mechanistic details have yet to be elucidated. In previous studies, the recently defined centrosome- and cytoskeleton-associated protein CAP350 was found to inhibit PPARalpha activity by redirecting PPARalpha's nuclear distribution to the perinuclear area with punctate distribution.\", \"In this study, we similarly observed that MuRF1 altered the nuclear distribution of TR (Fig. 3A), which led to subsequent studies to delineate the underlying mechanistic relationship between MuRF1's anti-hypertrophic activity with T3 stimulation and MuRF1's reliance upon CAP350 for preventing hypertrophic growth (Fig. 6C).\", \"The present studies illustrate that MuRF1's attenuation of T3-induced cardiomyocyte growth is dependent on TRalpha mono-ubiquitination of the ligand-binding domain (Fig. 7) and interaction with CAP350 (with significantly smaller cardiomyocytes with MuRF1 and CAP350 present, Fig. 6C, group 6).\", 'Our working hypothesis on the role of MuRF1 in this process is that MuRF1 mono-ubiquitinates TRalpha to enhance its affinity for CAP350, which sequesters TRalpha-Ub in nuclear compartments making it inaccessible to DNA promoters to diminish transcriptional activity.', 'This is an emerging, common theme, further illustrated with the estrogen receptor (ER), when the BRCA1 ubiquitin ligase mono-ubiquitinates it to inhibit its activity and the suppression of DNA polymerase Eta, when mono-ubiquitinated by the Pirh2 ubiquitin ligase. The present studies demonstrate a novel role for CAP350 activity in the heart and an emerging role in regulating TRalpha activity that has not been previously reported.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 13 occurrence(s) (each occurrence is at most 3 sentences long) for gene 74081 : CAP350.\n",
            "\n",
            "Gene(19185)\n",
            "  Symbol          : Psmd4\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : proteasome (prosome, macropain) 26S subunit, non-ATPase, 4\n",
            "  Also Known As   : Af1, Mcb1, angiocidin\n",
            "  In-text Name    : AF1\n",
            "  Occurrences     : ['Four full-length TRalpha constructs were created with lysine residues mutated to arginine (KR) (Supplementary Fig. S3): TRalpha A/B (AF1, N-terminus), TRalpha C (DNA-binding), TRalpha D (hinge), and TRalpha E/F (ligand-binding). These TRalpha constructs were then expressed in cardiomyocytes and assessed for their response to MuRF1 alterations in localization by confocal microscopy ().']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 19185 : AF1.\n",
            "\n",
            "Gene(18158)\n",
            "  Symbol          : Nppb\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : natriuretic peptide type B\n",
            "  Also Known As   : BNF, BNP, Iso-ANP\n",
            "  In-text Name    : BNP\n",
            "  Occurrences     : ['Significant reductions in systolic function over time were observed (Fig. 8C), consistent with an increase in pathological hypertrophy, identified by increases in BNP, alpha1-skeletal muscle actin, and ANF (data not shown). The observed effects on TH-induced physiological hypertrophy in our mouse models were most likely attributable to MuRF1 activity as circulating concentrations of T3/T4 were not altered in MuRF1-/- or MuRF1Tg  mice (Supplementary Fig.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 18158 : BNP.\n",
            "\n",
            "Gene(230899)\n",
            "  Symbol          : Nppa\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : natriuretic peptide type A\n",
            "  Also Known As   : ANP, Anf, CDD, Pnd\n",
            "  In-text Name    : ANF\n",
            "  Occurrences     : ['Significant reductions in systolic function over time were observed (Fig. 8C), consistent with an increase in pathological hypertrophy, identified by increases in BNP, alpha1-skeletal muscle actin, and ANF (data not shown). The observed effects on TH-induced physiological hypertrophy in our mouse models were most likely attributable to MuRF1 activity as circulating concentrations of T3/T4 were not altered in MuRF1-/- or MuRF1Tg  mice (Supplementary Fig.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 230899 : ANF.\n",
            "\n",
            "Gene(66998)\n",
            "  Symbol          : Psmd5\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : proteasome (prosome, macropain) 26S subunit, non-ATPase, 5\n",
            "  Also Known As   : 1500032A03Rik, S5b, mKIAA0072\n",
            "  In-text Name    : S5A and B\n",
            "  Occurrences     : ['The observed effects on TH-induced physiological hypertrophy in our mouse models were most likely attributable to MuRF1 activity as circulating concentrations of T3/T4 were not altered in MuRF1-/- or MuRF1Tg  mice (Supplementary Fig. S5A and B). Furthermore, MCT8 and -10 expressions (transporters involved in the uptake and efflux of T3) were not significantly different across the genotypic groups (Supplementary Fig.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 66998 : S5A and B.\n",
            "\n",
            "Gene(20502)\n",
            "  Symbol          : Slc16a2\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : solute carrier family 16 (monocarboxylic acid transporters), member 2\n",
            "  Also Known As   : Mct8, Xpct\n",
            "  In-text Name    : MCT8\n",
            "  Occurrences     : ['S5A and B). Furthermore, MCT8 and -10 expressions (transporters involved in the uptake and efflux of T3) were not significantly different across the genotypic groups (Supplementary Fig. S5C and D).']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 20502 : MCT8.\n",
            "\n",
            "Gene(26416)\n",
            "  Symbol          : Mapk14\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : mitogen-activated protein kinase 14\n",
            "  Also Known As   : CSBP2, Crk1, Csbp1, Mxi2, PRKM14, PRKM15, p38, p38-alpha, p38MAPK, p38a, p38alpha\n",
            "  In-text Name    : p38\n",
            "  Occurrences     : ['Understanding how the thyroid hormone receptor alpha is regulated is critical to our understanding of heart disease. Inhibiting TRalpha in myocardial infarction experimentally impairs post-ischemic performance due to activation of p38 and an increased phospholamban. Conversely, T3 therapy protects the heart against ischemia:reperfusion injury in a TRalpha-dependent manner.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 26416 : p38.\n",
            "\n",
            "Gene(11835)\n",
            "  Symbol          : Ar\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : androgen receptor\n",
            "  Also Known As   : Tfm\n",
            "  In-text Name    : androgen receptor\n",
            "  Occurrences     : ['Some nuclear hormone receptors are degraded in both the presence and absence of their cognate ligands, such as the estrogen, progesterone, and glucocorticoid receptors ( ). Alternatively, the androgen receptor and the PPARalpha are targeted to the proteasome in the absence of ligand. The closely related receptors that heterodimerize with PPARs, such as RARgamma and RXR, are also degraded by the ubiquitin proteasome system with fewer details available.', 'Proteasome activity is required for androgen receptor transcriptional activity via regulation of androgen receptor nuclear translocation and interaction with coregulators in prostate cancer cells']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 11835 : androgen receptor.\n",
            "\n",
            "Gene(16476)\n",
            "  Symbol          : Jun\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : jun proto-oncogene\n",
            "  Also Known As   : AP-1, Junc, c-jun\n",
            "  In-text Name    : cJun\n",
            "  Occurrences     : ['In pressure overload-induced cardiac hypertrophy, MuRF1 inhibits pathological hypertrophy by inhibiting the serum response transcription factor. In physiological hypertrophy induced by exercise (IGF-1), MuRF1 transcriptionally inhibits cardiomyocyte growth by inhibiting cJun-mediated increases in IGF-1 signaling proteins, by preferentially recognizing phospho-cJun, poly-ubiquitinating, and targeting it for proteasome-dependent degradation in vivo. Metabolically, MuRF1 has been found to ubiquitinate oxidized creatine kinase, mediating its turnover in vivo.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 16476 : cJun.\n",
            "\n",
            "Gene(17246)\n",
            "  Symbol          : Mdm2\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : transformed mouse 3T3 cell double minute 2\n",
            "  Also Known As   : 1700007J15Rik, Mdm-2\n",
            "  In-text Name    : MDM2\n",
            "  Occurrences     : ['Protein mono-ubiquitination that results in an enhanced affinity to other proteins is a common molecular process noted in the biology of cancer. An established example is the ubiquitin ligase MDM2, which mediates p53 mono-ubiquitination. In this system, p53 mono-ubiquitination alters its conformation to preferentially bind CRM1 (a subunit of the nuclear export apparatus), which then enhances p53 nuclear export resulting in activity inhibition.', 'Mechanistic studies of MDM2-mediated ubiquitination in p53 regulation', 'Human mdm2 mediates multiple mono-ubiquitination of p53 by a mechanism requiring enzyme isomerization']\n",
            "  Document ID     : PMC5453669\n",
            "There are 3 occurrence(s) (each occurrence is at most 3 sentences long) for gene 17246 : MDM2.\n",
            "\n",
            "Gene(22060)\n",
            "  Symbol          : Trp53-ps\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : transformation related protein 53, pseudogene\n",
            "  Also Known As   : p53\n",
            "  In-text Name    : p53\n",
            "  Occurrences     : ['Protein mono-ubiquitination that results in an enhanced affinity to other proteins is a common molecular process noted in the biology of cancer. An established example is the ubiquitin ligase MDM2, which mediates p53 mono-ubiquitination. In this system, p53 mono-ubiquitination alters its conformation to preferentially bind CRM1 (a subunit of the nuclear export apparatus), which then enhances p53 nuclear export resulting in activity inhibition.', 'This is an emerging, common theme, further illustrated with the estrogen receptor (ER), when the BRCA1 ubiquitin ligase mono-ubiquitinates it to inhibit its activity and the suppression of DNA polymerase Eta, when mono-ubiquitinated by the Pirh2 ubiquitin ligase.', 'Mechanistic studies of MDM2-mediated ubiquitination in p53 regulation', 'Human mdm2 mediates multiple mono-ubiquitination of p53 by a mechanism requiring enzyme isomerization', 'The multiple levels of regulation by p53 ubiquitination']\n",
            "  Document ID     : PMC5453669\n",
            "There are 5 occurrence(s) (each occurrence is at most 3 sentences long) for gene 22060 : p53.\n",
            "\n",
            "Gene(103573)\n",
            "  Symbol          : Xpo1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : exportin 1\n",
            "  Also Known As   : Crm1, Exp1\n",
            "  In-text Name    : CRM1\n",
            "  Occurrences     : ['An established example is the ubiquitin ligase MDM2, which mediates p53 mono-ubiquitination. In this system, p53 mono-ubiquitination alters its conformation to preferentially bind CRM1 (a subunit of the nuclear export apparatus), which then enhances p53 nuclear export resulting in activity inhibition. This is an emerging, common theme, further illustrated with the estrogen receptor (ER), when the BRCA1 ubiquitin ligase mono-ubiquitinates it to inhibit its activity and the suppression of DNA polymerase Eta, when mono-ubiquitinated by the Pirh2 ubiquitin ligase.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 103573 : CRM1.\n",
            "\n",
            "Gene(13982)\n",
            "  Symbol          : Esr1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : estrogen receptor 1 (alpha)\n",
            "  Also Known As   : ER, ER-alpha, ERa, ERalpha, ESR, Estr, Estra, Nr3a1\n",
            "  In-text Name    : estrogen receptor\n",
            "  Occurrences     : ['In this system, p53 mono-ubiquitination alters its conformation to preferentially bind CRM1 (a subunit of the nuclear export apparatus), which then enhances p53 nuclear export resulting in activity inhibition. This is an emerging, common theme, further illustrated with the estrogen receptor (ER), when the BRCA1 ubiquitin ligase mono-ubiquitinates it to inhibit its activity and the suppression of DNA polymerase Eta, when mono-ubiquitinated by the Pirh2 ubiquitin ligase. The present studies demonstrate a novel role for CAP350 activity in the heart and an emerging role in regulating TRalpha activity that has not been previously reported.', 'BRCA1 regulates acetylation and ubiquitination of estrogen receptor-alpha']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 13982 : estrogen receptor.\n",
            "\n",
            "Gene(12189)\n",
            "  Symbol          : Brca1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : breast cancer 1, early onset\n",
            "  Also Known As   : BRCA1\n",
            "  In-text Name    : BRCA1\n",
            "  Occurrences     : ['In this system, p53 mono-ubiquitination alters its conformation to preferentially bind CRM1 (a subunit of the nuclear export apparatus), which then enhances p53 nuclear export resulting in activity inhibition. This is an emerging, common theme, further illustrated with the estrogen receptor (ER), when the BRCA1 ubiquitin ligase mono-ubiquitinates it to inhibit its activity and the suppression of DNA polymerase Eta, when mono-ubiquitinated by the Pirh2 ubiquitin ligase. The present studies demonstrate a novel role for CAP350 activity in the heart and an emerging role in regulating TRalpha activity that has not been previously reported.', 'BRCA1 regulates acetylation and ubiquitination of estrogen receptor-alpha']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 12189 : BRCA1.\n",
            "\n",
            "Gene(80905)\n",
            "  Symbol          : Polh\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : polymerase (DNA directed), eta (RAD 30 related)\n",
            "  Also Known As   : RAD30A, XPV\n",
            "  In-text Name    : DNA polymerase Eta\n",
            "  Occurrences     : ['In this system, p53 mono-ubiquitination alters its conformation to preferentially bind CRM1 (a subunit of the nuclear export apparatus), which then enhances p53 nuclear export resulting in activity inhibition. This is an emerging, common theme, further illustrated with the estrogen receptor (ER), when the BRCA1 ubiquitin ligase mono-ubiquitinates it to inhibit its activity and the suppression of DNA polymerase Eta, when mono-ubiquitinated by the Pirh2 ubiquitin ligase. The present studies demonstrate a novel role for CAP350 activity in the heart and an emerging role in regulating TRalpha activity that has not been previously reported.', 'Pirh2 E3 ubiquitin ligase monoubiquitinates DNA polymerase eta to suppress translesion DNA synthesis']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 80905 : DNA polymerase Eta.\n",
            "\n",
            "Gene(68098)\n",
            "  Symbol          : Rchy1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : ring finger and CHY zinc finger domain containing 1\n",
            "  Also Known As   : ARNIP, CHIMP, Pirh2, Zfp363, Znf363\n",
            "  In-text Name    : Pirh2\n",
            "  Occurrences     : ['In this system, p53 mono-ubiquitination alters its conformation to preferentially bind CRM1 (a subunit of the nuclear export apparatus), which then enhances p53 nuclear export resulting in activity inhibition. This is an emerging, common theme, further illustrated with the estrogen receptor (ER), when the BRCA1 ubiquitin ligase mono-ubiquitinates it to inhibit its activity and the suppression of DNA polymerase Eta, when mono-ubiquitinated by the Pirh2 ubiquitin ligase. The present studies demonstrate a novel role for CAP350 activity in the heart and an emerging role in regulating TRalpha activity that has not been previously reported.', 'Pirh2 E3 ubiquitin ligase monoubiquitinates DNA polymerase eta to suppress translesion DNA synthesis']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 68098 : Pirh2.\n",
            "\n",
            "Gene(6955)\n",
            "  Symbol          : TRA\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : T cell receptor alpha locus\n",
            "  Also Known As   : IMD7, TCRA, TRA@\n",
            "  In-text Name    : TRalpha\n",
            "  Occurrences     : ['Interestingly, SUMO modification alters cellular processes, including nuclear transport, transcriptional regulation, apoptosis, and protein stability without degrading its substrate. SUMOylated TRalpha has been reported in both liver (HepG2) and pituitary (GH3) cell lines (Liu et al.2012), mediated by the SUMO ligase PIAS1. In the absence of ligand, PIAS1 SUMOylates TRalpha, whereas the SUMO ligase PIASx SUMOylates TRbeta only in the presence of ligand.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 6955 : TRalpha.\n",
            "\n",
            "Gene(300772)\n",
            "  Symbol          : Pias1\n",
            "  Organism        : Rattus norvegicus\n",
            "  Full Name       : protein inhibitor of activated STAT, 1\n",
            "  Also Known As   : PIAS1\n",
            "  In-text Name    : PIAS1\n",
            "  Occurrences     : ['Interestingly, SUMO modification alters cellular processes, including nuclear transport, transcriptional regulation, apoptosis, and protein stability without degrading its substrate. SUMOylated TRalpha has been reported in both liver (HepG2) and pituitary (GH3) cell lines (Liu et al.2012), mediated by the SUMO ligase PIAS1. In the absence of ligand, PIAS1 SUMOylates TRalpha, whereas the SUMO ligase PIASx SUMOylates TRbeta only in the presence of ligand.', 'Both SUMO1 and SUMO3 (two different types of SUMO modification) have been reported in T3-dependent gene regulation, closely matching TR activity and cofactor recruitment to the TRE in ChIP assays.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 300772 : PIAS1.\n",
            "\n",
            "Gene(22218)\n",
            "  Symbol          : Sumo1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : small ubiquitin-like modifier 1\n",
            "  Also Known As   : GMP1, PIC1, SENTRIN, SMT3, SMT3H3, SMTP3, SUMO-1, Smt3C, Ubl1\n",
            "  In-text Name    : SUMO1\n",
            "  Occurrences     : ['In the absence of ligand, PIAS1 SUMOylates TRalpha, whereas the SUMO ligase PIASx SUMOylates TRbeta only in the presence of ligand. Both SUMO1 and SUMO3 (two different types of SUMO modification) have been reported in T3-dependent gene regulation, closely matching TR activity and cofactor recruitment to the TRE in ChIP assays.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 22218 : SUMO1.\n",
            "\n",
            "Gene(20610)\n",
            "  Symbol          : Sumo3\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : small ubiquitin-like modifier 3\n",
            "  Also Known As   : 2810014B19Rik, D10Ertd345e, SMT3A, SUMO-3, Smt3h1\n",
            "  In-text Name    : SUMO3\n",
            "  Occurrences     : ['In the absence of ligand, PIAS1 SUMOylates TRalpha, whereas the SUMO ligase PIASx SUMOylates TRbeta only in the presence of ligand. Both SUMO1 and SUMO3 (two different types of SUMO modification) have been reported in T3-dependent gene regulation, closely matching TR activity and cofactor recruitment to the TRE in ChIP assays.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 20610 : SUMO3.\n",
            "\n",
            "Gene(2149)\n",
            "  Symbol          : F2R\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : coagulation factor II thrombin receptor\n",
            "  Also Known As   : CF2R, HTR, PAR-1, PAR1, TR\n",
            "  In-text Name    : TR\n",
            "  Occurrences     : ['TH induces molecular changes that benefit the failing heart, including the stimulation of pro-survival pathways such as Akt and mTOR that is accompanied by improved cardiac function, paralleling changes induced by exercise, mediated by IGF-1, to drive increases in contractile proteins critical to cardiac function and performance. Exercise has also been shown to increase myocardial TR expression and T3 levels in a small cohort of HF patients with ventricular assist devices, underscoring the role for TH in promoting positive adaptive remodeling of the injured heart. Exogenous approaches to supplementing TH also hold promise as low-dose T3 replacement has been shown to restore cardiac T3 levels and ameliorate dysfunction in preclinical models of heart failure.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 2149 : TR.\n",
            "\n",
            "Gene(100035352)\n",
            "  Symbol          : Org10\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : organ weight 10\n",
            "  Also Known As   : ORG17.2\n",
            "  In-text Name    : org/10\n",
            "  Occurrences     : ['This is linked to the online version of the paper at http://dx.doi.org/10.1530/JME-15-0283.']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 100035352 : org/10.\n",
            "\n",
            "Gene(18754)\n",
            "  Symbol          : Prkce\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : protein kinase C, epsilon\n",
            "  Also Known As   : 5830406C15Rik, PKC[e], PKCepsilon, Pkce\n",
            "  In-text Name    : PKC{epsilon\n",
            "  Occurrences     : ['Muscle ring finger protein-1 inhibits PKC{epsilon} activation and prevents cardiomyocyte hypertrophy']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 18754 : PKC{epsilon.\n",
            "\n",
            "Gene(18821)\n",
            "  Symbol          : Pln\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : phospholamban\n",
            "  Also Known As   : Plb\n",
            "  In-text Name    : cardiac phospholamban\n",
            "  Occurrences     : ['Repression of cardiac phospholamban gene expression is mediated by thyroid hormone receptor-{alpha}1 and involves targeted covalent histone modifications']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 18821 : cardiac phospholamban.\n",
            "\n",
            "Gene(56543)\n",
            "  Symbol          : Kcnd3\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : potassium voltage-gated channel, Shal-related family, member 3\n",
            "  Also Known As   : Kncd3, Kv4.3\n",
            "  In-text Name    : KCND3\n",
            "  Occurrences     : ['Divergent regulation of cardiac KCND3 potassium channel expression by the thyroid hormone receptors alpha1 and beta1']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 56543 : KCND3.\n",
            "\n",
            "Gene(12044)\n",
            "  Symbol          : Bcl2a1a\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : B cell leukemia/lymphoma 2 related protein A1a\n",
            "  Also Known As   : A1, Bcl2a1, Bfl-1, Hbpa1\n",
            "  In-text Name    : alpha1 and beta\n",
            "  Occurrences     : ['Divergent regulation of cardiac KCND3 potassium channel expression by the thyroid hormone receptors alpha1 and beta1']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 12044 : alpha1 and beta.\n",
            "\n",
            "Gene(14182)\n",
            "  Symbol          : Fgfr1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : fibroblast growth factor receptor 1\n",
            "  Also Known As   : Eask, FGFR-I, FLG, Fgfr-1, Flt-2, Fr1, Hspy, MFR, bFGF-R-1, c-fgr\n",
            "  In-text Name    : FGFR1\n",
            "  Occurrences     : ['Myeloproliferative disorder FOP-FGFR1 fusion kinase recruits phosphoinositide-3 kinase and phospholipase Cgamma at the centrosome']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 14182 : FGFR1.\n",
            "\n",
            "Gene(67731)\n",
            "  Symbol          : Fbxo32\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : F-box protein 32\n",
            "  Also Known As   : 4833442G10Rik, ATROGIN1, Gm20361, MAFbx\n",
            "  In-text Name    : Atrogin-1\n",
            "  Occurrences     : ['Atrogin-1/muscle atrophy F-box inhibits calcineurin-dependent cardiac hypertrophy by participating in an SCF ubiquitin ligase complex']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 67731 : Atrogin-1.\n",
            "\n",
            "Gene(19365)\n",
            "  Symbol          : Rad52\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : RAD52 homolog, DNA repair protein\n",
            "  Also Known As   : Rad52yh\n",
            "  In-text Name    : Rad52\n",
            "  Occurrences     : ['Colocalization of multiple DNA double-strand breaks at a single Rad52 repair centre']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 19365 : Rad52.\n",
            "\n",
            "Gene(16509)\n",
            "  Symbol          : Kcne1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : potassium voltage-gated channel, Isk-related subfamily, member 1\n",
            "  Also Known As   : Isk, MinK, nmf190\n",
            "  In-text Name    : KCNE1\n",
            "  Occurrences     : ['Thyroid hormone receptor alpha can control action potential duration in mouse ventricular myocytes through the KCNE1 ion channel subunit']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 16509 : KCNE1.\n",
            "\n",
            "Gene(26413)\n",
            "  Symbol          : Mapk1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : mitogen-activated protein kinase 1\n",
            "  Also Known As   : 9030612K14Rik, ERK, Erk2, MAPK2, PRKM2, Prkm1, p41mapk, p42mapk\n",
            "  In-text Name    : ERK\n",
            "  Occurrences     : ['Dose-dependent effects of thyroid hormone on post-ischemic cardiac performance: Potential involvement of Akt and ERK signalings', 'Thyroid hormone changes cardiomyocyte shape and geometry via ERK signaling pathway: Potential therapeutic implications in reversing cardiac remodeling?']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 26413 : ERK.\n",
            "\n",
            "Gene(3479)\n",
            "  Symbol          : IGF1\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : insulin like growth factor 1\n",
            "  Also Known As   : IGF, IGF-I, IGFI, MGF\n",
            "  In-text Name    : IGF1\n",
            "  Occurrences     : ['Thyroid hormone receptor and IGF1/IGFR systems: Possible relations in the human heart']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 3479 : IGF1.\n",
            "\n",
            "Gene(3480)\n",
            "  Symbol          : IGF1R\n",
            "  Organism        : Homo sapiens\n",
            "  Full Name       : insulin like growth factor 1 receptor\n",
            "  Also Known As   : CD221, IGFIR, IGFR, JTK13\n",
            "  In-text Name    : IGFR\n",
            "  Occurrences     : ['Thyroid hormone receptor and IGF1/IGFR systems: Possible relations in the human heart']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 3480 : IGFR.\n",
            "\n",
            "Gene(18750)\n",
            "  Symbol          : Prkca\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : protein kinase C, alpha\n",
            "  Also Known As   : Pkca\n",
            "  In-text Name    : PKCalpha\n",
            "  Occurrences     : ['PKCalpha activation of p120-catenin serine 879 phospho-switch disassembles VE-cadherin junctions and disrupts vascular integrity']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 18750 : PKCalpha.\n",
            "\n",
            "Gene(12388)\n",
            "  Symbol          : Ctnnd1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : catenin delta 1\n",
            "  Also Known As   : Catns, Ctnnd, P120, mKIAA0384\n",
            "  In-text Name    : p120-catenin\n",
            "  Occurrences     : ['PKCalpha activation of p120-catenin serine 879 phospho-switch disassembles VE-cadherin junctions and disrupts vascular integrity']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 12388 : p120-catenin.\n",
            "\n",
            "Gene(12562)\n",
            "  Symbol          : Cdh5\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : cadherin 5\n",
            "  Also Known As   : 7B4, Cd144, VE-Cad, VECD, VEcad, Vec\n",
            "  In-text Name    : VE-cadherin\n",
            "  Occurrences     : ['PKCalpha activation of p120-catenin serine 879 phospho-switch disassembles VE-cadherin junctions and disrupts vascular integrity']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 12562 : VE-cadherin.\n",
            "\n",
            "Gene(66949)\n",
            "  Symbol          : Trim59\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : tripartite motif-containing 59\n",
            "  Also Known As   : 2310035M22Rik, 2700022F13Rik, Mrf1, TSBF1\n",
            "  In-text Name    : RING Finger-1\n",
            "  Occurrences     : ['Muscle RING Finger-1 Attenuates IGF-1-dependent cardiomyocyte hypertrophy by inhibiting JNK signaling', 'Muscle ring finger 1 mediates cardiac atrophy in vivo']\n",
            "  Document ID     : PMC5453669\n",
            "There are 2 occurrence(s) (each occurrence is at most 3 sentences long) for gene 66949 : RING Finger-1.\n",
            "\n",
            "Gene(26419)\n",
            "  Symbol          : Mapk8\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : mitogen-activated protein kinase 8\n",
            "  Also Known As   : JNK, JNK1, Prkm8, SAPK1\n",
            "  In-text Name    : JNK\n",
            "  Occurrences     : ['Muscle RING Finger-1 Attenuates IGF-1-dependent cardiomyocyte hypertrophy by inhibiting JNK signaling']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 26419 : JNK.\n",
            "\n",
            "Gene(13555)\n",
            "  Symbol          : E2f1\n",
            "  Organism        : Mus musculus\n",
            "  Full Name       : E2F transcription factor 1\n",
            "  Also Known As   : E2F-1, Tg(Wnt1-cre)2Sor, mKIAA4009\n",
            "  In-text Name    : E2F1\n",
            "  Occurrences     : ['Muscle ring finger 1 and muscle ring finger 2 are necessary but functionally redundant during developmental cardiac growth and regulate E2F1-mediated gene expression in vivo']\n",
            "  Document ID     : PMC5453669\n",
            "There are 1 occurrence(s) (each occurrence is at most 3 sentences long) for gene 13555 : E2F1.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!pip install biopython\n",
        "import xml.etree.ElementTree as ET\n",
        "from Bio import Entrez\n",
        "import re\n",
        "import time\n",
        "\n",
        "class Gene:\n",
        "    def __init__(self, gene_id):\n",
        "        self.gene_id = gene_id\n",
        "        self.occurrences = []  # list to store snippet(s) where the gene was mentioned\n",
        "        self.symbol = None\n",
        "        self.organism = None\n",
        "        self.full_name = None\n",
        "        self.also_known_as = None\n",
        "        self.name_from_article = None\n",
        "        self.doc_id = None\n",
        "\n",
        "    def add_occurrence(self, snippet):\n",
        "        if snippet not in self.occurrences:  # Avoid duplicates\n",
        "            self.occurrences.append(snippet)\n",
        "\n",
        "    def set_name_from_article(self, name_from_article):\n",
        "        \"\"\"Sets the temporary name of the gene. This is the name according to the article\"\"\"\n",
        "        self.name_from_article = name_from_article\n",
        "\n",
        "    def get_name_from_article(self):\n",
        "        return self.name_from_article\n",
        "\n",
        "    def get_occurrences(self):\n",
        "        return self.occurrences\n",
        "\n",
        "    def get_also_known_as(self):\n",
        "        return self.also_known_as\n",
        "\n",
        "    def get_gene_id(self):\n",
        "        return self.gene_id\n",
        "    \n",
        "    def get_organsim(self):\n",
        "        return self.organism\n",
        "    \n",
        "    def get_doc_id(self):\n",
        "        return self.doc_id\n",
        "    \n",
        "    def set_doc_id(self, doc_id):\n",
        "        self.doc_id = doc_id\n",
        "\n",
        "    def update_info(self, symbol, organism, full_name, also_known_as):\n",
        "        self.symbol = symbol\n",
        "        self.organism = organism\n",
        "        self.full_name = full_name\n",
        "        self.also_known_as = also_known_as\n",
        "\n",
        "    def __repr__(self):\n",
        "        return (f\"Gene({self.gene_id})\\n\"\n",
        "                f\"  Symbol          : {self.symbol}\\n\"\n",
        "                f\"  Organism        : {self.organism}\\n\"\n",
        "                f\"  Full Name       : {self.full_name}\\n\"\n",
        "                f\"  Also Known As   : {self.also_known_as}\\n\"\n",
        "                f\"  In-text Name    : {self.name_from_article}\\n\"\n",
        "                f\"  Occurrences     : {self.occurrences}\\n\"\n",
        "                f\"  Document ID     : {self.doc_id}\")\n",
        "\n",
        "def parse_xml_file(xml_path):\n",
        "    \"\"\"Parses the XML file and returns a gene dictionary keyed by gene ID.\"\"\"\n",
        "    tree = ET.parse(xml_path)\n",
        "    root = tree.getroot()\n",
        "    gene_dict = {}\n",
        "\n",
        "    for document in root.findall('document'):\n",
        "        \n",
        "        # get the document ID (pmid or pmcid)\n",
        "        doc_id_elem = document.find('id')\n",
        "        if doc_id_elem is not None:\n",
        "            doc_id = doc_id_elem.text\n",
        "        else:\n",
        "            doc_id = None\n",
        "\n",
        "\n",
        "        # get the gene annotations in the document from the passages\n",
        "        for passage in document.findall('passage'):       \n",
        "            # Also check for 'article-id_pmc'\n",
        "            article_id_pmc_elem = passage.find(\"infon[@key='article-id_pmc']\")\n",
        "            if article_id_pmc_elem is not None:\n",
        "                pmc_value = article_id_pmc_elem.text.strip()\n",
        "                if not pmc_value.startswith(\"PMC\"):\n",
        "                    pmc_value = \"PMC\" + pmc_value\n",
        "                doc_id = pmc_value     \n",
        "            section_type_elem = passage.find(\"infon[@key='section_type']\")\n",
        "            if section_type_elem is not None and section_type_elem.text and section_type_elem.text.upper() in (\"METHODS\", \"FIG\", \"TABLE\"):\n",
        "                continue  # Skip passages under METHODS\n",
        "            # Get the full passage text.\n",
        "            passage_text_elem = passage.find(\"text\")\n",
        "            passage_text = passage_text_elem.text if passage_text_elem is not None else \"\"\n",
        "            # Determine the starting offset for this passage.\n",
        "            passage_offset_elem = passage.find(\"offset\")\n",
        "            passage_offset = int(passage_offset_elem.text) if passage_offset_elem is not None else 0\n",
        "            # Split the passage into sentences using regex.\n",
        "            sentences = re.split(r'(?<=[.!?])\\s+', passage_text)\n",
        "            # Compute start indices for each sentence within the passage text.\n",
        "            start_indices = []\n",
        "            current_index = passage_offset\n",
        "            for sentence in sentences:\n",
        "                start_indices.append(current_index)\n",
        "                current_index += len(sentence) + 1  # account for the delimiter space\n",
        "\n",
        "            # Dictionary to store annotations per gene in this passage.\n",
        "            # For each gene_id, we collect the sentence indices and a temporary in-text name.\n",
        "            gene_annotations = {}\n",
        "            for annotation in passage.findall('annotation'):\n",
        "                ann_type = annotation.find(\"infon[@key='type']\")\n",
        "                if ann_type is not None and ann_type.text == \"Gene\":\n",
        "                    # Check for both 'identifier' and 'NCBI Gene' keys.\n",
        "                    gene_id_elem = annotation.find(\"infon[@key='identifier']\")\n",
        "                    if gene_id_elem is None:\n",
        "                        gene_id_elem = annotation.find(\"infon[@key='NCBI Gene']\")\n",
        "                    gene_id = gene_id_elem.text if gene_id_elem is not None else None\n",
        "\n",
        "                    # Handle multiple IDs by keeping only the first if needed.\n",
        "                    if gene_id and \";\" in gene_id:\n",
        "                        gene_id = gene_id.split(\";\")[0].strip()\n",
        "\n",
        "                    # Extract the gene name from the annotation text.\n",
        "                    in_text_gene_name_elem = annotation.find(\"text\")\n",
        "                    in_text_gene_name = in_text_gene_name_elem.text if in_text_gene_name_elem is not None else None\n",
        "\n",
        "                    # Extract the annotation location (offset) to find the sentence.\n",
        "                    location_elem = annotation.find(\"location\")\n",
        "                    ann_offset = int(location_elem.attrib.get('offset', 0)) if location_elem is not None else 0\n",
        "\n",
        "                    # Determine which sentence contains the annotation based on its offset.\n",
        "                    sentence_index = None\n",
        "                    for i, start in enumerate(start_indices):\n",
        "                        if start <= ann_offset < start + len(sentences[i]):\n",
        "                            sentence_index = i\n",
        "                            break\n",
        "\n",
        "                    if sentence_index is not None and gene_id:\n",
        "                        if gene_id not in gene_annotations:\n",
        "                            gene_annotations[gene_id] = {'indices': [], 'name': in_text_gene_name}\n",
        "                        gene_annotations[gene_id]['indices'].append(sentence_index)\n",
        "\n",
        "            # For each gene in this passage, process the annotation indices.\n",
        "            for gene_id, data in gene_annotations.items():\n",
        "                # Remove duplicate indices and sort.\n",
        "                indices = sorted(set(data['indices']))\n",
        "                candidate_windows = []\n",
        "                # For each annotation occurrence, form a candidate window:\n",
        "                # one sentence before and one sentence after the annotated sentence.\n",
        "                for idx in indices:\n",
        "                    start_sentence = max(0, idx - 1)\n",
        "                    end_sentence = min(len(sentences), idx + 2)  # end index is non-inclusive\n",
        "                    candidate_windows.append((start_sentence, end_sentence))\n",
        "                # Merge overlapping candidate windows.\n",
        "                merged_windows = []\n",
        "                for window in sorted(candidate_windows):\n",
        "                    if not merged_windows:\n",
        "                        merged_windows.append(window)\n",
        "                    else:\n",
        "                        last = merged_windows[-1]\n",
        "                        # If windows overlap or touch, merge them.\n",
        "                        if window[0] <= last[1]:\n",
        "                            merged_windows[-1] = (last[0], max(last[1], window[1]))\n",
        "                        else:\n",
        "                            merged_windows.append(window)\n",
        "                # For each merged window, if it spans more than 3 sentences, split it into chunks of 3 sentences maximum.\n",
        "                for window in merged_windows:\n",
        "                    start, end = window\n",
        "                    window_size = end - start\n",
        "                    if window_size <= 3:\n",
        "                        snippet = \" \".join(sentences[start:end])\n",
        "                        if gene_id in gene_dict:\n",
        "                            gene_dict[gene_id].add_occurrence(snippet)\n",
        "                        else:\n",
        "                            gene_obj = Gene(gene_id)\n",
        "                            gene_obj.add_occurrence(snippet)\n",
        "                            gene_obj.set_doc_id(doc_id)\n",
        "                            gene_obj.set_name_from_article(data['name'])\n",
        "                            gene_dict[gene_id] = gene_obj\n",
        "                    else:\n",
        "                        # Split the merged window into non-overlapping chunks of at most 3 sentences.\n",
        "                        for i in range(start, end, 3):\n",
        "                            chunk_end = min(i + 3, end)\n",
        "                            snippet = \" \".join(sentences[i:chunk_end])\n",
        "                            if snippet.strip():\n",
        "                                if gene_id in gene_dict:\n",
        "                                    gene_dict[gene_id].add_occurrence(snippet)\n",
        "                                else:\n",
        "                                    gene_obj = Gene(gene_id)\n",
        "                                    gene_obj.set_doc_id(doc_id)\n",
        "                                    gene_obj.add_occurrence(snippet)\n",
        "                                    gene_obj.set_name_from_article(data['name'])\n",
        "                                    gene_dict[gene_id] = gene_obj\n",
        "    return gene_dict\n",
        "\n",
        "def fetch_and_update_gene_info(gene_dict):\n",
        "    \"\"\"Retrieves gene information from NCBI and updates the genes in gene_dict.\"\"\"\n",
        "    gene_ids = list(gene_dict.keys())\n",
        "    if gene_ids:\n",
        "        # Post the gene IDs to NCBI.\n",
        "        handle = Entrez.epost(db=\"gene\", id=\",\".join(gene_ids))\n",
        "        result = Entrez.read(handle)\n",
        "        handle.close()\n",
        "\n",
        "        webenv = result[\"WebEnv\"]\n",
        "        query_key = result[\"QueryKey\"]\n",
        "\n",
        "        handle = Entrez.esummary(db=\"gene\", webenv=webenv, query_key=query_key)\n",
        "        record = Entrez.read(handle)\n",
        "        handle.close()\n",
        "\n",
        "        for docsum in record[\"DocumentSummarySet\"][\"DocumentSummary\"]:\n",
        "            gene_id = docsum.attributes[\"uid\"]\n",
        "            symbol = docsum.get('NomenclatureSymbol', 'No symbol')\n",
        "            organism = docsum.get('Organism', {}).get('ScientificName', 'No organism')\n",
        "            full_name = docsum.get('NomenclatureName', gene_dict[gene_id].get_name_from_article())\n",
        "            also_known_as = docsum.get('OtherAliases', gene_dict[gene_id].get_name_from_article())\n",
        "\n",
        "            if full_name == '':\n",
        "                full_name = gene_dict[gene_id].get_name_from_article()\n",
        "            if also_known_as == '':\n",
        "                also_known_as = gene_dict[gene_id].get_name_from_article()\n",
        "\n",
        "            if gene_id in gene_dict:\n",
        "                gene_dict[gene_id].update_info(symbol, organism, full_name, also_known_as)\n",
        "            time.sleep(0.5)  # Pause briefly to avoid overwhelming NCBI servers.\n",
        "\n",
        "# Set your email (and API key if available)\n",
        "Entrez.email = \"email here\"\n",
        "# Entrez.api_key = \"your_api_key\"\n",
        "\n",
        "xml_path = \"./output/PMC5453669.xml\"\n",
        "gene_dict = parse_xml_file(xml_path)\n",
        "fetch_and_update_gene_info(gene_dict)  # Optionally fetch gene info from NCBI\n",
        "\n",
        "print(\"There are {} genes in the paper.\".format(len(gene_dict)))\n",
        "if len(gene_dict) == 0:\n",
        "    print(\"No genes found in the paper.\")\n",
        "else:\n",
        "    for gene in gene_dict.values():\n",
        "        print(gene)\n",
        "        print(f\"There are {len(gene.get_occurrences())} occurrence(s) (each occurrence is at most 3 sentences long) for gene {gene.get_gene_id()} : {gene.get_name_from_article()}.\")\n",
        "        print(\"\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "33A7yYBLXZes"
      },
      "source": [
        "# 3.5 Prompt Engineering\n",
        "- Creates the prompt using the Gene dictionary (from Gene chunking)\n",
        "- batch_size: allows you to have more than 1 Gene in the prompt such that you have something like the following. Ideally keep it at batch_size=1 to ensure that the GO terms generated are exclusive to the Gene being analyzed.\n",
        "\n",
        "```\n",
        "# batch_size > 1 has the basic format:\n",
        "Basic Prompt\n",
        "Gene 1\n",
        "Gene 2\n",
        "...\n",
        "# batch_size = 1 has the basic format:\n",
        "Basic Prompt\n",
        "Gene 1\n",
        "```\n",
        "- batch_index: allows you to iterate through the gene dictionary so that if you have a paper with x amount of genes, you can just increment  the index and run the inference x amount of times.\n",
        "- direct: ignore this. This is for the future.\n",
        "- customized_prompt: ignore this as well. This is for the future."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UsRz3Wb5YTQl",
        "outputId": "dc245a37-b770-4af2-f071-f0b973bda774"
      },
      "outputs": [],
      "source": [
        "def make_go_term_prompt(gene_dict, batch_size=1, batch_index=0, direct=False, customized_prompt=None):\n",
        "    \"\"\"\n",
        "    Create a prompt for Llama to output Gene Ontology (GO) terms (without IDs) for a batch of genes\n",
        "    from the gene dictionary. The output is expected to be categorized into three sections for each gene:\n",
        "      1. Biological Processes\n",
        "      2. Cellular Components\n",
        "      3. Molecular Functions\n",
        "\n",
        "    For each gene, the output should follow the format below without mixing results across different genes.\n",
        "\n",
        "    :param gene_dict: A dictionary where each key is a gene ID and each value is a Gene object that\n",
        "                      contains attributes such as symbol, organism, full_name, and occurrences.\n",
        "    :param batch_size: The number of genes to include in this prompt batch.\n",
        "    :param batch_index: The index (0-based) of the batch to process. For instance, if batch_size=1,\n",
        "                        batch_index=2 will include the 3rd gene in the dictionary.\n",
        "    :param direct: If True, use a more direct instruction for GO term extraction.\n",
        "    :param customized_prompt: If provided, use this custom prompt text in place of the default.\n",
        "    :return: A string containing the complete prompt.\n",
        "    \"\"\"\n",
        "\n",
        "    # Instruction blocks\n",
        "\n",
        "    # context = \"\"\"You are an efficient and insightful assistant to a geneticist.\"\"\"\n",
        "\n",
        "    general_instructions = \"\"\"Analyze the following gene information and text snippets under \"Occurrences\".\n",
        "Identify all potentially relevant Gene Ontology (GO) terms using only evidence explicitly stated in the text.\n",
        "Do not include any GO terms based on inferences or assumptions not explicitly supported by the text.\n",
        "\n",
        "After completing your analysis, for each Gene Ontology (GO) term you list, provide a short one-line justification referencing the text as well as a relationship between the gene and the term.\n",
        "If the text does not explicitly support a GO term (even if it might be inferred), do not include it. If no GO term is supported for a particular gene, say “None found.”\n",
        "\n",
        "Be concise, do not use unnecessary words.\n",
        "Be factual, do not editorialize.\n",
        "Be specific, avoid overly general statements like \"it is involved in many cellular processes\".\n",
        "\n",
        "\"\"\"\n",
        "\n",
        "    example = \"\"\"To help you in your work, I am providing an example of gene information and the corresponding example analysis output structured in JSON.\n",
        "\n",
        "Example gene information is:\n",
        "Gene: myogenin\n",
        "Organism: Mus musculus\n",
        "Occurrences:\n",
        "- Activation of NF-kappaB increases the expression of the inducible nitric oxide synthase (iNOS) in muscle cells that sequester HuR (RNA-binding protein) by preventing transcription of MyoD. Denervation-induced atrophy is related to the upregulation of specific histone deacetylases (HDAC), able to repress a negative regulator of myogenin, with an increase in MuRF1 expression and muscle wasting. AMPK (protein kinase activated by 5'adenosine monophosphate) is also involved in muscle atrophy.\n",
        "\n",
        "Example anaylysis output is:\n",
        "[\n",
        "  {\n",
        "    \"gene\": \"myogenin\",\n",
        "    \"go_data\": [\n",
        "      {\n",
        "        \"term\": \"DNA-binding transcription activator activity\",\n",
        "        \"relationship\": \"enables\",\n",
        "        \"namespace\": \"molecular_function\",\n",
        "        \"justification\": \"Myogenin is a key transcription factor in muscle differentiation that binds to E-box sequences in muscle-specific genes and activates their transcription through RNA polymerase II.\"\n",
        "      },\n",
        "      {\n",
        "        \"term\": \"skeletal muscle cell differentiation\",\n",
        "        \"relationship\": \"involved_in\",\n",
        "        \"namespace\": \"biological_process\",\n",
        "        \"justification\": \"Myogenin is one of the four myogenic regulatory factors (MRFs) that drive skeletal muscle differentiation by promoting the transition from myoblasts to myotubes.\"\n",
        "      },\n",
        "      {\n",
        "        \"term\": \"nucleus\",\n",
        "        \"relationship\": \"located_in\",\n",
        "        \"namespace\": \"cellular_component\",\n",
        "        \"justification\": \"As a transcription factor, myogenin localizes to the nucleus to regulate the expression of muscle-specific genes.\"\n",
        "      }\n",
        "    ]\n",
        "  }\n",
        "]\n",
        "\"\"\"\n",
        "\n",
        "\n",
        "    # Assemble the prompt text based on the chosen mode\n",
        "    if direct:\n",
        "        # prompt_text = context\n",
        "        # prompt_text += direct_instructions\n",
        "        # prompt_text += format_instructions\n",
        "        # prompt_text += example_output\n",
        "        pass\n",
        "    elif customized_prompt:\n",
        "        # prompt_text = context\n",
        "        # prompt_text += customized_prompt\n",
        "        # prompt_text += format_instructions\n",
        "        # prompt_text += example_output\n",
        "        pass\n",
        "    else:\n",
        "        prompt_text = general_instructions\n",
        "        prompt_text += example\n",
        "        prompt_text += \"\\nHere is the gene information:\\n\"\n",
        "\n",
        "    # Convert the gene dictionary to a list and determine the batch slice.\n",
        "    gene_items = list(gene_dict.items())\n",
        "    start_index = batch_index * batch_size\n",
        "    end_index = start_index + batch_size\n",
        "    batch_genes = gene_items[start_index:end_index]\n",
        "\n",
        "    # Add details and occurrence context for each gene in the selected batch.\n",
        "    for gene_id, gene_obj in batch_genes:\n",
        "        # prompt_text += f\"\\nGene ID: {gene_id}\\n\"\n",
        "        # prompt_text += f\"Symbol: {gene_obj.symbol}\\n\"\n",
        "        # prompt_text += f\"Organism: {gene_obj.organism}\\n\"\n",
        "        # prompt_text += f\"Full Name: {gene_obj.full_name}\\n\"\n",
        "        prompt_text += f\"\\nGene: {gene_obj.get_name_from_article()}\\n\"\n",
        "        if gene_obj.get_organsim() != None:\n",
        "          # add organism if available\n",
        "          prompt_text += f\"Organism: {gene_obj.get_organsim()}\\n\"\n",
        "        prompt_text += \"Occurrences:\\n\"\n",
        "        for occ in gene_obj.occurrences:\n",
        "            prompt_text += f\"- {occ}\\n\"\n",
        "\n",
        "    return prompt_text\n",
        "\n",
        "# Example usage:\n",
        "# Assuming gene_dict is defined and contains Gene objects with attributes: symbol, organism, full_name, and occurrences.\n",
        "# To process only one gene at a time:\n",
        "# prompt = make_go_term_prompt(gene_dict, batch_size=1, batch_index=0)\n",
        "# print(prompt)\n",
        "\n",
        "# prompt = make_go_term_prompt(gene_dict, batch_size=116, batch_index=0)\n",
        "# print(prompt)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6oaKSo6iXGOc"
      },
      "source": [
        "# 4. Inference (Local)\n",
        "The two cells below rely on local inference meaning that you must have the model locally. Skip them if youre using the API like GROQ"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hzFESXHRXBHG"
      },
      "outputs": [],
      "source": [
        "print(\"Loading LLaMA model.\")\n",
        "# model_dir = \"/content/drive/MyDrive/Llama 3.2-3B-Instruct-model\"\n",
        "model_dir = \"/content/drive/MyDrive/GOLLM/Llama 3.2-3B-Instruct-model\"\n",
        "device = torch.device(\"cuda\")\n",
        "print(torch.cuda.is_available())\n",
        "\n",
        "tokenizer = AutoTokenizer.from_pretrained(model_dir, local_files_only=True)\n",
        "model = AutoModelForCausalLM.from_pretrained(model_dir, local_files_only=True).to(device)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "IzBYJajjYYly"
      },
      "outputs": [],
      "source": [
        "def run_inference(prompt):\n",
        "    \"\"\"\n",
        "    Run inference on the provided prompt using the specified model and tokenizer.\n",
        "\n",
        "    :param prompt: The prompt string to send to the model.\n",
        "    :return: The generated text.\n",
        "    \"\"\"\n",
        "\n",
        "    # Encode the prompt and move to the appropriate device\n",
        "    inputs = tokenizer(prompt, return_tensors=\"pt\").to(\"cuda\")\n",
        "    attention_mask = inputs[\"attention_mask\"]\n",
        "\n",
        "    # Greedy or minimal variation == When: You want consistent, short answers, and you don’t mind if it’s a bit “blunt.”\n",
        "    # outputs = model.generate(\n",
        "    #     input_ids=inputs[\"input_ids\"],\n",
        "    #     attention_mask=attention_mask,\n",
        "    #     max_new_tokens=150,\n",
        "    #     do_sample=False,     # no sampling\n",
        "    #     num_beams=1,         # purely greedy\n",
        "    #     no_repeat_ngram_size=3,\n",
        "    #     pad_token_id=tokenizer.eos_token_id\n",
        "    # )\n",
        "\n",
        "    # Beam Search (Higher Quality / Less Randomness) == You want a more “global optimum” text.\n",
        "    # outputs = model.generate(\n",
        "    #     input_ids=inputs[\"input_ids\"],\n",
        "    #     attention_mask=attention_mask,\n",
        "    #     max_new_tokens=200,\n",
        "    #     do_sample=False,\n",
        "    #     num_beams=4,      # search multiple beams\n",
        "    #     length_penalty=1.0,  # see if you want to encourage or discourage long outputs\n",
        "    #     no_repeat_ngram_size=3,\n",
        "    #     pad_token_id=tokenizer.eos_token_id\n",
        "    # )\n",
        "\n",
        "    # Beam SearchV2\n",
        "    outputs = model.generate(\n",
        "        input_ids=inputs[\"input_ids\"],\n",
        "        attention_mask=attention_mask,\n",
        "        max_new_tokens=400,       # Enough tokens for a concise but thorough answer\n",
        "        do_sample=False,          # Turn off sampling; we want a more deterministic, stable answer\n",
        "        num_beams=3,              # Explore multiple beams for higher-quality completions\n",
        "        length_penalty=0.1,       # 1.0 means \"neutral\" length preference (>=1.0 encourages longer outputs)\n",
        "        no_repeat_ngram_size=4,   # Helps avoid repeating the same phrase\n",
        "        early_stopping=True,      # Stops as soon as the best beam is complete\n",
        "        pad_token_id=tokenizer.eos_token_id,\n",
        "        temperature=None,\n",
        "        top_p=None,\n",
        "    )\n",
        "    # Sampling (More Creative / Less Deterministic)\n",
        "    # outputs = model.generate(\n",
        "    #     input_ids=inputs[\"input_ids\"],\n",
        "    #     attention_mask=attention_mask,\n",
        "    #     max_new_tokens=200,\n",
        "    #     do_sample=True,       # sampling\n",
        "    #     temperature=0.7,      # moderate creativity\n",
        "    #     top_p=0.9,            # nucleus sampling\n",
        "    #     no_repeat_ngram_size=3,\n",
        "    #     pad_token_id=tokenizer.eos_token_id\n",
        "    # )\n",
        "\n",
        "    # Decode the output and return the generated text\n",
        "    generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)\n",
        "    completion = generated_text[len(prompt):].strip()\n",
        "    return completion\n",
        "\n",
        "for i in range(3):\n",
        "    start_time = time.time()\n",
        "    output = run_inference(test_prompt)\n",
        "    print(output)\n",
        "\n",
        "    # with open(f'/content/drive/MyDrive/GOLLM/Inference Outputs/output_{i+1}.txt', 'w') as file:\n",
        "    #     file.write(output)\n",
        "\n",
        "    print(f\"-->Model loaded on {device} named {torch.cuda.get_device_name(0)} and it took {time.time()-start_time:.2f} seconds.\")\n",
        "    print(\"-\"*50)\n",
        "\n",
        "# prompt = \"what is 2+2\"\n",
        "# print(\"Running inference...\")\n",
        "# start_time = time.time()\n",
        "# output = run_inference(prompt)\n",
        "# print(f\"Model loaded on {device} in {time.time()-start_time:.2f} seconds.\")\n",
        "# print(output)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gSz3gKoaY7yu"
      },
      "source": [
        "# 4v2. GROQ API Inference\n",
        "- need API Key\n",
        "- uses llama 3.3-70b-versatile"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "J_k_yNbOwpkb",
        "outputId": "16a5c4d1-e0bd-4c8b-a736-77a7dd2e2130"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com\n",
            "Requirement already satisfied: Groq in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (0.18.0)\n",
            "Requirement already satisfied: anyio<5,>=3.5.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from Groq) (4.8.0)\n",
            "Requirement already satisfied: distro<2,>=1.7.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from Groq) (1.9.0)\n",
            "Requirement already satisfied: httpx<1,>=0.23.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from Groq) (0.28.1)\n",
            "Requirement already satisfied: pydantic<3,>=1.9.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from Groq) (2.10.6)\n",
            "Requirement already satisfied: sniffio in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from Groq) (1.3.1)\n",
            "Requirement already satisfied: typing-extensions<5,>=4.10 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from Groq) (4.12.2)\n",
            "Requirement already satisfied: idna>=2.8 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from anyio<5,>=3.5.0->Groq) (3.10)\n",
            "Requirement already satisfied: certifi in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from httpx<1,>=0.23.0->Groq) (2024.12.14)\n",
            "Requirement already satisfied: httpcore==1.* in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from httpx<1,>=0.23.0->Groq) (1.0.7)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from httpcore==1.*->httpx<1,>=0.23.0->Groq) (0.14.0)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from pydantic<3,>=1.9.0->Groq) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.27.2 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from pydantic<3,>=1.9.0->Groq) (2.27.2)\n"
          ]
        }
      ],
      "source": [
        "# using GROQ API\n",
        "!pip install Groq\n",
        "from groq import Groq\n",
        "\n",
        "client = Groq(\n",
        "    api_key=\"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\"\n",
        ")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pmOgXgUc_DAC"
      },
      "source": [
        "## Run it again if it stops. It will skip documents that already exists."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6eOnOkZsztbK",
        "outputId": "f9dbfdd0-0f1e-4238-cfb7-7cca62af4790"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Skipping gene 433766: output file already exists.\n",
            "Skipping gene 14685: output file already exists.\n",
            "Skipping gene 11651: output file already exists.\n",
            "Finished processing 61 genes. Final merged JSON saved to c:\\Users\\aivan\\Desktop\\BIOIN 401\\GOLLM\\Llama3.2-3B-Instruct-Inference\\llama_inferences\\llama_output_PMC5453669.json\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "import json\n",
        "import tempfile\n",
        "import shutil \n",
        "\n",
        "# Create a temporary directory to store individual JSON outputs.\n",
        "temp_dir = os.path.join(os.getcwd(), \"temp_inferences\")\n",
        "if not os.path.exists(temp_dir):\n",
        "    os.makedirs(temp_dir)\n",
        "# Create a  directory to store individual llama JSON outputs.\n",
        "llama_dir = os.path.join(os.getcwd(), \"llama_inferences\")\n",
        "if not os.path.exists(llama_dir):\n",
        "    os.makedirs(llama_dir)\n",
        "\n",
        "# retrieve the doc_id from the gene_dict of any gene (since all genes in the gene_dict are from the same document)\n",
        "doc_id = None\n",
        "for gene in gene_dict.values():\n",
        "    if gene.get_doc_id() is not None:\n",
        "        doc_id = gene.get_doc_id()\n",
        "        break\n",
        "\n",
        "# Create a subdirectory under temp_inferences named after doc_id.\n",
        "# If doc_id is None, fall back to a placeholder name, e.g. \"unknown_doc_id\".\n",
        "temp_doc_dir = os.path.join(temp_dir, doc_id if doc_id else \"unknown_doc_id\")\n",
        "if not os.path.exists(temp_doc_dir):\n",
        "    os.makedirs(temp_doc_dir)\n",
        "\n",
        "# List of gene items from gene_dict\n",
        "gene_items = list(gene_dict.items())\n",
        "final_results = []\n",
        "\n",
        "# Process each gene one-by-one (batch_size=1)\n",
        "for idx, (gene_id, gene_obj) in enumerate(gene_items):\n",
        "    output_filename = os.path.join(temp_doc_dir, f\"gene_{gene_id}.json\")\n",
        "\n",
        "    # Check if the gene's output file already exists\n",
        "    if os.path.exists(output_filename):\n",
        "        print(f\"Skipping gene {gene_id}: output file already exists.\")\n",
        "        try:\n",
        "            with open(output_filename, \"r\", encoding=\"utf-8\") as f:\n",
        "                gene_output = f.read()\n",
        "            final_results.append(json.loads(gene_output))\n",
        "        except json.JSONDecodeError:\n",
        "            print(f\"Warning: Existing output for gene {gene_id} is not valid JSON.\")\n",
        "        continue\n",
        "\n",
        "    # Generate a prompt for a single gene.\n",
        "    prompt = make_go_term_prompt(gene_dict, batch_size=1, batch_index=idx)\n",
        "    # Use your GROQ API client to request the inference.\n",
        "    chat_completion = client.chat.completions.create(\n",
        "        messages=[\n",
        "            {\"role\": \"user\", \"content\": prompt},\n",
        "        ],\n",
        "        model=\"llama-3.3-70b-versatile\",\n",
        "        temperature=0,\n",
        "        max_completion_tokens=4096,\n",
        "        top_p=0.5,\n",
        "        stream=False,\n",
        "        response_format={\"type\": \"json_object\"},\n",
        "        stop=None,\n",
        "    )\n",
        "\n",
        "    # Extract the JSON output from the response.\n",
        "    gene_output = chat_completion.choices[0].message.content\n",
        "\n",
        "    # Add the output to the final results list.\n",
        "    try:\n",
        "        model_json = json.loads(gene_output)\n",
        "\n",
        "        # Construct a new dict to ensure key order: gene, geneID, go_data\n",
        "        ordered_model_json = {\n",
        "            \"gene\": model_json.get(\"gene\", \"\"),\n",
        "            \"geneID\": gene_id,\n",
        "            \"go_data\": model_json.get(\"go_data\", [])\n",
        "        }\n",
        "\n",
        "        # Add reference(s) to each GO entry\n",
        "        for entry in ordered_model_json[\"go_data\"]:\n",
        "            entry[\"reference(s)\"] = [gene_obj.get_doc_id()]\n",
        "\n",
        "        # Save the updated JSON in the temp folder\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            json.dump(ordered_model_json, f, indent=2)\n",
        "\n",
        "        final_results.append(ordered_model_json)\n",
        "    except json.JSONDecodeError:\n",
        "        print(f\"Warning: Output for gene {gene_id} is not valid JSON.\")\n",
        "\n",
        "# Merge all outputs into one finalized JSON file, using doc_id in the filename\n",
        "llama_output_path = os.path.join(llama_dir, f\"llama_output_{doc_id}.json\")\n",
        "with open(llama_output_path, \"w\", encoding=\"utf-8\") as f:\n",
        "    json.dump(final_results, f, indent=2)\n",
        "\n",
        "print(f\"Finished processing {len(gene_items)} genes. Final merged JSON saved to {llama_output_path}\")\n",
        "\n",
        "# Remove the temporary directory\n",
        "# shutil.rmtree(temp_dir)\n",
        "# print(f\"Removed temporary folder: {temp_dir}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "P58Ns1lwdbbg"
      },
      "source": [
        "# 5. Normalize\n",
        "- use an embedding model (SAPBERT) to ensure that the terms generated by Llama are actual GO terms. You will need the Go.obo file from the [Gene Ontology Consortium](https://geneontology.org/docs/download-ontology/)\n",
        "  - Essentially, for each go term created by Llama, Sapbert will do a similarity search against the Go.obo and retrieve the actual term + GO:ID"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "n5lKdeAZ1Ssh",
        "outputId": "31f1d594-43c0-4f83-e46c-4873128724d0"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com\n",
            "Requirement already satisfied: sentence-transformers in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (3.4.1)\n",
            "Requirement already satisfied: transformers<5.0.0,>=4.41.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from sentence-transformers) (4.48.2)\n",
            "Requirement already satisfied: tqdm in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from sentence-transformers) (4.67.1)\n",
            "Requirement already satisfied: torch>=1.11.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from sentence-transformers) (2.5.1+cu124)\n",
            "Requirement already satisfied: scikit-learn in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from sentence-transformers) (1.6.1)\n",
            "Requirement already satisfied: scipy in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from sentence-transformers) (1.15.1)\n",
            "Requirement already satisfied: huggingface-hub>=0.20.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from sentence-transformers) (0.27.1)\n",
            "Requirement already satisfied: Pillow in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from sentence-transformers) (10.2.0)\n",
            "Requirement already satisfied: filelock in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (3.17.0)\n",
            "Requirement already satisfied: fsspec>=2023.5.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (2024.12.0)\n",
            "Requirement already satisfied: packaging>=20.9 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (24.2)\n",
            "Requirement already satisfied: pyyaml>=5.1 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (6.0.2)\n",
            "Requirement already satisfied: requests in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (2.32.3)\n",
            "Requirement already satisfied: typing-extensions>=3.7.4.3 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from huggingface-hub>=0.20.0->sentence-transformers) (4.12.2)\n",
            "Requirement already satisfied: networkx in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from torch>=1.11.0->sentence-transformers) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from torch>=1.11.0->sentence-transformers) (3.1.5)\n",
            "Requirement already satisfied: sympy==1.13.1 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from torch>=1.11.0->sentence-transformers) (1.13.1)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from sympy==1.13.1->torch>=1.11.0->sentence-transformers) (1.3.0)\n",
            "Requirement already satisfied: colorama in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from tqdm->sentence-transformers) (0.4.6)\n",
            "Requirement already satisfied: numpy>=1.17 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (1.26.4)\n",
            "Requirement already satisfied: regex!=2019.12.17 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (2024.11.6)\n",
            "Requirement already satisfied: tokenizers<0.22,>=0.21 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (0.21.0)\n",
            "Requirement already satisfied: safetensors>=0.4.1 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers) (0.5.2)\n",
            "Requirement already satisfied: joblib>=1.2.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from scikit-learn->sentence-transformers) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from scikit-learn->sentence-transformers) (3.5.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from jinja2->torch>=1.11.0->sentence-transformers) (3.0.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers) (2024.12.14)\n",
            "Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com\n",
            "Requirement already satisfied: obonet in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (1.1.0)\n",
            "Requirement already satisfied: networkx in c:\\users\\aivan\\anaconda3\\envs\\bioin311c1\\lib\\site-packages (from obonet) (3.4.2)\n"
          ]
        }
      ],
      "source": [
        "# Install required packages (run these in your environment if not already installed)\n",
        "!pip install sentence-transformers\n",
        "!pip install obonet"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8TPMPZKF1P1X",
        "outputId": "80689670-c815-40d7-8bb8-d202d1fc88c4"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "c:\\Users\\aivan\\anaconda3\\envs\\BIOIN311C1\\Lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
            "  from .autonotebook import tqdm as notebook_tqdm\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Loaded 40267 GO terms from C:/Users/aivan/Desktop/BIOIN 401/Go Files/go.obo\n"
          ]
        }
      ],
      "source": [
        "import obonet\n",
        "import json\n",
        "from sentence_transformers import SentenceTransformer, util\n",
        "import numpy as np\n",
        "\n",
        "##########################################\n",
        "# 1. Load and parse the GO.obo file\n",
        "##########################################\n",
        "\n",
        "def load_go_ontology(obo_filepath):\n",
        "    \"\"\"\n",
        "    Parse the GO.obo file using obonet.\n",
        "    Returns a dictionary mapping GO IDs to their term names.\n",
        "    \"\"\"\n",
        "    graph = obonet.read_obo(obo_filepath)\n",
        "    go_terms = {}\n",
        "    for go_id, data in graph.nodes(data=True):\n",
        "        # Only include nodes that have a 'name' field\n",
        "        if 'name' in data:\n",
        "            go_terms[go_id] = data\n",
        "    return go_terms\n",
        "\n",
        "# set the path to your local go.obo (or something else later) file\n",
        "obo_file = \"C:/Users/aivan/Desktop/BIOIN 401/Go Files/go.obo\"\n",
        "go_terms_dict = load_go_ontology(obo_file)\n",
        "print(f\"Loaded {len(go_terms_dict)} GO terms from {obo_file}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 340,
          "referenced_widgets": [
            "e4718c03421b41a2a665c26293acd697",
            "7d97b347b7cb40df81041035931d7f27",
            "b85f6bb59d024282b5552ff665da4548",
            "f1219528a46a4a8e94bd1c8de9e6e16a",
            "1b8c25318a2b4b64bab3496083679c85",
            "53bb7f0fbad34a818e266d0ea44ffaf1",
            "bc435a026437452bb820c14d1ce5d3ee",
            "47571c7570f14ab58f96e5d30129da1b",
            "ec42f776af1b48a982636f0c122ab7e6",
            "de6e3ef0670042378384c41ecd695f1a",
            "ec7200455fa34863bec2ae60f9f4aad5",
            "618fd6a5462e4c5296b560f888b5d638",
            "331dbfdb6d7849c6a7b6d43bd33a7325",
            "54a0cadcc9ef406db05d36b690d7891c",
            "2443ed122977428c98248537123b674b",
            "c3911d8aca764dd38fad4c1dded8198f",
            "8eb65674477e477091f2501704df7eb8",
            "6d29b58c64cc4f869d6944699149a60e",
            "c939f017049643b18d5142514937af73",
            "6c463c9a4bef4b5194113acf7ced731e",
            "f7dcd748d0944da09948f4372366e8a8",
            "466f8ef6ba644d31a8586abaf9830448",
            "6ac927f5e2a2453d9aae4b5c474cf96d",
            "7593152124e04332b709dd30ec87c5a5",
            "4a743c94312c4b5db25087a747f6b879",
            "59dc6c4216cb4c5aad8c3764cf3d3f6d",
            "3117723c389b4e47a97d6e582078e5c9",
            "c6b5cf46fbeb41adb8da935949b808c1",
            "d4b8ff5330d14128ba304020f7fe8412",
            "ea731973ab9c44caa479fb7934fd3b4f",
            "097046dc7c4f46f5866068c82b706aca",
            "cc7573c531c84e61999e2161710173f3",
            "9775c0cd37bc43dcb7953f8b15343332",
            "1675d70c8a5e426691d0af5e6ba333f3",
            "784d7bf8c3be48b8ad5a3b6e6954690f",
            "730fe49f91204e2cbd3b58d24442be0a",
            "792eb1f8daf34e2590fda2f60e3ff684",
            "3e0f9cdf58ef4af1a0b16f9795747dfc",
            "1845ddb7ff634d969fa2538209fc4e0a",
            "9b03af45365c42419bb39779b82be47c",
            "d598fdb3d94a4095ba5a950cef321f1b",
            "39794838a2164151a5edf298a63f8147",
            "e7eaa149a903445db37dfdf719e14f9e",
            "4af857d8e5a94c278ae0fb3d7e092ac5",
            "769b37c671334690b6a7e5ba9059343e",
            "bdf351e0ccbc4aa18e168f593d1659cf",
            "8901a8884437441c934c53a98a1dac06",
            "32ef3e232304493d8b7e129c679f4f6f",
            "77472246b6b0497fbe67b8a3529040e5",
            "61de99be2c2e4afe90ff06e3bb30a2b7",
            "873ea6832a0447a19938fe764cb7ac58",
            "b483173c5fa74a72bb0a88206d3206a3",
            "8039b35f8f654a5e93ab69a0d1d8f3ea",
            "b1c01cfdd57e40e7bc4de9baf81d9f45",
            "82a3774cc5cd4acfb1822d9cb74d6f45"
          ]
        },
        "id": "zfuUI3PF1boL",
        "outputId": "e257d393-bdbb-4695-d2ac-d8f2ee4ecbea"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "No sentence-transformers model found with name cambridgeltl/SapBERT-from-PubMedBERT-fulltext. Creating a new one with mean pooling.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Loaded embedding model: cambridgeltl/SapBERT-from-PubMedBERT-fulltext\n"
          ]
        }
      ],
      "source": [
        "##########################################\n",
        "# 2. Load the SAPBERT embedding model (or an alternative)\n",
        "##########################################\n",
        "\n",
        "# Here we use a SAPBERT model available from Hugging Face.\n",
        "# may choose another model (e.g., \"sentence-transformers/all-MiniLM-L6-v2\")\n",
        "model_name = \"cambridgeltl/SapBERT-from-PubMedBERT-fulltext\"  # SAPBERT model fine-tuned on UMLS\n",
        "sapbert_model = SentenceTransformer(model_name)\n",
        "print(f\"Loaded embedding model: {model_name}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OM1So4sX2JLT",
        "outputId": "8477c5ed-2546-409c-8284-f1d957453bde"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Embedded GO terms with additional metadata.\n"
          ]
        }
      ],
      "source": [
        "def embed_go_terms(go_terms, model):\n",
        "    \"\"\"\n",
        "    Given a dictionary of GO terms (GO ID -> term data) and an embedding model,\n",
        "    returns:\n",
        "      - a list of GO IDs,\n",
        "      - a list of GO term names (in the same order),\n",
        "      - a tensor of embeddings for the GO term names,\n",
        "      - a list of GO namespaces,\n",
        "      - a list of GO relationships (if available, otherwise default to 'unknown')\n",
        "    \"\"\"\n",
        "    go_ids = []\n",
        "    go_names = []\n",
        "    go_namespaces = []\n",
        "    go_relationships = []\n",
        "\n",
        "    for go_id, data in go_terms.items():\n",
        "        if 'name' in data:\n",
        "            go_ids.append(go_id)\n",
        "            go_names.append(data['name'])\n",
        "            go_namespaces.append(data.get('namespace', 'unknown'))\n",
        "            # Some GO entries might not have a \"relationship\" field; default to 'unknown'\n",
        "            go_relationships.append(data.get('relationship', 'unknown'))\n",
        "\n",
        "    go_embeddings = model.encode(go_names, convert_to_tensor=True)\n",
        "    return go_ids, go_names, go_embeddings, go_namespaces, go_relationships\n",
        "\n",
        "# Reload GO ontology using the updated load_go_ontology:\n",
        "# go_terms_dict = load_go_ontology(obo_file) # already loaded in the previous cell\n",
        "go_ids, go_names, go_embeddings, go_namespaces, go_relationships = embed_go_terms(go_terms_dict, sapbert_model)\n",
        "print(\"Embedded GO terms with additional metadata.\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "G0yR2eF6228J",
        "outputId": "3bfeca3b-8998-4598-f8f4-56505af53917"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Example: Llama term 'muscle cell differentiation' matched with 'muscle cell differentiation' (GO:0042692) with score 1.00\n"
          ]
        }
      ],
      "source": [
        "import torch\n",
        "\n",
        "def get_best_matching_go_term(llama_term, go_ids, go_names, go_embeddings, model, threshold=0.6):\n",
        "    \"\"\"\n",
        "    Given a GO term from Llama, compute its embedding and find the best matching GO term\n",
        "    from the provided GO ontology embeddings.\n",
        "\n",
        "    Parameters:\n",
        "      - llama_term: The GO term string generated by Llama.\n",
        "      - go_ids, go_names, go_embeddings: Outputs from embed_go_terms().\n",
        "      - model: The embedding model.\n",
        "      - threshold: Minimum cosine similarity required to accept the match.\n",
        "\n",
        "    Returns:\n",
        "      - best_go_id: The GO:ID of the best match (or None if score is below threshold)\n",
        "      - best_go_name: The GO term name of the best match.\n",
        "      - best_score: The cosine similarity score.\n",
        "    \"\"\"\n",
        "    # Encode the query using the model\n",
        "    query_embedding = model.encode(llama_term, convert_to_tensor=True)\n",
        "    # Compute cosine similarities between the query and all GO term embeddings\n",
        "    cos_scores = torch.squeeze(util.cos_sim(query_embedding, go_embeddings))\n",
        "    # Use torch.argmax to find the best matching index\n",
        "    best_idx = torch.argmax(cos_scores)\n",
        "    best_score = cos_scores[best_idx].item()\n",
        "\n",
        "    if best_score >= threshold:\n",
        "        return go_ids[best_idx], go_names[best_idx], best_score, best_idx\n",
        "    else:\n",
        "        return None, None, best_score, None\n",
        "\n",
        "# Test the matching function with an example Llama-generated term:\n",
        "example_llama_term = \"muscle cell differentiation\"\n",
        "best_go_id, best_go_name, score, best_idx = get_best_matching_go_term(example_llama_term, go_ids, go_names, go_embeddings, sapbert_model)\n",
        "print(f\"Example: Llama term '{example_llama_term}' matched with '{best_go_name}' ({best_go_id}) with score {score:.2f}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zp0nhP0_3pHz",
        "outputId": "331efcf4-a0ca-4146-ceab-abf18abd3981"
      },
      "outputs": [],
      "source": [
        "import json\n",
        "\n",
        "def parse_llama_output(llama_output_path):\n",
        "    \"\"\"\n",
        "    Parses Llama's output JSON file.\n",
        "    Returns the parsed JSON data (a list of gene entries).\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with open(llama_output_path, 'r') as f:\n",
        "            data = json.load(f)\n",
        "        # If the JSON is a single dictionary, convert it to a list\n",
        "        if isinstance(data, dict):\n",
        "            data = [data]\n",
        "        return data\n",
        "    except Exception as e:\n",
        "        print(f\"Error parsing JSON: {e}\")\n",
        "        return []\n",
        "\n",
        "# print(parse_llama_output(\"./llama_inferences/llama_output_22118460))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rpyVuCmM_EtU",
        "outputId": "c8d869f2-162d-4661-cf94-91f28390fa6e"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "there are 61 genes with matched GO terms.\n",
            "Gene: MuRF1 (433766)\n",
            "  Llama Term: ubiquitin-protein transferase activity\n",
            "  Matched GO Term: ubiquitin-protein transferase activity (GO:0004842)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: MuRF1 catalyzes TRalpha mono-ubiquitination.\n",
            "\n",
            "  Llama Term: negative regulation of transcription by RNA polymerase II\n",
            "  Matched GO Term: negative regulation of transcription by RNA polymerase II (GO:0000122)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['negatively_regulates GO:0006366']\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 inhibits TRalpha activity by mono-ubiquitinating it.\n",
            "\n",
            "  Llama Term: protein ubiquitination\n",
            "  Matched GO Term: protein ubiquitination (GO:0016567)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 mono-ubiquitinates TRalpha to inhibit its activity.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: MuRF1 sequesters TRalpha in the nucleus to inhibit its activity.\n",
            "\n",
            "  Llama Term: negative regulation of cardiac hypertrophy\n",
            "  Matched GO Term: negative regulation of cardiac muscle hypertrophy (GO:0010614)\n",
            "  Similarity: 0.95\n",
            "  Relationship: ['negatively_regulates GO:0003300']\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 inhibits T3-induced cardiomyocyte hypertrophy.\n",
            "\n",
            "  Llama Term: regulation of fatty acid oxidation\n",
            "  Matched GO Term: regulation of fatty acid oxidation (GO:0046320)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0019395']\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 inhibits PPARalpha-regulated fatty acid oxidation.\n",
            "\n",
            "Gene: TRalpha (14685)\n",
            "  Llama Term: thyroid hormone receptor activity\n",
            "  Matched GO Term: nuclear receptor activity (GO:0004879)\n",
            "  Similarity: 0.83\n",
            "  Relationship: ['part_of GO:0030522']\n",
            "  Namespace: molecular_function\n",
            "  Justification: TRalpha is a thyroid hormone receptor that mediates the effects of T3 on cardiac function and gene expression.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: TRalpha is a nuclear receptor that localizes to the nucleus to regulate gene expression.\n",
            "\n",
            "  Llama Term: regulation of cardiac hypertrophy\n",
            "  Matched GO Term: regulation of cardiac muscle hypertrophy (GO:0010611)\n",
            "  Similarity: 0.96\n",
            "  Relationship: ['regulates GO:0003300']\n",
            "  Namespace: biological_process\n",
            "  Justification: TRalpha activity is critical for regulating cardiac hypertrophy in response to T3.\n",
            "\n",
            "  Llama Term: protein ubiquitination\n",
            "  Matched GO Term: protein ubiquitination (GO:0016567)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: TRalpha is mono-ubiquitinated by MuRF1, which inhibits its activity.\n",
            "\n",
            "  Llama Term: ligand-activated transcription factor activity\n",
            "  Matched GO Term: ligand-modulated transcription activator activity (GO:0141097)\n",
            "  Similarity: 0.85\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TRalpha is activated by T3 to regulate gene expression.\n",
            "\n",
            "Gene: Akt (11651)\n",
            "  Llama Term: cell survival\n",
            "  Matched GO Term: cell death (GO:0008219)\n",
            "  Similarity: 0.70\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: T3 activates pro-survival pathways including Akt.\n",
            "\n",
            "  Llama Term: heart process\n",
            "  Matched GO Term: heart process (GO:0003015)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Treatment with T3 after myocardial infarction is cardioprotective and promotes elements of physiological hypertrophic response after cardiac injury, activating pro-survival pathways including Akt.\n",
            "\n",
            "  Llama Term: intracellular signal transduction\n",
            "  Matched GO Term: intracellular signal transduction (GO:0035556)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['occurs_in GO:0005622']\n",
            "  Namespace: biological_process\n",
            "  Justification: Akt is activated by T3, indicating involvement in intracellular signal transduction pathways.\n",
            "\n",
            "  Llama Term: protein kinase activity\n",
            "  Matched GO Term: protein kinase activity (GO:0004672)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Akt is a protein kinase.\n",
            "\n",
            "Gene: mTOR (56717)\n",
            "  Llama Term: cellular response to hormone stimulus\n",
            "  Matched GO Term: cellular response to hormone stimulus (GO:0032870)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: T3 activates pro-survival pathways including Akt and mTOR.\n",
            "\n",
            "  Llama Term: regulation of cardiac muscle cell growth\n",
            "  Matched GO Term: regulation of cardiac muscle tissue growth (GO:0055021)\n",
            "  Similarity: 0.94\n",
            "  Relationship: ['regulates GO:0055017']\n",
            "  Namespace: biological_process\n",
            "  Justification: Treatment with T3 after myocardial infarction is cardioprotective and promotes elements of physiological hypertrophic response after cardiac injury.\n",
            "\n",
            "  Llama Term: protein kinase binding\n",
            "  Matched GO Term: protein kinase binding (GO:0019901)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: mTOR is activated by T3, which also activates pro-survival pathways including Akt.\n",
            "\n",
            "  Llama Term: cell growth\n",
            "  Matched GO Term: cell growth (GO:0016049)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: T3 activates many of the physiological changes that would benefit the heart, activating pro-survival pathways Akt and mTOR.\n",
            "\n",
            "Gene: TH (21823)\n",
            "  Llama Term: hormone activity\n",
            "  Matched GO Term: hormone activity (GO:0005179)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Thyroid hormone (TH) is recognized for its role in cellular metabolism and growth.\n",
            "\n",
            "  Llama Term: cellular metabolism\n",
            "  Matched GO Term: metabolic process (GO:0008152)\n",
            "  Similarity: 0.85\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Thyroid hormone (TH) is recognized for its role in cellular metabolism and growth.\n",
            "\n",
            "  Llama Term: heart development\n",
            "  Matched GO Term: heart development (GO:0007507)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0072359']\n",
            "  Namespace: biological_process\n",
            "  Justification: TH participates in homeostasis of the heart.\n",
            "\n",
            "  Llama Term: cardiac muscle cell differentiation\n",
            "  Matched GO Term: cardiac muscle cell differentiation (GO:0055007)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0048738']\n",
            "  Namespace: biological_process\n",
            "  Justification: T3 is involved in directing myocyte alignment and geometry.\n",
            "\n",
            "  Llama Term: extracellular matrix organization\n",
            "  Matched GO Term: extracellular matrix organization (GO:0030198)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: T3 is involved in extracellular matrix remodeling.\n",
            "\n",
            "  Llama Term: regulation of cardiac muscle contraction\n",
            "  Matched GO Term: regulation of cardiac muscle contraction (GO:0055117)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0060048']\n",
            "  Namespace: biological_process\n",
            "  Justification: T3 mediates direct influences on cardiac function by affecting calcium flux and calcium handling proteins.\n",
            "\n",
            "  Llama Term: thyroid hormone receptor binding\n",
            "  Matched GO Term: nuclear thyroid hormone receptor binding (GO:0046966)\n",
            "  Similarity: 0.96\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TH is a ligand for the thyroid hormone receptor.\n",
            "\n",
            "  Llama Term: cellular response to hormone stimulus\n",
            "  Matched GO Term: cellular response to hormone stimulus (GO:0032870)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Treatment with TH following myocardial infarction is cardioprotective.\n",
            "\n",
            "Gene: IGF-1 (16000)\n",
            "  Llama Term: regulation of cardiac muscle hypertrophy\n",
            "  Matched GO Term: regulation of cardiac muscle hypertrophy (GO:0010611)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0003300']\n",
            "  Namespace: biological_process\n",
            "  Justification: IGF-1 is mentioned to induce physiological hypertrophic response after cardiac injury and is involved in exercise-mediated physiological hypertrophy.\n",
            "\n",
            "  Llama Term: positive regulation of cardiomyocyte hypertrophy\n",
            "  Matched GO Term: positive regulation of cardiac muscle hypertrophy (GO:0010613)\n",
            "  Similarity: 0.94\n",
            "  Relationship: ['positively_regulates GO:0003300']\n",
            "  Namespace: biological_process\n",
            "  Justification: IGF-1 signaling is increased in physiological hypertrophy and MuRF1 inhibits IGF-1-dependent cardiomyocyte hypertrophy.\n",
            "\n",
            "  Llama Term: cell growth\n",
            "  Matched GO Term: cell growth (GO:0016049)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: IGF-1 is involved in driving increases in contractile proteins critical to cardiac function and performance, implying a role in cell growth.\n",
            "\n",
            "Gene: RARgamma (19411)\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: RARgamma is a nuclear receptor, implying it is located in the nucleus.\n",
            "\n",
            "  Llama Term: protein heterodimerization activity\n",
            "  Matched GO Term: protein heterodimerization activity (GO:0046982)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: RARgamma heterodimerizes with RXR, indicating it enables protein heterodimerization activity.\n",
            "\n",
            "  Llama Term: ubiquitin-proteasome system\n",
            "  Matched GO Term: proteasome-mediated ubiquitin-dependent protein catabolic process (GO:0043161)\n",
            "  Similarity: 0.79\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: RARgamma is degraded by the ubiquitin proteasome system, indicating involvement in this process.\n",
            "\n",
            "Gene: PPARalpha (19013)\n",
            "  Llama Term: fatty acid oxidation\n",
            "  Matched GO Term: fatty acid oxidation (GO:0019395)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo.\n",
            "\n",
            "  Llama Term: nuclear export\n",
            "  Matched GO Term: nuclear export (GO:0051168)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 inhibited PPARalpha activity by enhancing nuclear export without promoting PPARalpha degradation.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: PPARalpha is redirected to the perinuclear area with punctate distribution by CAP350.\n",
            "\n",
            "Gene: PPARbeta/delta (19015)\n",
            "  Llama Term: fatty acid oxidation\n",
            "  Matched GO Term: fatty acid oxidation (GO:0019395)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: PPARbeta/delta activity is related to fatty acid oxidation, as MuRF3 inhibits PPARbeta activity in vivo.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: As a nuclear receptor, PPARbeta/delta is expected to be located in the nucleus.\n",
            "\n",
            "Gene: MuRF2 (381485)\n",
            "  Llama Term: ubiquitin-protein transferase activity\n",
            "  Matched GO Term: ubiquitin-protein transferase activity (GO:0004842)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: MuRF2 inhibits PPARgamma1 through mono-ubiquitination.\n",
            "\n",
            "  Llama Term: negative regulation of fatty acid oxidation\n",
            "  Matched GO Term: negative regulation of fatty acid oxidation (GO:0046322)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['negatively_regulates GO:0019395']\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF2 inhibits PPARgamma1, which regulates fatty acid oxidation.\n",
            "\n",
            "  Llama Term: protein ubiquitination\n",
            "  Matched GO Term: protein ubiquitination (GO:0016567)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF2 mono-ubiquitinates PPARgamma1.\n",
            "\n",
            "  Llama Term: cytoplasm\n",
            "  Matched GO Term: cytoplasm (GO:0005737)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005622']\n",
            "  Namespace: cellular_component\n",
            "  Justification: MuRF2 is a muscle-specific ubiquitin ligase, which is typically located in the cytoplasm.\n",
            "\n",
            "Gene: MuRF3 (58522)\n",
            "  Llama Term: ubiquitin-protein transferase activity\n",
            "  Matched GO Term: ubiquitin-protein transferase activity (GO:0004842)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: MuRF3 inhibits PPARbeta activity in vivo by mono-ubiquitination of substrates.\n",
            "\n",
            "  Llama Term: protein ubiquitination\n",
            "  Matched GO Term: protein ubiquitination (GO:0016567)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF3 inhibits PPARbeta activity in vivo through mono-ubiquitination of substrates.\n",
            "\n",
            "  Llama Term: fatty acid oxidation\n",
            "  Matched GO Term: fatty acid oxidation (GO:0019395)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF3 inhibits PPARbeta activity, which is involved in regulating fatty acid oxidation.\n",
            "\n",
            "Gene: ubiquitin ligase (84585)\n",
            "  Llama Term: ubiquitin protein ligase activity\n",
            "  Matched GO Term: ubiquitin protein ligase activity (GO:0061630)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: The muscle-specific ubiquitin ligase MuRF1 has recently been shown to specifically inhibit PPARalpha, but not PPARbeta/delta or PPARgamma1, activity in vivo through mono-ubiquitination.\n",
            "\n",
            "  Llama Term: protein monoubiquitination\n",
            "  Matched GO Term: protein monoubiquitination (GO:0006513)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 was found to inhibit PPARalpha in the cardiomyocyte through mono-ubiquitination that enhanced nuclear export without promoting PPARalpha degradation.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: MuRF1 inhibits PPARalpha-regulated fatty acid oxidation by mono-ubiquitinating lysines adjacent to a newly identified nuclear export sequence in PPARalpha, implying nuclear localization.\n",
            "\n",
            "  Llama Term: regulation of fatty acid oxidation\n",
            "  Matched GO Term: regulation of fatty acid oxidation (GO:0046320)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0019395']\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 inhibited PPARalpha-regulated fatty acid oxidation both in vitro and in vivo by mono-ubiquitination.\n",
            "\n",
            "  Llama Term: negative regulation of transcription by RNA polymerase II\n",
            "  Matched GO Term: negative regulation of transcription by RNA polymerase II (GO:0000122)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['negatively_regulates GO:0006366']\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 inhibits PPARalpha activity, which is a transcription factor, implying a role in regulating transcription.\n",
            "\n",
            "Gene: MuRF1 (140939)\n",
            "  Llama Term: negative regulation of cardiac hypertrophy\n",
            "  Matched GO Term: negative regulation of cardiac muscle hypertrophy (GO:0010614)\n",
            "  Similarity: 0.95\n",
            "  Relationship: ['negatively_regulates GO:0003300']\n",
            "  Namespace: biological_process\n",
            "  Justification: MuRF1 attenuates T3-induced cardiac hypertrophy.\n",
            "\n",
            "  Llama Term: muscle cell\n",
            "  Matched GO Term: contractile muscle fiber (GO:0043292)\n",
            "  Similarity: 0.84\n",
            "  Relationship: ['part_of GO:0005737']\n",
            "  Namespace: cellular_component\n",
            "  Justification: MuRF1 is expressed in cardiomyocytes.\n",
            "\n",
            "Gene: CSA (71991)\n",
            "  Llama Term: cardiac muscle cell hypertrophy\n",
            "  Matched GO Term: cardiac muscle hypertrophy (GO:0003300)\n",
            "  Similarity: 0.93\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: T3 induced cardiomyocyte hypertrophy, reflected by increased CSA.\n",
            "\n",
            "Gene: TH (25085)\n",
            "  Llama Term: cardiac muscle cell differentiation\n",
            "  Matched GO Term: cardiac muscle cell differentiation (GO:0055007)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0048738']\n",
            "  Namespace: biological_process\n",
            "  Justification: TH-induced physiological hypertrophy in cardiomyocytes\n",
            "\n",
            "  Llama Term: atrial cardiac muscle cell development\n",
            "  Matched GO Term: atrial cardiac muscle cell development (GO:0055014)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0055011']\n",
            "  Namespace: biological_process\n",
            "  Justification: Thyroid hormone replacement therapy attenuates atrial remodeling\n",
            "\n",
            "  Llama Term: heart regeneration\n",
            "  Matched GO Term: cardiac muscle tissue regeneration (GO:0061026)\n",
            "  Similarity: 0.89\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Thyroid hormone replacement therapy reduces atrial fibrillation inducibility in a rat myocardial infarction-heart failure model\n",
            "\n",
            "Gene: TRalpha1 (104218)\n",
            "  Llama Term: thyroid hormone receptor activity\n",
            "  Matched GO Term: nuclear receptor activity (GO:0004879)\n",
            "  Similarity: 0.83\n",
            "  Relationship: ['part_of GO:0030522']\n",
            "  Namespace: molecular_function\n",
            "  Justification: TRalpha1 stimulates growth in cardiomyocytes in response to T3.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: TRalpha1 localizes to the nucleus, where its activity is regulated by MuRF1.\n",
            "\n",
            "  Llama Term: regulation of cardiomyocyte hypertrophy\n",
            "  Matched GO Term: regulation of cardiac muscle hypertrophy (GO:0010611)\n",
            "  Similarity: 0.95\n",
            "  Relationship: ['regulates GO:0003300']\n",
            "  Namespace: biological_process\n",
            "  Justification: TRalpha1 activity is inhibited by MuRF1, which in turn inhibits T3-induced cardiomyocyte hypertrophy.\n",
            "\n",
            "  Llama Term: response to ischemia\n",
            "  Matched GO Term: response to ischemia (GO:0002931)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Acute T3 treatment protects the heart against ischemia-reperfusion injury via TRalpha1 receptor.\n",
            "\n",
            "Gene: TRalpha2 (21833)\n",
            "  Llama Term: thyroid hormone receptor activity\n",
            "  Matched GO Term: nuclear receptor activity (GO:0004879)\n",
            "  Similarity: 0.83\n",
            "  Relationship: ['part_of GO:0030522']\n",
            "  Namespace: molecular_function\n",
            "  Justification: TRalpha2 is a thyroid hormone receptor.\n",
            "\n",
            "  Llama Term: heart development\n",
            "  Matched GO Term: heart development (GO:0007507)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0072359']\n",
            "  Namespace: biological_process\n",
            "  Justification: Regulation of thyroid hormone receptor alpha is critical to understanding heart disease.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: As a thyroid hormone receptor, TRalpha2 is expected to be located in the nucleus.\n",
            "\n",
            "Gene: TRbeta1 (211770)\n",
            "  Llama Term: thyroid hormone receptor activity\n",
            "  Matched GO Term: nuclear receptor activity (GO:0004879)\n",
            "  Similarity: 0.83\n",
            "  Relationship: ['part_of GO:0030522']\n",
            "  Namespace: molecular_function\n",
            "  Justification: TRbeta1 is a thyroid receptor isoform.\n",
            "\n",
            "  Llama Term: cardiomyocyte\n",
            "  Matched GO Term: cardiac myofibril (GO:0097512)\n",
            "  Similarity: 0.70\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: TRbeta1 is present in cardiomyocytes.\n",
            "\n",
            "Gene: sarco/endoplasmic reticulum Ca2 -ATPase (53313)\n",
            "  Llama Term: calcium ion transmembrane transport\n",
            "  Matched GO Term: calcium ion transmembrane transport (GO:0070588)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: sarco/endoplasmic reticulum Ca2+-ATPase is affected by MuRF1, which impacts systolic function and contractility, suggesting a role in calcium transport.\n",
            "\n",
            "  Llama Term: sarcoplasmic reticulum\n",
            "  Matched GO Term: sarcoplasmic reticulum (GO:0016529)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0016528']\n",
            "  Namespace: cellular_component\n",
            "  Justification: The gene is named sarco/endoplasmic reticulum Ca2+-ATPase, indicating its location in the sarcoplasmic reticulum.\n",
            "\n",
            "Gene: desmin (13346)\n",
            "  Llama Term: cytoskeleton\n",
            "  Matched GO Term: cytoskeleton (GO:0005856)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: Desmin is co-localized with nuclear lamin beta1 and TRalpha1, suggesting its association with the cytoskeletal structure.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: Desmin co-localizes with nuclear proteins lamin beta1 and TRalpha1, indicating its presence in the nucleus.\n",
            "\n",
            "Gene: myc (17869)\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Myc-labeled MuRF1 interacted directly with TRalpha.\n",
            "\n",
            "Gene: TRbeta (56321)\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: TRbeta is a thyroid hormone receptor, which is known to be located in the nucleus.\n",
            "\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TRbeta interacts with SUMO ligase PIASx, which SUMOylates TRbeta in the presence of ligand.\n",
            "\n",
            "  Llama Term: regulation of transcription by RNA polymerase II\n",
            "  Matched GO Term: regulation of transcription by RNA polymerase II (GO:0006357)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0006366']\n",
            "  Namespace: biological_process\n",
            "  Justification: TRbeta is involved in T3-dependent gene regulation, closely matching TR activity and cofactor recruitment to the TRE in ChIP assays.\n",
            "\n",
            "Gene: CAP350 (74081)\n",
            "  Llama Term: nuclear receptor activity regulation\n",
            "  Matched GO Term: nuclear receptor activity (GO:0004879)\n",
            "  Similarity: 0.86\n",
            "  Relationship: ['part_of GO:0030522']\n",
            "  Namespace: molecular_function\n",
            "  Justification: CAP350 has been identified as a protein involved in regulating nuclear receptor activity, including PPAR.\n",
            "\n",
            "  Llama Term: centrosome\n",
            "  Matched GO Term: centrosome (GO:0005813)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['has_part GO:0005814']\n",
            "  Namespace: cellular_component\n",
            "  Justification: CAP350 is defined as a centrosome-associated protein.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: CAP350 has been found to localize to the nucleus and affect nuclear distribution of TRalpha.\n",
            "\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CAP350 interacts with TRalpha and MuRF1, indicating a role in protein binding.\n",
            "\n",
            "Gene: AF1 (19185)\n",
            "  Llama Term: transcription factor activity\n",
            "  Matched GO Term: DNA-binding transcription factor activity (GO:0003700)\n",
            "  Similarity: 0.91\n",
            "  Relationship: ['has_part GO:0000976', 'part_of GO:0006355']\n",
            "  Namespace: molecular_function\n",
            "  Justification: AF1 is part of the TRalpha A/B construct, which is associated with the N-terminus, a common location for transcription factor activity.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: As part of a transcription factor construct, AF1 is likely located in the nucleus.\n",
            "\n",
            "Gene: BNP (18158)\n",
            "  Llama Term: heart development\n",
            "  Matched GO Term: heart development (GO:0007507)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0072359']\n",
            "  Namespace: biological_process\n",
            "  Justification: BNP increase is associated with pathological hypertrophy.\n",
            "\n",
            "  Llama Term: cardiac muscle cell differentiation\n",
            "  Matched GO Term: cardiac muscle cell differentiation (GO:0055007)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0048738']\n",
            "  Namespace: biological_process\n",
            "  Justification: BNP is identified alongside alpha1-skeletal muscle actin, suggesting a role in cardiac cell differentiation.\n",
            "\n",
            "  Llama Term: secretion\n",
            "  Matched GO Term: secretion (GO:0046903)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: BNP increase is observed in circulating concentrations, implying secretion.\n",
            "\n",
            "Gene: ANF (230899)\n",
            "  Llama Term: heart development\n",
            "  Matched GO Term: heart development (GO:0007507)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0072359']\n",
            "  Namespace: biological_process\n",
            "  Justification: ANF is identified in the context of pathological hypertrophy, implying a role in heart development.\n",
            "\n",
            "  Llama Term: atrial cardiac muscle cell differentiation\n",
            "  Matched GO Term: atrial cardiac muscle cell differentiation (GO:0055011)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: ANF expression is associated with cardiac hypertrophy, suggesting involvement in atrial cardiac muscle cell differentiation.\n",
            "\n",
            "  Llama Term: secretion\n",
            "  Matched GO Term: secretion (GO:0046903)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: ANF is known to be secreted, as implied by its association with cardiac function and regulation.\n",
            "\n",
            "Gene: S5A and B (66998)\n",
            "Gene: MCT8 (20502)\n",
            "  Llama Term: thyroid hormone transmembrane transporter activity\n",
            "  Matched GO Term: thyroid hormone transmembrane transporter activity (GO:0015349)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0070327']\n",
            "  Namespace: molecular_function\n",
            "  Justification: MCT8 is involved in the uptake and efflux of T3.\n",
            "\n",
            "  Llama Term: thyroid hormone transport\n",
            "  Matched GO Term: thyroid hormone transport (GO:0070327)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MCT8 is involved in the uptake and efflux of T3.\n",
            "\n",
            "Gene: p38 (26416)\n",
            "  Llama Term: response to ischemia\n",
            "  Matched GO Term: response to ischemia (GO:0002931)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Activation of p38 impairs post-ischemic performance.\n",
            "\n",
            "  Llama Term: protein phosphorylation\n",
            "  Matched GO Term: protein phosphorylation (GO:0006468)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Activation of p38 increases phospholamban phosphorylation.\n",
            "\n",
            "Gene: androgen receptor (11835)\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: Androgen receptor is targeted to the proteasome in the absence of ligand and is also involved in nuclear translocation.\n",
            "\n",
            "  Llama Term: transcription regulator activity\n",
            "  Matched GO Term: transcription regulator activity (GO:0140110)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Proteasome activity is required for androgen receptor transcriptional activity.\n",
            "\n",
            "  Llama Term: prostate cancer cell differentiation\n",
            "  Matched GO Term: epithelial cell differentiation involved in prostate gland development (GO:0060742)\n",
            "  Similarity: 0.80\n",
            "  Relationship: ['part_of GO:0030850']\n",
            "  Namespace: biological_process\n",
            "  Justification: Androgen receptor transcriptional activity is regulated in prostate cancer cells.\n",
            "\n",
            "Gene: cJun (16476)\n",
            "  Llama Term: regulation of cell growth\n",
            "  Matched GO Term: regulation of cell growth (GO:0001558)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0016049']\n",
            "  Namespace: biological_process\n",
            "  Justification: cJun-mediated increases in IGF-1 signaling proteins are involved in cardiomyocyte growth.\n",
            "\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: MuRF1 preferentially recognizes phospho-cJun.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: As a transcription factor, cJun is expected to localize to the nucleus.\n",
            "\n",
            "Gene: MDM2 (17246)\n",
            "  Llama Term: ubiquitin-protein transferase activity\n",
            "  Matched GO Term: ubiquitin-protein transferase activity (GO:0004842)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: MDM2 mediates p53 mono-ubiquitination as a ubiquitin ligase.\n",
            "\n",
            "  Llama Term: protein ubiquitination\n",
            "  Matched GO Term: protein ubiquitination (GO:0016567)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: MDM2 mediates p53 mono-ubiquitination, altering its conformation and enhancing nuclear export.\n",
            "\n",
            "Gene: p53 (22060)\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: p53 mono-ubiquitination alters its conformation to preferentially bind CRM1.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: p53 nuclear export results in activity inhibition, implying it is normally located in the nucleus.\n",
            "\n",
            "  Llama Term: regulation of cellular process\n",
            "  Matched GO Term: regulation of cellular process (GO:0050794)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0009987']\n",
            "  Namespace: biological_process\n",
            "  Justification: The multiple levels of regulation by p53 ubiquitination imply p53 is involved in regulating cellular processes.\n",
            "\n",
            "Gene: CRM1 (103573)\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: CRM1 preferentially binds p53 after p53 mono-ubiquitination.\n",
            "\n",
            "  Llama Term: nuclear export\n",
            "  Matched GO Term: nuclear export (GO:0051168)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: CRM1 enhances p53 nuclear export resulting in activity inhibition.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: CRM1 is part of the nuclear export apparatus and interacts with p53 in the nucleus.\n",
            "\n",
            "Gene: estrogen receptor (13982)\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: Estrogen receptor (ER) is mono-ubiquitinated by the BRCA1 ubiquitin ligase, implying nuclear localization.\n",
            "\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Estrogen receptor (ER) interacts with BRCA1 ubiquitin ligase, indicating protein binding capability.\n",
            "\n",
            "  Llama Term: regulation of transcription by RNA polymerase II\n",
            "  Matched GO Term: regulation of transcription by RNA polymerase II (GO:0006357)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0006366']\n",
            "  Namespace: biological_process\n",
            "  Justification: Estrogen receptor (ER) is implied to have a role in transcriptional regulation, given its activity is inhibited by mono-ubiquitination.\n",
            "\n",
            "Gene: BRCA1 (12189)\n",
            "  Llama Term: ubiquitin-protein transferase activity\n",
            "  Matched GO Term: ubiquitin-protein transferase activity (GO:0004842)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: BRCA1 mono-ubiquitinates the estrogen receptor to inhibit its activity.\n",
            "\n",
            "  Llama Term: protein ubiquitination\n",
            "  Matched GO Term: protein ubiquitination (GO:0016567)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: BRCA1 acts as a ubiquitin ligase to mono-ubiquitinate the estrogen receptor.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: BRCA1 regulates acetylation and ubiquitination of estrogen receptor-alpha, implying nuclear localization.\n",
            "\n",
            "Gene: DNA polymerase Eta (80905)\n",
            "  Llama Term: DNA-directed DNA polymerase activity\n",
            "  Matched GO Term: DNA-directed DNA polymerase activity (GO:0003887)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Pirh2 E3 ubiquitin ligase monoubiquitinates DNA polymerase eta to suppress translesion DNA synthesis, indicating DNA polymerase activity.\n",
            "\n",
            "  Llama Term: translesion DNA synthesis\n",
            "  Matched GO Term: translesion synthesis (GO:0019985)\n",
            "  Similarity: 0.92\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Pirh2 E3 ubiquitin ligase monoubiquitinates DNA polymerase eta to suppress translesion DNA synthesis.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: The regulation of DNA polymerase Eta by nuclear proteins like p53 and Pirh2 suggests its nuclear localization.\n",
            "\n",
            "Gene: Pirh2 (68098)\n",
            "  Llama Term: ubiquitin-protein transferase activity\n",
            "  Matched GO Term: ubiquitin-protein transferase activity (GO:0004842)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Pirh2 acts as an E3 ubiquitin ligase to monoubiquitinate DNA polymerase eta.\n",
            "\n",
            "  Llama Term: negative regulation of translesion DNA synthesis\n",
            "  Matched GO Term: negative regulation of error-prone translesion synthesis (GO:1904332)\n",
            "  Similarity: 0.81\n",
            "  Relationship: ['negatively_regulates GO:0042276']\n",
            "  Namespace: biological_process\n",
            "  Justification: Pirh2 suppresses translesion DNA synthesis by monoubiquitinating DNA polymerase eta.\n",
            "\n",
            "  Llama Term: protein ubiquitination\n",
            "  Matched GO Term: protein ubiquitination (GO:0016567)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Pirh2 is involved in the ubiquitination of target proteins, such as DNA polymerase eta.\n",
            "\n",
            "Gene: TRalpha (6955)\n",
            "  Llama Term: nuclear transport\n",
            "  Matched GO Term: nuclear transport (GO:0051169)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: SUMO modification of TRalpha alters nuclear transport.\n",
            "\n",
            "  Llama Term: transcriptional regulation\n",
            "  Matched GO Term: regulation of gene expression (GO:0010468)\n",
            "  Similarity: 0.84\n",
            "  Relationship: ['regulates GO:0010467']\n",
            "  Namespace: biological_process\n",
            "  Justification: SUMO modification of TRalpha alters transcriptional regulation.\n",
            "\n",
            "  Llama Term: apoptosis\n",
            "  Matched GO Term: apoptotic process (GO:0006915)\n",
            "  Similarity: 0.97\n",
            "  Relationship: ['has_part GO:0097190', 'has_part GO:0097194']\n",
            "  Namespace: biological_process\n",
            "  Justification: SUMO modification of TRalpha alters apoptosis.\n",
            "\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: TRalpha interacts with the SUMO ligase PIAS1.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: TRalpha is SUMOylated in the nucleus by PIAS1.\n",
            "\n",
            "Gene: PIAS1 (300772)\n",
            "  Llama Term: SUMO ligase activity\n",
            "  Matched GO Term: SUMO ligase activity (GO:0061665)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: PIAS1 is a SUMO ligase that mediates SUMOylation of TRalpha.\n",
            "\n",
            "  Llama Term: protein sumoylation\n",
            "  Matched GO Term: protein sumoylation (GO:0016925)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: PIAS1 SUMOylates TRalpha, indicating involvement in protein sumoylation.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: PIAS1 mediates SUMOylation of nuclear receptors like TRalpha, suggesting nuclear localization.\n",
            "\n",
            "Gene: SUMO1 (22218)\n",
            "  Llama Term: protein sumoylation\n",
            "  Matched GO Term: protein sumoylation (GO:0016925)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: SUMO1 is involved in SUMOylation of proteins such as TRalpha and TRbeta.\n",
            "\n",
            "  Llama Term: regulation of transcription by RNA polymerase II\n",
            "  Matched GO Term: regulation of transcription by RNA polymerase II (GO:0006357)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0006366']\n",
            "  Namespace: biological_process\n",
            "  Justification: SUMO1 modification is reported in T3-dependent gene regulation, matching TR activity and cofactor recruitment to the TRE.\n",
            "\n",
            "Gene: SUMO3 (20610)\n",
            "  Llama Term: protein sumoylation\n",
            "  Matched GO Term: protein sumoylation (GO:0016925)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: SUMO3 is involved in SUMO modification, a type of post-translational modification.\n",
            "\n",
            "  Llama Term: transcription regulation by TR\n",
            "  Matched GO Term: transcription regulator activity (GO:0140110)\n",
            "  Similarity: 0.76\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: SUMO3 has been reported in T3-dependent gene regulation, closely matching TR activity.\n",
            "\n",
            "Gene: TR (2149)\n",
            "  Llama Term: thyroid hormone receptor activity\n",
            "  Matched GO Term: nuclear receptor activity (GO:0004879)\n",
            "  Similarity: 0.83\n",
            "  Relationship: ['part_of GO:0030522']\n",
            "  Namespace: molecular_function\n",
            "  Justification: TH induces molecular changes, implying TR is involved in thyroid hormone receptor activity.\n",
            "\n",
            "  Llama Term: heart process\n",
            "  Matched GO Term: heart process (GO:0003015)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: TR expression is increased in the myocardium, suggesting involvement in heart processes.\n",
            "\n",
            "  Llama Term: nucleus\n",
            "  Matched GO Term: nucleus (GO:0005634)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: As a receptor, TR is expected to be located in the nucleus to regulate gene expression.\n",
            "\n",
            "Gene: org/10 (100035352)\n",
            "Gene: PKC{epsilon (18754)\n",
            "  Llama Term: protein kinase C activity\n",
            "  Matched GO Term: activation of protein kinase C activity (GO:1990051)\n",
            "  Similarity: 0.88\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: PKC{epsilon} is implied to have protein kinase activity as it is referred to as PKC.\n",
            "\n",
            "  Llama Term: cardiomyocyte hypertrophy\n",
            "  Matched GO Term: cardiac muscle hypertrophy (GO:0003300)\n",
            "  Similarity: 0.91\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: PKC{epsilon} activation is associated with cardiomyocyte hypertrophy, which is prevented by Muscle ring finger protein-1.\n",
            "\n",
            "Gene: cardiac phospholamban (18821)\n",
            "  Llama Term: regulation of cardiac muscle contraction\n",
            "  Matched GO Term: regulation of cardiac muscle contraction (GO:0055117)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0060048']\n",
            "  Namespace: biological_process\n",
            "  Justification: Cardiac phospholamban is implied to be involved in cardiac muscle contraction through its regulation.\n",
            "\n",
            "  Llama Term: heart\n",
            "  Matched GO Term: heart process (GO:0003015)\n",
            "  Similarity: 0.81\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: The term 'cardiac' in cardiac phospholamban implies expression in the heart.\n",
            "\n",
            "Gene: KCND3 (56543)\n",
            "  Llama Term: voltage-gated potassium channel activity\n",
            "  Matched GO Term: voltage-gated potassium channel activity (GO:0005249)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: KCND3 is a potassium channel\n",
            "\n",
            "  Llama Term: regulation of ion transport\n",
            "  Matched GO Term: regulation of transmembrane transport (GO:0034762)\n",
            "  Similarity: 0.82\n",
            "  Relationship: ['regulates GO:0055085']\n",
            "  Namespace: biological_process\n",
            "  Justification: KCND3 is regulated by thyroid hormone receptors\n",
            "\n",
            "  Llama Term: plasma membrane\n",
            "  Matched GO Term: plasma membrane (GO:0005886)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0071944']\n",
            "  Namespace: cellular_component\n",
            "  Justification: KCND3 is a potassium channel, which is typically located in the plasma membrane\n",
            "\n",
            "Gene: alpha1 and beta (12044)\n",
            "  Llama Term: thyroid hormone receptor activity\n",
            "  Matched GO Term: nuclear receptor activity (GO:0004879)\n",
            "  Similarity: 0.83\n",
            "  Relationship: ['part_of GO:0030522']\n",
            "  Namespace: molecular_function\n",
            "  Justification: alpha1 and beta1 are thyroid hormone receptors\n",
            "\n",
            "  Llama Term: regulation of potassium channel activity\n",
            "  Matched GO Term: potassium channel regulator activity (GO:0015459)\n",
            "  Similarity: 0.84\n",
            "  Relationship: ['regulates GO:0005267']\n",
            "  Namespace: molecular_function\n",
            "  Justification: alpha1 and beta1 regulate cardiac KCND3 potassium channel expression\n",
            "\n",
            "Gene: FGFR1 (14182)\n",
            "  Llama Term: receptor tyrosine kinase activity\n",
            "  Matched GO Term: transmembrane receptor protein tyrosine kinase activity (GO:0004714)\n",
            "  Similarity: 0.89\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: FGFR1 is a fibroblast growth factor receptor with tyrosine kinase activity.\n",
            "\n",
            "  Llama Term: centrosome\n",
            "  Matched GO Term: centrosome (GO:0005813)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['has_part GO:0005814']\n",
            "  Namespace: cellular_component\n",
            "  Justification: FOP-FGFR1 fusion kinase recruits phosphoinositide-3 kinase and phospholipase Cgamma at the centrosome.\n",
            "\n",
            "Gene: Atrogin-1 (67731)\n",
            "  Llama Term: ubiquitin-protein transferase activity\n",
            "  Matched GO Term: ubiquitin-protein transferase activity (GO:0004842)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: Atrogin-1 participates in an SCF ubiquitin ligase complex.\n",
            "\n",
            "  Llama Term: negative regulation of cardiac hypertrophy\n",
            "  Matched GO Term: negative regulation of cardiac muscle hypertrophy (GO:0010614)\n",
            "  Similarity: 0.95\n",
            "  Relationship: ['negatively_regulates GO:0003300']\n",
            "  Namespace: biological_process\n",
            "  Justification: Atrogin-1 inhibits calcineurin-dependent cardiac hypertrophy.\n",
            "\n",
            "  Llama Term: SCF ubiquitin ligase complex\n",
            "  Matched GO Term: SCF ubiquitin ligase complex (GO:0019005)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: Atrogin-1 participates in an SCF ubiquitin ligase complex.\n",
            "\n",
            "Gene: Rad52 (19365)\n",
            "  Llama Term: DNA repair\n",
            "  Matched GO Term: DNA repair (GO:0006281)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Rad52 is found at a repair centre for DNA double-strand breaks.\n",
            "\n",
            "  Llama Term: double-strand break repair\n",
            "  Matched GO Term: double-strand break repair (GO:0006302)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Rad52 is involved in the repair of DNA double-strand breaks.\n",
            "\n",
            "Gene: KCNE1 (16509)\n",
            "  Llama Term: voltage-gated potassium channel activity\n",
            "  Matched GO Term: voltage-gated potassium channel activity (GO:0005249)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: KCNE1 is an ion channel subunit.\n",
            "\n",
            "  Llama Term: regulation of action potential duration\n",
            "  Matched GO Term: regulation of action potential (GO:0098900)\n",
            "  Similarity: 0.81\n",
            "  Relationship: ['regulates GO:0001508']\n",
            "  Namespace: biological_process\n",
            "  Justification: Thyroid hormone receptor alpha controls action potential duration through KCNE1.\n",
            "\n",
            "  Llama Term: plasma membrane\n",
            "  Matched GO Term: plasma membrane (GO:0005886)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0071944']\n",
            "  Namespace: cellular_component\n",
            "  Justification: As an ion channel subunit, KCNE1 is expected to be located in the plasma membrane.\n",
            "\n",
            "Gene: ERK (26413)\n",
            "  Llama Term: signal transduction\n",
            "  Matched GO Term: signal transduction (GO:0007165)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0007154', 'part_of GO:0023052', 'part_of GO:0051716']\n",
            "  Namespace: biological_process\n",
            "  Justification: ERK is involved in signaling pathway.\n",
            "\n",
            "  Llama Term: cardiac muscle cell differentiation\n",
            "  Matched GO Term: cardiac muscle cell differentiation (GO:0055007)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0048738']\n",
            "  Namespace: biological_process\n",
            "  Justification: Thyroid hormone changes cardiomyocyte shape and geometry via ERK signaling pathway.\n",
            "\n",
            "  Llama Term: thyroid hormone signaling pathway\n",
            "  Matched GO Term: thyroid hormone receptor signaling pathway (GO:0002154)\n",
            "  Similarity: 0.95\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Dose-dependent effects of thyroid hormone on post-ischemic cardiac performance: Potential involvement of Akt and ERK signalings\n",
            "\n",
            "Gene: IGF1 (3479)\n",
            "  Llama Term: growth factor activity\n",
            "  Matched GO Term: growth factor activity (GO:0008083)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: IGF1 is mentioned in relation to the IGF1/IGFR system, implying its role as a growth factor.\n",
            "\n",
            "  Llama Term: heart development\n",
            "  Matched GO Term: heart development (GO:0007507)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0072359']\n",
            "  Namespace: biological_process\n",
            "  Justification: IGF1 is mentioned in the context of possible relations in the human heart, suggesting involvement in heart development.\n",
            "\n",
            "Gene: IGFR (3480)\n",
            "  Llama Term: receptor activity\n",
            "  Matched GO Term: peptide receptor activity (GO:0001653)\n",
            "  Similarity: 0.81\n",
            "  Relationship: ['has_part GO:0042277']\n",
            "  Namespace: molecular_function\n",
            "  Justification: IGFR is mentioned in relation to the IGF1/IGFR system, implying it has receptor activity.\n",
            "\n",
            "Gene: PKCalpha (18750)\n",
            "  Llama Term: protein serine/threonine kinase activity\n",
            "  Matched GO Term: protein serine/threonine kinase activity (GO:0004674)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: PKCalpha activates p120-catenin by phosphorylating serine 879.\n",
            "\n",
            "  Llama Term: cell-cell adhesion\n",
            "  Matched GO Term: cell-cell adhesion (GO:0098609)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: PKCalpha activation disrupts vascular integrity by disassembling VE-cadherin junctions.\n",
            "\n",
            "  Llama Term: cytoplasm\n",
            "  Matched GO Term: cytoplasm (GO:0005737)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['part_of GO:0005622']\n",
            "  Namespace: cellular_component\n",
            "  Justification: PKCalpha is implied to be in the cytoplasm where it can interact with and phosphorylate p120-catenin.\n",
            "\n",
            "Gene: p120-catenin (12388)\n",
            "  Llama Term: cell junction disassembly\n",
            "  Matched GO Term: cell junction disassembly (GO:0150146)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: PKCalpha activation of p120-catenin disassembles VE-cadherin junctions\n",
            "\n",
            "  Llama Term: cell-cell junction\n",
            "  Matched GO Term: cell-cell junction (GO:0005911)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: p120-catenin is involved in the disassembly of VE-cadherin junctions\n",
            "\n",
            "  Llama Term: protein binding\n",
            "  Matched GO Term: protein binding (GO:0005515)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: p120-catenin interacts with VE-cadherin\n",
            "\n",
            "Gene: VE-cadherin (12562)\n",
            "  Llama Term: cell-cell adhesion\n",
            "  Matched GO Term: cell-cell adhesion (GO:0098609)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: VE-cadherin junctions are disassembled by PKCalpha activation, disrupting vascular integrity, indicating involvement in cell-cell adhesion.\n",
            "\n",
            "  Llama Term: cell junction\n",
            "  Matched GO Term: cell junction (GO:0030054)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: cellular_component\n",
            "  Justification: VE-cadherin forms junctions that are disassembled by PKCalpha activation, indicating location at cell junctions.\n",
            "\n",
            "Gene: RING Finger-1 (66949)\n",
            "  Llama Term: inhibition of JNK signaling\n",
            "  Matched GO Term: negative regulation of JNK cascade (GO:0046329)\n",
            "  Similarity: 0.93\n",
            "  Relationship: ['negatively_regulates GO:0007254']\n",
            "  Namespace: biological_process\n",
            "  Justification: RING Finger-1 Attenuates IGF-1-dependent cardiomyocyte hypertrophy by inhibiting JNK signaling\n",
            "\n",
            "  Llama Term: cardiac atrophy\n",
            "  Matched GO Term: cardiac muscle atrophy (GO:0014899)\n",
            "  Similarity: 0.94\n",
            "  Relationship: unknown\n",
            "  Namespace: biological_process\n",
            "  Justification: Muscle ring finger 1 mediates cardiac atrophy in vivo\n",
            "\n",
            "Gene: JNK (26419)\n",
            "  Llama Term: protein kinase activity\n",
            "  Matched GO Term: protein kinase activity (GO:0004672)\n",
            "  Similarity: 1.00\n",
            "  Relationship: unknown\n",
            "  Namespace: molecular_function\n",
            "  Justification: JNK is involved in signaling, implying it has protein kinase activity.\n",
            "\n",
            "  Llama Term: regulation of cardiomyocyte hypertrophy\n",
            "  Matched GO Term: regulation of cardiac muscle hypertrophy (GO:0010611)\n",
            "  Similarity: 0.95\n",
            "  Relationship: ['regulates GO:0003300']\n",
            "  Namespace: biological_process\n",
            "  Justification: JNK signaling is inhibited to attenuate IGF-1-dependent cardiomyocyte hypertrophy.\n",
            "\n",
            "Gene: E2F1 (13555)\n",
            "  Llama Term: regulation of gene expression\n",
            "  Matched GO Term: regulation of gene expression (GO:0010468)\n",
            "  Similarity: 1.00\n",
            "  Relationship: ['regulates GO:0010467']\n",
            "  Namespace: biological_process\n",
            "  Justification: E2F1-mediated gene expression is regulated by muscle ring finger 1 and muscle ring finger 2.\n",
            "\n"
          ]
        }
      ],
      "source": [
        "def auto_match_go_terms(llama_data, go_ids, go_names, go_embeddings, go_namespaces, go_relationships, model, threshold=0.6):\n",
        "    \"\"\"\n",
        "    Given parsed Llama JSON data, match each Llama-generated GO term to the best official GO term.\n",
        "    \"\"\"\n",
        "    matched_go_terms = {}\n",
        "\n",
        "    # If the JSON is nested under \"genes\", extract that list.\n",
        "    if isinstance(llama_data, list) and len(llama_data) > 0 and \"genes\" in llama_data[0]:\n",
        "        gene_entries = llama_data[0][\"genes\"]\n",
        "    else:\n",
        "        gene_entries = llama_data\n",
        "\n",
        "    for gene_entry in gene_entries:\n",
        "        gene_name = gene_entry.get(\"gene\", \"UnknownGene\")\n",
        "        gene_id = gene_entry.get(\"geneID\", \"UnknownID\")     # using id instead of name for the dictionary since some genes have the same name\n",
        "        go_data = gene_entry.get(\"go_data\", [])\n",
        "        unique_matches = {}\n",
        "\n",
        "        for entry in go_data:\n",
        "            llama_term = entry.get(\"term\")\n",
        "            if not llama_term:\n",
        "                continue\n",
        "\n",
        "            best_go_id, best_go_name, best_score, best_idx = get_best_matching_go_term(\n",
        "                llama_term, go_ids, go_names, go_embeddings, model, threshold=threshold\n",
        "            )\n",
        "\n",
        "            if best_go_id and best_idx is not None:\n",
        "                official_relationship = go_relationships[best_idx]\n",
        "                official_namespace = go_namespaces[best_idx]\n",
        "\n",
        "                if best_go_id in unique_matches:\n",
        "                    if best_score > unique_matches[best_go_id][\"similarity\"]:\n",
        "                        unique_matches[best_go_id] = {\n",
        "                            \"llama_term\": llama_term,\n",
        "                            \"matched_go_id\": best_go_id,\n",
        "                            \"matched_go_name\": best_go_name,\n",
        "                            \"similarity\": best_score,\n",
        "                            \"relationship\": official_relationship,\n",
        "                            \"namespace\": official_namespace,\n",
        "                            \"justification\": entry.get(\"justification\")\n",
        "                        }\n",
        "                else:\n",
        "                    unique_matches[best_go_id] = {\n",
        "                        \"llama_term\": llama_term,\n",
        "                        \"matched_go_id\": best_go_id,\n",
        "                        \"matched_go_name\": best_go_name,\n",
        "                        \"similarity\": best_score,\n",
        "                        \"relationship\": official_relationship,\n",
        "                        \"namespace\": official_namespace,\n",
        "                        \"justification\": entry.get(\"justification\")\n",
        "                    }\n",
        "\n",
        "        matched_go_terms[gene_id] = list(unique_matches.values())\n",
        "\n",
        "    return matched_go_terms\n",
        "\n",
        "def replace_terms_with_go(llama_data, matched_go_terms, similarity_threshold=0.65):\n",
        "    \"\"\"\n",
        "    Replaces each llama_data entry's GO term with the matched GO term info from matched_go_terms.\n",
        "    \"\"\"\n",
        "    # If llama_data is nested under \"genes\", re-point gene_entries\n",
        "    if isinstance(llama_data, list) and len(llama_data) > 0 and \"genes\" in llama_data[0]:\n",
        "        gene_entries = llama_data[0][\"genes\"]\n",
        "    else:\n",
        "        gene_entries = llama_data\n",
        "\n",
        "    for gene_entry in gene_entries:\n",
        "        gene_id = gene_entry.get(\"geneID\")\n",
        "        if gene_id in matched_go_terms:\n",
        "            new_go_data = []\n",
        "            for match in matched_go_terms[gene_id]:\n",
        "                # Skip if the match similarity is below threshold\n",
        "                if match[\"similarity\"] < similarity_threshold:\n",
        "                    continue\n",
        "                new_entry = {\n",
        "                    \"term\": match[\"matched_go_name\"],      # replace Llama term\n",
        "                    \"go_id\": match[\"matched_go_id\"],       # new field\n",
        "                    \"relationship\": match[\"relationship\"], # official relationship\n",
        "                    \"namespace\": match[\"namespace\"],\n",
        "                    \"justification\": match[\"justification\"],\n",
        "                    # carry over references if you want:\n",
        "                    \"reference(s)\": gene_entry[\"go_data\"][0].get(\"reference(s)\", [])\n",
        "                }\n",
        "                new_go_data.append(new_entry)\n",
        "            gene_entry[\"go_data\"] = new_go_data\n",
        "    return llama_data\n",
        "\n",
        "# Create a  directory to store PROCESSED llama JSON outputs.\n",
        "processed_dir = os.path.join(os.getcwd(), \"processed_inferences\")\n",
        "if not os.path.exists(processed_dir):\n",
        "    os.makedirs(processed_dir)\n",
        "\n",
        "# Usage example:\n",
        "# 1. Load the Llama output JSON file and parse it\n",
        "llama_output_path = \"./llama_inferences/llama_output_PMC5453669.json\"\n",
        "llama_data = parse_llama_output(llama_output_path)\n",
        "\n",
        "# 2. Match the Llama terms to official GO terms\n",
        "matched_terms = auto_match_go_terms(\n",
        "    llama_data, go_ids, go_names, go_embeddings, go_namespaces, go_relationships, sapbert_model, threshold=0.6\n",
        ")\n",
        "\n",
        "# 3. Replace llama terms with the official GO terms\n",
        "updated_data = replace_terms_with_go(llama_data, matched_terms)\n",
        "\n",
        "# 4. save the processed JSON output to a new file\n",
        "procesed_output_path = os.path.join(processed_dir, f\"processed_output_{doc_id}.json\")\n",
        "with open(procesed_output_path, \"w\", encoding=\"utf-8\") as f:\n",
        "    json.dump(updated_data, f, indent=2)\n",
        "\n",
        "print(\"there are {} genes with matched GO terms.\".format(len(matched_terms)))\n",
        "for gene_id, matches in matched_terms.items():\n",
        "    print(f\"Gene: {gene_dict[gene_id].get_name_from_article()} ({gene_id})\")\n",
        "    for match in matches:\n",
        "        print(f\"  Llama Term: {match['llama_term']}\")\n",
        "        print(f\"  Matched GO Term: {match['matched_go_name']} ({match['matched_go_id']})\")\n",
        "        print(f\"  Similarity: {match['similarity']:.2f}\")\n",
        "        print(f\"  Relationship: {match['relationship']}\")\n",
        "        print(f\"  Namespace: {match['namespace']}\")\n",
        "        print(f\"  Justification: {match['justification']}\\n\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "52hJV2NbJMzp"
      },
      "source": [
        "# 6. Evaluate"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6DBtiTOtJgFO"
      },
      "source": [
        "6.1. Counting and Plotting Similarity Distributions"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 487
        },
        "id": "sPcTb_-UJLvq",
        "outputId": "4754b2f1-b823-432a-f485-1a874be690ee"
      },
      "outputs": [],
      "source": [
        "import matplotlib.pyplot as plt\n",
        "from collections import Counter\n",
        "\n",
        "# Suppose matched_terms is your dictionary of gene predictions (from auto_match_go_terms)\n",
        "# where each value is a list of dictionaries with a key 'similarity'\n",
        "similarity_scores = []\n",
        "for gene, matches in matched_terms.items():\n",
        "    for match in matches:\n",
        "        similarity_scores.append(match['similarity'])\n",
        "\n",
        "# Bin the similarity scores, e.g., rounded to one decimal place.\n",
        "binned_scores = [round(score, 1) for score in similarity_scores]\n",
        "score_counts = Counter(binned_scores)\n",
        "\n",
        "# Prepare data for the bar plot.\n",
        "bins = sorted(score_counts.keys())\n",
        "counts = [score_counts[b] for b in bins]\n",
        "\n",
        "plt.figure(figsize=(8, 5))\n",
        "plt.bar(bins, counts, width=0.08, edgecolor='black')\n",
        "plt.xlabel('Cosine Similarity')\n",
        "plt.ylabel('Count')\n",
        "plt.title('Distribution of GO Term Similarity Scores')\n",
        "plt.xticks(bins)\n",
        "plt.show()\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ONKEXAOtZtjO"
      },
      "source": [
        "## 6.2. Scoring against NCBI Gene database\n",
        "\n",
        "Need to make a new scoring system: possibly True Positive (what we found), Nobel Positive (NP), Not Detected\n",
        "\n",
        "- for cases where we have Go terms that are alike like \"Wnt signaling pathway\" and \"canonical Wnt signaling pathway\", we can embed the terms and check if they are similar. If they are, then we can consider them as the same term. If not, then we can consider them as different terms."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# term = \"Cul4A-RING E3 ubiquitin ligase complex\"\n",
        "# example_llama_term = \"ubiquitin ligase complex\"\n",
        "# def compute_semantic_similarity(term1, term2, model):\n",
        "#     \"\"\"\n",
        "#     Returns a cosine similarity score between two text strings, using\n",
        "#     a SentenceTransformer model.\n",
        "#     \"\"\"\n",
        "#     emb1 = model.encode(term1, convert_to_tensor=True)\n",
        "#     emb2 = model.encode(term2, convert_to_tensor=True)\n",
        "#     cos_sim = util.cos_sim(emb1, emb2)\n",
        "#     return float(cos_sim[0][0])  # scalar\n",
        "\n",
        "# # Compute the semantic similarity between the Llama term and the official GO term\n",
        "# res = compute_semantic_similarity(example_llama_term, term, sapbert_model)\n",
        "# print(f\"Semantic similarity between '{example_llama_term}' and '{term}': {res:.2f}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The code below takes a bit of time due to:\n",
        "- api calls to retrieve the gene ontology terms from NCBI Gene database for each gene"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "=== Checking Gene: MuRF1 (ID=433766) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "LocalTerm='protein ubiquitination' | BestNCBI='protein monoubiquitination' | Score=0.87 | PMIDs=['26862156']\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='negative regulation of transcription by RNA polymerase II' | Score=0.3723091185092926| PMIDs=['26862156']\n",
            "LocalTerm='regulation of fatty acid oxidation' | Score=0.4209190011024475| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='ubiquitin-protein transferase activity' | BestNCBI='ubiquitin-protein transferase activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='negative regulation of cardiac muscle hypertrophy' | BestNCBI='negative regulation of cardiac muscle hypertrophy' (diff. references) | Score=0.9999999403953552 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for MuRF1 (433766) ==\n",
            "TP = 1, NP = 2, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TRalpha (ID=14685) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='nuclear receptor activity' | Score=0.5340927243232727| PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | Score=0.6704468131065369| PMIDs=['26862156']\n",
            "LocalTerm='regulation of cardiac muscle hypertrophy' | Score=0.38464248180389404| PMIDs=['26862156']\n",
            "LocalTerm='protein ubiquitination' | Score=0.3942946791648865| PMIDs=['26862156']\n",
            "LocalTerm='ligand-modulated transcription activator activity' | Score=0.5715138912200928| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TRalpha (14685) ==\n",
            "TP = 0, NP = 5, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Akt (ID=11651) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='cell death' | Score=0.6167064905166626| PMIDs=['26862156']\n",
            "LocalTerm='heart process' | Score=0.4735707640647888| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='intracellular signal transduction' | BestNCBI='intracellular signal transduction' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='protein kinase activity' | BestNCBI='protein kinase activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Akt (11651) ==\n",
            "TP = 0, NP = 2, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: mTOR (ID=56717) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cellular response to hormone stimulus' | BestNCBI='cellular response to insulin stimulus' (diff. references) | Score=0.7497140169143677 | PMIDs=['26862156']\n",
            "LocalTerm='regulation of cardiac muscle tissue growth' | BestNCBI='positive regulation of cell growth involved in cardiac muscle cell development' (diff. references) | Score=0.7121151089668274 | PMIDs=['26862156']\n",
            "LocalTerm='protein kinase binding' | BestNCBI='protein kinase binding' (diff. references) | Score=1.0000001192092896 | PMIDs=['26862156']\n",
            "LocalTerm='cell growth' | BestNCBI='positive regulation of cell growth' (diff. references) | Score=0.7911554574966431 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for mTOR (56717) ==\n",
            "TP = 0, NP = 0, DCU = 4, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TH (ID=21823) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='hormone activity' | Score=0.45052847266197205| PMIDs=['26862156']\n",
            "LocalTerm='extracellular matrix organization' | Score=0.5932707190513611| PMIDs=['26862156']\n",
            "LocalTerm='nuclear thyroid hormone receptor binding' | Score=0.5986411571502686| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='metabolic process' | BestNCBI='fatty acid metabolic process' (diff. references) | Score=0.7032467126846313 | PMIDs=['26862156']\n",
            "LocalTerm='heart development' | BestNCBI='heart development' (diff. references) | Score=0.9999999403953552 | PMIDs=['26862156']\n",
            "LocalTerm='cardiac muscle cell differentiation' | BestNCBI='heart development' (diff. references) | Score=0.7718674540519714 | PMIDs=['26862156']\n",
            "LocalTerm='regulation of cardiac muscle contraction' | BestNCBI='regulation of heart contraction' (diff. references) | Score=0.9631195068359375 | PMIDs=['26862156']\n",
            "LocalTerm='cellular response to hormone stimulus' | BestNCBI='cellular response to growth factor stimulus' (diff. references) | Score=0.763095498085022 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TH (21823) ==\n",
            "TP = 0, NP = 3, DCU = 5, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: IGF-1 (ID=16000) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='regulation of cardiac muscle hypertrophy' | BestNCBI='positive regulation of cardiac muscle hypertrophy' (diff. references) | Score=0.8371738195419312 | PMIDs=['26862156']\n",
            "LocalTerm='positive regulation of cardiac muscle hypertrophy' | BestNCBI='positive regulation of cardiac muscle hypertrophy' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='cell growth' | BestNCBI='positive regulation of cell growth' (diff. references) | Score=0.7911554574966431 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for IGF-1 (16000) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: RARgamma (ID=19411) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='protein heterodimerization activity' | Score=0.6373044848442078| PMIDs=['26862156']\n",
            "LocalTerm='proteasome-mediated ubiquitin-dependent protein catabolic process' | Score=0.46197545528411865| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for RARgamma (19411) ==\n",
            "TP = 0, NP = 2, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: PPARalpha (ID=19013) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='nuclear export' | Score=0.41224461793899536| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='fatty acid oxidation' | BestNCBI='fatty acid metabolic process' (diff. references) | Score=0.7996388077735901 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for PPARalpha (19013) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: PPARbeta/delta (ID=19015) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='fatty acid oxidation' | BestNCBI='fatty acid oxidation' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for PPARbeta/delta (19015) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: MuRF2 (ID=381485) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='negative regulation of fatty acid oxidation' | Score=0.20995573699474335| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='ubiquitin-protein transferase activity' | BestNCBI='ubiquitin protein ligase activity' (diff. references) | Score=0.818665623664856 | PMIDs=['26862156']\n",
            "LocalTerm='protein ubiquitination' | BestNCBI='protein K48-linked ubiquitination' (diff. references) | Score=0.7497677803039551 | PMIDs=['26862156']\n",
            "LocalTerm='cytoplasm' | BestNCBI='cytoplasm' (diff. references) | Score=0.9999999403953552 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for MuRF2 (381485) ==\n",
            "TP = 0, NP = 1, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: MuRF3 (ID=58522) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='protein ubiquitination' | Score=0.3293459713459015| PMIDs=['26862156']\n",
            "LocalTerm='fatty acid oxidation' | Score=0.28796064853668213| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='ubiquitin-protein transferase activity' | BestNCBI='ubiquitin protein ligase activity' (diff. references) | Score=0.818665623664856 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for MuRF3 (58522) ==\n",
            "TP = 0, NP = 2, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: ubiquitin ligase (ID=84585) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='nucleus' | Score=0.6517293453216553| PMIDs=['26862156']\n",
            "LocalTerm='regulation of fatty acid oxidation' | Score=0.30507951974868774| PMIDs=['26862156']\n",
            "LocalTerm='negative regulation of transcription by RNA polymerase II' | Score=0.24362178146839142| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='ubiquitin protein ligase activity' | BestNCBI='ubiquitin protein ligase activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='protein monoubiquitination' | BestNCBI='protein polyubiquitination' (diff. references) | Score=0.8478775024414062 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for ubiquitin ligase (84585) ==\n",
            "TP = 0, NP = 3, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: MuRF1 (ID=140939) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='negative regulation of cardiac muscle hypertrophy' | BestNCBI='negative regulation of cardiac muscle hypertrophy' (diff. references) | Score=0.9999999403953552 | PMIDs=['26862156']\n",
            "LocalTerm='contractile muscle fiber' | BestNCBI='contractile muscle fiber' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for MuRF1 (140939) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CSA (ID=71991) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='cardiac muscle hypertrophy' | Score=0.3114530146121979| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CSA (71991) ==\n",
            "TP = 0, NP = 1, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TH (ID=25085) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='cardiac muscle tissue regeneration' | Score=0.619160532951355| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cardiac muscle cell differentiation' | BestNCBI='heart development' (diff. references) | Score=0.7718674540519714 | PMIDs=['26862156']\n",
            "LocalTerm='atrial cardiac muscle cell development' | BestNCBI='heart development' (diff. references) | Score=0.7348788976669312 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TH (25085) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TRalpha1 (ID=104218) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='nuclear receptor activity' | Score=-1.0| PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | Score=-1.0| PMIDs=['26862156']\n",
            "LocalTerm='regulation of cardiac muscle hypertrophy' | Score=-1.0| PMIDs=['26862156']\n",
            "LocalTerm='response to ischemia' | Score=-1.0| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TRalpha1 (104218) ==\n",
            "TP = 0, NP = 4, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TRalpha2 (ID=21833) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='heart development' | Score=0.6544169187545776| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='nuclear receptor activity' | BestNCBI='nuclear receptor activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TRalpha2 (21833) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TRbeta1 (ID=211770) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='nuclear receptor activity' | Score=0.4600355625152588| PMIDs=['26862156']\n",
            "LocalTerm='cardiac myofibril' | Score=0.34791868925094604| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TRbeta1 (211770) ==\n",
            "TP = 0, NP = 2, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: sarco/endoplasmic reticulum Ca2+-ATPase (ID=53313) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='calcium ion transmembrane transport' | BestNCBI='calcium ion transmembrane transport' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='sarcoplasmic reticulum' | BestNCBI='sarcoplasmic reticulum' (diff. references) | Score=1.0000001192092896 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for sarco/endoplasmic reticulum Ca2+-ATPase (53313) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: desmin (ID=13346) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cytoskeleton' | BestNCBI='cytoskeleton' (diff. references) | Score=1.0000001192092896 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for desmin (13346) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: myc (ID=17869) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='protein binding' | BestNCBI='protein binding' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for myc (17869) ==\n",
            "TP = 0, NP = 0, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TRbeta (ID=56321) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of transcription by RNA polymerase II' | Score=0.49213045835494995| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='protein binding' | BestNCBI='protein binding' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TRbeta (56321) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CAP350 (ID=74081) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='nuclear receptor activity' | Score=0.5340927243232727| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='centrosome' | BestNCBI='centrosome' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleoplasm' (diff. references) | Score=0.7647878527641296 | PMIDs=['26862156']\n",
            "LocalTerm='protein binding' | BestNCBI='protein binding' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CAP350 (74081) ==\n",
            "TP = 0, NP = 1, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: AF1 (ID=19185) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='DNA-binding transcription factor activity' | Score=0.5334563851356506| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for AF1 (19185) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: BNP (ID=18158) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='secretion' | Score=0.5080999732017517| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='heart development' | BestNCBI='coronary vasculature development' (diff. references) | Score=0.7823546528816223 | PMIDs=['26862156']\n",
            "LocalTerm='cardiac muscle cell differentiation' | BestNCBI='coronary vasculature development' (diff. references) | Score=0.6868568658828735 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for BNP (18158) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: ANF (ID=230899) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='heart development' | Score=0.43331411480903625| PMIDs=['26862156']\n",
            "LocalTerm='atrial cardiac muscle cell differentiation' | Score=0.48001518845558167| PMIDs=['26862156']\n",
            "LocalTerm='secretion' | Score=0.5080999732017517| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for ANF (230899) ==\n",
            "TP = 0, NP = 3, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: S5A and B (ID=66998) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for S5A and B (66998) ==\n",
            "TP = 0, NP = 0, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: MCT8 (ID=20502) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='thyroid hormone transmembrane transporter activity' | BestNCBI='thyroid hormone transmembrane transporter activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='thyroid hormone transport' | BestNCBI='thyroid hormone transport' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for MCT8 (20502) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: p38 (ID=26416) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='response to ischemia' | BestNCBI='response to insulin' (diff. references) | Score=0.7278741598129272 | PMIDs=['26862156']\n",
            "LocalTerm='protein phosphorylation' | BestNCBI='protein phosphorylation' (diff. references) | Score=1.0000001192092896 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for p38 (26416) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: androgen receptor (ID=11835) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='transcription regulator activity' | BestNCBI='DNA-binding transcription factor activity' (diff. references) | Score=0.7397277355194092 | PMIDs=['26862156']\n",
            "LocalTerm='epithelial cell differentiation involved in prostate gland development' | BestNCBI='epithelial cell differentiation involved in prostate gland development' (diff. references) | Score=0.9999998807907104 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for androgen receptor (11835) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: cJun (ID=16476) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='regulation of cell growth' | BestNCBI='regulation of cell cycle' (diff. references) | Score=0.7136453986167908 | PMIDs=['26862156']\n",
            "LocalTerm='protein binding' | BestNCBI='protein binding' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for cJun (16476) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: MDM2 (ID=17246) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='ubiquitin-protein transferase activity' | BestNCBI='ubiquitin-protein transferase activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='protein ubiquitination' | BestNCBI='protein ubiquitination' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for MDM2 (17246) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: p53 (ID=22060) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='protein binding' | Score=-1.0| PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | Score=-1.0| PMIDs=['26862156']\n",
            "LocalTerm='regulation of cellular process' | Score=-1.0| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for p53 (22060) ==\n",
            "TP = 0, NP = 3, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: CRM1 (ID=103573) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='protein binding' | BestNCBI='protein binding' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='nuclear export' | BestNCBI='protein export from nucleus' (diff. references) | Score=0.9191309213638306 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for CRM1 (103573) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: estrogen receptor (ID=13982) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='protein binding' | BestNCBI='protein binding' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='regulation of transcription by RNA polymerase II' | BestNCBI='regulation of transcription by RNA polymerase II' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for estrogen receptor (13982) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: BRCA1 (ID=12189) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='ubiquitin-protein transferase activity' | BestNCBI='ubiquitin-protein transferase activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='protein ubiquitination' | BestNCBI='protein ubiquitination' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for BRCA1 (12189) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: DNA polymerase Eta (ID=80905) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='DNA-directed DNA polymerase activity' | BestNCBI='DNA-directed DNA polymerase activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='translesion synthesis' | BestNCBI='error-prone translesion synthesis' (diff. references) | Score=0.8004177808761597 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for DNA polymerase Eta (80905) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Pirh2 (ID=68098) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='negative regulation of error-prone translesion synthesis' | Score=0.4970044493675232| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='ubiquitin-protein transferase activity' | BestNCBI='ubiquitin protein ligase activity' (diff. references) | Score=0.818665623664856 | PMIDs=['26862156']\n",
            "LocalTerm='protein ubiquitination' | BestNCBI='protein ubiquitination' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Pirh2 (68098) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TRalpha (ID=6955) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='nuclear transport' | Score=0.4474779963493347| PMIDs=['26862156']\n",
            "LocalTerm='regulation of gene expression' | Score=0.3716229796409607| PMIDs=['26862156']\n",
            "LocalTerm='apoptotic process' | Score=0.4523211419582367| PMIDs=['26862156']\n",
            "LocalTerm='protein binding' | Score=0.5049641132354736| PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | Score=0.26845353841781616| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TRalpha (6955) ==\n",
            "TP = 0, NP = 5, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: PIAS1 (ID=300772) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='SUMO ligase activity' | BestNCBI='SUMO ligase activity' (diff. references) | Score=0.9999999403953552 | PMIDs=['26862156']\n",
            "LocalTerm='protein sumoylation' | BestNCBI='protein sumoylation' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | BestNCBI='nucleus' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for PIAS1 (300772) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: SUMO1 (ID=22218) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of transcription by RNA polymerase II' | Score=0.6381694078445435| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='protein sumoylation' | BestNCBI='protein sumoylation' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for SUMO1 (22218) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: SUMO3 (ID=20610) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='transcription regulator activity' | Score=0.4534270465373993| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='protein sumoylation' | BestNCBI='protein sumoylation' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for SUMO3 (20610) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: TR (ID=2149) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='nuclear receptor activity' | Score=0.6049193143844604| PMIDs=['26862156']\n",
            "LocalTerm='heart process' | Score=0.4251892566680908| PMIDs=['26862156']\n",
            "LocalTerm='nucleus' | Score=0.48052528500556946| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for TR (2149) ==\n",
            "TP = 0, NP = 3, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: org/10 (ID=100035352) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for org/10 (100035352) ==\n",
            "TP = 0, NP = 0, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: PKC{epsilon (ID=18754) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='activation of protein kinase C activity' | Score=0.47128355503082275| PMIDs=['26862156']\n",
            "LocalTerm='cardiac muscle hypertrophy' | Score=0.3066774308681488| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for PKC{epsilon (18754) ==\n",
            "TP = 0, NP = 2, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: cardiac phospholamban (ID=18821) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='heart process' | Score=0.6066915392875671| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='regulation of cardiac muscle contraction' | BestNCBI='regulation of heart contraction' (diff. references) | Score=0.9631195068359375 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for cardiac phospholamban (18821) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: KCND3 (ID=56543) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of transmembrane transport' | Score=0.5110597610473633| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='voltage-gated potassium channel activity' | BestNCBI='voltage-gated potassium channel activity' (diff. references) | Score=1.0000001192092896 | PMIDs=['26862156']\n",
            "LocalTerm='plasma membrane' | BestNCBI='membrane' (diff. references) | Score=0.8591228723526001 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for KCND3 (56543) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: alpha1 and beta (ID=12044) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='nuclear receptor activity' | Score=0.5340927243232727| PMIDs=['26862156']\n",
            "LocalTerm='potassium channel regulator activity' | Score=0.6381615400314331| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for alpha1 and beta (12044) ==\n",
            "TP = 0, NP = 2, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: FGFR1 (ID=14182) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='centrosome' | Score=0.6156402230262756| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='transmembrane receptor protein tyrosine kinase activity' | BestNCBI='protein tyrosine kinase activity' (diff. references) | Score=0.7916347980499268 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for FGFR1 (14182) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Atrogin-1 (ID=67731) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='ubiquitin-protein transferase activity' | Score=0.38717415928840637| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='negative regulation of cardiac muscle hypertrophy' | BestNCBI='negative regulation of cardiac muscle hypertrophy' (diff. references) | Score=0.9999999403953552 | PMIDs=['26862156']\n",
            "LocalTerm='SCF ubiquitin ligase complex' | BestNCBI='SCF ubiquitin ligase complex' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Atrogin-1 (67731) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: Rad52 (ID=19365) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='DNA repair' | BestNCBI='DNA damage response' (diff. references) | Score=0.7743324041366577 | PMIDs=['26862156']\n",
            "LocalTerm='double-strand break repair' | BestNCBI='double-strand break repair via homologous recombination' (diff. references) | Score=0.8471412062644958 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for Rad52 (19365) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: KCNE1 (ID=16509) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='voltage-gated potassium channel activity' | BestNCBI='voltage-gated potassium channel activity' (diff. references) | Score=1.0000001192092896 | PMIDs=['26862156']\n",
            "LocalTerm='regulation of action potential' | BestNCBI='membrane repolarization during action potential' (diff. references) | Score=0.7258103489875793 | PMIDs=['26862156']\n",
            "LocalTerm='plasma membrane' | BestNCBI='plasma membrane' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for KCNE1 (16509) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: ERK (ID=26413) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='signal transduction' | BestNCBI='intracellular signal transduction' (diff. references) | Score=0.9288160800933838 | PMIDs=['26862156']\n",
            "LocalTerm='cardiac muscle cell differentiation' | BestNCBI='cardiac neural crest cell development involved in heart development' (diff. references) | Score=0.8268816471099854 | PMIDs=['26862156']\n",
            "LocalTerm='thyroid hormone receptor signaling pathway' | BestNCBI='steroid hormone receptor signaling pathway' (diff. references) | Score=0.8511527180671692 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for ERK (26413) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: IGF1 (ID=3479) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='growth factor activity' | BestNCBI='growth factor activity' (diff. references) | Score=0.9999998807907104 | PMIDs=['26862156']\n",
            "LocalTerm='heart development' | BestNCBI='muscle organ development' (diff. references) | Score=0.7112216949462891 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for IGF1 (3479) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: IGFR (ID=3480) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='peptide receptor activity' | Score=0.593866765499115| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for IGFR (3480) ==\n",
            "TP = 0, NP = 1, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: PKCalpha (ID=18750) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='protein serine/threonine kinase activity' | BestNCBI='protein serine/threonine kinase activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='cell-cell adhesion' | BestNCBI='cell adhesion' (diff. references) | Score=0.9700780510902405 | PMIDs=['26862156']\n",
            "LocalTerm='cytoplasm' | BestNCBI='cytoplasm' (diff. references) | Score=0.9999999403953552 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for PKCalpha (18750) ==\n",
            "TP = 0, NP = 0, DCU = 3, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: p120-catenin (ID=12388) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='cell junction disassembly' | Score=0.6365375518798828| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cell-cell junction' | BestNCBI='cell-cell junction' (diff. references) | Score=0.9999999403953552 | PMIDs=['26862156']\n",
            "LocalTerm='protein binding' | BestNCBI='protein binding' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for p120-catenin (12388) ==\n",
            "TP = 0, NP = 1, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: VE-cadherin (ID=12562) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='cell-cell adhesion' | BestNCBI='cell-cell adhesion' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "LocalTerm='cell junction' | BestNCBI='cell junction' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for VE-cadherin (12562) ==\n",
            "TP = 0, NP = 0, DCU = 2, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: RING Finger-1 (ID=66949) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='negative regulation of JNK cascade' | Score=0.5058084726333618| PMIDs=['26862156']\n",
            "LocalTerm='cardiac muscle atrophy' | Score=0.26624611020088196| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for RING Finger-1 (66949) ==\n",
            "TP = 0, NP = 2, DCU = 0, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: JNK (ID=26419) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "LocalTerm='regulation of cardiac muscle hypertrophy' | Score=0.5357162356376648| PMIDs=['26862156']\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='protein kinase activity' | BestNCBI='protein kinase activity' (diff. references) | Score=1.0 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for JNK (26419) ==\n",
            "TP = 0, NP = 1, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n",
            "\n",
            "=== Checking Gene: E2F1 (ID=13555) ===\n",
            "-- NDs --\n",
            "\n",
            "-- TPs --\n",
            "\n",
            "-- NPs --\n",
            "\n",
            "-- DCUs --\n",
            "LocalTerm='regulation of gene expression' | BestNCBI='positive regulation of gene expression' (diff. references) | Score=0.8302247524261475 | PMIDs=['26862156']\n",
            "\n",
            "-- ND --\n",
            "\n",
            "== Per-gene Summary for E2F1 (13555) ==\n",
            "TP = 0, NP = 0, DCU = 1, ND = 0\n",
            "--------------------------------------------------------\n"
          ]
        }
      ],
      "source": [
        "import json\n",
        "from sentence_transformers import SentenceTransformer, util\n",
        "from Bio import Entrez\n",
        "import xml.etree.ElementTree as ET\n",
        "import time\n",
        "import requests\n",
        "\n",
        "###############################################################################\n",
        "# CONFIGURATION\n",
        "###############################################################################\n",
        "Entrez.email = \"your_email@example.com\"  # Use a valid email address\n",
        "\n",
        "SIMILARITY_THRESHOLD = 0.68\n",
        "MARGIN = 0.1  # The margin by which we'll allow a same-paper match to be \"close enough\"\n",
        "\n",
        "###############################################################################\n",
        "# 1) Utility: map from local \"namespace\" to NCBI's GO Category strings\n",
        "###############################################################################\n",
        "def map_namespace_to_ncbi_category(ns):\n",
        "    if ns == \"biological_process\":\n",
        "        return \"Process\"\n",
        "    elif ns == \"molecular_function\":\n",
        "        return \"Function\"\n",
        "    elif ns == \"cellular_component\":\n",
        "        return \"Component\"\n",
        "    return None\n",
        "\n",
        "###############################################################################\n",
        "# 2) Fetch the GO annotations from NCBI Gene for a given gene_id\n",
        "###############################################################################\n",
        "def fetch_gene_go_terms(gene_id):\n",
        "    \"\"\"\n",
        "    Fetch the NCBI Gene record (XML) for `gene_id` and parse out GO annotations.\n",
        "    Returns a dict of lists, keyed by: \"Function\", \"Process\", \"Component\".\n",
        "    \"\"\"\n",
        "    # Sleep to respect NCBI usage guidelines\n",
        "    time.sleep(0.3)\n",
        "\n",
        "    handle = Entrez.efetch(db=\"gene\", id=gene_id, retmode=\"xml\")\n",
        "    xml_data = handle.read()\n",
        "    handle.close()\n",
        "\n",
        "    root = ET.fromstring(xml_data)\n",
        "    go_annotations = {}\n",
        "\n",
        "    for gene_commentary in root.findall(\".//Gene-commentary\"):\n",
        "        heading_elem = gene_commentary.find(\"Gene-commentary_heading\")\n",
        "        if heading_elem is not None and heading_elem.text == \"GeneOntology\":\n",
        "            for category_block in gene_commentary.findall(\".//Gene-commentary_comment/Gene-commentary\"):\n",
        "                label_elem = category_block.find(\"Gene-commentary_label\")\n",
        "                if label_elem is None:\n",
        "                    continue\n",
        "                category_label = label_elem.text  # \"Function\", \"Process\", or \"Component\"\n",
        "\n",
        "                if category_label not in go_annotations:\n",
        "                    go_annotations[category_label] = []\n",
        "\n",
        "                inner_comment = category_block.find(\"Gene-commentary_comment\")\n",
        "                if inner_comment is None:\n",
        "                    continue\n",
        "\n",
        "                for go_term_block in inner_comment.findall(\"Gene-commentary\"):\n",
        "                    dbtag_db = go_term_block.find(\".//Dbtag_db\")\n",
        "                    if dbtag_db is not None and dbtag_db.text == \"GO\":\n",
        "                        # Parse the GO ID\n",
        "                        object_id_elem = go_term_block.find(\".//Object-id_id\")\n",
        "                        if object_id_elem is not None:\n",
        "                            numeric_id = object_id_elem.text\n",
        "                            padded_id = numeric_id.zfill(7)\n",
        "                            go_id = f\"GO:{padded_id}\"\n",
        "                        else:\n",
        "                            go_id = \"GO:unknown\"\n",
        "\n",
        "                        # Parse the GO term name\n",
        "                        anchor_elem = go_term_block.find(\".//Other-source_anchor\")\n",
        "                        go_term_name = anchor_elem.text if anchor_elem is not None else \"Unknown Term\"\n",
        "\n",
        "                        # Evidence code\n",
        "                        evidence_code = None\n",
        "                        post_text_elem = go_term_block.find(\".//Other-source_post-text\")\n",
        "                        if post_text_elem is not None:\n",
        "                            post_text_str = post_text_elem.text\n",
        "                            if post_text_str and \"evidence: \" in post_text_str:\n",
        "                                evidence_code = post_text_str.split(\"evidence:\")[-1].strip()\n",
        "                            else:\n",
        "                                evidence_code = post_text_str\n",
        "\n",
        "                        # PubMed IDs\n",
        "                        pubmed_ids = []\n",
        "                        refs_block = go_term_block.find(\"Gene-commentary_refs\")\n",
        "                        if refs_block is not None:\n",
        "                            for pub in refs_block.findall(\"Pub\"):\n",
        "                                pmid_elem = pub.find(\".//PubMedId\")\n",
        "                                if pmid_elem is not None:\n",
        "                                    pubmed_ids.append(pmid_elem.text)\n",
        "\n",
        "                        go_annotations[category_label].append({\n",
        "                            \"go_id\": go_id,\n",
        "                            \"term\": go_term_name,\n",
        "                            \"evidence\": evidence_code,\n",
        "                            \"pubmed_ids\": pubmed_ids\n",
        "                        })\n",
        "\n",
        "    return go_annotations\n",
        "\n",
        "###############################################################################\n",
        "# 3) Utility: compute cosine similarity using Sentence-BERT\n",
        "###############################################################################\n",
        "def compute_semantic_similarity(term1, term2, model):\n",
        "    emb1 = model.encode(term1, convert_to_tensor=True)\n",
        "    emb2 = model.encode(term2, convert_to_tensor=True)\n",
        "    cos_sim = util.cos_sim(emb1, emb2)\n",
        "    return float(cos_sim[0][0])\n",
        "\n",
        "###############################################################################\n",
        "# 4) Utility: convert a PMCID to PMID using NCBI E-utilities\n",
        "###############################################################################\n",
        "pmc_cache = {} # Cache for PMCID to PMID conversion\n",
        "\n",
        "def pmcid_to_pmid(pmcid):\n",
        "    \"\"\"\n",
        "    Convert a PMCID (e.g. 'PMC123456') to a numeric PMID string using NCBI E-utilities.\n",
        "    Returns None if not found or conversion fails.\n",
        "    \"\"\"\n",
        "    # cache since we dont want to hit the same PMCID multiple times\n",
        "    if pmcid in pmc_cache:\n",
        "        return pmc_cache[pmcid]\n",
        "    \n",
        "    base = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/\"\n",
        "    url = base + f\"esummary.fcgi?db=pmc&id={pmcid[3:]}&retmode=xml\"\n",
        "    # Note pmcid[3:] strips \"PMC\" prefix if we pass \"PMC123456\" to the 'id=' param\n",
        "\n",
        "    resp = requests.get(url)\n",
        "    if resp.status_code != 200:\n",
        "        return None\n",
        "\n",
        "    root = ET.fromstring(resp.text)\n",
        "\n",
        "    for item in root.findall(\".//Item\"):\n",
        "        if item.get(\"Name\") == \"pmid\":\n",
        "            pmid = item.text\n",
        "            pmc_cache[pmcid] = pmid\n",
        "            return pmid\n",
        "\n",
        "    return None\n",
        "\n",
        "def convert_local_pmcids_to_pmids(local_data):\n",
        "    for gene_entry in local_data:\n",
        "        for local_ann in gene_entry.get(\"go_data\", []):\n",
        "            refs = local_ann.get(\"reference(s)\", [])\n",
        "            new_refs = []\n",
        "            for r in refs:\n",
        "                # If r starts with 'PMC', attempt to convert\n",
        "                if r.startswith(\"PMC\"):\n",
        "                    time.sleep(0.3)  # Respect NCBI usage guidelines\n",
        "                    pmid = pmcid_to_pmid(r)\n",
        "                    if pmid:\n",
        "                        new_refs.append(pmid)\n",
        "                    else:\n",
        "                        # Conversion failed or none found, fallback\n",
        "                        new_refs.append(r)\n",
        "                else:\n",
        "                    new_refs.append(r)\n",
        "            local_ann[\"reference(s)\"] = new_refs\n",
        "\n",
        "###############################################################################\n",
        "# 5) Main comparison & classification logic with tie-break margin\n",
        "###############################################################################\n",
        "def compare_json_vs_ncbi(json_path, model):\n",
        "    \"\"\"\n",
        "    For each local annotation (term + references), we:\n",
        "      1) Gather all possible NCBI annotations in the same GO category\n",
        "      2) Find best match overall & best match among same-paper (i.e., shared PMIDs)\n",
        "      3) If best_same_score >= best_overall_score - MARGIN, prefer the same-paper match\n",
        "      4) Classify that chosen match as TP if it shares references, else DCU (if >= threshold),\n",
        "         or NP if below threshold.\n",
        "    \"\"\"\n",
        "    with open(json_path, \"r\") as f:\n",
        "        local_data = json.load(f)\n",
        "\n",
        "    # Convert PMCIDs in local_data to numeric PMIDs since NCBI uses PMIDs even thorugh GO annotations are curated from PMC\n",
        "    convert_local_pmcids_to_pmids(local_data)\n",
        "\n",
        "    all_comparisons = []\n",
        "\n",
        "    for gene_entry in local_data:\n",
        "        gene_name = gene_entry[\"gene\"]\n",
        "        gene_id = gene_entry[\"geneID\"]\n",
        "        local_go_data = gene_entry.get(\"go_data\", [])\n",
        "\n",
        "        print(f\"\\n=== Checking Gene: {gene_name} (ID={gene_id}) ===\")\n",
        "\n",
        "        ncbi_annotations = fetch_gene_go_terms(gene_id)\n",
        "\n",
        "        # We'll keep track of official GO annotations (for ND detection)\n",
        "        official_by_pmid = set()\n",
        "        for category_label, ann_list in ncbi_annotations.items():\n",
        "            for ann in ann_list:\n",
        "                for pmid in ann[\"pubmed_ids\"]:\n",
        "                    official_by_pmid.add((category_label, ann[\"term\"], pmid))\n",
        "\n",
        "        matched_officials = set()\n",
        "        comparisons_this_gene = []\n",
        "\n",
        "        # Per-gene counters\n",
        "        count_tp = 0\n",
        "        count_np = 0\n",
        "        count_dcu = 0\n",
        "\n",
        "        #\n",
        "        # 1) CLASSIFY EACH LOCAL ANNOTATION (TP, NP, DCU)\n",
        "        #\n",
        "        for local_ann in local_go_data:\n",
        "            local_term = local_ann[\"term\"]\n",
        "            local_refs = set(local_ann.get(\"reference(s)\", []))\n",
        "            local_ns = local_ann.get(\"namespace\", None)\n",
        "            target_category = map_namespace_to_ncbi_category(local_ns)\n",
        "\n",
        "            classification = None\n",
        "            chosen_match_info = None\n",
        "            chosen_score = -1.0\n",
        "\n",
        "            best_score_all = -1.0\n",
        "            best_info_all = None\n",
        "            best_score_same = -1.0\n",
        "            best_info_same = None\n",
        "\n",
        "            if target_category and target_category in ncbi_annotations:\n",
        "                possible_anns = ncbi_annotations[target_category]\n",
        "\n",
        "                # (A) Find best match overall\n",
        "                for ann in possible_anns:\n",
        "                    sc = compute_semantic_similarity(local_term, ann[\"term\"], model)\n",
        "                    if sc > best_score_all:\n",
        "                        best_score_all = sc\n",
        "                        best_info_all = ann\n",
        "\n",
        "                # (B) Find best match among same-paper\n",
        "                same_paper_anns = [\n",
        "                    ann for ann in possible_anns\n",
        "                    if set(ann[\"pubmed_ids\"]).intersection(local_refs)\n",
        "                ]\n",
        "                for ann in same_paper_anns:\n",
        "                    sc = compute_semantic_similarity(local_term, ann[\"term\"], model)\n",
        "                    if sc > best_score_same:\n",
        "                        best_score_same = sc\n",
        "                        best_info_same = ann\n",
        "\n",
        "                # (C) Tie-break\n",
        "                if best_score_same >= 0:\n",
        "                    # There's at least one same-paper annotation\n",
        "                    if best_score_same >= best_score_all - MARGIN:\n",
        "                        chosen_score = best_score_same\n",
        "                        chosen_match_info = best_info_same\n",
        "                    else:\n",
        "                        chosen_score = best_score_all\n",
        "                        chosen_match_info = best_info_all\n",
        "                else:\n",
        "                    chosen_score = best_score_all\n",
        "                    chosen_match_info = best_info_all\n",
        "\n",
        "                # (D) Classify\n",
        "                if chosen_score >= SIMILARITY_THRESHOLD and chosen_match_info is not None:\n",
        "                    overlap = set(chosen_match_info[\"pubmed_ids\"]).intersection(local_refs)\n",
        "                    if overlap:\n",
        "                        classification = \"TP\"\n",
        "                        count_tp += 1\n",
        "                        # Mark the official annotation as matched\n",
        "                        for pmid in chosen_match_info[\"pubmed_ids\"]:\n",
        "                            if pmid in local_refs:\n",
        "                                matched_officials.add((target_category, chosen_match_info[\"term\"], pmid))\n",
        "                    else:\n",
        "                        classification = \"DCU\"\n",
        "                        count_dcu += 1\n",
        "                else:\n",
        "                    classification = \"NP\"\n",
        "                    count_np += 1\n",
        "            else:\n",
        "                # No recognized category => NP\n",
        "                classification = \"NP\"\n",
        "                count_np += 1\n",
        "\n",
        "            rec = {\n",
        "                \"gene\": gene_name,\n",
        "                \"geneID\": gene_id,\n",
        "                \"local_term\": local_term,\n",
        "                \"namespace\": local_ns,\n",
        "                \"references_used\": list(local_refs),\n",
        "                \"classification\": classification,\n",
        "                \"best_score\": chosen_score,\n",
        "            }\n",
        "            if chosen_match_info:\n",
        "                rec[\"best_ncbi_match\"] = chosen_match_info[\"term\"]\n",
        "                rec[\"best_ncbi_go_id\"] = chosen_match_info[\"go_id\"]\n",
        "                rec[\"best_ncbi_pubmed_ids\"] = chosen_match_info[\"pubmed_ids\"]\n",
        "            else:\n",
        "                rec[\"best_ncbi_match\"] = None\n",
        "                rec[\"best_ncbi_go_id\"] = None\n",
        "                rec[\"best_ncbi_pubmed_ids\"] = []\n",
        "\n",
        "            comparisons_this_gene.append(rec)\n",
        "\n",
        "        #\n",
        "        # 2) DETECT ND (NOT DETECTED) AND STORE AS ADDITIONAL RECORDS\n",
        "        #\n",
        "        # ND means: official NCBI annotation referencing the local paper(s),\n",
        "        # but never matched as a TP above.\n",
        "        #\n",
        "        undetected = official_by_pmid - matched_officials\n",
        "\n",
        "        # The local paper set is all refs mentioned in local_go_data\n",
        "        all_local_pmids_for_gene = set()\n",
        "        for ann in local_go_data:\n",
        "            all_local_pmids_for_gene |= set(ann.get(\"reference(s)\", []))\n",
        "\n",
        "        actual_nd = []\n",
        "        for (cat_label, term_text, pmid) in undetected:\n",
        "            if pmid in all_local_pmids_for_gene:\n",
        "                actual_nd.append((cat_label, term_text, pmid))\n",
        "\n",
        "        count_nd = len(actual_nd)\n",
        "\n",
        "        # Storing ND in comparisons_this_gene\n",
        "        # For each official annotation that is ND, we add a new record\n",
        "        for (cat_label, ncbi_term, pmid) in actual_nd:\n",
        "            rec = {\n",
        "                \"gene\": gene_name,\n",
        "                \"geneID\": gene_id,\n",
        "                \"local_term\": None,  # No local annotation matched\n",
        "                \"namespace\": cat_label,\n",
        "                \"references_used\": [pmid],  # or empty if you prefer\n",
        "                \"classification\": \"ND\",\n",
        "                \"best_score\": None,\n",
        "                # We might store the official annotation in 'best_ncbi_match'\n",
        "                \"best_ncbi_match\": ncbi_term,\n",
        "                \"best_ncbi_go_id\": None,  # We didn't parse that here, but could\n",
        "                \"best_ncbi_pubmed_ids\": [pmid],\n",
        "            }\n",
        "            comparisons_this_gene.append(rec)\n",
        "\n",
        "        #\n",
        "        # 3) PRINT RESULTS\n",
        "        #\n",
        "        print(\"-- NDs --\")\n",
        "        for nd_item in actual_nd:\n",
        "            print(f\"Category={nd_item[0]}, Term='{nd_item[1]}', PMID={nd_item[2]}\")\n",
        "\n",
        "        # Summaries\n",
        "        tps_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"TP\"]\n",
        "        nps_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"NP\"]\n",
        "        dcus_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"DCU\"]\n",
        "        nds_for_gene = [x for x in comparisons_this_gene if x[\"classification\"] == \"ND\"]\n",
        "\n",
        "        print(\"\\n-- TPs --\")\n",
        "        for item in tps_for_gene:\n",
        "            print(f\"LocalTerm='{item['local_term']}' | BestNCBI='{item['best_ncbi_match']}' \"\n",
        "                  f\"| Score={item['best_score']:.2f} | PMIDs={item['references_used']}\")\n",
        "\n",
        "        print(\"\\n-- NPs --\")\n",
        "        for item in nps_for_gene:\n",
        "            print(f\"LocalTerm='{item['local_term']}' | Score={item['best_score']}\"\n",
        "                  f\"| PMIDs={item['references_used']}\")\n",
        "\n",
        "        print(\"\\n-- DCUs --\")\n",
        "        for item in dcus_for_gene:\n",
        "            print(f\"LocalTerm='{item['local_term']}' | BestNCBI='{item['best_ncbi_match']}' \"\n",
        "                  f\"(diff. references) | Score={item['best_score']} | PMIDs={item['references_used']}\")\n",
        "\n",
        "        print(\"\\n-- ND --\")\n",
        "        for item in nds_for_gene:\n",
        "            print(f\"NCBITerm='{item['best_ncbi_match']}' | PMID={item['best_ncbi_pubmed_ids']}\")\n",
        "        \n",
        "        print(f\"\\n== Per-gene Summary for {gene_name} ({gene_id}) ==\")\n",
        "        print(f\"TP = {count_tp}, NP = {count_np}, DCU = {count_dcu}, ND = {count_nd}\")\n",
        "        print(\"--------------------------------------------------------\")\n",
        "\n",
        "        all_comparisons.extend(comparisons_this_gene)\n",
        "\n",
        "    return all_comparisons\n",
        "\n",
        "###############################################################################\n",
        "# 6) Example usage\n",
        "###############################################################################\n",
        "json_file_path = \"./processed_inferences/processed_output_PMC5453669.json\"\n",
        "\n",
        "try:\n",
        "    comparison_results = compare_json_vs_ncbi(json_file_path, sapbert_model)\n",
        "except requests.HTTPError as e:\n",
        "    print(f\"HTTPError: {e.code} - {e.reason}\")\n",
        "    print(\"Please check the gene ID or ensure that the NCBI API is accessible.\")\n",
        "\n",
        "# # Now each item in comparison_results has classification in {TP, NP, DCU, ND}\n",
        "\n",
        "# # You can also sum ND from the final list if you want a global ND count\n",
        "# tp_global = sum(1 for r in comparison_results if r[\"classification\"] == \"TP\")\n",
        "# np_global = sum(1 for r in comparison_results if r[\"classification\"] == \"NP\")\n",
        "# dcu_global = sum(1 for r in comparison_results if r[\"classification\"] == \"DCU\")\n",
        "# nd_global = sum(1 for r in comparison_results if r[\"classification\"] == \"ND\")\n",
        "\n",
        "# print(\"\\n=== Global Summary Across All Genes ===\")\n",
        "# print(f\"Total entries: {len(comparison_results)}\")\n",
        "# print(f\"TP = {tp_global}, NP = {np_global}, DCU = {dcu_global}, ND = {nd_global}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Will put TP, ND, DCU and NP in a excel file in the output folder\n",
        "change the sheet_name to the name of the file you want to save it as in the excel. After that, open excel and copy the sheet to the main sheet. Then delete the other sheet."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import os\n",
        "from openpyxl import load_workbook\n",
        "from collections import defaultdict\n",
        "\n",
        "###############################################################################\n",
        "# 1) Aggregate classification_counts from comparison_results\n",
        "###############################################################################\n",
        "classification_counts = defaultdict(lambda: {\"TP\": 0, \"NP\": 0, \"DCU\": 0, \"ND\": 0})\n",
        "\n",
        "for row in comparison_results:\n",
        "    gene_id = row[\"geneID\"]\n",
        "    gene_name = row[\"gene\"]\n",
        "    classification = row[\"classification\"]\n",
        "    \n",
        "    # For each PMID in references_used, increment the classification count\n",
        "    for pmid in row[\"references_used\"]:\n",
        "        key = (gene_id, gene_name, pmid)\n",
        "        classification_counts[key][classification] += 1\n",
        "\n",
        "records_for_df = []\n",
        "for (gid, gname, pmid), counts in classification_counts.items():\n",
        "    records_for_df.append({\n",
        "        \"geneID\": gid,\n",
        "        \"gene\": gname,\n",
        "        \"pmid\": pmid,\n",
        "        \"TP\": counts[\"TP\"],\n",
        "        \"ND\": counts[\"ND\"],\n",
        "        \"NP\": counts[\"NP\"],\n",
        "        \"DCU\": counts[\"DCU\"]\n",
        "    })\n",
        "\n",
        "df = pd.DataFrame(\n",
        "    records_for_df,\n",
        "    columns=[\"geneID\", \"gene\", \"pmid\", \"TP\", \"ND\", \"NP\", \"DCU\"]\n",
        ")\n",
        "\n",
        "###############################################################################\n",
        "# 2) Write to Excel, appending a NEW SHEET for each paper\n",
        "###############################################################################\n",
        "output_path = r\"C:\\Users\\aivan\\OneDrive\\Documents\\Evaluation.xlsx\"\n",
        "\n",
        "# You can change this to something unique for each paper you analyze.\n",
        "sheet_name = \"PMC5453669\"\n",
        "\n",
        "# If the file already exists, open in append mode. Otherwise, create a new file.\n",
        "mode = \"a\" if os.path.exists(output_path) else \"w\"\n",
        "\n",
        "if mode == \"w\":\n",
        "    # Create a new file and write data\n",
        "    df.to_excel(output_path, sheet_name=sheet_name, index=False)\n",
        "else:\n",
        "    # If the file already exists, append a new sheet\n",
        "    with pd.ExcelWriter(output_path, engine=\"openpyxl\", mode=\"a\", if_sheet_exists=\"new\") as writer:\n",
        "        df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
        "\n",
        "print(f\"Finished! Data written to sheet '{sheet_name}' in file: {output_path}\")\n"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "collapsed_sections": [
        "6oaKSo6iXGOc"
      ],
      "gpuType": "T4",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "BIOIN311C1",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.11"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "097046dc7c4f46f5866068c82b706aca": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "1675d70c8a5e426691d0af5e6ba333f3": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_784d7bf8c3be48b8ad5a3b6e6954690f",
              "IPY_MODEL_730fe49f91204e2cbd3b58d24442be0a",
              "IPY_MODEL_792eb1f8daf34e2590fda2f60e3ff684"
            ],
            "layout": "IPY_MODEL_3e0f9cdf58ef4af1a0b16f9795747dfc"
          }
        },
        "1845ddb7ff634d969fa2538209fc4e0a": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "1b8c25318a2b4b64bab3496083679c85": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "2443ed122977428c98248537123b674b": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_f7dcd748d0944da09948f4372366e8a8",
            "placeholder": "​",
            "style": "IPY_MODEL_466f8ef6ba644d31a8586abaf9830448",
            "value": " 438M/438M [00:01&lt;00:00, 251MB/s]"
          }
        },
        "3117723c389b4e47a97d6e582078e5c9": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "32ef3e232304493d8b7e129c679f4f6f": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_b1c01cfdd57e40e7bc4de9baf81d9f45",
            "placeholder": "​",
            "style": "IPY_MODEL_82a3774cc5cd4acfb1822d9cb74d6f45",
            "value": " 112/112 [00:00&lt;00:00, 9.27kB/s]"
          }
        },
        "331dbfdb6d7849c6a7b6d43bd33a7325": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_8eb65674477e477091f2501704df7eb8",
            "placeholder": "​",
            "style": "IPY_MODEL_6d29b58c64cc4f869d6944699149a60e",
            "value": "model.safetensors: 100%"
          }
        },
        "39794838a2164151a5edf298a63f8147": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "3e0f9cdf58ef4af1a0b16f9795747dfc": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "466f8ef6ba644d31a8586abaf9830448": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "47571c7570f14ab58f96e5d30129da1b": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "4a743c94312c4b5db25087a747f6b879": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_ea731973ab9c44caa479fb7934fd3b4f",
            "max": 198,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_097046dc7c4f46f5866068c82b706aca",
            "value": 198
          }
        },
        "4af857d8e5a94c278ae0fb3d7e092ac5": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "53bb7f0fbad34a818e266d0ea44ffaf1": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "54a0cadcc9ef406db05d36b690d7891c": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c939f017049643b18d5142514937af73",
            "max": 437955508,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_6c463c9a4bef4b5194113acf7ced731e",
            "value": 437955508
          }
        },
        "59dc6c4216cb4c5aad8c3764cf3d3f6d": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_cc7573c531c84e61999e2161710173f3",
            "placeholder": "​",
            "style": "IPY_MODEL_9775c0cd37bc43dcb7953f8b15343332",
            "value": " 198/198 [00:00&lt;00:00, 14.3kB/s]"
          }
        },
        "618fd6a5462e4c5296b560f888b5d638": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_331dbfdb6d7849c6a7b6d43bd33a7325",
              "IPY_MODEL_54a0cadcc9ef406db05d36b690d7891c",
              "IPY_MODEL_2443ed122977428c98248537123b674b"
            ],
            "layout": "IPY_MODEL_c3911d8aca764dd38fad4c1dded8198f"
          }
        },
        "61de99be2c2e4afe90ff06e3bb30a2b7": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "6ac927f5e2a2453d9aae4b5c474cf96d": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_7593152124e04332b709dd30ec87c5a5",
              "IPY_MODEL_4a743c94312c4b5db25087a747f6b879",
              "IPY_MODEL_59dc6c4216cb4c5aad8c3764cf3d3f6d"
            ],
            "layout": "IPY_MODEL_3117723c389b4e47a97d6e582078e5c9"
          }
        },
        "6c463c9a4bef4b5194113acf7ced731e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "6d29b58c64cc4f869d6944699149a60e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "730fe49f91204e2cbd3b58d24442be0a": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_d598fdb3d94a4095ba5a950cef321f1b",
            "max": 226150,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_39794838a2164151a5edf298a63f8147",
            "value": 226150
          }
        },
        "7593152124e04332b709dd30ec87c5a5": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_c6b5cf46fbeb41adb8da935949b808c1",
            "placeholder": "​",
            "style": "IPY_MODEL_d4b8ff5330d14128ba304020f7fe8412",
            "value": "tokenizer_config.json: 100%"
          }
        },
        "769b37c671334690b6a7e5ba9059343e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_bdf351e0ccbc4aa18e168f593d1659cf",
              "IPY_MODEL_8901a8884437441c934c53a98a1dac06",
              "IPY_MODEL_32ef3e232304493d8b7e129c679f4f6f"
            ],
            "layout": "IPY_MODEL_77472246b6b0497fbe67b8a3529040e5"
          }
        },
        "77472246b6b0497fbe67b8a3529040e5": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "784d7bf8c3be48b8ad5a3b6e6954690f": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_1845ddb7ff634d969fa2538209fc4e0a",
            "placeholder": "​",
            "style": "IPY_MODEL_9b03af45365c42419bb39779b82be47c",
            "value": "vocab.txt: 100%"
          }
        },
        "792eb1f8daf34e2590fda2f60e3ff684": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_e7eaa149a903445db37dfdf719e14f9e",
            "placeholder": "​",
            "style": "IPY_MODEL_4af857d8e5a94c278ae0fb3d7e092ac5",
            "value": " 226k/226k [00:00&lt;00:00, 1.03MB/s]"
          }
        },
        "7d97b347b7cb40df81041035931d7f27": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_53bb7f0fbad34a818e266d0ea44ffaf1",
            "placeholder": "​",
            "style": "IPY_MODEL_bc435a026437452bb820c14d1ce5d3ee",
            "value": "config.json: 100%"
          }
        },
        "8039b35f8f654a5e93ab69a0d1d8f3ea": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "82a3774cc5cd4acfb1822d9cb74d6f45": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "873ea6832a0447a19938fe764cb7ac58": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "8901a8884437441c934c53a98a1dac06": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_b483173c5fa74a72bb0a88206d3206a3",
            "max": 112,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_8039b35f8f654a5e93ab69a0d1d8f3ea",
            "value": 112
          }
        },
        "8eb65674477e477091f2501704df7eb8": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "9775c0cd37bc43dcb7953f8b15343332": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "9b03af45365c42419bb39779b82be47c": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "b1c01cfdd57e40e7bc4de9baf81d9f45": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b483173c5fa74a72bb0a88206d3206a3": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b85f6bb59d024282b5552ff665da4548": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_47571c7570f14ab58f96e5d30129da1b",
            "max": 462,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_ec42f776af1b48a982636f0c122ab7e6",
            "value": 462
          }
        },
        "bc435a026437452bb820c14d1ce5d3ee": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "bdf351e0ccbc4aa18e168f593d1659cf": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_61de99be2c2e4afe90ff06e3bb30a2b7",
            "placeholder": "​",
            "style": "IPY_MODEL_873ea6832a0447a19938fe764cb7ac58",
            "value": "special_tokens_map.json: 100%"
          }
        },
        "c3911d8aca764dd38fad4c1dded8198f": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "c6b5cf46fbeb41adb8da935949b808c1": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "c939f017049643b18d5142514937af73": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "cc7573c531c84e61999e2161710173f3": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "d4b8ff5330d14128ba304020f7fe8412": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "d598fdb3d94a4095ba5a950cef321f1b": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "de6e3ef0670042378384c41ecd695f1a": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "e4718c03421b41a2a665c26293acd697": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_7d97b347b7cb40df81041035931d7f27",
              "IPY_MODEL_b85f6bb59d024282b5552ff665da4548",
              "IPY_MODEL_f1219528a46a4a8e94bd1c8de9e6e16a"
            ],
            "layout": "IPY_MODEL_1b8c25318a2b4b64bab3496083679c85"
          }
        },
        "e7eaa149a903445db37dfdf719e14f9e": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ea731973ab9c44caa479fb7934fd3b4f": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ec42f776af1b48a982636f0c122ab7e6": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "ec7200455fa34863bec2ae60f9f4aad5": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "f1219528a46a4a8e94bd1c8de9e6e16a": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_de6e3ef0670042378384c41ecd695f1a",
            "placeholder": "​",
            "style": "IPY_MODEL_ec7200455fa34863bec2ae60f9f4aad5",
            "value": " 462/462 [00:00&lt;00:00, 40.9kB/s]"
          }
        },
        "f7dcd748d0944da09948f4372366e8a8": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        }
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
